
<!doctype html>
<html lang="es" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://training.nextflow.io/0.dev/es/nf4_science/genomics/02_joint_calling/">
      
      
        <link rel="prev" href="../01_per_sample_variant_calling/">
      
      
        <link rel="next" href="../03_modules/">
      
      
      <link rel="icon" href="../../../../assets/img/nextflow-icon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Part 2: Joint calling on a cohort - training.nextflow.io</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/fonts.css">
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-244N3GEN75"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-244N3GEN75",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-244N3GEN75",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>if("undefined"!=typeof __md_analytics){var consent=__md_get("__consent");consent&&consent.analytics&&__md_analytics()}</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#part-2-joint-calling-on-a-cohort" class="md-skip">
          Saltar a contenido
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
            
<div style="text-align: center">
  Nextflow Training Week (2025-Q1):
  <strong>March 10-14, 2025</strong>
  <a
    class="md-button"
    href="https://community.seqera.io/t/nextflow-training-week-2025-q1/1775/6"
    target="_blank"
  >
    More info
    <span class="twemoji"
      ><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path
          d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"
        ></path></svg
    ></span>
  </a>
</div>
<style type="text/css">
  .md-banner {
    display: none;
  }
</style>

<script>
  (function () {
    function hideAnnouncement() {
      var announcementDate = new Date(
        "2025-03-15T00:00:00Z",
      );
      var currentDate = new Date();
      var banner = document.querySelector(".md-banner");
      if (banner && currentDate <= announcementDate) {
        banner.style.display = "block";
      }
    }

    // Run on initial load
    hideAnnouncement();

    // Hook into XHR to run after each page update
    var originalOpen = XMLHttpRequest.prototype.open;
    XMLHttpRequest.prototype.open = function () {
      this.addEventListener("load", function () {
        setTimeout(hideAnnouncement, 0); // Ensure it runs after the request completes
      });
      return originalOpen.apply(this, arguments);
    };
  })();
</script>
 
          </div>
          
        </aside>
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Cabecera">
    <a href="../../../" title="training.nextflow.io" class="md-header__button md-logo" aria-label="training.nextflow.io" data-md-component="logo">
      
  <img src="../../../../assets/img/nextflow_logo_dark.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            training.nextflow.io
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Part 2: Joint calling on a cohort
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
      <div class="md-header__option">
  <div class="md-select">
    
    <button class="md-header__button md-icon" aria-label="Seleccionar idioma">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12.87 15.07-2.54-2.51.03-.03A17.5 17.5 0 0 0 14.07 6H17V4h-7V2H8v2H1v2h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2zm-2.62 7 1.62-4.33L19.12 17z"/></svg>
    </button>
    <div class="md-select__inner">
      <ul class="md-select__list">
        
          <li class="md-select__item">
            <a href="../../../../nf4_science/genomics/02_joint_calling/" hreflang="en" class="md-select__link">
              English
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../pt/nf4_science/genomics/02_joint_calling/" hreflang="pt" class="md-select__link">
              Português
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="./" hreflang="es" class="md-select__link">
              Español
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../fr/nf4_science/genomics/02_joint_calling/" hreflang="fr" class="md-select__link">
              Français
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../it/nf4_science/genomics/02_joint_calling/" hreflang="it" class="md-select__link">
              Italiano
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../ko/nf4_science/genomics/02_joint_calling/" hreflang="ko" class="md-select__link">
              Korean
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</div>
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Búsqueda" placeholder="Búsqueda" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Buscar">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Compartir" aria-label="Compartir" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Limpiar" aria-label="Limpiar" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Inicializando búsqueda
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/nextflow-io/training" title="Ir al repositorio" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    nextflow-io/training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navegación" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../" title="training.nextflow.io" class="md-nav__button md-logo" aria-label="training.nextflow.io" data-md-component="logo">
      
  <img src="../../../../assets/img/nextflow_logo_dark.png" alt="logo">

    </a>
    training.nextflow.io
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/nextflow-io/training" title="Ir al repositorio" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    nextflow-io/training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Environment setup
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Environment setup
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Environment Setup
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/01_setup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    GitHub Codespaces
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/02_local/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/03_devcontainer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local installation using VSCode Devcontainers extension
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Hello Nextflow
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Hello Nextflow
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hello Nextflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/01_hello_world/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Hello World
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/02_hello_channels/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Hello Channels
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/03_hello_workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Hello Workflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/04_hello_modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 4: Hello Modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/05_hello_containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 5: Hello Containers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/06_hello_config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 6: Hello Config
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Hello nf-core
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Hello nf-core
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hello nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/01_run_demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Run a demo pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/02_rewrite_hello/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Rewrite Hello for nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Nextflow for Genomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Nextflow for Genomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow for Genomics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../01_per_sample_variant_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Per-sample variant calling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Part 2: Joint calling on a cohort
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Part 2: Joint calling on a cohort
    
  </span>
  

      </a>
      
          

<nav class="md-nav md-nav--secondary" aria-label="On this page">
       
  <label class="md-nav__title" for="__toc">
    <span class="md-nav__icon md-icon"></span>
    On this page
  </label>
  <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
     <li class="md-nav__item">
  <a href="#method-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Method overview
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#0-warmup-run-samtools-and-gatk-directly" class="md-nav__link">
    <span class="md-ellipsis">
      0. Warmup: Run Samtools and GATK directly
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0. Warmup: Run Samtools and GATK directly">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#01-index-a-bam-input-file-with-samtools" class="md-nav__link">
    <span class="md-ellipsis">
      0.1. Index a BAM input file with Samtools
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.1. Index a BAM input file with Samtools">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#011-spin-up-the-samtools-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.1. Spin up the Samtools container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#012-run-the-indexing-command-for-the-three-samples" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.2. Run the indexing command for the three samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#013-exit-the-samtools-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.3. Exit the Samtools container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#02-call-variants-with-gatk-haplotypecaller-in-gvcf-mode" class="md-nav__link">
    <span class="md-ellipsis">
      0.2. Call variants with GATK HaplotypeCaller in GVCF mode
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.2. Call variants with GATK HaplotypeCaller in GVCF mode">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#021-spin-up-the-gatk-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.1. Spin up the GATK container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#022-run-the-variant-calling-command-with-the-gvcf-option" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.2. Run the variant calling command with the GVCF option
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#023-repeat-the-process-on-the-other-two-samples" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.3. Repeat the process on the other two samples
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#03-run-joint-genotyping" class="md-nav__link">
    <span class="md-ellipsis">
      0.3. Run joint genotyping
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.3. Run joint genotyping">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#031-combine-all-the-per-sample-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      0.3.1. Combine all the per-sample GVCFs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#032-run-the-joint-genotyping-analysis-proper" class="md-nav__link">
    <span class="md-ellipsis">
      0.3.2. Run the joint genotyping analysis proper
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#033-exit-the-gatk-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.3.3. Exit the GATK container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#1-modify-the-per-sample-variant-calling-step-to-produce-a-gvcf" class="md-nav__link">
    <span class="md-ellipsis">
      1. Modify the per-sample variant calling step to produce a GVCF
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Modify the per-sample variant calling step to produce a GVCF">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-tell-haplotypecaller-to-emit-a-gvcf-and-update-the-output-extension" class="md-nav__link">
    <span class="md-ellipsis">
      1.1. Tell HaplotypeCaller to emit a GVCF and update the output extension
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-run-the-pipeline-to-verify-that-you-can-generate-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      1.2. Run the pipeline to verify that you can generate GVCFs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-update-the-output-file-extension-in-the-process-outputs-block-too" class="md-nav__link">
    <span class="md-ellipsis">
      1.2. Update the output file extension in the process outputs block too
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-run-the-pipeline-again" class="md-nav__link">
    <span class="md-ellipsis">
      1.3. Run the pipeline again
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_1" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_1" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#2-collect-and-combine-the-gvcf-data-across-all-samples" class="md-nav__link">
    <span class="md-ellipsis">
      2. Collect and combine the GVCF data across all samples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Collect and combine the GVCF data across all samples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-define-the-process-that-will-combine-the-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      2.1. Define the process that will combine the GVCFs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-add-a-cohort_name-parameter-with-a-default-value" class="md-nav__link">
    <span class="md-ellipsis">
      2.2. Add a cohort_name parameter with a default value
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-gather-the-outputs-of-gatk_haplotypecaller-across-samples" class="md-nav__link">
    <span class="md-ellipsis">
      2.3. Gather the outputs of GATK_HAPLOTYPECALLER across samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-add-a-call-to-the-workflow-block-to-run-gatk_genomicsdb" class="md-nav__link">
    <span class="md-ellipsis">
      2.4. Add a call to the workflow block to run GATK_GENOMICSDB
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#25-run-the-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      2.5. Run the workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#26-construct-a-command-line-with-a-separate-v-argument-for-each-input-gvcf" class="md-nav__link">
    <span class="md-ellipsis">
      2.6. Construct a command line with a separate -V argument for each input GVCF
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#27-run-the-workflow-to-verify-that-it-generates-the-genomicsdb-output-as-expected" class="md-nav__link">
    <span class="md-ellipsis">
      2.7. Run the workflow to verify that it generates the GenomicsDB output as expected
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_2" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_2" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#3-run-the-joint-genotyping-step-as-part-of-the-same-process" class="md-nav__link">
    <span class="md-ellipsis">
      3. Run the joint genotyping step as part of the same process
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Run the joint genotyping step as part of the same process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31-rename-the-process-from-gatk_genomicsdb-to-gatk_jointgenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      3.1. Rename the process from GATK_GENOMICSDB to GATK_JOINTGENOTYPING
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-add-the-joint-genotyping-command-to-the-gatk_jointgenotyping-process" class="md-nav__link">
    <span class="md-ellipsis">
      3.2. Add the joint genotyping command to the GATK_JOINTGENOTYPING process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33-add-the-reference-genome-files-to-the-gatk_jointgenotyping-process-input-definitions" class="md-nav__link">
    <span class="md-ellipsis">
      3.3. Add the reference genome files to the GATK_JOINTGENOTYPING process input definitions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34-update-the-process-output-definition-to-emit-the-vcf-of-cohort-level-variant-calls" class="md-nav__link">
    <span class="md-ellipsis">
      3.4. Update the process output definition to emit the VCF of cohort-level variant calls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#35-update-the-process-call-from-gatk_genomicsdb-to-gatk_jointgenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      3.5. Update the process call from GATK_GENOMICSDB to GATK_JOINTGENOTYPING
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#36-run-the-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      3.6. Run the workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_3" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_3" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li> 
  </ul>
  
</nav>
<nav class="md-nav md-nav--secondary sidebar_resources" aria-label="Resources">
  <label class="md-nav__title">Resources</label>
  <ul class="md-nav__list">
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/docs/latest/index.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M17.5 14.33c.79 0 1.63.08 2.5.24v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.33-4.5.99v-1.69c1.17-.53 2.67-.8 4.5-.8M13 12.46c1.29-.53 2.79-.79 4.5-.79.79 0 1.63.07 2.5.23v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.34-4.5.99m4.5-3.65c-1.9 0-3.39.32-4.5 1V9.84c1.23-.56 2.73-.84 4.5-.84.79 0 1.63.08 2.5.23v1.55c-.74-.19-1.59-.28-2.5-.28m3.5 8V7c-1.04-.33-2.21-.5-3.5-.5-2.05 0-3.88.5-5.5 1.5v11.5c1.62-1 3.45-1.5 5.5-1.5 1.19 0 2.36.16 3.5.5m-3.5-14c2.35 0 4.19.5 5.5 1.5v14.56c0 .12-.05.24-.16.35-.11.09-.23.17-.34.17-.11 0-.19-.02-.25-.05-1.28-.69-2.87-1.03-4.75-1.03-2.05 0-3.88.5-5.5 1.5-1.34-1-3.17-1.5-5.5-1.5-1.66 0-3.25.36-4.75 1.07-.03.01-.07.01-.12.03-.04.01-.08.02-.13.02-.11 0-.23-.04-.34-.12a.475.475 0 0 1-.16-.35V6c1.34-1 3.18-1.5 5.5-1.5 2.33 0 4.16.5 5.5 1.5 1.34-1 3.17-1.5 5.5-1.5Z"
          ></path>
        </svg>
        Nextflow Docs
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://www.nextflow.io/slack-invite.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M6 15a2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2h2v2m1 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-5m2-8a2 2 0 0 1-2-2 2 2 0 0 1 2-2 2 2 0 0 1 2 2v2H9m0 1a2 2 0 0 1 2 2 2 2 0 0 1-2 2H4a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5m8 2a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2h-2v-2m-1 0a2 2 0 0 1-2 2 2 2 0 0 1-2-2V5a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5m-2 8a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2v-2h2m0-1a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-5Z"
          ></path>
        </svg>
        Nextflow Slack
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="80 50 350 350"
        >
          <style type="text/css">
            .st0 {
              fill: "gray";
            }
          </style>
          <path
            class="st0"
            d="M124.5,171.5c39.4,2.6,71.1,34.3,73.5,73.8h46.1c-2.6-65-54.7-117.2-119.6-119.9V171.5z"
          />
          <path
            class="st0"
            d="M198.3,255.8c-2.6,39.5-34.3,71.1-73.8,73.5v46.1c65-2.6,117.2-54.7,119.9-119.6H198.3z"
          />
          <path
            class="st0"
            d="M300.7,245c2.6-39.5,34.3-71.1,73.8-73.5v-46.1c-65,2.6-117.2,54.7-119.9,119.6H300.7z"
          />
          <path
            class="st0"
            d="M374.5,329.3c-39.4-2.6-71.1-34.3-73.5-73.8h-46.1c2.6,65,54.7,117.2,119.6,119.9V329.3z"
          />
        </svg>
        Nextflow Homepage
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://seqera.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <?xml version="1.0" encoding="utf-8"?>
        <!-- Generator: Adobe Illustrator 28.3.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          x="0px"
          y="0px"
          viewBox="80 50 350 350"
          style="enable-background: new 0 0 500 500"
          xml:space="preserve"
        >
          <style type="text/css">
            .st0 {
              clip-path: url(#SVGID_00000006699037935518914080000000086571810102047141_);
            }
            .st1 {
              fill: "gray";
            }
          </style>
          <g>
            <defs>
              <rect id="SVGID_1_" x="125" y="125" width="250" height="250" />
            </defs>
            <clipPath
              id="SVGID_00000183214694142566821210000012389151436317298101_"
            >
              <use xlink:href="#SVGID_1_" style="overflow: visible" />
            </clipPath>
            <g
              style="
                clip-path: url(#SVGID_00000183214694142566821210000012389151436317298101_);
              "
            >
              <path
                class="st1"
                d="M375,333l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.2-0.9l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7L365,308l-0.6-0.6l-0.6-0.6l-0.6-0.6l-0.6-0.6L362,305
                      l-0.6-0.6l-0.7-0.6l-0.6-0.5l-1.2,0.7l-0.8,0.5l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.9,0.4l-0.9,0.4l-0.9,0.3l-0.9,0.3
                      l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0.1
                      l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0l-0.9,0h-57.5l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.3l0.1,0.3l0.1,0.4l0.1,0.4
                      l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4l-0.1,0.3l-0.2,0.3l-0.2,0.3
                      l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.1l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0H125V375h209.7l0.9,0l0.9,0l0.9,0l0.8-0.1
                      l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3
                      l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7
                      l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8
                      l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0-0.9l0-0.9l0-0.8L375,333z"
              />
              <path
                class="st1"
                d="M375,260.4l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.1-0.8l-0.1-0.8l-0.2-0.8l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.7
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.2-0.2h0h-89.8l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.4l0.1,0.3
                      l0.1,0.4l0.1,0.4l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4
                      l-0.2,0.4l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0h-82.5h-111H125v32.2
                      h122.8h86.9h0.9l0.9,0l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3
                      l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.4l0.8-0.4l0.2-0.1l0.6-0.3l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.5-0.7l0.5-0.7l0.5-0.7
                      l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7l0.4-0.7l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8
                      l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0-0.8l0-0.9l0-0.9
                      L375,260.4z"
              />
              <path
                class="st1"
                d="M165.3,197.6l-0.9,0l-0.8,0l-0.9,0l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.2,0.1l-0.6,0.3l-0.8,0.4
                      l-0.8,0.4l-0.8,0.4l-0.7,0.4l-0.7,0.5l-0.7,0.5L142,205l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.8l0,0.9l0.1,0.8l0.1,0.8
                      l0.1,0.9l0.1,0.9l0.1,0.8l0.2,0.9l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.2,0.2h89.8l-59.7-16.4l-0.4-0.1l-0.3-0.1
                      l-0.4-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3
                      l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4
                      l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3
                      l0.3-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.4-0.2l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0
                      h0.4h82.5h111H375v-32.3H252.2H165.3z"
              />
              <path
                class="st1"
                d="M165.3,125l-0.9,0l-0.8,0l-0.9,0.1l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4
                      l-0.8,0.4l-0.7,0.5l-0.7,0.4l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.9l0,0.8l0.1,0.9l0.1,0.9
                      l0.1,0.8l0.1,0.8l0.1,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.6,0.6l0.6,0.6
                      l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.7,0.6l0.6,0.5h0l1.1-0.7l0.8-0.5l0.8-0.5l0.8-0.4l0.8-0.4l0.9-0.4l0.9-0.4l0.9-0.4
                      l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.1l0.9-0.1l0.9-0.1l0.9-0.1
                      l0.9-0.1l0.9-0.1l0.9,0l0.9,0l0.9,0h57.5l-59.7-16.4l-0.4-0.1l-0.3-0.1l-0.4-0.1l-0.3-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2
                      l-0.3-0.2l-0.3-0.2l-0.3-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.3l-0.1-0.4
                      l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3
                      l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2
                      l0.3-0.2l0.4-0.1l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0l0.4,0H375V125H165.3z"
              />
            </g>
          </g>
        </svg>
        Seqera
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://github.com/nextflow-io/training"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"
          ></path>
        </svg>
        Training on GitHub
      </a>
    </li>
  </ul>
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../03_modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Moving code into modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../04_testing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 4: Adding tests
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Nextflow for RNAseq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Nextflow for RNAseq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow for RNAseq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/01_method/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Method overview and manual testing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/02_single-sample/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Single-sample implementation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/03_multi-sample/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Multi-sample paired-end implementation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Side Quests
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Side Quests
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Side Quests
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/workflows_of_workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflows of Workflows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/splitting_and_grouping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Splitting and Grouping
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/nf-test/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Testing with nf-test
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/nf-core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Fundamentals Training
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Fundamentals Training
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Index
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basic concepts
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/rnaseq_pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Simple RNA-Seq workflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Dependencies and containers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/channels/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Channels
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/processes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Processes
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/operators/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Operators
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/groovy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Groovy introduction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Modularization
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/executors/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Deployment scenarios
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/seqera_platform/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Seqera Platform
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/cache_and_resume/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cache and resume
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/debugging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troubleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Advanced Training
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Advanced Training
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Advanced Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/operators/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Operator Tour
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Metadata Propagation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/grouping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Grouping and Splitting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/groovy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Groovy Imports
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/structure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow Structure
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/configuration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/summary/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Summary
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/support/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Support
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../help/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Help
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                      

<nav class="md-nav md-nav--secondary" aria-label="On this page">
       
  <label class="md-nav__title" for="__toc">
    <span class="md-nav__icon md-icon"></span>
    On this page
  </label>
  <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
     <li class="md-nav__item">
  <a href="#method-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Method overview
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#0-warmup-run-samtools-and-gatk-directly" class="md-nav__link">
    <span class="md-ellipsis">
      0. Warmup: Run Samtools and GATK directly
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0. Warmup: Run Samtools and GATK directly">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#01-index-a-bam-input-file-with-samtools" class="md-nav__link">
    <span class="md-ellipsis">
      0.1. Index a BAM input file with Samtools
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.1. Index a BAM input file with Samtools">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#011-spin-up-the-samtools-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.1. Spin up the Samtools container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#012-run-the-indexing-command-for-the-three-samples" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.2. Run the indexing command for the three samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#013-exit-the-samtools-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.3. Exit the Samtools container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#02-call-variants-with-gatk-haplotypecaller-in-gvcf-mode" class="md-nav__link">
    <span class="md-ellipsis">
      0.2. Call variants with GATK HaplotypeCaller in GVCF mode
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.2. Call variants with GATK HaplotypeCaller in GVCF mode">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#021-spin-up-the-gatk-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.1. Spin up the GATK container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#022-run-the-variant-calling-command-with-the-gvcf-option" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.2. Run the variant calling command with the GVCF option
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#023-repeat-the-process-on-the-other-two-samples" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.3. Repeat the process on the other two samples
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#03-run-joint-genotyping" class="md-nav__link">
    <span class="md-ellipsis">
      0.3. Run joint genotyping
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.3. Run joint genotyping">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#031-combine-all-the-per-sample-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      0.3.1. Combine all the per-sample GVCFs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#032-run-the-joint-genotyping-analysis-proper" class="md-nav__link">
    <span class="md-ellipsis">
      0.3.2. Run the joint genotyping analysis proper
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#033-exit-the-gatk-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.3.3. Exit the GATK container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#1-modify-the-per-sample-variant-calling-step-to-produce-a-gvcf" class="md-nav__link">
    <span class="md-ellipsis">
      1. Modify the per-sample variant calling step to produce a GVCF
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Modify the per-sample variant calling step to produce a GVCF">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-tell-haplotypecaller-to-emit-a-gvcf-and-update-the-output-extension" class="md-nav__link">
    <span class="md-ellipsis">
      1.1. Tell HaplotypeCaller to emit a GVCF and update the output extension
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-run-the-pipeline-to-verify-that-you-can-generate-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      1.2. Run the pipeline to verify that you can generate GVCFs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-update-the-output-file-extension-in-the-process-outputs-block-too" class="md-nav__link">
    <span class="md-ellipsis">
      1.2. Update the output file extension in the process outputs block too
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-run-the-pipeline-again" class="md-nav__link">
    <span class="md-ellipsis">
      1.3. Run the pipeline again
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_1" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_1" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#2-collect-and-combine-the-gvcf-data-across-all-samples" class="md-nav__link">
    <span class="md-ellipsis">
      2. Collect and combine the GVCF data across all samples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Collect and combine the GVCF data across all samples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-define-the-process-that-will-combine-the-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      2.1. Define the process that will combine the GVCFs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-add-a-cohort_name-parameter-with-a-default-value" class="md-nav__link">
    <span class="md-ellipsis">
      2.2. Add a cohort_name parameter with a default value
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-gather-the-outputs-of-gatk_haplotypecaller-across-samples" class="md-nav__link">
    <span class="md-ellipsis">
      2.3. Gather the outputs of GATK_HAPLOTYPECALLER across samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-add-a-call-to-the-workflow-block-to-run-gatk_genomicsdb" class="md-nav__link">
    <span class="md-ellipsis">
      2.4. Add a call to the workflow block to run GATK_GENOMICSDB
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#25-run-the-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      2.5. Run the workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#26-construct-a-command-line-with-a-separate-v-argument-for-each-input-gvcf" class="md-nav__link">
    <span class="md-ellipsis">
      2.6. Construct a command line with a separate -V argument for each input GVCF
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#27-run-the-workflow-to-verify-that-it-generates-the-genomicsdb-output-as-expected" class="md-nav__link">
    <span class="md-ellipsis">
      2.7. Run the workflow to verify that it generates the GenomicsDB output as expected
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_2" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_2" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#3-run-the-joint-genotyping-step-as-part-of-the-same-process" class="md-nav__link">
    <span class="md-ellipsis">
      3. Run the joint genotyping step as part of the same process
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Run the joint genotyping step as part of the same process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31-rename-the-process-from-gatk_genomicsdb-to-gatk_jointgenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      3.1. Rename the process from GATK_GENOMICSDB to GATK_JOINTGENOTYPING
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-add-the-joint-genotyping-command-to-the-gatk_jointgenotyping-process" class="md-nav__link">
    <span class="md-ellipsis">
      3.2. Add the joint genotyping command to the GATK_JOINTGENOTYPING process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33-add-the-reference-genome-files-to-the-gatk_jointgenotyping-process-input-definitions" class="md-nav__link">
    <span class="md-ellipsis">
      3.3. Add the reference genome files to the GATK_JOINTGENOTYPING process input definitions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34-update-the-process-output-definition-to-emit-the-vcf-of-cohort-level-variant-calls" class="md-nav__link">
    <span class="md-ellipsis">
      3.4. Update the process output definition to emit the VCF of cohort-level variant calls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#35-update-the-process-call-from-gatk_genomicsdb-to-gatk_jointgenotyping" class="md-nav__link">
    <span class="md-ellipsis">
      3.5. Update the process call from GATK_GENOMICSDB to GATK_JOINTGENOTYPING
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#36-run-the-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      3.6. Run the workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_3" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_3" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li> 
  </ul>
  
</nav>
<nav class="md-nav md-nav--secondary sidebar_resources" aria-label="Resources">
  <label class="md-nav__title">Resources</label>
  <ul class="md-nav__list">
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/docs/latest/index.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M17.5 14.33c.79 0 1.63.08 2.5.24v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.33-4.5.99v-1.69c1.17-.53 2.67-.8 4.5-.8M13 12.46c1.29-.53 2.79-.79 4.5-.79.79 0 1.63.07 2.5.23v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.34-4.5.99m4.5-3.65c-1.9 0-3.39.32-4.5 1V9.84c1.23-.56 2.73-.84 4.5-.84.79 0 1.63.08 2.5.23v1.55c-.74-.19-1.59-.28-2.5-.28m3.5 8V7c-1.04-.33-2.21-.5-3.5-.5-2.05 0-3.88.5-5.5 1.5v11.5c1.62-1 3.45-1.5 5.5-1.5 1.19 0 2.36.16 3.5.5m-3.5-14c2.35 0 4.19.5 5.5 1.5v14.56c0 .12-.05.24-.16.35-.11.09-.23.17-.34.17-.11 0-.19-.02-.25-.05-1.28-.69-2.87-1.03-4.75-1.03-2.05 0-3.88.5-5.5 1.5-1.34-1-3.17-1.5-5.5-1.5-1.66 0-3.25.36-4.75 1.07-.03.01-.07.01-.12.03-.04.01-.08.02-.13.02-.11 0-.23-.04-.34-.12a.475.475 0 0 1-.16-.35V6c1.34-1 3.18-1.5 5.5-1.5 2.33 0 4.16.5 5.5 1.5 1.34-1 3.17-1.5 5.5-1.5Z"
          ></path>
        </svg>
        Nextflow Docs
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://www.nextflow.io/slack-invite.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M6 15a2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2h2v2m1 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-5m2-8a2 2 0 0 1-2-2 2 2 0 0 1 2-2 2 2 0 0 1 2 2v2H9m0 1a2 2 0 0 1 2 2 2 2 0 0 1-2 2H4a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5m8 2a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2h-2v-2m-1 0a2 2 0 0 1-2 2 2 2 0 0 1-2-2V5a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5m-2 8a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2v-2h2m0-1a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-5Z"
          ></path>
        </svg>
        Nextflow Slack
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="80 50 350 350"
        >
          <style type="text/css">
            .st0 {
              fill: "gray";
            }
          </style>
          <path
            class="st0"
            d="M124.5,171.5c39.4,2.6,71.1,34.3,73.5,73.8h46.1c-2.6-65-54.7-117.2-119.6-119.9V171.5z"
          />
          <path
            class="st0"
            d="M198.3,255.8c-2.6,39.5-34.3,71.1-73.8,73.5v46.1c65-2.6,117.2-54.7,119.9-119.6H198.3z"
          />
          <path
            class="st0"
            d="M300.7,245c2.6-39.5,34.3-71.1,73.8-73.5v-46.1c-65,2.6-117.2,54.7-119.9,119.6H300.7z"
          />
          <path
            class="st0"
            d="M374.5,329.3c-39.4-2.6-71.1-34.3-73.5-73.8h-46.1c2.6,65,54.7,117.2,119.6,119.9V329.3z"
          />
        </svg>
        Nextflow Homepage
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://seqera.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <?xml version="1.0" encoding="utf-8"?>
        <!-- Generator: Adobe Illustrator 28.3.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          x="0px"
          y="0px"
          viewBox="80 50 350 350"
          style="enable-background: new 0 0 500 500"
          xml:space="preserve"
        >
          <style type="text/css">
            .st0 {
              clip-path: url(#SVGID_00000006699037935518914080000000086571810102047141_);
            }
            .st1 {
              fill: "gray";
            }
          </style>
          <g>
            <defs>
              <rect id="SVGID_1_" x="125" y="125" width="250" height="250" />
            </defs>
            <clipPath
              id="SVGID_00000183214694142566821210000012389151436317298101_"
            >
              <use xlink:href="#SVGID_1_" style="overflow: visible" />
            </clipPath>
            <g
              style="
                clip-path: url(#SVGID_00000183214694142566821210000012389151436317298101_);
              "
            >
              <path
                class="st1"
                d="M375,333l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.2-0.9l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7L365,308l-0.6-0.6l-0.6-0.6l-0.6-0.6l-0.6-0.6L362,305
                      l-0.6-0.6l-0.7-0.6l-0.6-0.5l-1.2,0.7l-0.8,0.5l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.9,0.4l-0.9,0.4l-0.9,0.3l-0.9,0.3
                      l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0.1
                      l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0l-0.9,0h-57.5l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.3l0.1,0.3l0.1,0.4l0.1,0.4
                      l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4l-0.1,0.3l-0.2,0.3l-0.2,0.3
                      l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.1l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0H125V375h209.7l0.9,0l0.9,0l0.9,0l0.8-0.1
                      l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3
                      l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7
                      l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8
                      l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0-0.9l0-0.9l0-0.8L375,333z"
              />
              <path
                class="st1"
                d="M375,260.4l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.1-0.8l-0.1-0.8l-0.2-0.8l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.7
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.2-0.2h0h-89.8l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.4l0.1,0.3
                      l0.1,0.4l0.1,0.4l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4
                      l-0.2,0.4l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0h-82.5h-111H125v32.2
                      h122.8h86.9h0.9l0.9,0l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3
                      l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.4l0.8-0.4l0.2-0.1l0.6-0.3l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.5-0.7l0.5-0.7l0.5-0.7
                      l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7l0.4-0.7l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8
                      l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0-0.8l0-0.9l0-0.9
                      L375,260.4z"
              />
              <path
                class="st1"
                d="M165.3,197.6l-0.9,0l-0.8,0l-0.9,0l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.2,0.1l-0.6,0.3l-0.8,0.4
                      l-0.8,0.4l-0.8,0.4l-0.7,0.4l-0.7,0.5l-0.7,0.5L142,205l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.8l0,0.9l0.1,0.8l0.1,0.8
                      l0.1,0.9l0.1,0.9l0.1,0.8l0.2,0.9l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.2,0.2h89.8l-59.7-16.4l-0.4-0.1l-0.3-0.1
                      l-0.4-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3
                      l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4
                      l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3
                      l0.3-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.4-0.2l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0
                      h0.4h82.5h111H375v-32.3H252.2H165.3z"
              />
              <path
                class="st1"
                d="M165.3,125l-0.9,0l-0.8,0l-0.9,0.1l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4
                      l-0.8,0.4l-0.7,0.5l-0.7,0.4l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.9l0,0.8l0.1,0.9l0.1,0.9
                      l0.1,0.8l0.1,0.8l0.1,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.6,0.6l0.6,0.6
                      l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.7,0.6l0.6,0.5h0l1.1-0.7l0.8-0.5l0.8-0.5l0.8-0.4l0.8-0.4l0.9-0.4l0.9-0.4l0.9-0.4
                      l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.1l0.9-0.1l0.9-0.1l0.9-0.1
                      l0.9-0.1l0.9-0.1l0.9,0l0.9,0l0.9,0h57.5l-59.7-16.4l-0.4-0.1l-0.3-0.1l-0.4-0.1l-0.3-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2
                      l-0.3-0.2l-0.3-0.2l-0.3-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.3l-0.1-0.4
                      l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3
                      l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2
                      l0.3-0.2l0.4-0.1l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0l0.4,0H375V125H165.3z"
              />
            </g>
          </g>
        </svg>
        Seqera
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://github.com/nextflow-io/training"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"
          ></path>
        </svg>
        Training on GitHub
      </a>
    </li>
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/nextflow-io/training/edit/master/docs/nf4_science/genomics/02_joint_calling.md" title="Editar esta página" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<h1 id="part-2-joint-calling-on-a-cohort">Part 2: Joint calling on a cohort<a class="headerlink" href="#part-2-joint-calling-on-a-cohort" title="Permanent link">&para;</a></h1>
<p>In the first part of this course, you built a variant calling pipeline that was completely linear and processed each sample's data independently of the others.
However, in a real genomics use case, you'll typically need to look at the variant calls of multiple samples together.</p>
<p>In this second part, we show you how to use channels and channel operators to implement joint variant calling with GATK, building on the pipeline from Part 1.</p>
<h3 id="method-overview">Method overview<a class="headerlink" href="#method-overview" title="Permanent link">&para;</a></h3>
<p>The GATK variant calling method we used in first part of this course simply generated variant calls per sample.
That's fine if you only want to look at the variants from each sample in isolation, but that yields limited information.
It's often more interesting to look at how variant calls differ across multiple samples, and to do so, GATK offers an alternative method called joint variant calling, which we demonstrate here.</p>
<p>Joint variant calling involves generating a special kind of variant output called GVCF (for Genomic VCF) for each sample, then combining the GVCF data from all the samples and finally, running a 'joint genotyping' statistical analysis.</p>
<p><img alt="Joint analysis" src="../../../../nf4_science/genomics/img/joint-calling.png" /></p>
<p>What's special about a sample's GVCF is that it contains records summarizing sequence data statistics about all positions in the targeted area of the genome, not just the positions where the program found evidence of variation.
This is critical for the joint genotyping calculation (<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890431-The-logic-of-joint-calling-for-germline-short-variants">further reading</a>).</p>
<p>The GVCF is produced by GATK HaplotypeCaller, the same tool we used in Part 1, with an additional parameter (<code>-ERC GVCF</code>).
Combining the GVCFs is done with GATK GenomicsDBImport, which combines the per-sample calls into a data store (analogous to a database), then the actual 'joint genotyping' analysis is done with GATK GenotypeGVCFs.</p>
<h3 id="workflow">Workflow<a class="headerlink" href="#workflow" title="Permanent link">&para;</a></h3>
<p>So to recap, in this part of the course, we're going to develop a workflow that does the following:</p>
<figure class="excalidraw">
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 981.7387923958568 1283.5356465603895" width="1963.4775847917135" height="2567.071293120779">
  <!-- svg-source:excalidraw -->

  <defs>
    <style class="style-fonts">
      @font-face {
        font-family: Virgil;
        src: url(https://unpkg.com/@excalidraw/excalidraw@undefined/dist/prod/Virgil-Regular-YHAB2VGJ.woff2);
          }
@font-face {
        font-family: Excalifont;
        src: url(https://unpkg.com/@excalidraw/excalidraw@undefined/dist/prod/Excalifont-Regular-CPKEUDVM.woff2);
          }
    </style>

  </defs>
  <rect x="0" y="0" width="981.7387923958568" height="1283.5356465603895" fill="#ffffff"></rect><g stroke-linecap="round" transform="translate(10 10) rotate(0 480.8693961979284 260.43542301409775)"><path d="M32 0 C315.92 1.05, 599.76 1.6, 929.74 0 M32 0 C385.4 0.53, 739.22 0.98, 929.74 0 M929.74 0 C949.12 0.23, 961.87 8.71, 961.74 32 M929.74 0 C950.01 -2.07, 960.02 9.88, 961.74 32 M961.74 32 C963.48 147.18, 963.23 263.9, 961.74 488.87 M961.74 32 C959.71 126.83, 959.92 221.45, 961.74 488.87 M961.74 488.87 C960.63 508.35, 952.15 519.42, 929.74 520.87 M961.74 488.87 C963.32 510.79, 949.16 520.06, 929.74 520.87 M929.74 520.87 C577.8 517.56, 224.63 516.75, 32 520.87 M929.74 520.87 C621.53 517.98, 313.31 518.16, 32 520.87 M32 520.87 C10.44 522.39, -1.62 510.02, 0 488.87 M32 520.87 C12.68 520.88, -1.4 512.04, 0 488.87 M0 488.87 C-0.99 396.14, -0.81 304.61, 0 32 M0 488.87 C-0.88 381.64, -1.01 274.77, 0 32 M0 32 C-0.02 11.65, 12.38 -1.08, 32 0 M0 32 C0.06 12.3, 10.34 2.03, 32 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g stroke-linecap="round" transform="translate(30.035148788527863 32.15135221324704) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C52.17 2.44, 83.37 4.1, 153.5 0 C166.04 -0.79, 170.84 3.38, 174.42 20.92 C175.18 37.07, 170.33 48.16, 174.42 62.75 C173.91 77.37, 167.35 86.91, 153.5 83.66 C107.36 78.9, 60.95 79.43, 20.92 83.66 C5.96 83.73, -0.91 78.57, 0 62.75 C-0.87 50.02, -3.47 38.79, 0 20.92 C0 8.7, 7.26 0.43, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C48 -2.37, 76.33 -1.5, 153.5 0 M20.92 0 C54.16 -1.23, 85.53 -0.82, 153.5 0 M153.5 0 C167.21 -1.87, 173.37 6.53, 174.42 20.92 M153.5 0 C167.01 -1.06, 173.07 7.8, 174.42 20.92 M174.42 20.92 C176.23 29.92, 176.26 39.63, 174.42 62.75 M174.42 20.92 C174.44 33.17, 174.53 43.24, 174.42 62.75 M174.42 62.75 C172.69 78.28, 167.49 82.74, 153.5 83.66 M174.42 62.75 C172.54 76.73, 167.23 84.43, 153.5 83.66 M153.5 83.66 C117.09 83.91, 80.29 84, 20.92 83.66 M153.5 83.66 C103.23 84.89, 53.56 84.61, 20.92 83.66 M20.92 83.66 C6.7 81.9, -0.87 77.17, 0 62.75 M20.92 83.66 C4.84 83.81, 0.87 76.34, 0 62.75 M0 62.75 C-0.13 51.17, 1.55 41.75, 0 20.92 M0 62.75 C0.16 49.93, 0.66 37.7, 0 20.92 M0 20.92 C1.17 8.03, 5.37 -0.24, 20.92 0 M0 20.92 C1.09 8.99, 4.73 2.24, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(95.75417849042685 61.48143033824704) rotate(0 21.489990234375 12.5)"><text x="21.489990234375" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">BAM</text></g><g stroke-linecap="round" transform="translate(289.5844035082173 31.059555338247037) rotate(0 109.333984375 41.830078125)"><path d="M4.91 5.59 C4.91 5.59, 4.91 5.59, 4.91 5.59 M4.91 5.59 C4.91 5.59, 4.91 5.59, 4.91 5.59 M1.11 22.16 C6.94 13.81, 13.66 5.33, 20.13 0.28 M1.11 22.16 C8.48 15.51, 14.46 7.62, 20.13 0.28 M0.58 34.96 C9.49 24.06, 19.08 15.64, 30.76 0.24 M0.58 34.96 C9.33 25.13, 16.39 16.31, 30.76 0.24 M0.72 47 C14.11 31.26, 30.55 14.48, 41.39 0.21 M0.72 47 C13.09 32.34, 25.74 20.16, 41.39 0.21 M0.85 59.04 C9.94 48.58, 21.24 35.85, 52.02 0.17 M0.85 59.04 C19.43 39.07, 35.95 17.48, 52.02 0.17 M1.64 70.33 C24.63 44.06, 42.97 20.29, 62.65 0.14 M1.64 70.33 C18.62 49.89, 36.98 27.58, 62.65 0.14 M5.71 77.84 C27.26 55.42, 46.82 30.2, 73.28 0.1 M5.71 77.84 C28.81 52.45, 51.5 26.26, 73.28 0.1 M12.4 82.33 C35.07 56.81, 57.38 30.69, 83.91 0.07 M12.4 82.33 C37.15 52.76, 62.87 23.57, 83.91 0.07 M21.72 83.81 C37.76 64.37, 55.38 47.34, 94.54 0.04 M21.72 83.81 C49.14 53.43, 75.3 21.3, 94.54 0.04 M32.35 83.77 C58.24 52.53, 86.93 19.66, 104.52 0.76 M32.35 83.77 C52.27 60.56, 72.26 38.04, 104.52 0.76 M42.33 84.49 C67.16 51.85, 94.14 21.89, 115.15 0.72 M42.33 84.49 C64.65 56.98, 87.21 30.81, 115.15 0.72 M52.96 84.46 C78.5 57.54, 102.11 29.18, 125.78 0.69 M52.96 84.46 C77.55 54.81, 102.31 27.56, 125.78 0.69 M63.59 84.42 C81.73 65.24, 100.2 44.45, 136.41 0.65 M63.59 84.42 C90 54.29, 117.36 23.93, 136.41 0.65 M74.22 84.39 C101.53 51.44, 132.65 17.55, 147.04 0.62 M74.22 84.39 C90.28 65.7, 105.96 49.96, 147.04 0.62 M84.85 84.36 C108.82 56.24, 134.95 27.65, 157.67 0.58 M84.85 84.36 C102.24 65.9, 118.66 46.24, 157.67 0.58 M95.48 84.32 C110.95 67.07, 126.41 49.21, 168.3 0.55 M95.48 84.32 C124.39 51.75, 151.93 21.28, 168.3 0.55 M106.11 84.29 C134.46 51.38, 160.28 21.63, 178.93 0.51 M106.11 84.29 C130.34 56.18, 153.53 27.69, 178.93 0.51 M116.74 84.25 C131.23 68.36, 146.91 48.98, 189.56 0.48 M116.74 84.25 C141.95 54.25, 168.66 23.45, 189.56 0.48 M127.37 84.22 C149.24 56.22, 173.25 30.45, 200.19 0.45 M127.37 84.22 C144.06 65.03, 162.22 44.34, 200.19 0.45 M138 84.18 C162.55 56.82, 188.29 25.58, 208.85 2.67 M138 84.18 C159.93 58.41, 181.18 31.95, 208.85 2.67 M148.63 84.15 C172.14 55.7, 199.45 27.27, 214.89 7.92 M148.63 84.15 C163.58 67.68, 179.51 49.75, 214.89 7.92 M159.26 84.11 C176.03 64.73, 196.42 40.93, 218.3 16.19 M159.26 84.11 C183.4 58.5, 204.38 31.72, 218.3 16.19 M169.89 84.08 C185.71 66.1, 200.84 47.58, 218.44 28.23 M169.89 84.08 C183.13 69.12, 196.09 52.61, 218.44 28.23 M180.52 84.05 C188.69 70.87, 202.77 61.45, 218.57 40.27 M180.52 84.05 C193.57 69.19, 207.78 53.01, 218.57 40.27 M191.15 84.01 C196.8 75.68, 204.83 70.22, 218.7 52.31 M191.15 84.01 C199.73 73.38, 208.23 64.12, 218.7 52.31 M203.09 82.47 C209.17 76.55, 216.67 67.67, 218.83 64.35 M203.09 82.47 C206.51 77.5, 209.58 73.12, 218.83 64.35 M4.93 78.17 C4.93 78.17, 4.93 78.17, 4.93 78.17 M4.93 78.17 C4.93 78.17, 4.93 78.17, 4.93 78.17 M22.25 82.63 C15.78 78.25, 12.46 73.02, 2.63 65.57 M22.25 82.63 C17.06 77.54, 11.42 71.81, 2.63 65.57 M34.29 82.49 C24.32 73.38, 13.85 65.26, 2.59 54.94 M34.29 82.49 C23.11 73.76, 11.54 62.56, 2.59 54.94 M46.33 82.36 C36.59 72.79, 28.31 65.17, 2.56 44.31 M46.33 82.36 C30.35 67.76, 11.69 53.44, 2.56 44.31 M58.37 82.23 C38.45 63.16, 16.54 46.34, 2.52 33.68 M58.37 82.23 C44.73 70.63, 31.41 57.26, 2.52 33.68 M70.41 82.09 C47.33 63.62, 23.36 43.21, 2.49 23.05 M70.41 82.09 C54.97 66.57, 37.54 53.61, 2.49 23.05 M83.21 82.62 C60.45 61, 36.07 40.04, 3.21 13.08 M83.21 82.62 C66.95 69.53, 51.02 55.36, 3.21 13.08 M95.25 82.48 C75.51 65.79, 54.1 47.49, 5.44 4.41 M95.25 82.48 C75.35 65.59, 55.59 47.67, 5.44 4.41 M107.29 82.35 C86.05 62.67, 67.52 47.03, 12.95 0.34 M107.29 82.35 C78.56 55.93, 50.37 31.55, 12.95 0.34 M119.33 82.22 C98.59 66.22, 80.59 45.62, 22.73 -1.76 M119.33 82.22 C85.01 51.39, 48.74 20.61, 22.73 -1.76 M131.37 82.08 C107.78 62.25, 83.44 40.63, 34.77 -1.89 M131.37 82.08 C107.54 62.05, 83.87 41.68, 34.77 -1.89 M144.17 82.61 C110.86 51.83, 74.16 22.55, 46.81 -2.02 M144.17 82.61 C115.22 58.21, 88.96 34.29, 46.81 -2.02 M156.21 82.47 C127.28 55.38, 94.19 30.56, 58.85 -2.16 M156.21 82.47 C120.69 52, 85.32 22.01, 58.85 -2.16 M168.25 82.34 C147.36 63.13, 123.8 43.05, 71.65 -1.63 M168.25 82.34 C139.96 57.93, 111.52 31.8, 71.65 -1.63 M180.29 82.21 C144.48 51.26, 104.3 19.68, 83.69 -1.77 M180.29 82.21 C157.23 63.62, 136.27 45.87, 83.69 -1.77 M193.09 82.73 C158.08 50.29, 121.98 21.84, 95.73 -1.9 M193.09 82.73 C169.3 62.08, 144.43 40.52, 95.73 -1.9 M201.35 79.32 C175.76 58.83, 153.79 35.47, 107.77 -2.03 M201.35 79.32 C181.73 61.84, 162.26 46.79, 107.77 -2.03 M212.64 78.53 C183.39 50.27, 149.89 23.16, 119.81 -2.17 M212.64 78.53 C183.44 52.26, 153.91 27.8, 119.81 -2.17 M217.13 71.83 C189.47 46.04, 162.45 23.76, 132.61 -1.64 M217.13 71.83 C187.71 48.51, 159.35 23.81, 132.61 -1.64 M218.61 62.52 C196.06 39.55, 169.02 21.26, 144.65 -1.78 M218.61 62.52 C202.55 49.91, 187.36 36.26, 144.65 -1.78 M218.57 51.89 C203.39 38.96, 189.57 25.17, 156.69 -1.91 M218.57 51.89 C203.11 37.7, 187.02 24.96, 156.69 -1.91 M218.54 41.26 C197.63 25.9, 180.36 7.76, 168.73 -2.04 M218.54 41.26 C199.81 25.92, 181.88 9.01, 168.73 -2.04 M218.5 30.63 C209.22 22.81, 199.59 14.41, 180.77 -2.18 M218.5 30.63 C209.32 23.88, 200.94 15.31, 180.77 -2.18 M218.47 20 C210.17 13.69, 202.98 6.43, 193.56 -1.65 M218.47 20 C209.47 12.93, 200.67 5.63, 193.56 -1.65" stroke="#ffec99" stroke-width="1" fill="none"></path><path d="M20.92 0 C61.76 1.98, 104.49 1.96, 197.75 0 M20.92 0 C74.85 0.46, 130.88 0.24, 197.75 0 M197.75 0 C211.01 -1.84, 218.24 6.31, 218.67 20.92 M197.75 0 C212.37 0.03, 220.66 6.14, 218.67 20.92 M218.67 20.92 C219.23 31.58, 218.21 40.1, 218.67 62.75 M218.67 20.92 C219.24 35.25, 219.46 50.93, 218.67 62.75 M218.67 62.75 C217.25 75.4, 210.64 84.12, 197.75 83.66 M218.67 62.75 C217.58 74.53, 210.08 85.67, 197.75 83.66 M197.75 83.66 C144.25 82.53, 90.8 83.1, 20.92 83.66 M197.75 83.66 C159.19 84.1, 122.5 83.58, 20.92 83.66 M20.92 83.66 C8.19 84.31, -0.92 75.81, 0 62.75 M20.92 83.66 C8.83 81.83, -1.21 76.05, 0 62.75 M0 62.75 C1.42 54.53, -0.32 45, 0 20.92 M0 62.75 C-0.83 50.56, 0.46 37.91, 0 20.92 M0 20.92 C-0.4 7.67, 5.48 -1.29, 20.92 0 M0 20.92 C0.37 8.77, 6.66 1.11, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(350.04843091300245 47.88963346324704) rotate(0 48.869956970214844 25)"><text x="48.869956970214844" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Samtools </text><text x="48.869956970214844" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">index</text></g><g stroke-linecap="round" transform="translate(313.7324744831734 186.98143033824704) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C53.87 -1.29, 87.07 -1.25, 153.5 0 C170.57 -0.1, 176.03 7.28, 174.42 20.92 C175.59 35.38, 174.51 45.05, 174.42 62.75 C172.85 77.99, 169.11 82.53, 153.5 83.66 C101.5 85.16, 49.57 87.82, 20.92 83.66 C10.19 84.75, 1.76 78.17, 0 62.75 C0.15 46.65, -1.52 39.09, 0 20.92 C2.33 6.44, 3.67 -1.91, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C50.72 0.76, 82.73 1.08, 153.5 0 M20.92 0 C69.33 -0.53, 118.43 0.3, 153.5 0 M153.5 0 C167.96 1.56, 174.97 6.67, 174.42 20.92 M153.5 0 C167.11 -1.36, 175.45 6.39, 174.42 20.92 M174.42 20.92 C172.26 31.78, 174.09 39.8, 174.42 62.75 M174.42 20.92 C174.97 34.67, 174.11 50.03, 174.42 62.75 M174.42 62.75 C173.95 77.75, 165.95 83.29, 153.5 83.66 M174.42 62.75 C175.11 76.99, 168.22 84.71, 153.5 83.66 M153.5 83.66 C117.37 85.39, 84.53 85.74, 20.92 83.66 M153.5 83.66 C103.9 83.43, 53.42 82.49, 20.92 83.66 M20.92 83.66 C5.85 85.13, 0.69 75.37, 0 62.75 M20.92 83.66 C6.33 85.57, 0.05 78.55, 0 62.75 M0 62.75 C-0.5 52.46, -0.72 41.51, 0 20.92 M0 62.75 C1.09 48.7, 0.42 33.7, 0 20.92 M0 20.92 C-0.41 6.01, 7.89 -0.86, 20.92 0 M0 20.92 C-2.09 9.23, 7.52 -0.15, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(349.9415238641419 216.31150846324704) rotate(0 50.99997055530548 12.5)"><text x="50.99997055530548" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">BAM index</text></g><g stroke-linecap="round" transform="translate(561.4249569996227 186.87205533824704) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C53.87 -4.1, 88.73 -0.01, 153.5 0 C165.12 -3.58, 174.02 4.98, 174.42 20.92 C172.46 31.72, 175.29 49.65, 174.42 62.75 C174.9 73.7, 165.85 82.36, 153.5 83.66 C126.52 84.35, 93.77 81.79, 20.92 83.66 C8.43 81.03, -2.87 78.51, 0 62.75 C0.3 49.29, -1.28 29.75, 0 20.92 C3.45 5.73, 10.55 0.5, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C57.78 0.13, 99.78 0.7, 153.5 0 M20.92 0 C66.25 0.36, 113.73 2, 153.5 0 M153.5 0 C168.57 -1.73, 174.09 6.2, 174.42 20.92 M153.5 0 C166.69 -0.96, 176.55 6.13, 174.42 20.92 M174.42 20.92 C172.63 32.24, 172.78 43.79, 174.42 62.75 M174.42 20.92 C174.18 31.35, 174.1 40.61, 174.42 62.75 M174.42 62.75 C175.43 78.07, 168.97 85.23, 153.5 83.66 M174.42 62.75 C174.93 78.63, 166.45 84.28, 153.5 83.66 M153.5 83.66 C116.31 84.17, 74.05 84.59, 20.92 83.66 M153.5 83.66 C110.88 84.55, 69.39 83.64, 20.92 83.66 M20.92 83.66 C5.55 81.81, 1.05 76.76, 0 62.75 M20.92 83.66 C4.97 83.94, 1.58 75.51, 0 62.75 M0 62.75 C-0.6 50.2, 0.81 34.76, 0 20.92 M0 62.75 C-0.05 47.24, -0.91 30.91, 0 20.92 M0 20.92 C0.57 6.23, 6.48 1.7, 20.92 0 M0 20.92 C-0.05 7.79, 8.1 -2.29, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(603.0140076126193 216.20213346324704) rotate(0 45.61996932327747 12.5)"><text x="45.61996932327747" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Intervals</text></g><g stroke-linecap="round" transform="translate(773.8492358254448 187.75486783824704) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C72.18 -4.21, 119.69 -0.07, 153.5 0 C170.32 2.79, 175.29 10.2, 174.42 20.92 C174.38 37.48, 173.22 48.85, 174.42 62.75 C177.98 73.82, 169.88 84.14, 153.5 83.66 C116.91 82.41, 82.49 82.73, 20.92 83.66 C7.18 83.98, 0.15 74.37, 0 62.75 C-2.67 52.18, -2.98 43.81, 0 20.92 C-2.31 6.43, 8.25 -0.96, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C72.56 -1.41, 123.27 0.25, 153.5 0 M20.92 0 C70.1 0.94, 120.16 0.75, 153.5 0 M153.5 0 C165.88 1.31, 174.4 7.8, 174.42 20.92 M153.5 0 C168.31 -1.39, 173.35 5.3, 174.42 20.92 M174.42 20.92 C173.37 31.08, 175.62 36.87, 174.42 62.75 M174.42 20.92 C174.66 30.39, 175.11 40.87, 174.42 62.75 M174.42 62.75 C175.15 75.3, 167.89 82.38, 153.5 83.66 M174.42 62.75 C176 76.6, 169.73 84.94, 153.5 83.66 M153.5 83.66 C106.16 84.42, 57.4 86.15, 20.92 83.66 M153.5 83.66 C116.59 84.14, 81.78 84.04, 20.92 83.66 M20.92 83.66 C6.04 85.29, 0.47 75.3, 0 62.75 M20.92 83.66 C9.21 84.85, 2.28 78.34, 0 62.75 M0 62.75 C0.23 53.63, -1.41 40.04, 0 20.92 M0 62.75 C0.48 47.62, -0.12 32.09, 0 20.92 M0 20.92 C-0.22 7.69, 8.15 -1.74, 20.92 0 M0 20.92 C1.17 5.38, 8.89 -2.12, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(789.8582954107285 204.58494596324704) rotate(0 71.1999603509903 25)"><text x="71.1999603509903" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Reference</text><text x="71.1999603509903" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">+ index &amp; dict</text></g><g stroke-linecap="round" transform="translate(291.1703410082173 341.94236783824704) rotate(0 109.333984375 42.38449770473267)"><path d="M4.98 5.66 C4.98 5.66, 4.98 5.66, 4.98 5.66 M4.98 5.66 C4.98 5.66, 4.98 5.66, 4.98 5.66 M3.81 19.21 C6.71 13.78, 11.54 8.89, 16.27 4.87 M3.81 19.21 C7.36 14.75, 9.93 12.21, 16.27 4.87 M3.94 31.25 C11.19 24.65, 15.44 18.54, 26.9 4.83 M3.94 31.25 C12.16 20.78, 21.38 11.94, 26.9 4.83 M4.07 43.29 C14.93 31.77, 25.68 19, 37.53 4.8 M4.07 43.29 C14.04 33.71, 21.29 22.29, 37.53 4.8 M3.55 56.09 C17.23 43.52, 28.99 30.91, 48.16 4.77 M3.55 56.09 C21.22 35.33, 38.41 14.73, 48.16 4.77 M1.71 70.39 C13.67 54.24, 27.14 42.36, 58.79 4.73 M1.71 70.39 C13.49 56.18, 26.12 40.9, 58.79 4.73 M5.78 77.9 C28.53 49.59, 56.78 19.53, 69.42 4.7 M5.78 77.9 C21.08 59.8, 37.46 42.27, 69.42 4.7 M10.51 84.66 C23.9 66.56, 38.16 50.91, 80.05 4.66 M10.51 84.66 C37.48 52.22, 66.66 21.47, 80.05 4.66 M17.86 88.4 C42.86 58.37, 70.56 29.37, 90.68 4.63 M17.86 88.4 C34.45 69.77, 51.5 49.63, 90.68 4.63 M27.18 89.88 C55.84 56.48, 83.84 24.17, 101.31 4.59 M27.18 89.88 C48.07 66.43, 68.08 42.26, 101.31 4.59 M37.81 89.84 C57.88 64.7, 82.92 39.74, 111.94 4.56 M37.81 89.84 C63.59 60.84, 88.94 31.08, 111.94 4.56 M48.44 89.81 C75.59 56.3, 104.16 25.03, 121.92 5.28 M48.44 89.81 C74.34 59.23, 100.47 27.08, 121.92 5.28 M59.07 89.77 C87.53 58.46, 116.44 26.11, 132.55 5.24 M59.07 89.77 C84.77 61.04, 109.4 33.68, 132.55 5.24 M69.7 89.74 C98.14 59.36, 121.93 29.56, 143.18 5.21 M69.7 89.74 C97.54 55.64, 127.62 24.85, 143.18 5.21 M80.33 89.7 C97.58 71.99, 111.7 50.8, 153.81 5.18 M80.33 89.7 C99.48 65.69, 119.82 42.24, 153.81 5.18 M90.96 89.67 C105.93 72.79, 119.6 54.61, 164.44 5.14 M90.96 89.67 C113.01 65.13, 134.44 39.82, 164.44 5.14 M101.59 89.63 C117.19 69.82, 134.58 53.47, 175.07 5.11 M101.59 89.63 C127.13 60.04, 153.99 31.39, 175.07 5.11 M112.22 89.6 C134.62 62.6, 161.92 33.05, 185.7 5.07 M112.22 89.6 C131.47 69, 150.72 46.09, 185.7 5.07 M122.85 89.56 C143.16 65.81, 163.58 44.24, 199.61 1.26 M122.85 89.56 C141.37 67.16, 162.85 43.75, 199.61 1.26 M132.82 90.29 C155.3 62.71, 175.75 40.45, 208.92 2.74 M132.82 90.29 C154.91 66.9, 176.12 43.34, 208.92 2.74 M143.45 90.25 C157.67 72.57, 174.82 56.01, 214.96 7.99 M143.45 90.25 C159.24 70.7, 176.22 51.85, 214.96 7.99 M154.08 90.22 C175.73 62.35, 197.9 39.27, 218.38 16.25 M154.08 90.22 C176.8 64.05, 200.3 38.99, 218.38 16.25 M164.71 90.18 C182.44 65.62, 204.87 43.07, 218.51 28.3 M164.71 90.18 C175.84 75.6, 189.1 63.14, 218.51 28.3 M175.34 90.15 C186.62 77.23, 203.59 59.06, 218.64 40.34 M175.34 90.15 C189.57 73.01, 203.1 57.49, 218.64 40.34 M185.97 90.11 C196.33 78.83, 203.99 68.3, 218.78 52.38 M185.97 90.11 C196.39 77.97, 205.59 67.95, 218.78 52.38 M199.88 86.3 C204.29 80.32, 210.53 74.24, 218.91 64.42 M199.88 86.3 C204.8 81.18, 209.8 74.87, 218.91 64.42 M5.2 79.38 C5.2 79.38, 5.2 79.38, 5.2 79.38 M5.2 79.38 C5.2 79.38, 5.2 79.38, 5.2 79.38 M22.52 83.84 C17.23 77.06, 11.88 72.52, 2.14 66.13 M22.52 83.84 C17.91 78.53, 11.47 74.37, 2.14 66.13 M34.56 83.71 C24.01 71.97, 11.6 61.36, 2.11 55.5 M34.56 83.71 C25.63 76.81, 17.95 69.46, 2.11 55.5 M46.6 83.57 C27.4 66.85, 13.25 54.29, 2.07 44.87 M46.6 83.57 C32.5 70.31, 16.39 56.54, 2.07 44.87 M58.64 83.44 C38.75 66.91, 16.16 45.86, 2.04 34.24 M58.64 83.44 C46.49 73.5, 34.95 62.02, 2.04 34.24 M70.68 83.31 C53.53 65.99, 32.91 52.35, 2 23.61 M70.68 83.31 C49.78 65.46, 28 47.77, 2 23.61 M83.48 83.83 C61.78 62.17, 34.58 40.95, 2.72 13.63 M83.48 83.83 C61.09 64.81, 40.7 45.42, 2.72 13.63 M95.52 83.7 C72.78 64.64, 50.39 44.88, 4.95 4.97 M95.52 83.7 C66.21 58.97, 37.33 36, 4.95 4.97 M107.56 83.56 C76.65 56.53, 47.27 32.9, 12.47 0.9 M107.56 83.56 C85.18 63.83, 64.31 44.52, 12.47 0.9 M119.6 83.43 C96.22 64.23, 72.16 43.27, 22.24 -1.2 M119.6 83.43 C97.89 65.03, 76.32 46.91, 22.24 -1.2 M131.64 83.3 C92.09 49.32, 54.17 17.75, 34.28 -1.33 M131.64 83.3 C104.71 59.39, 76.53 35.75, 34.28 -1.33 M144.44 83.82 C118.86 63.02, 95.56 40.15, 46.32 -1.47 M144.44 83.82 C123.52 64.73, 101.77 46.18, 46.32 -1.47 M156.48 83.69 C124.58 56.16, 93.73 28.48, 59.12 -0.94 M156.48 83.69 C132.08 63.47, 109.13 41.39, 59.12 -0.94 M168.52 83.56 C149.14 67.16, 126.52 50.01, 71.16 -1.08 M168.52 83.56 C146.47 66.06, 125.58 47.24, 71.16 -1.08 M180.56 83.42 C150.49 54.53, 115.4 28.04, 83.2 -1.21 M180.56 83.42 C143.79 49.58, 103.74 16.18, 83.2 -1.21 M192.6 83.29 C154.95 52.82, 121.59 20.56, 95.24 -1.34 M192.6 83.29 C166.3 58.8, 137.51 36.47, 95.24 -1.34 M201.62 80.53 C175.82 60.23, 154.49 40.97, 107.28 -1.48 M201.62 80.53 C182.54 62.48, 163.51 47.39, 107.28 -1.48 M212.15 79.09 C180.04 55.68, 151.72 28.56, 120.08 -0.95 M212.15 79.09 C177.78 49.78, 143.15 20.6, 120.08 -0.95 M217.4 73.05 C188.8 48.07, 158.64 22.14, 132.12 -1.09 M217.4 73.05 C182.58 44.37, 148.85 14.82, 132.12 -1.09 M218.88 63.73 C194.17 42.32, 166.34 19.28, 144.16 -1.22 M218.88 63.73 C188.62 37.34, 159.94 11.82, 144.16 -1.22 M218.84 53.1 C200.31 35.97, 178.73 18.95, 156.2 -1.35 M218.84 53.1 C199.75 35.62, 179.63 19.46, 156.2 -1.35 M218.81 42.47 C204.51 29.74, 192.21 19.36, 168.24 -1.49 M218.81 42.47 C198.47 25.72, 179.81 9.14, 168.24 -1.49 M218.77 31.84 C209.24 26.5, 199.95 16.04, 181.04 -0.96 M218.77 31.84 C206.77 21.42, 195.49 10.81, 181.04 -0.96 M218.74 21.21 C214.69 17.97, 205.7 10.85, 193.08 -1.1 M218.74 21.21 C211.13 14.05, 204.11 7.43, 193.08 -1.1" stroke="#ffec99" stroke-width="1" fill="none"></path><path d="M21.19 0 C73.73 0.01, 127.76 -0.58, 197.48 0 M21.19 0 C60.27 0.7, 96.63 1.87, 197.48 0 M197.48 0 C213 -0.47, 218.81 7.55, 218.67 21.19 M197.48 0 C210.35 -1.65, 220.84 6.92, 218.67 21.19 M218.67 21.19 C219.06 34.26, 218.06 44.46, 218.67 63.58 M218.67 21.19 C217.92 31.91, 219.08 44.03, 218.67 63.58 M218.67 63.58 C217.87 75.85, 213.14 83.74, 197.48 84.77 M218.67 63.58 C218.09 76.95, 209.67 84.34, 197.48 84.77 M197.48 84.77 C141.13 83.1, 87.69 85.21, 21.19 84.77 M197.48 84.77 C126.67 82.18, 58.67 82.8, 21.19 84.77 M21.19 84.77 C6.72 86.72, -1.14 77.51, 0 63.58 M21.19 84.77 C6.45 84.62, 2.18 78.47, 0 63.58 M0 63.58 C-1.74 44.76, 0.12 28.7, 0 21.19 M0 63.58 C0.02 53.4, -0.78 43.21, 0 21.19 M0 21.19 C-0.55 8.06, 8.74 0.95, 21.19 0 M0 21.19 C0.75 7.83, 4.86 0.79, 21.19 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(325.7243907087159 359.3268655429797) rotate(0 74.77993467450142 25)"><text x="74.77993467450142" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GATK </text><text x="74.77993467450142" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">HaplotypeCaller</text></g><g stroke-linecap="round" transform="translate(313.78544267977986 490.34080533824704) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C68.74 0.97, 113.08 2.59, 153.5 0 C166.35 -1.31, 174.25 6.32, 174.42 20.92 C172.52 34.83, 171.41 57.67, 174.42 62.75 C171.7 74.25, 167.3 82.67, 153.5 83.66 C126.4 84.38, 97.04 82.15, 20.92 83.66 C6.59 83.14, 0.94 78.84, 0 62.75 C-1.18 45.69, -2.83 34.39, 0 20.92 C2 3.42, 3.62 0.99, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C66.29 -0.96, 114.98 -1.31, 153.5 0 M20.92 0 C52.15 -0.3, 82.47 -0.88, 153.5 0 M153.5 0 C166.82 0.22, 176 8.68, 174.42 20.92 M153.5 0 C169.12 -0.36, 175.64 8.74, 174.42 20.92 M174.42 20.92 C172.87 37.16, 172.23 49.78, 174.42 62.75 M174.42 20.92 C175.08 32.33, 174.77 41.95, 174.42 62.75 M174.42 62.75 C176.27 74.88, 167.49 81.99, 153.5 83.66 M174.42 62.75 C172.65 77.61, 169.15 85.95, 153.5 83.66 M153.5 83.66 C126.01 85.16, 95.39 84.74, 20.92 83.66 M153.5 83.66 C111.67 83.47, 69.78 82.69, 20.92 83.66 M20.92 83.66 C7.37 85.15, 1.17 74.94, 0 62.75 M20.92 83.66 C7.73 85.59, -0.39 77.89, 0 62.75 M0 62.75 C1.36 53.92, -1.65 44.26, 0 20.92 M0 62.75 C-0.14 48.01, 0.16 32.6, 0 20.92 M0 20.92 C-1.53 7.76, 7.52 0.79, 20.92 0 M0 20.92 C-0.98 8.04, 8.83 -0.88, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(334.87448824128273 519.670883463247) rotate(0 66.11997437477112 12.5)"><text x="66.11997437477112" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GVCF + index</text></g><g stroke-linecap="round"><g transform="translate(116.97655503852786 128.67869596324704) rotate(0 79.43591774469584 126.99154703200844)"><path d="M-1.11 -0.86 C0.32 35.35, -18.94 175.22, 8.77 217.61 C36.48 260, 138.85 247.33, 165.18 253.49 M0.51 1.3 C1.77 37.69, -19.37 176.91, 7.92 219.17 C35.21 261.42, 138.23 249.27, 164.23 254.85" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(116.97655503852786 128.67869596324704) rotate(0 79.43591774469584 126.99154703200844)"><path d="M140.31 262.11 C146.46 260.33, 155.57 257.13, 164.23 254.85 M140.31 262.11 C147.33 259.38, 154.74 257.72, 164.23 254.85" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(116.97655503852786 128.67869596324704) rotate(0 79.43591774469584 126.99154703200844)"><path d="M141.24 245.03 C147.14 248.68, 155.95 250.91, 164.23 254.85 M141.24 245.03 C148.14 247.55, 155.27 251.12, 164.23 254.85" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(400.26450742638235 126.15760381316147) rotate(0 0.1301044309677195 24.439764024615755)"><path d="M0.11 0.11 C0.21 8.11, 0.77 40.46, 0.76 48.48 M-0.5 -0.31 C-0.46 8.02, 0.21 41.03, 0.49 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(400.26450742638235 126.15760381316147) rotate(0 0.1301044309677195 24.439764024615755)"><path d="M-8.38 26.65 C-5.19 34.57, -2.43 40.67, 0.49 49.19 M-8.38 26.65 C-5.76 32.15, -4.23 38.16, 0.49 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(400.26450742638235 126.15760381316147) rotate(0 0.1301044309677195 24.439764024615755)"><path d="M8.18 26.22 C5.97 34.3, 3.32 40.54, 0.49 49.19 M8.18 26.22 C6.55 31.88, 3.83 38.01, 0.49 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(400.79651421561266 282.2131943987241) rotate(0 0.22226737518252548 24.26726710375675)"><path d="M-0.53 -0.42 C-0.38 7.69, 0.8 40.76, 0.98 48.95 M0.19 0.55 C0.32 8.5, 0.82 40.37, 0.82 48.25" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(400.79651421561266 282.2131943987241) rotate(0 0.22226737518252548 24.26726710375675)"><path d="M-7.67 25.57 C-5.74 33.61, -1.41 39.8, 0.82 48.25 M-7.67 25.57 C-5.05 33.11, -1.41 41.5, 0.82 48.25" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(400.79651421561266 282.2131943987241) rotate(0 0.22226737518252548 24.26726710375675)"><path d="M8.89 25.41 C5.69 33.42, 4.88 39.67, 0.82 48.25 M8.89 25.41 C5.63 33.07, 3.38 41.52, 0.82 48.25" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(400.93349365228926 432.98312841639597) rotate(0 0.3785265712478463 24.535335349170907)"><path d="M-0.13 0.24 C0.05 8.23, 0.78 40.43, 0.89 48.47 M0.8 -0.11 C0.96 8.21, 0.78 40.99, 0.69 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(400.93349365228926 432.98312841639597) rotate(0 0.3785265712478463 24.535335349170907)"><path d="M-7.42 26.36 C-4.73 33.24, -2.96 38.64, 0.69 49.19 M-7.42 26.36 C-5.36 33.54, -2.44 40.5, 0.69 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(400.93349365228926 432.98312841639597) rotate(0 0.3785265712478463 24.535335349170907)"><path d="M9.15 26.49 C7.44 33.31, 4.81 38.67, 0.69 49.19 M9.15 26.49 C6.19 33.56, 4.1 40.48, 0.69 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(637.0531381216375 279.855213477077) rotate(0 -57.891527346304485 46.60797353495241)"><path d="M-0.48 -0.87 C-3.73 11.25, 1.07 58.23, -18.35 73.84 C-37.76 89.44, -100.55 89.67, -116.98 92.77 M1.47 1.29 C-2.06 13.55, 0.33 56.53, -19.46 71.99 C-39.24 87.46, -101.14 90.51, -117.25 94.08" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(637.0531381216375 279.855213477077) rotate(0 -57.891527346304485 46.60797353495241)"><path d="M-95.17 82.36 C-105.19 86.36, -111.55 93.06, -117.25 94.08 M-95.17 82.36 C-101.31 85.02, -108.77 88.37, -117.25 94.08" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(637.0531381216375 279.855213477077) rotate(0 -57.891527346304485 46.60797353495241)"><path d="M-92.8 99.3 C-103.91 96.63, -111.21 96.67, -117.25 94.08 M-92.8 99.3 C-99.8 96.89, -107.97 95.16, -117.25 94.08" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(845.0587329975301 284.21712113042065) rotate(0 -162.67930168177395 56.56735605023431)"><path d="M0.06 -0.75 C-6.92 15.88, 11.27 81.89, -42.9 100.71 C-97.06 119.54, -277.98 110.37, -324.93 112.22 M-1.36 1.47 C-7.87 18.16, 13.49 83.24, -40.52 101.86 C-94.53 120.48, -278.19 111.45, -325.42 113.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(845.0587329975301 284.21712113042065) rotate(0 -162.67930168177395 56.56735605023431)"><path d="M-301.89 104.74 C-308.21 105.52, -313.34 109.32, -325.42 113.19 M-301.89 104.74 C-307.42 106.79, -313.09 109.26, -325.42 113.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(845.0587329975301 284.21712113042065) rotate(0 -162.67930168177395 56.56735605023431)"><path d="M-301.97 121.84 C-308.4 118.89, -313.51 118.96, -325.42 113.19 M-301.97 121.84 C-307.6 119.7, -313.25 117.99, -325.42 113.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(216.7744256961314 75.55969650441648) rotate(0 29.336689349718768 0.02151378544505178)"><path d="M0.67 1.02 C10.56 1.31, 49.51 0.86, 59.12 0.69 M-0.44 0.51 C9.28 0.55, 47.8 -1.21, 57.97 -1.07" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(216.7744256961314 75.55969650441648) rotate(0 29.336689349718768 0.02151378544505178)"><path d="M34.65 7.92 C42.96 4.66, 48.94 0.78, 57.97 -1.07 M34.65 7.92 C41.47 5.72, 48.8 3.07, 57.97 -1.07" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(216.7744256961314 75.55969650441648) rotate(0 29.336689349718768 0.02151378544505178)"><path d="M34.32 -9.17 C42.58 -7.65, 48.65 -6.74, 57.97 -1.07 M34.32 -9.17 C41.4 -6.21, 48.82 -3.71, 57.97 -1.07" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g transform="translate(427.28441033483523 608.4006578107978) rotate(0 119.65793700516224 17.5)"><text x="0" y="24.668" font-family="Excalifont, Segoe UI Emoji" font-size="28px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">x multiple samples</text></g><g stroke-linecap="round" transform="translate(309.8036672251775 701.7855233736506) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C58.69 -0.58, 102.4 3.62, 153.5 0 C169.35 0.54, 174.25 6.75, 174.42 20.92 C174.06 31.82, 171.76 49.31, 174.42 62.75 C171.44 77.69, 166.04 81.2, 153.5 83.66 C118.99 87.98, 79.82 83.96, 20.92 83.66 C8.45 86.73, 3.02 73.53, 0 62.75 C2.97 49.92, 2.8 39.79, 0 20.92 C2.09 5.73, 5.11 -3.09, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C64.26 0.26, 107.25 1.48, 153.5 0 M20.92 0 C58.08 -0.03, 95.3 1.16, 153.5 0 M153.5 0 C165.54 -0.45, 175.12 6.79, 174.42 20.92 M153.5 0 C168.36 2.1, 175.35 8.33, 174.42 20.92 M174.42 20.92 C172.6 37.38, 175.58 53.65, 174.42 62.75 M174.42 20.92 C174.92 30.57, 173.08 37.72, 174.42 62.75 M174.42 62.75 C173.67 76.16, 166.73 84.44, 153.5 83.66 M174.42 62.75 C175.37 76.92, 167.86 84.65, 153.5 83.66 M153.5 83.66 C121.41 82.01, 85.62 83, 20.92 83.66 M153.5 83.66 C106.94 82.48, 59.27 81.53, 20.92 83.66 M20.92 83.66 C6.23 85.56, -0.54 75.11, 0 62.75 M20.92 83.66 C8.06 81.57, 1.99 76.01, 0 62.75 M0 62.75 C0.44 47.26, -0.69 35.54, 0 20.92 M0 62.75 C-0.79 49.14, -0.63 36.4, 0 20.92 M0 20.92 C-1.29 7.53, 5.29 0.82, 20.92 0 M0 20.92 C-1.61 6.26, 7.51 -0.97, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(330.89271278668036 731.1156014986506) rotate(0 66.11997437477112 12.5)"><text x="66.11997437477112" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GVCF + index</text></g><g stroke-linecap="round" transform="translate(329.7791772392834 724.7337375089094) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C55.52 0.8, 88.18 1.15, 153.5 0 C167.35 -1.13, 177.72 4.94, 174.42 20.92 C170.86 35.06, 171.41 42.26, 174.42 62.75 C177.85 74.7, 163.85 83.7, 153.5 83.66 C108.99 84.99, 69.84 84.12, 20.92 83.66 C4.37 80.89, -1.07 78.6, 0 62.75 C0.31 52.33, 1.59 32.15, 0 20.92 C1.21 5.86, 6.63 -2.05, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C72.72 -1.91, 119.26 0.87, 153.5 0 M20.92 0 C71.72 0.38, 124.45 1.23, 153.5 0 M153.5 0 C167.56 0.83, 172.5 5.96, 174.42 20.92 M153.5 0 C169.43 -0.47, 172.21 5.93, 174.42 20.92 M174.42 20.92 C175.54 31.43, 174.3 42.35, 174.42 62.75 M174.42 20.92 C173.52 38.1, 174.4 54.12, 174.42 62.75 M174.42 62.75 C175.29 78.49, 169.19 84.04, 153.5 83.66 M174.42 62.75 C172.58 76.82, 165.38 85.96, 153.5 83.66 M153.5 83.66 C125.14 82.91, 101.76 83.92, 20.92 83.66 M153.5 83.66 C108.92 83.72, 63.04 82.71, 20.92 83.66 M20.92 83.66 C8.68 81.83, 1.34 75.51, 0 62.75 M20.92 83.66 C5.29 82.03, -1.63 74.91, 0 62.75 M0 62.75 C0.15 54.24, 0.39 42.85, 0 20.92 M0 62.75 C-0.78 47.89, 0.98 32.12, 0 20.92 M0 20.92 C-1.82 5.83, 8.02 -1.78, 20.92 0 M0 20.92 C-2.11 8.96, 7.29 -0.21, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(350.86822280078627 754.0638156339094) rotate(0 66.11997437477112 12.5)"><text x="66.11997437477112" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GVCF + index</text></g><g stroke-linecap="round" transform="translate(352.6052981695668 747.2838216279436) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C51.37 -4.83, 75.92 -5.19, 153.5 0 C168.51 2.71, 175.59 8.27, 174.42 20.92 C171.42 33.98, 171.26 56.26, 174.42 62.75 C173.64 73.12, 164.21 86.11, 153.5 83.66 C114.93 79.59, 73.66 82.22, 20.92 83.66 C9.61 84.72, 1.14 77.82, 0 62.75 C2.59 49.53, -0.9 41.21, 0 20.92 C-2.97 7.73, 10.16 2.4, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C58.25 -0.03, 96.84 1.05, 153.5 0 M20.92 0 C69.73 -0.06, 119.5 1.16, 153.5 0 M153.5 0 C167.61 -1.83, 173.64 5.38, 174.42 20.92 M153.5 0 C169.29 -1.71, 174.13 8.35, 174.42 20.92 M174.42 20.92 C172.95 28.65, 175.59 39.4, 174.42 62.75 M174.42 20.92 C173.84 28.79, 173.76 37.06, 174.42 62.75 M174.42 62.75 C172.9 75.42, 168.98 83.87, 153.5 83.66 M174.42 62.75 C172.95 77.56, 167.33 83.19, 153.5 83.66 M153.5 83.66 C101.39 84.56, 53.19 84.62, 20.92 83.66 M153.5 83.66 C109.18 82.04, 66.1 82.77, 20.92 83.66 M20.92 83.66 C6.5 84.64, -1.26 78.11, 0 62.75 M20.92 83.66 C7.3 84.48, -1.18 76.68, 0 62.75 M0 62.75 C1.55 47.13, 1.73 32.28, 0 20.92 M0 62.75 C0.85 47.49, -0.82 31.03, 0 20.92 M0 20.92 C1.22 7.89, 8.93 1.49, 20.92 0 M0 20.92 C-0.23 5.92, 7.08 0.23, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(373.6943437310697 776.6138997529436) rotate(0 66.11997437477112 12.5)"><text x="66.11997437477112" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GVCF + index</text></g><g stroke-linecap="round"><g transform="translate(402.82760984364415 591.8180966675825) rotate(0 0.05654684182462688 45.69977301972045)"><path d="M-1.18 -0.65 C-1.46 14.7, -1.24 76.63, -0.85 92.05 M0.4 1.63 C0.42 16.63, 1.31 75.6, 1.42 90.3" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(402.82760984364415 591.8180966675825) rotate(0 0.05654684182462688 45.69977301972045)"><path d="M-7.4 66.91 C-5.2 74.3, 0.36 86.41, 1.42 90.3 M-7.4 66.91 C-3.5 74.59, -1.45 82.2, 1.42 90.3" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(402.82760984364415 591.8180966675825) rotate(0 0.05654684182462688 45.69977301972045)"><path d="M9.7 66.71 C5.53 74.01, 4.72 86.19, 1.42 90.3 M9.7 66.71 C7.69 74.51, 3.83 82.19, 1.42 90.3" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round" transform="translate(621.1891155157545 882.4930887867315) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C74.98 1.79, 122.94 -3.74, 153.5 0 C170.41 -2.27, 177.04 3.94, 174.42 20.92 C176.53 33.79, 175.84 47.25, 174.42 62.75 C172.08 80.02, 169.18 85.9, 153.5 83.66 C104.35 83.84, 47.82 85.52, 20.92 83.66 C7.25 80.53, -1.55 76.12, 0 62.75 C-1.77 49.72, -1.33 36.28, 0 20.92 C-2.6 5.76, 3.4 3.05, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C48.68 0.33, 78.58 0.92, 153.5 0 M20.92 0 C64.13 -0.59, 105.41 -0.37, 153.5 0 M153.5 0 C167.62 1.98, 173.94 6.99, 174.42 20.92 M153.5 0 C168.54 0, 173.37 7.24, 174.42 20.92 M174.42 20.92 C174.15 36.56, 175.23 52.83, 174.42 62.75 M174.42 20.92 C173.83 37.02, 173.76 54.83, 174.42 62.75 M174.42 62.75 C173.34 77.92, 168.24 84.34, 153.5 83.66 M174.42 62.75 C173.79 75.88, 165.24 82.26, 153.5 83.66 M153.5 83.66 C113.72 83.16, 74.1 81.44, 20.92 83.66 M153.5 83.66 C116.31 83.38, 78.19 82.45, 20.92 83.66 M20.92 83.66 C5.42 81.93, 1.8 78.04, 0 62.75 M20.92 83.66 C7.46 84.87, 1.81 76.52, 0 62.75 M0 62.75 C-1.74 46.05, 1.7 30.7, 0 20.92 M0 62.75 C-0.46 52, 0.11 39.37, 0 20.92 M0 20.92 C-1.91 5.08, 8.2 -1.62, 20.92 0 M0 20.92 C-1.09 5.51, 7.54 2.16, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(641.2581854895317 899.3231669117315) rotate(0 67.1399499624967 25)"><text x="67.13994996249676" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GenomicsDB</text><text x="67.13994996249676" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">variant store</text></g><g stroke-linecap="round" transform="translate(601.2405951657298 722.6037647004479) rotate(0 109.33398437499994 42.50000000000006)"><path d="M3.19 8.07 C3.19 8.07, 3.19 8.07, 3.19 8.07 M3.19 8.07 C3.19 8.07, 3.19 8.07, 3.19 8.07 M0.7 23.13 C9.2 15.6, 15.37 4.5, 20.38 0.49 M0.7 23.13 C9.09 15.65, 14.93 5.94, 20.38 0.49 M0.83 35.17 C9.94 23.22, 20.55 10.38, 31.01 0.46 M0.83 35.17 C11.61 21.68, 22.48 9.41, 31.01 0.46 M0.96 47.21 C8.68 38.49, 20.53 28.27, 41.64 0.42 M0.96 47.21 C15.81 31.63, 29.92 14.44, 41.64 0.42 M1.1 59.25 C17.85 43.23, 29.15 26.51, 52.27 0.39 M1.1 59.25 C14.96 41.69, 30.26 24.92, 52.27 0.39 M1.23 71.29 C21 51.73, 38.18 31.31, 62.9 0.35 M1.23 71.29 C14.26 55.37, 27.83 39.55, 62.9 0.35 M5.3 78.81 C21.6 56.94, 40.2 38.02, 73.53 0.32 M5.3 78.81 C20.31 63.21, 33.84 45.65, 73.53 0.32 M11.99 83.3 C32.98 59.59, 52.82 34.5, 84.16 0.28 M11.99 83.3 C40.38 52.09, 66.1 21.48, 84.16 0.28 M20.65 85.53 C49.15 53.02, 76.23 21.62, 94.79 0.25 M20.65 85.53 C38.78 66.77, 55.2 47.97, 94.79 0.25 M31.28 85.5 C52.1 61.66, 71.15 38.05, 105.42 0.21 M31.28 85.5 C49.36 64.86, 66.1 44.79, 105.42 0.21 M41.91 85.46 C59.5 64.95, 76.9 46.65, 116.05 0.18 M41.91 85.46 C55.86 67.61, 73.17 50.69, 116.05 0.18 M52.54 85.43 C67.62 68.78, 81.22 50.64, 126.68 0.14 M52.54 85.43 C72.64 60.55, 93.1 36.74, 126.68 0.14 M63.17 85.39 C90.84 53.97, 114 23.03, 137.31 0.11 M63.17 85.39 C84.27 61.64, 104.01 40.62, 137.31 0.11 M73.8 85.36 C102.06 51.99, 127.95 21.56, 147.94 0.08 M73.8 85.36 C89.63 65.62, 106.64 46.77, 147.94 0.08 M84.44 85.32 C107.02 62.7, 124.1 38.05, 157.91 0.8 M84.44 85.32 C103.56 63.47, 122.22 40.88, 157.91 0.8 M95.07 85.29 C120.19 54.95, 149.02 22.39, 168.54 0.76 M95.07 85.29 C115.89 63.61, 133.96 40.44, 168.54 0.76 M105.7 85.25 C127.83 64.13, 147.12 41.77, 179.17 0.73 M105.7 85.25 C126.4 60.73, 149.84 34.34, 179.17 0.73 M116.33 85.22 C146.58 52.53, 174.42 16.87, 189.8 0.69 M116.33 85.22 C144.29 52.41, 172.6 18.85, 189.8 0.69 M126.3 85.94 C143.64 68.1, 158.51 48.37, 200.43 0.66 M126.3 85.94 C143.71 64.03, 163.79 43.06, 200.43 0.66 M136.93 85.91 C153.85 63.72, 174.2 44.59, 209.1 2.89 M136.93 85.91 C158.48 61.45, 181.4 35.55, 209.1 2.89 M147.56 85.87 C173.7 55.4, 197.02 29.35, 215.13 8.14 M147.56 85.87 C161.02 68.91, 175.59 53.52, 215.13 8.14 M158.19 85.84 C168.24 70.82, 183.69 56.82, 218.55 16.4 M158.19 85.84 C174.73 65.23, 193.27 45.45, 218.55 16.4 M168.82 85.8 C185.87 63.76, 202.3 46.96, 218.68 28.44 M168.82 85.8 C179.46 73.51, 192.52 59.2, 218.68 28.44 M179.45 85.77 C190.61 72.66, 206.45 57.46, 218.81 40.49 M179.45 85.77 C192.92 70.21, 207.11 54.18, 218.81 40.49 M190.08 85.73 C198.2 73.84, 208.18 63.97, 218.95 52.53 M190.08 85.73 C197.67 78.3, 203.53 70.84, 218.95 52.53 M201.37 84.94 C205.31 78.72, 212.46 72.78, 219.08 64.57 M201.37 84.94 C206.3 79.24, 211.3 74.15, 219.08 64.57 M3.31 76.98 C3.31 76.98, 3.31 76.98, 3.31 76.98 M3.31 76.98 C3.31 76.98, 3.31 76.98, 3.31 76.98 M24.41 84.72 C16.14 77.22, 9.06 71.15, 1.01 64.38 M24.41 84.72 C17.4 77.45, 9.32 70.8, 1.01 64.38 M36.45 84.58 C23.62 75.72, 14.4 65.82, 0.98 53.75 M36.45 84.58 C22.61 73.43, 9.55 61.22, 0.98 53.75 M49.24 85.11 C36.91 76.9, 25.77 63.94, 0.94 43.12 M49.24 85.11 C38.58 75.35, 26.76 66.03, 0.94 43.12 M61.28 84.97 C45.56 70.52, 31.62 58.33, 0.91 32.49 M61.28 84.97 C43.63 70.51, 27.31 56.51, 0.91 32.49 M73.32 84.84 C57.98 71.07, 39.38 57.26, 0.87 21.86 M73.32 84.84 C51.15 64.94, 28.53 45.35, 0.87 21.86 M85.37 84.71 C59.54 60.07, 29.64 36.31, 1.59 11.89 M85.37 84.71 C69.05 70.78, 52.07 55.47, 1.59 11.89 M97.41 84.58 C75.18 61.29, 50.03 40.9, 6.09 5.19 M97.41 84.58 C68.9 59.27, 40.63 35.79, 6.09 5.19 M110.2 85.1 C86.29 63.05, 59.35 42.63, 13.6 1.12 M110.2 85.1 C83.15 62.98, 56 37.73, 13.6 1.12 M122.24 84.96 C101.11 68.93, 82.39 48.65, 24.13 -0.32 M122.24 84.96 C83.72 50.8, 43.57 16.64, 24.13 -0.32 M134.28 84.83 C110.29 63.27, 86.4 39.57, 36.93 0.2 M134.28 84.83 C96.72 52.21, 61.13 20.4, 36.93 0.2 M146.32 84.7 C109.62 53.9, 72.39 22.66, 48.97 0.07 M146.32 84.7 C113.81 55.38, 80.25 27.21, 48.97 0.07 M158.37 84.57 C125.31 57.48, 93.18 30.15, 61.01 -0.07 M158.37 84.57 C120.42 54.52, 83.94 21.34, 61.01 -0.07 M171.16 85.09 C145.72 61.63, 123.21 42.14, 73.05 -0.2 M171.16 85.09 C137.43 55.61, 101.88 25.18, 73.05 -0.2 M183.2 84.96 C145.67 49.54, 106.92 17.04, 85.09 -0.33 M183.2 84.96 C154.65 62.05, 128.51 38.78, 85.09 -0.33 M195.24 84.82 C161.23 56.8, 128.73 28.35, 97.89 0.19 M195.24 84.82 C170.36 63.06, 147.09 42.1, 97.89 0.19 M205.02 82.72 C173.15 55.68, 139.64 26.33, 109.93 0.06 M205.02 82.72 C173.85 56.24, 143.15 30.08, 109.93 0.06 M212.53 78.65 C181.48 52.12, 155.14 28.47, 121.97 -0.08 M212.53 78.65 C182.23 51.36, 152.83 25.13, 121.97 -0.08 M217.78 72.61 C200.49 55.66, 185.47 42.33, 134.01 -0.21 M217.78 72.61 C195.92 53.05, 173.66 34.21, 134.01 -0.21 M219.26 63.3 C195.23 43.48, 174.79 23.31, 146.05 -0.34 M219.26 63.3 C200.07 46.72, 182.72 30.56, 146.05 -0.34 M219.22 52.67 C199.41 33.68, 177.66 15.95, 158.84 0.18 M219.22 52.67 C200.54 36.69, 182.59 21.51, 158.84 0.18 M219.19 42.04 C201.88 25.34, 183.52 10.43, 170.89 0.05 M219.19 42.04 C201.94 27.38, 184.09 12.22, 170.89 0.05 M219.15 31.41 C209.22 22.01, 199.18 13.54, 182.93 -0.09 M219.15 31.41 C207.47 20.24, 194.64 9.47, 182.93 -0.09 M219.12 20.78 C209.31 14.51, 201.42 6.28, 194.97 -0.22 M219.12 20.78 C210.42 13.14, 203.82 5.72, 194.97 -0.22" stroke="#ffec99" stroke-width="1" fill="none"></path><path d="M21.25 0 C61.3 2.45, 101.11 1.53, 197.42 0 M21.25 0 C88.34 0.59, 155.78 0.58, 197.42 0 M197.42 0 C212.87 1.03, 220.5 5.2, 218.67 21.25 M197.42 0 C213.54 -1.24, 219.99 8.86, 218.67 21.25 M218.67 21.25 C218.87 39.06, 220.77 51.52, 218.67 63.75 M218.67 21.25 C219.01 37.59, 219.39 53.78, 218.67 63.75 M218.67 63.75 C218.41 77.27, 210.33 84.46, 197.42 85 M218.67 63.75 C217.01 75.86, 209.94 84.41, 197.42 85 M197.42 85 C147.64 88.18, 94.5 86.46, 21.25 85 M197.42 85 C140.4 84.1, 84.98 83.95, 21.25 85 M21.25 85 C5.77 84.55, 0.77 77.32, 0 63.75 M21.25 85 C6.09 87.09, 2.08 77.82, 0 63.75 M0 63.75 C0.29 49.19, -0.2 33.62, 0 21.25 M0 63.75 C0.58 47.4, -0.38 32.13, 0 21.25 M0 21.25 C-1.74 7.65, 8.56 -0.88, 21.25 0 M0 21.25 C1.8 8.22, 5.68 2.28, 21.25 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(620.7646582760814 740.1037647004479) rotate(0 89.80992126464838 25)"><text x="89.80992126464844" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GATK </text><text x="89.80992126464844" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GenomicsDBImport</text></g><g stroke-linecap="round"><g transform="translate(543.183165362838 764.239098919755) rotate(0 19.898706537638873 0.9641304987845842)"><path d="M0.28 0.53 C7.03 0.75, 33.38 0.9, 40.04 0.87 M-0.24 0.33 C6.47 0.69, 32.84 1.53, 39.54 1.6" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(543.183165362838 764.239098919755) rotate(0 19.898706537638873 0.9641304987845842)"><path d="M20.52 8.03 C25.57 6.38, 28.76 5.4, 39.54 1.6 M20.52 8.03 C24.54 6.72, 28.93 5.18, 39.54 1.6" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(543.183165362838 764.239098919755) rotate(0 19.898706537638873 0.9641304987845842)"><path d="M20.83 -5.7 C25.77 -4.42, 28.89 -2.46, 39.54 1.6 M20.83 -5.7 C24.76 -4.03, 29.09 -2.59, 39.54 1.6" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(708.0635290596329 820.5643813804256) rotate(0 0.012863403921187455 23.977137729086053)"><path d="M0.23 0.03 C0.27 8.09, 0.24 39.86, 0.36 47.94 M-0.32 -0.43 C-0.36 7.71, -0.32 40.12, -0.12 48.38" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(708.0635290596329 820.5643813804256) rotate(0 0.012863403921187455 23.977137729086053)"><path d="M-8.65 25.71 C-6 32.63, -4.31 40.84, -0.12 48.38 M-8.65 25.71 C-5.78 32.42, -3.77 39.84, -0.12 48.38" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(708.0635290596329 820.5643813804256) rotate(0 0.012863403921187455 23.977137729086053)"><path d="M7.91 25.53 C5.38 32.54, 1.89 40.8, -0.12 48.38 M7.91 25.53 C5.7 32.29, 2.61 39.77, -0.12 48.38" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(878.0993942008804 282.8571356518237) rotate(0 -20.347661301520816 392.067855342782)"><path d="M1.02 -0.69 C1.06 110.71, 7.5 538.43, 0.11 669.2 C-7.27 799.98, -36.33 764.54, -43.29 783.93 M0.09 1.56 C0.08 112.57, 7.1 537.04, -0.31 667.58 C-7.72 798.12, -36.99 765.51, -44.35 784.82" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(878.0993942008804 282.8571356518237) rotate(0 -20.347661301520816 392.067855342782)"><path d="M-31.44 763.42 C-36.16 771.36, -37.15 774.91, -44.35 784.82 M-31.44 763.42 C-33.65 768.21, -36.7 772.91, -44.35 784.82" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(878.0993942008804 282.8571356518237) rotate(0 -20.347661301520816 392.067855342782)"><path d="M-20.7 776.73 C-28.6 780.93, -32.68 780.65, -44.35 784.82 M-20.7 776.73 C-25.53 778.51, -31.06 780.11, -44.35 784.82" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round" transform="translate(600.1364479207297 1034.6328013487562) rotate(0 109.33398437499994 42.5)"><path d="M3.19 8.07 C3.19 8.07, 3.19 8.07, 3.19 8.07 M3.19 8.07 C3.19 8.07, 3.19 8.07, 3.19 8.07 M0.7 23.13 C6.93 17.22, 12.55 9.53, 20.38 0.49 M0.7 23.13 C6.7 15.81, 14.12 7.18, 20.38 0.49 M0.83 35.17 C12.23 25.84, 18.26 14.14, 31.01 0.46 M0.83 35.17 C8.28 25.35, 15.71 17.53, 31.01 0.46 M0.96 47.21 C15.47 30.88, 30.34 13.94, 41.64 0.42 M0.96 47.21 C15.55 30.82, 28.35 13.62, 41.64 0.42 M1.1 59.25 C18.27 41.21, 35.19 21.42, 52.27 0.39 M1.1 59.25 C17.95 40.22, 34.55 21.68, 52.27 0.39 M1.23 71.29 C22.14 45.03, 45.19 21.13, 62.9 0.35 M1.23 71.29 C19.95 49.1, 38.92 28.25, 62.9 0.35 M5.3 78.81 C18.83 60.77, 35.58 44.17, 73.53 0.32 M5.3 78.81 C23.97 58.65, 40.16 37.38, 73.53 0.32 M11.99 83.3 C34.89 58.67, 57.55 30.42, 84.16 0.28 M11.99 83.3 C35.26 55.44, 59.72 28.64, 84.16 0.28 M20.65 85.53 C49.92 52.31, 79.32 19.12, 94.79 0.25 M20.65 85.53 C39.32 64.1, 57.82 42.44, 94.79 0.25 M31.28 85.5 C58.19 54.02, 86.26 20.56, 105.42 0.21 M31.28 85.5 C51.05 62.46, 68.43 40.66, 105.42 0.21 M41.91 85.46 C64.08 56.55, 85.95 33.48, 116.05 0.18 M41.91 85.46 C66.57 56.55, 92.23 25.64, 116.05 0.18 M52.54 85.43 C69.02 67.16, 85.84 49.48, 126.68 0.14 M52.54 85.43 C76.82 57.51, 101.16 30.12, 126.68 0.14 M63.17 85.39 C86.54 58.87, 107.89 34.04, 137.31 0.11 M63.17 85.39 C87.08 56.61, 112.33 28, 137.31 0.11 M73.8 85.36 C103.17 51.55, 129.68 18.93, 147.94 0.08 M73.8 85.36 C92.22 62.83, 111.3 42.3, 147.94 0.08 M84.44 85.32 C112.83 49.83, 143.04 16.87, 157.91 0.8 M84.44 85.32 C102.77 64.66, 121.81 41.56, 157.91 0.8 M95.07 85.29 C111.38 66.74, 127.68 44.97, 168.54 0.76 M95.07 85.29 C121.81 53.95, 150.1 20.29, 168.54 0.76 M105.7 85.25 C124.65 62.57, 141.66 42.45, 179.17 0.73 M105.7 85.25 C127.67 59.88, 151.47 32.43, 179.17 0.73 M116.33 85.22 C143.19 56.24, 168.5 25.35, 189.8 0.69 M116.33 85.22 C140.23 58.21, 163.2 28.42, 189.8 0.69 M126.3 85.94 C143.85 63.48, 164.12 38.6, 200.43 0.66 M126.3 85.94 C151.2 58.64, 174.72 32.68, 200.43 0.66 M136.93 85.91 C162.78 55.07, 187.95 24.34, 209.1 2.89 M136.93 85.91 C160.02 60.24, 182.71 33.21, 209.1 2.89 M147.56 85.87 C163.13 63.71, 182.52 47.64, 215.13 8.14 M147.56 85.87 C166.59 64.68, 184.93 42.3, 215.13 8.14 M158.19 85.84 C177.69 65.69, 195.12 42.78, 218.55 16.4 M158.19 85.84 C173.79 65.49, 190.41 48.02, 218.55 16.4 M168.82 85.8 C186.14 66.16, 203.62 44.54, 218.68 28.44 M168.82 85.8 C178.89 73.36, 190.88 60.33, 218.68 28.44 M179.45 85.77 C190.51 73.08, 198.33 63.35, 218.81 40.49 M179.45 85.77 C192.13 70.47, 205.17 55.83, 218.81 40.49 M190.08 85.73 C200.23 74.97, 208.05 62.15, 218.95 52.53 M190.08 85.73 C200.52 75.14, 208.65 64.1, 218.95 52.53 M201.37 84.94 C206.93 78.21, 214.64 71.04, 219.08 64.57 M201.37 84.94 C206.4 78.84, 210.47 74.97, 219.08 64.57 M3.31 76.98 C3.31 76.98, 3.31 76.98, 3.31 76.98 M3.31 76.98 C3.31 76.98, 3.31 76.98, 3.31 76.98 M24.41 84.72 C17.91 80.19, 13.32 77.07, 1.01 64.38 M24.41 84.72 C15.4 76.42, 5.5 69.31, 1.01 64.38 M36.45 84.58 C27.76 76.81, 19.22 69.21, 0.98 53.75 M36.45 84.58 C23.59 71.45, 7.87 59.51, 0.98 53.75 M49.24 85.11 C34.84 72.81, 23.78 61.47, 0.94 43.12 M49.24 85.11 C33.16 71.89, 16.8 58.49, 0.94 43.12 M61.28 84.97 C41.53 64.73, 21.99 47.4, 0.91 32.49 M61.28 84.97 C39.47 67.11, 19.46 47.17, 0.91 32.49 M73.32 84.84 C46.48 61.7, 22 42.16, 0.87 21.86 M73.32 84.84 C47.51 61.39, 21.42 39.17, 0.87 21.86 M85.37 84.71 C60.82 64.62, 39.65 45, 1.59 11.89 M85.37 84.71 C52.44 57.16, 20.98 28.38, 1.59 11.89 M97.41 84.58 C74.16 66.72, 55.34 48.59, 6.09 5.19 M97.41 84.58 C78.87 67.54, 58.16 50.64, 6.09 5.19 M110.2 85.1 C72.92 52.74, 35.45 20.21, 13.6 1.12 M110.2 85.1 C80.4 58.02, 48.81 30.1, 13.6 1.12 M122.24 84.96 C84.53 53.12, 44.1 16.39, 24.13 -0.32 M122.24 84.96 C99.21 64.64, 75.76 43.01, 24.13 -0.32 M134.28 84.83 C112.23 62.65, 87.51 41.79, 36.93 0.2 M134.28 84.83 C114.15 68.79, 95.29 52.58, 36.93 0.2 M146.32 84.7 C122.24 64.66, 104.52 47.82, 48.97 0.07 M146.32 84.7 C124.9 65.19, 102.19 45.19, 48.97 0.07 M158.37 84.57 C131.31 63.07, 103.29 38.34, 61.01 -0.07 M158.37 84.57 C139.14 67.08, 118.94 49.62, 61.01 -0.07 M171.16 85.09 C133.63 52.58, 96.92 20.72, 73.05 -0.2 M171.16 85.09 C137.26 54.72, 100.51 24.69, 73.05 -0.2 M183.2 84.96 C151.97 60.07, 123.31 33.65, 85.09 -0.33 M183.2 84.96 C151.48 57.57, 118.43 27.78, 85.09 -0.33 M195.24 84.82 C171.56 63.62, 148.69 46.59, 97.89 0.19 M195.24 84.82 C165.66 61.11, 138.35 35.91, 97.89 0.19 M205.02 82.72 C179.68 58.58, 151.05 36.72, 109.93 0.06 M205.02 82.72 C173.42 55.01, 141.7 28.12, 109.93 0.06 M212.53 78.65 C188.9 55.76, 164.06 36.36, 121.97 -0.08 M212.53 78.65 C190.36 57.59, 166.53 39.17, 121.97 -0.08 M217.78 72.61 C189.59 47.57, 159.97 25.29, 134.01 -0.21 M217.78 72.61 C189.47 47.19, 160.01 22.2, 134.01 -0.21 M219.26 63.3 C197 45.02, 180.38 31.32, 146.05 -0.34 M219.26 63.3 C193.88 39.9, 169.25 19.55, 146.05 -0.34 M219.22 52.67 C202.1 36.6, 182.72 22.11, 158.84 0.18 M219.22 52.67 C204.03 38.29, 187.97 25.44, 158.84 0.18 M219.19 42.04 C204.76 31.51, 193.51 19.21, 170.89 0.05 M219.19 42.04 C206.65 30.54, 193.78 21.06, 170.89 0.05 M219.15 31.41 C212.76 25.08, 203.39 17.6, 182.93 -0.09 M219.15 31.41 C206.05 19.96, 191.86 6.95, 182.93 -0.09 M219.12 20.78 C213.43 14.78, 205.89 11.02, 194.97 -0.22 M219.12 20.78 C210.77 12.86, 202 4.65, 194.97 -0.22" stroke="#ffec99" stroke-width="1" fill="none"></path><path d="M21.25 0 C75.16 0.81, 128.66 -1.22, 197.42 0 M21.25 0 C67 -1, 113.65 0.33, 197.42 0 M197.42 0 C213.58 -1.88, 217.11 6.6, 218.67 21.25 M197.42 0 C212.1 -0.82, 217.9 7.39, 218.67 21.25 M218.67 21.25 C218.05 37.54, 219.09 51.68, 218.67 63.75 M218.67 21.25 C218.86 34.21, 218.43 48.7, 218.67 63.75 M218.67 63.75 C218.72 78.67, 211.39 86.7, 197.42 85 M218.67 63.75 C218.83 76.02, 209.99 82.93, 197.42 85 M197.42 85 C146.59 84.81, 93.44 82.97, 21.25 85 M197.42 85 C160.83 84.24, 124.15 82.46, 21.25 85 M21.25 85 C6.66 86.99, -0.05 77.76, 0 63.75 M21.25 85 C7.38 84.25, 2.13 78.08, 0 63.75 M0 63.75 C-0.25 52.72, -0.16 39.67, 0 21.25 M0 63.75 C-0.42 50.69, -0.07 38.96, 0 21.25 M0 21.25 C-1.21 6.23, 7.7 1.37, 21.25 0 M0 21.25 C1.44 6.7, 8.69 -0.9, 21.25 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(634.3404863023062 1052.1328013487562) rotate(0 75.1299459934234 25)"><text x="75.12994599342346" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GATK </text><text x="75.12994599342346" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GenotypeGVCFs</text></g><g stroke-linecap="round"><g transform="translate(709.0880503013873 1130.7422941087596) rotate(0 0.2608347071280832 23.913571700817442)"><path d="M-0.4 -0.19 C-0.27 7.83, 0.86 40.54, 0.92 48.59 M0.4 -0.77 C0.48 7.32, 0.63 39.48, 0.73 47.71" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(709.0880503013873 1130.7422941087596) rotate(0 0.2608347071280832 23.913571700817442)"><path d="M-7.74 25.01 C-5.2 33.36, -1.8 39.04, 0.73 47.71 M-7.74 25.01 C-5.24 30.15, -4.02 36.15, 0.73 47.71" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(709.0880503013873 1130.7422941087596) rotate(0 0.2608347071280832 23.913571700817442)"><path d="M8.83 24.88 C5.92 33.21, 3.88 38.94, 0.73 47.71 M8.83 24.88 C7.21 30.12, 4.31 36.15, 0.73 47.71" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(708.2100948409675 976.3420814315709) rotate(0 0.3216333956127073 24.37765970970554)"><path d="M-0.14 -0.36 C-0.07 7.63, 0.63 40.29, 0.72 48.42 M0.79 0.65 C0.79 8.96, 0.55 41.05, 0.42 49.11" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(708.2100948409675 976.3420814315709) rotate(0 0.3216333956127073 24.37765970970554)"><path d="M-7.61 26.26 C-5.39 34.66, -1.99 41.96, 0.42 49.11 M-7.61 26.26 C-4.86 33.22, -2.77 39.61, 0.42 49.11" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(708.2100948409675 976.3420814315709) rotate(0 0.3216333956127073 24.37765970970554)"><path d="M8.96 26.45 C5.45 34.76, 3.12 41.99, 0.42 49.11 M8.96 26.45 C6.75 33.39, 3.87 39.74, 0.42 49.11" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round" transform="translate(621.067022310778 1189.8754903103895) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C50.13 1.52, 88.63 1.54, 153.5 0 C169.23 -1.29, 171.26 7.24, 174.42 20.92 C177.28 36.78, 172.82 51.79, 174.42 62.75 C174.92 75.17, 164.83 85.5, 153.5 83.66 C128.2 86.99, 96.46 82.08, 20.92 83.66 C7.25 83.4, -3.1 79.24, 0 62.75 C-0.78 45.94, 3.68 36.87, 0 20.92 C1.85 6.66, 3.63 -2.48, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C55.69 -0.66, 89.05 1.2, 153.5 0 M20.92 0 C58.28 0.31, 96.41 0.68, 153.5 0 M153.5 0 C168.79 -1.76, 174.16 6.23, 174.42 20.92 M153.5 0 C166.84 1, 172.99 6.93, 174.42 20.92 M174.42 20.92 C175.79 35.04, 173.83 51.54, 174.42 62.75 M174.42 20.92 C173.67 33.42, 173.96 46.23, 174.42 62.75 M174.42 62.75 C176 78.25, 165.84 84.6, 153.5 83.66 M174.42 62.75 C174.79 75.77, 166.63 82.68, 153.5 83.66 M153.5 83.66 C119.97 81.9, 88.09 82.04, 20.92 83.66 M153.5 83.66 C118.93 83.11, 83.87 81.71, 20.92 83.66 M20.92 83.66 C7.64 83.38, -0.6 76.13, 0 62.75 M20.92 83.66 C5.46 82.78, -0.53 75.95, 0 62.75 M0 62.75 C-2.11 52.53, -0.34 41.88, 0 20.92 M0 62.75 C0.52 53.54, -0.02 43.95, 0 20.92 M0 20.92 C0.64 8.2, 6.89 0.99, 20.92 0 M0 20.92 C-1.57 4.76, 7.39 -0.24, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(651.6360894086312 1206.7055684353895) rotate(0 56.63995283842081 25)"><text x="56.63995283842087" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Joint-called</text><text x="56.63995283842087" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">VCF</text></g><g stroke-linecap="round"><g transform="translate(528.6342933469583 723.6557347863668) rotate(0 27.258060338754092 13.922532988289447)"><path d="M-0.71 -0.54 C8.44 4, 45.99 22.42, 55.23 27.15 M1.11 1.8 C10.06 6.49, 44.8 23.82, 54.15 28.38" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(528.6342933469583 723.6557347863668) rotate(0 27.258060338754092 13.922532988289447)"><path d="M29.3 25.64 C36.51 26.27, 44.53 26.36, 54.15 28.38 M29.3 25.64 C37.48 25.62, 45.53 27.14, 54.15 28.38" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(528.6342933469583 723.6557347863668) rotate(0 27.258060338754092 13.922532988289447)"><path d="M36.88 10.31 C41.61 15.89, 47.16 20.97, 54.15 28.38 M36.88 10.31 C42.7 15.05, 48.41 21.28, 54.15 28.38" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(545.0425584156804 805.7607609325732) rotate(0 20.329742944411294 -13.782084528743042)"><path d="M-0.35 -0.48 C6.24 -5.09, 33.27 -23.44, 40.01 -28.03 M0.47 0.47 C7.18 -4.06, 34.23 -22.89, 41.01 -27.58" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(545.0425584156804 805.7607609325732) rotate(0 20.329742944411294 -13.782084528743042)"><path d="M26.88 -7.59 C31.59 -16.15, 37.1 -23.84, 41.01 -27.58 M26.88 -7.59 C29.65 -12.11, 33.26 -16.11, 41.01 -27.58" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(545.0425584156804 805.7607609325732) rotate(0 20.329742944411294 -13.782084528743042)"><path d="M17.34 -21.35 C25.72 -24.52, 34.94 -26.84, 41.01 -27.58 M17.34 -21.35 C22.13 -22.92, 27.8 -23.96, 41.01 -27.58" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g transform="translate(425.76323401603327 447.0206387203026) rotate(0 54.66996765136719 12.5)"><text x="0" y="17.619999999999997" font-family="Excalifont, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GVCF mode</text></g></svg>
</figure>

<ol>
<li>Generate an index file for each BAM input file using Samtools</li>
<li>Run the GATK HaplotypeCaller on each BAM input file to generate a GVCF of per-sample genomic variant calls</li>
<li>Collect all the GVCFs and combine them into a GenomicsDB data store</li>
<li>Run joint genotyping on the combined GVCF data store to produce a cohort-level VCF</li>
</ol>
<p>We'll apply this to the same dataset as in Part 1.</p>
<hr />
<h2 id="0-warmup-run-samtools-and-gatk-directly">0. Warmup: Run Samtools and GATK directly<a class="headerlink" href="#0-warmup-run-samtools-and-gatk-directly" title="Permanent link">&para;</a></h2>
<p>Just like previously, we want to try out the commands manually before we attempt to wrap them in a workflow.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Make sure you're in the correct working directory:
 <code>cd /workspaces/training/nf4-science/genomics</code></p>
</div>
<h3 id="01-index-a-bam-input-file-with-samtools">0.1. Index a BAM input file with Samtools<a class="headerlink" href="#01-index-a-bam-input-file-with-samtools" title="Permanent link">&para;</a></h3>
<p>This first step is the same as in Part 1, so it should feel very familiar, but this time we need to do it for all three samples.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We've technically already generated index files for the three samples through our pipeline, so we could go fish those out of the results directory. However, it's cleaner to just redo this manually, and it'll only take a minute.</p>
</div>
<h4 id="011-spin-up-the-samtools-container-interactively">0.1.1. Spin up the Samtools container interactively<a class="headerlink" href="#011-spin-up-the-samtools-container-interactively" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>docker<span class="w"> </span>run<span class="w"> </span>-it<span class="w"> </span>-v<span class="w"> </span>./data:/data<span class="w"> </span>community.wave.seqera.io/library/samtools:1.20--b5dfbd93de237464
</code></pre></div>
<h4 id="012-run-the-indexing-command-for-the-three-samples">0.1.2. Run the indexing command for the three samples<a class="headerlink" href="#012-run-the-indexing-command-for-the-three-samples" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>samtools<span class="w"> </span>index<span class="w"> </span>/data/bam/reads_mother.bam
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>samtools<span class="w"> </span>index<span class="w"> </span>/data/bam/reads_father.bam
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>samtools<span class="w"> </span>index<span class="w"> </span>/data/bam/reads_son.bam
</code></pre></div>
<p>Just like previously, this should produce the index files in the same directory as the corresponding BAM files.</p>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="go">data/bam/</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="go">├── reads_father.bam</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="go">├── reads_father.bam.bai</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="go">├── reads_mother.bam</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="go">├── reads_mother.bam.bai</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="go">├── reads_son.bam</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a><span class="go">└── reads_son.bam.bai</span>
</code></pre></div>
<p>Now that we have index files for all three samples, we can proceed to generating the GVCFs for each of them.</p>
<h4 id="013-exit-the-samtools-container">0.1.3. Exit the Samtools container<a class="headerlink" href="#013-exit-the-samtools-container" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="nb">exit</span>
</code></pre></div>
<h3 id="02-call-variants-with-gatk-haplotypecaller-in-gvcf-mode">0.2. Call variants with GATK HaplotypeCaller in GVCF mode<a class="headerlink" href="#02-call-variants-with-gatk-haplotypecaller-in-gvcf-mode" title="Permanent link">&para;</a></h3>
<p>This second step is very similar to what we did Part 1: Hello Genomics, but we are now going to run GATK in 'GVCF mode'.</p>
<h4 id="021-spin-up-the-gatk-container-interactively">0.2.1. Spin up the GATK container interactively<a class="headerlink" href="#021-spin-up-the-gatk-container-interactively" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>docker<span class="w"> </span>run<span class="w"> </span>-it<span class="w"> </span>-v<span class="w"> </span>./data:/data<span class="w"> </span>community.wave.seqera.io/library/gatk4:4.5.0.0--730ee8817e436867
</code></pre></div>
<h4 id="022-run-the-variant-calling-command-with-the-gvcf-option">0.2.2. Run the variant calling command with the GVCF option<a class="headerlink" href="#022-run-the-variant-calling-command-with-the-gvcf-option" title="Permanent link">&para;</a></h4>
<p>In order to produce a genomic VCF (GVCF), we add the <code>-ERC GVCF</code> option to the base command, which switches on the HaplotypeCaller's GVCF mode.</p>
<p>We also change the file extension for the output file from <code>.vcf</code> to <code>.g.vcf</code>.
This is technically not a requirement, but it is a strongly recommended convention.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>gatk<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="w">        </span>-R<span class="w"> </span>/data/ref/ref.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="w">        </span>-I<span class="w"> </span>/data/bam/reads_mother.bam<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="w">        </span>-O<span class="w"> </span>reads_mother.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a><span class="w">        </span>-L<span class="w"> </span>/data/ref/intervals.bed<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a><span class="w">        </span>-ERC<span class="w"> </span>GVCF
</code></pre></div>
<p>This creates the GVCF output file <code>reads_mother.g.vcf</code> in the current working directory in the container.</p>
<p>If you <code>cat</code> it to view the contents, you'll see it's much longer than the equivalent VCF we generated in Part 1. You can't even scroll up to the start of the file, and most of the lines look quite different from what we saw in the VCF in Part 1.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Output</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-6-1674">1674</a></span>
<span class="normal"><a href="#__codelineno-6-1675">1675</a></span>
<span class="normal"><a href="#__codelineno-6-1676">1676</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-6-1674" name="__codelineno-6-1674"></a><span class="go">20_10037292_10066351    14714   .       T       &lt;NON_REF&gt;       .       .       END=14718       GT:DP:GQ:MIN_DP:PL       0/0:37:99:37:0,99,1192</span>
<a id="__codelineno-6-1675" name="__codelineno-6-1675"></a><span class="go">20_10037292_10066351    14719   .       T       &lt;NON_REF&gt;       .       .       END=14719       GT:DP:GQ:MIN_DP:PL       0/0:36:82:36:0,82,1087</span>
<a id="__codelineno-6-1676" name="__codelineno-6-1676"></a><span class="go">20_10037292_10066351    14720   .       T       &lt;NON_REF&gt;       .       .       END=14737       GT:DP:GQ:MIN_DP:PL       0/0:42:99:37:0,100,1160</span>
</code></pre></div></td></tr></table></div>
<p>These represent non-variant regions where the variant caller found no evidence of variation, so it captured some statistics describing its level of confidence in the absence of variation. This makes it possible to distinguish between two very different case figures: (1) there is good quality data showing that the sample is homozygous-reference, and (2) there is not enough good data available to make a determination either way.</p>
<p>In a GVCF, there are typically lots of such non-variant lines, with a smaller number of variant records sprinkled among them. Try running <code>head -176</code> on the GVCF to load in just the first 176 lines of the file to find an actual variant call.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Output</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-7-174">174</a></span>
<span class="normal"><a href="#__codelineno-7-175">175</a></span>
<span class="normal"><a href="#__codelineno-7-176">176</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-7-174" name="__codelineno-7-174"></a><span class="go">20_10037292_10066351    3479    .       T       &lt;NON_REF&gt;       .       .       END=3479        GT:DP:GQ:MIN_DP:PL       0/0:34:36:34:0,36,906</span>
<a id="__codelineno-7-175" name="__codelineno-7-175"></a><span class="go">20_10037292_10066351    3480    .       C       CT,&lt;NON_REF&gt;    503.03  .       DP=23;ExcessHet=0.0000;MLEAC=2,0;MLEAF=1.00,0.00;RAW_MQandDP=82800,23    GT:AD:DP:GQ:PL:SB       1/1:0,18,0:18:54:517,54,0,517,54,517:0,0,7,11</span>
<a id="__codelineno-7-176" name="__codelineno-7-176"></a><span class="go">20_10037292_10066351    3481    .       T       &lt;NON_REF&gt;       .       .       END=3481        GT:DP:GQ:MIN_DP:PL       0/0:21:51:21:0,51,765</span>
</code></pre></div></td></tr></table></div>
<p>The second line shows the first variant record in the file, which corresponds to the first variant in the VCF file we looked at in Part 1.</p>
<p>Just like the original VCF was, the output GVCF file is also accompanied by an index file, called <code>reads_mother.g.vcf.idx</code>.</p>
<h4 id="023-repeat-the-process-on-the-other-two-samples">0.2.3. Repeat the process on the other two samples<a class="headerlink" href="#023-repeat-the-process-on-the-other-two-samples" title="Permanent link">&para;</a></h4>
<p>In order to test the joint genotyping step, we need GVCFs for all three samples, so let's generate those manually now.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>gatk<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="w">        </span>-R<span class="w"> </span>/data/ref/ref.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a><span class="w">        </span>-I<span class="w"> </span>/data/bam/reads_father.bam<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="w">        </span>-O<span class="w"> </span>reads_father.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a><span class="w">        </span>-L<span class="w"> </span>/data/ref/intervals.bed<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a><span class="w">        </span>-ERC<span class="w"> </span>GVCF
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>gatk<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a><span class="w">        </span>-R<span class="w"> </span>/data/ref/ref.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="w">        </span>-I<span class="w"> </span>/data/bam/reads_son.bam<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a><span class="w">        </span>-O<span class="w"> </span>reads_son.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a><span class="w">        </span>-L<span class="w"> </span>/data/ref/intervals.bed<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="w">        </span>-ERC<span class="w"> </span>GVCF
</code></pre></div>
<p>Once this completes, you should have three files ending in <code>.g.vcf</code> in your work directory (one per sample) and their respective index files ending in <code>.g.vcf.idx</code>.</p>
<h3 id="03-run-joint-genotyping">0.3. Run joint genotyping<a class="headerlink" href="#03-run-joint-genotyping" title="Permanent link">&para;</a></h3>
<p>Now that we have all the GVCFs, we can finally try out the joint genotyping approach to generating variant calls for a cohort of samples.
As a reminder, it's a two-step method that consists of combining the data from all the GVCFs into a data store, then running the joint genotyping analysis proper to generate the final VCF of joint-called variants.</p>
<h4 id="031-combine-all-the-per-sample-gvcfs">0.3.1. Combine all the per-sample GVCFs<a class="headerlink" href="#031-combine-all-the-per-sample-gvcfs" title="Permanent link">&para;</a></h4>
<p>This first step uses another GATK tool, called GenomicsDBImport, to combine the data from all the GVCFs into a GenomicsDB data store.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>gatk<span class="w"> </span>GenomicsDBImport<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="w">    </span>-V<span class="w"> </span>reads_mother.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="w">    </span>-V<span class="w"> </span>reads_father.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="w">    </span>-V<span class="w"> </span>reads_son.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a><span class="w">    </span>-L<span class="w"> </span>/data/ref/intervals.bed<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-10-6" name="__codelineno-10-6" href="#__codelineno-10-6"></a><span class="w">    </span>--genomicsdb-workspace-path<span class="w"> </span>family_trio_gdb
</code></pre></div>
<p>The output of this step is effectively a directory containing a set of further nested directories holding the combined variant data in the form of multiple different files.
You can poke around it but you'll quickly see this data store format is not intended to be read directly by humans.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>GATK includes tools that make it possible to inspect and extract variant call data from the data store as needed.</p>
</div>
<h4 id="032-run-the-joint-genotyping-analysis-proper">0.3.2. Run the joint genotyping analysis proper<a class="headerlink" href="#032-run-the-joint-genotyping-analysis-proper" title="Permanent link">&para;</a></h4>
<p>This second step uses yet another GATK tool, called GenotypeGVCFs, to recalculate variant statistics and individual genotypes in light of the data available across all samples in the cohort.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>gatk<span class="w"> </span>GenotypeGVCFs<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="w">    </span>-R<span class="w"> </span>/data/ref/ref.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="w">    </span>-V<span class="w"> </span>gendb://family_trio_gdb<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="w">    </span>-O<span class="w"> </span>family_trio.vcf
</code></pre></div>
<p>This creates the VCF output file <code>family_trio.vcf</code> in the current working directory in the container.
It's another reasonably small file so you can <code>cat</code> this file to view its contents, and scroll up to find the first few variant lines.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">family_trio.vcf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-12-40">40</a></span>
<span class="normal"><a href="#__codelineno-12-41">41</a></span>
<span class="normal"><a href="#__codelineno-12-42">42</a></span>
<span class="normal"><a href="#__codelineno-12-43">43</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-12-40" name="__codelineno-12-40"></a><span class="gp">#</span>CHROM<span class="w">  </span>POS<span class="w">     </span>ID<span class="w">      </span>REF<span class="w">     </span>ALT<span class="w">     </span>QUAL<span class="w">    </span>FILTER<span class="w">  </span>INFO<span class="w">    </span>FORMAT<span class="w">  </span>reads_father<span class="w">    </span>reads_mother<span class="w">    </span>reads_son
<a id="__codelineno-12-41" name="__codelineno-12-41"></a><span class="go">20_10037292_10066351    3480    .       C       CT      1625.89 .       AC=5;AF=0.833;AN=6;BaseQRankSum=0.220;DP=85;ExcessHet=0.0000;FS=2.476;MLEAC=5;MLEAF=0.833;MQ=60.00;MQRankSum=0.00;QD=21.68;ReadPosRankSum=-1.147e+00;SOR=0.487    GT:AD:DP:GQ:PL  0/1:15,16:31:99:367,0,375       1/1:0,18:18:54:517,54,0 1/1:0,26:26:78:756,78,0</span>
<a id="__codelineno-12-42" name="__codelineno-12-42"></a><span class="go">20_10037292_10066351    3520    .       AT      A       1678.89 .       AC=5;AF=0.833;AN=6;BaseQRankSum=1.03;DP=80;ExcessHet=0.0000;FS=2.290;MLEAC=5;MLEAF=0.833;MQ=60.00;MQRankSum=0.00;QD=22.39;ReadPosRankSum=0.701;SOR=0.730 GT:AD:DP:GQ:PL   0/1:18,13:31:99:296,0,424       1/1:0,18:18:54:623,54,0 1/1:0,26:26:78:774,78,0</span>
<a id="__codelineno-12-43" name="__codelineno-12-43"></a><span class="go">20_10037292_10066351    3529    .       T       A       154.29  .       AC=1;AF=0.167;AN=6;BaseQRankSum=-5.440e-01;DP=104;ExcessHet=0.0000;FS=1.871;MLEAC=1;MLEAF=0.167;MQ=60.00;MQRankSum=0.00;QD=7.71;ReadPosRankSum=-1.158e+00;SOR=1.034       GT:AD:DP:GQ:PL  0/0:44,0:44:99:0,112,1347       0/1:12,8:20:99:163,0,328        0/0:39,0:39:99:0,105,1194</span>
</code></pre></div></td></tr></table></div>
<p>This looks more like the original VCF we generated in Part 1, except this time we have genotype-level information for all three samples.
The last three columns in the file are the genotype blocks for the samples, listed in alphabetical order.</p>
<p>If we look at the genotypes called for our test family trio for the very first variant, we see that the father is heterozygous-variant (<code>0/1</code>), and the mother and son are both homozygous-variant (<code>1/1</code>).</p>
<p>That is ultimately the information we're looking to extract from the dataset! So let's go wrap all this into a Nextflow workflow so we can do this at scale.</p>
<h4 id="033-exit-the-gatk-container">0.3.3. Exit the GATK container<a class="headerlink" href="#033-exit-the-gatk-container" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="nb">exit</span>
</code></pre></div>
<h3 id="takeaway">Takeaway<a class="headerlink" href="#takeaway" title="Permanent link">&para;</a></h3>
<p>You know how to run the individual commands involved in joint variant calling in the terminal to verify that they will produce the information you want.</p>
<h3 id="whats-next">What's next?<a class="headerlink" href="#whats-next" title="Permanent link">&para;</a></h3>
<p>Wrap these commands into an actual pipeline.</p>
<hr />
<h2 id="1-modify-the-per-sample-variant-calling-step-to-produce-a-gvcf">1. Modify the per-sample variant calling step to produce a GVCF<a class="headerlink" href="#1-modify-the-per-sample-variant-calling-step-to-produce-a-gvcf" title="Permanent link">&para;</a></h2>
<p>The good news is that we don't need to start all over, since we already wrote a workflow that does some of this work in Part 1.
However, that pipeline produces VCF files, whereas now we want GVCF files in order to do the joint genotyping.
So we need to start by switching on the GVCF variant calling mode and updating the output file extension.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For convenience, we are going to work with a fresh copy of the GATK workflow as it stands at the end of Part 1, but under a different name: <code>genomics-2.nf</code>.</p>
</div>
<h3 id="11-tell-haplotypecaller-to-emit-a-gvcf-and-update-the-output-extension">1.1. Tell HaplotypeCaller to emit a GVCF and update the output extension<a class="headerlink" href="#11-tell-haplotypecaller-to-emit-a-gvcf-and-update-the-output-extension" title="Permanent link">&para;</a></h3>
<p>Let's open the <code>genomics-2.nf</code> file in the code editor.
It should look very familiar, but feel free to run it if you want to satisfy yourself that it runs as expected.</p>
<p>We're going to start by making two changes:</p>
<ul>
<li>Add the <code>-ERC GVCF</code> parameter to the GATK HaplotypeCaller command;</li>
<li>Update the output file path to use the corresponding <code>.g.vcf</code> extension, as per GATK convention.</li>
</ul>
<p>Make sure you add a backslash (<code>\</code>) at the end of the previous line when you add <code>-ERC GVCF</code>.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-14-56">56</a></span>
<span class="normal"><a href="#__codelineno-14-57">57</a></span>
<span class="normal"><a href="#__codelineno-14-58">58</a></span>
<span class="normal"><a href="#__codelineno-14-59">59</a></span>
<span class="normal"><a href="#__codelineno-14-60">60</a></span>
<span class="normal"><a href="#__codelineno-14-61">61</a></span>
<span class="normal"><a href="#__codelineno-14-62">62</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-14-56" name="__codelineno-14-56"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-14-57" name="__codelineno-14-57"></a><span class="s2">    gatk HaplotypeCaller \</span>
<a id="__codelineno-14-58" name="__codelineno-14-58"></a><span class="s2">        -R ${ref_fasta} \</span>
<a id="__codelineno-14-59" name="__codelineno-14-59"></a><span class="hll"><span class="s2">        -I ${input_bam} \</span>
</span><a id="__codelineno-14-60" name="__codelineno-14-60"></a><span class="s2">        -O ${input_bam}.vcf \</span>
<a id="__codelineno-14-61" name="__codelineno-14-61"></a><span class="s2">        -L ${interval_list}</span>
<a id="__codelineno-14-62" name="__codelineno-14-62"></a><span class="s2">    &quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-15-56">56</a></span>
<span class="normal"><a href="#__codelineno-15-57">57</a></span>
<span class="normal"><a href="#__codelineno-15-58">58</a></span>
<span class="normal"><a href="#__codelineno-15-59">59</a></span>
<span class="normal"><a href="#__codelineno-15-60">60</a></span>
<span class="normal"><a href="#__codelineno-15-61">61</a></span>
<span class="normal"><a href="#__codelineno-15-62">62</a></span>
<span class="normal"><a href="#__codelineno-15-63">63</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-15-56" name="__codelineno-15-56"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-15-57" name="__codelineno-15-57"></a><span class="s2">    gatk HaplotypeCaller \</span>
<a id="__codelineno-15-58" name="__codelineno-15-58"></a><span class="s2">        -R ${ref_fasta} \</span>
<a id="__codelineno-15-59" name="__codelineno-15-59"></a><span class="hll"><span class="s2">        -I ${input_bam} \</span>
</span><a id="__codelineno-15-60" name="__codelineno-15-60"></a><span class="s2">        -O ${input_bam}.g.vcf \</span>
<a id="__codelineno-15-61" name="__codelineno-15-61"></a><span class="hll"><span class="s2">        -L ${interval_list} \</span>
</span><a id="__codelineno-15-62" name="__codelineno-15-62"></a><span class="s2">        -ERC GVCF</span>
<a id="__codelineno-15-63" name="__codelineno-15-63"></a><span class="s2">    &quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p>And that's all it takes to switch HaplotypeCaller to generating GVCFs instead of VCFs, right?</p>
<h3 id="12-run-the-pipeline-to-verify-that-you-can-generate-gvcfs">1.2. Run the pipeline to verify that you can generate GVCFs<a class="headerlink" href="#12-run-the-pipeline-to-verify-that-you-can-generate-gvcfs" title="Permanent link">&para;</a></h3>
<p>The Nextflow execution command is the same as before, save for the workflow filename itself.
Make sure to update that appropriately.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-2.nf
</code></pre></div>
<p>And the output is... all red! Oh no.</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>
<a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="go"> ┃ Launching `genomics-2.nf` [nice_gates] DSL2 - revision: 43c7de9890</span>
<a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a>
<a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a><span class="go">executor &gt;  local (6)</span>
<a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a><span class="go">[a1/0b5d00] SAMTOOLS_INDEX (3)       | 3 of 3 ✔</span>
<a id="__codelineno-17-7" name="__codelineno-17-7" href="#__codelineno-17-7"></a><span class="go">[89/4d0c70] GATK_HAPLOTYPECALLER (3) | 0 of 3</span>
<a id="__codelineno-17-8" name="__codelineno-17-8" href="#__codelineno-17-8"></a><span class="go">ERROR ~ Error executing process &gt; &#39;GATK_HAPLOTYPECALLER (2)&#39;</span>
<a id="__codelineno-17-9" name="__codelineno-17-9" href="#__codelineno-17-9"></a>
<a id="__codelineno-17-10" name="__codelineno-17-10" href="#__codelineno-17-10"></a><span class="go">Caused by:</span>
<a id="__codelineno-17-11" name="__codelineno-17-11" href="#__codelineno-17-11"></a><span class="go">  Missing output file(s) `reads_mother.bam.vcf` expected by process `GATK_HAPLOTYPECALLER (2)`</span>
<a id="__codelineno-17-12" name="__codelineno-17-12" href="#__codelineno-17-12"></a>
<a id="__codelineno-17-13" name="__codelineno-17-13" href="#__codelineno-17-13"></a><span class="go">Command executed:</span>
<a id="__codelineno-17-14" name="__codelineno-17-14" href="#__codelineno-17-14"></a>
<a id="__codelineno-17-15" name="__codelineno-17-15" href="#__codelineno-17-15"></a><span class="go">  gatk HaplotypeCaller         -R ref.fasta         -I reads_mother.bam         -O reads_mother.bam.g.vcf         -L intervals.bed         -ERC GVCF</span>
</code></pre></div>
<p>The command that was executed is correct, so we were right that that was enough to change the GATK tool's behavior.
But look at that line about the missing output file. Notice anything?</p>
<p>That's right, we forgot to tell Nextflow to expect a new file name. Oops.</p>
<h3 id="12-update-the-output-file-extension-in-the-process-outputs-block-too">1.2. Update the output file extension in the process outputs block too<a class="headerlink" href="#12-update-the-output-file-extension-in-the-process-outputs-block-too" title="Permanent link">&para;</a></h3>
<p>Because it's not enough to just change the file extension in the tool command itself, you also have to tell Nextflow that the expected output filename has changed.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-18-52">52</a></span>
<span class="normal"><a href="#__codelineno-18-53">53</a></span>
<span class="normal"><a href="#__codelineno-18-54">54</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-18-52" name="__codelineno-18-52"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-18-53" name="__codelineno-18-53"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.vcf&quot;</span><span class="w">     </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">vcf</span>
<a id="__codelineno-18-54" name="__codelineno-18-54"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.vcf.idx&quot;</span><span class="w"> </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">idx</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-19-52">52</a></span>
<span class="normal"><a href="#__codelineno-19-53">53</a></span>
<span class="normal"><a href="#__codelineno-19-54">54</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-19-52" name="__codelineno-19-52"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-19-53" name="__codelineno-19-53"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.g.vcf&quot;</span><span class="w">     </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">vcf</span>
<a id="__codelineno-19-54" name="__codelineno-19-54"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.g.vcf.idx&quot;</span><span class="w"> </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">idx</span>
</code></pre></div></td></tr></table></div>
<h3 id="13-run-the-pipeline-again">1.3. Run the pipeline again<a class="headerlink" href="#13-run-the-pipeline-again" title="Permanent link">&para;</a></h3>
<p>Let's run it with <code>-resume</code> this time.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-2.nf<span class="w"> </span>-resume
</code></pre></div>
<p>Ah, this time it works.</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a>
<a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a><span class="go"> ┃ Launching `genomics-2.nf` [elated_carlsson] DSL2 - revision: 6a5786a6fa</span>
<a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a>
<a id="__codelineno-21-5" name="__codelineno-21-5" href="#__codelineno-21-5"></a><span class="go">executor &gt;  local (3)</span>
<a id="__codelineno-21-6" name="__codelineno-21-6" href="#__codelineno-21-6"></a><span class="go">[47/f7fac1] SAMTOOLS_INDEX (2)       | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-21-7" name="__codelineno-21-7" href="#__codelineno-21-7"></a><span class="go">[ce/096ac6] GATK_HAPLOTYPECALLER (1) | 3 of 3 ✔</span>
</code></pre></div>
<p>The Nextflow output itself doesn't look any different (compared to a successful run in normal VCF mode), but now we can find the <code>.g.vcf</code> files and their respective index files, for all three samples, in the <code>results_genomics</code> directory.</p>
<div class="highlight"><span class="filename">Directory contents (symlinks truncated)</span><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a><span class="go">results_genomics/</span>
<a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a><span class="go">├── reads_father.bam -&gt; */47/f7fac1*/reads_father.bam</span>
<a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a><span class="go">├── reads_father.bam.bai -&gt; */47/f7fac1*/reads_father.bam.bai</span>
<a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a><span class="go">├── reads_father.bam.g.vcf -&gt; */cb/ad7430*/reads_father.bam.g.vcf</span>
<a id="__codelineno-22-5" name="__codelineno-22-5" href="#__codelineno-22-5"></a><span class="go">├── reads_father.bam.g.vcf.idx -&gt; */cb/ad7430*/reads_father.bam.g.vcf.idx</span>
<a id="__codelineno-22-6" name="__codelineno-22-6" href="#__codelineno-22-6"></a><span class="go">├── reads_mother.bam -&gt; */a2/56a3a8*/reads_mother.bam</span>
<a id="__codelineno-22-7" name="__codelineno-22-7" href="#__codelineno-22-7"></a><span class="go">├── reads_mother.bam.bai -&gt; */a2/56a3a8*/reads_mother.bam.bai</span>
<a id="__codelineno-22-8" name="__codelineno-22-8" href="#__codelineno-22-8"></a><span class="go">├── reads_mother.bam.g.vcf -&gt; */ce/096ac6*/reads_mother.bam.g.vcf</span>
<a id="__codelineno-22-9" name="__codelineno-22-9" href="#__codelineno-22-9"></a><span class="go">├── reads_mother.bam.g.vcf.idx -&gt; */ce/096ac6*/reads_mother.bam.g.vcf.idx</span>
<a id="__codelineno-22-10" name="__codelineno-22-10" href="#__codelineno-22-10"></a><span class="go">├── reads_son.bam -&gt; */a1/0b5d00*/reads_son.bam</span>
<a id="__codelineno-22-11" name="__codelineno-22-11" href="#__codelineno-22-11"></a><span class="go">├── reads_son.bam.bai -&gt; */a1/0b5d00*/reads_son.bam.bai</span>
<a id="__codelineno-22-12" name="__codelineno-22-12" href="#__codelineno-22-12"></a><span class="go">├── reads_son.bam.g.vcf -&gt; */c2/6b6563*/reads_son.bam.g.vcf</span>
<a id="__codelineno-22-13" name="__codelineno-22-13" href="#__codelineno-22-13"></a><span class="go">└── reads_son.bam.g.vcf.idx -&gt; */c2/6b6563*/reads_son.bam.g.vcf.idx</span>
</code></pre></div>
<p>If you open one of the GVCF files and scroll through it, you can verify that GATK HaplotypeCaller produced GVCF files as requested.</p>
<h3 id="takeaway_1">Takeaway<a class="headerlink" href="#takeaway_1" title="Permanent link">&para;</a></h3>
<p>Okay, this one was minimal in terms of Nextflow learning...
But it was a nice opportunity to reiterate the importance of the process output block!</p>
<h3 id="whats-next_1">What's next?<a class="headerlink" href="#whats-next_1" title="Permanent link">&para;</a></h3>
<p>Learn to collect the contents of a channel and pass them on to the next process as a single input.</p>
<hr />
<h2 id="2-collect-and-combine-the-gvcf-data-across-all-samples">2. Collect and combine the GVCF data across all samples<a class="headerlink" href="#2-collect-and-combine-the-gvcf-data-across-all-samples" title="Permanent link">&para;</a></h2>
<p>We now need to combine the data from all the per-sample GVCFs into a form that supports the joint genotyping analysis we want to do.</p>
<h3 id="21-define-the-process-that-will-combine-the-gvcfs">2.1. Define the process that will combine the GVCFs<a class="headerlink" href="#21-define-the-process-that-will-combine-the-gvcfs" title="Permanent link">&para;</a></h3>
<p>As a reminder of what we did earlier in the warmup section, combining the GVCFs is a job for the GATK tool GenomicsDBImport, which will produce a data store in the so-called GenomicsDB format.</p>
<p>Let's write a new process to define how that's going to work, based on the command we used earlier in the warmup section.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-23-66">66</a></span>
<span class="normal"><a href="#__codelineno-23-67">67</a></span>
<span class="normal"><a href="#__codelineno-23-68">68</a></span>
<span class="normal"><a href="#__codelineno-23-69">69</a></span>
<span class="normal"><a href="#__codelineno-23-70">70</a></span>
<span class="normal"><a href="#__codelineno-23-71">71</a></span>
<span class="normal"><a href="#__codelineno-23-72">72</a></span>
<span class="normal"><a href="#__codelineno-23-73">73</a></span>
<span class="normal"><a href="#__codelineno-23-74">74</a></span>
<span class="normal"><a href="#__codelineno-23-75">75</a></span>
<span class="normal"><a href="#__codelineno-23-76">76</a></span>
<span class="normal"><a href="#__codelineno-23-77">77</a></span>
<span class="normal"><a href="#__codelineno-23-78">78</a></span>
<span class="normal"><a href="#__codelineno-23-79">79</a></span>
<span class="normal"><a href="#__codelineno-23-80">80</a></span>
<span class="normal"><a href="#__codelineno-23-81">81</a></span>
<span class="normal"><a href="#__codelineno-23-82">82</a></span>
<span class="normal"><a href="#__codelineno-23-83">83</a></span>
<span class="normal"><a href="#__codelineno-23-84">84</a></span>
<span class="normal"><a href="#__codelineno-23-85">85</a></span>
<span class="normal"><a href="#__codelineno-23-86">86</a></span>
<span class="normal"><a href="#__codelineno-23-87">87</a></span>
<span class="normal"><a href="#__codelineno-23-88">88</a></span>
<span class="normal"><a href="#__codelineno-23-89">89</a></span>
<span class="normal"><a href="#__codelineno-23-90">90</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-23-66" name="__codelineno-23-66"></a><span class="cm">/*</span>
<a id="__codelineno-23-67" name="__codelineno-23-67"></a><span class="cm"> * Combine GVCFs into GenomicsDB datastore</span>
<a id="__codelineno-23-68" name="__codelineno-23-68"></a><span class="cm"> */</span>
<a id="__codelineno-23-69" name="__codelineno-23-69"></a><span class="n">process</span><span class="w"> </span><span class="n">GATK_GENOMICSDB</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-23-70" name="__codelineno-23-70"></a>
<a id="__codelineno-23-71" name="__codelineno-23-71"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s2">&quot;community.wave.seqera.io/library/gatk4:4.5.0.0--730ee8817e436867&quot;</span>
<a id="__codelineno-23-72" name="__codelineno-23-72"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">outdir</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;symlink&#39;</span>
<a id="__codelineno-23-73" name="__codelineno-23-73"></a>
<a id="__codelineno-23-74" name="__codelineno-23-74"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-23-75" name="__codelineno-23-75"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">all_gvcfs</span>
<a id="__codelineno-23-76" name="__codelineno-23-76"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">all_idxs</span>
<a id="__codelineno-23-77" name="__codelineno-23-77"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">interval_list</span>
<a id="__codelineno-23-78" name="__codelineno-23-78"></a><span class="w">        </span><span class="n">val</span><span class="w"> </span><span class="n">cohort_name</span>
<a id="__codelineno-23-79" name="__codelineno-23-79"></a>
<a id="__codelineno-23-80" name="__codelineno-23-80"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-23-81" name="__codelineno-23-81"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${cohort_name}_gdb&quot;</span>
<a id="__codelineno-23-82" name="__codelineno-23-82"></a>
<a id="__codelineno-23-83" name="__codelineno-23-83"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-23-84" name="__codelineno-23-84"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-23-85" name="__codelineno-23-85"></a><span class="s2">    gatk GenomicsDBImport \</span>
<a id="__codelineno-23-86" name="__codelineno-23-86"></a><span class="s2">        -V ${all_gvcfs} \</span>
<a id="__codelineno-23-87" name="__codelineno-23-87"></a><span class="s2">        -L ${interval_list} \</span>
<a id="__codelineno-23-88" name="__codelineno-23-88"></a><span class="s2">        --genomicsdb-workspace-path ${cohort_name}_gdb</span>
<a id="__codelineno-23-89" name="__codelineno-23-89"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-23-90" name="__codelineno-23-90"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<p>What do you think, looks reasonable?</p>
<p>Let's wire it up and see what happens.</p>
<h3 id="22-add-a-cohort_name-parameter-with-a-default-value">2.2. Add a <code>cohort_name</code> parameter with a default value<a class="headerlink" href="#22-add-a-cohort_name-parameter-with-a-default-value" title="Permanent link">&para;</a></h3>
<p>We need to provide an arbitrary name for the cohort.
Later in the training series you'll learn how to use sample metadata for this sort of thing, but for now we just declare a CLI parameter using <code>params</code> and give it a default value for convenience.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-24-16">16</a></span>
<span class="normal"><a href="#__codelineno-24-17">17</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-24-16" name="__codelineno-24-16"></a><span class="c1">// Base name for final output file</span>
<a id="__codelineno-24-17" name="__codelineno-24-17"></a><span class="n">params</span><span class="o">.</span><span class="na">cohort_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;family_trio&quot;</span>
</code></pre></div></td></tr></table></div>
<h3 id="23-gather-the-outputs-of-gatk_haplotypecaller-across-samples">2.3. Gather the outputs of GATK_HAPLOTYPECALLER across samples<a class="headerlink" href="#23-gather-the-outputs-of-gatk_haplotypecaller-across-samples" title="Permanent link">&para;</a></h3>
<p>If we were to just plug the output channel from the <code>GATK_HAPLOTYPECALLER</code> process as is, Nextflow would call the process on each sample GVCF separately.
However, we want to bundle all three GVCFs (and their index files) in such a way that Nextflow hands all of them together to a single process call.</p>
<p>Good news: we can do that using the <code>collect()</code> channel operator. Let's add the following lines to the <code>workflow</code> body, right after the call to GATK_HAPLOTYPECALLER:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-25-118">118</a></span>
<span class="normal"><a href="#__codelineno-25-119">119</a></span>
<span class="normal"><a href="#__codelineno-25-120">120</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-25-118" name="__codelineno-25-118"></a><span class="c1">// Collect variant calling outputs across samples</span>
<a id="__codelineno-25-119" name="__codelineno-25-119"></a><span class="n">all_gvcfs_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GATK_HAPLOTYPECALLER</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">vcf</span><span class="o">.</span><span class="na">collect</span><span class="o">()</span>
<a id="__codelineno-25-120" name="__codelineno-25-120"></a><span class="n">all_idxs_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">GATK_HAPLOTYPECALLER</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">idx</span><span class="o">.</span><span class="na">collect</span><span class="o">()</span>
</code></pre></div></td></tr></table></div>
<p>Does that seem a bit complicated? Let's break this down and translate it into plain language.</p>
<ol>
<li>We're taking the output channel from the <code>GATK_HAPLOTYPECALLER</code> process, referred to using the <code>.out</code> property.</li>
<li>Each 'element' coming out of the channel is a pair of files: the GVCF and its index file, in that order because that's the order they're listed in the process output block. Conveniently, because in the last session we named the outputs of this process (using <code>emit:</code>), we can pick out the GVCFs on one hand by adding <code>.vcf</code> and the index files on the other by adding <code>.idx</code> after the <code>.out</code> property. If we had not named those outputs, we would have had to refer to them by <code>.out[0]</code> and <code>.out[1]</code>, respectively.</li>
<li>We append the <code>collect()</code> channel operator to bundle all the GVCF files together into a single element in a new channel called <code>all_gvcfs_ch</code>, and do the same with the index files to form the new channel called <code>all_idxs_ch</code>.</li>
</ol>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you're having a hard time envisioning exactly what is happening here, remember that you can use the <code>view()</code> operator to inspect the contents of channels before and after applying channel operators.</p>
</div>
<p>The resulting <code>all_gvcfs_ch</code> and <code>all_idxs_ch</code> channels are what we're going to plug into the <code>GATK_GENOMICSDB</code> process we just wrote.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case you were wondering, we collect the GVCFs and their index files separately because the GATK GenomicsDBImport command only wants to see the GVCF file paths. Fortunately, since Nextflow will stage all the files together for execution, we don't have to worry about the order of files like we did for BAMs and their index in Part 1.</p>
</div>
<h3 id="24-add-a-call-to-the-workflow-block-to-run-gatk_genomicsdb">2.4. Add a call to the workflow block to run GATK_GENOMICSDB<a class="headerlink" href="#24-add-a-call-to-the-workflow-block-to-run-gatk_genomicsdb" title="Permanent link">&para;</a></h3>
<p>We've got a process, and we've got input channels. We just need to add the process call.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-26-122">122</a></span>
<span class="normal"><a href="#__codelineno-26-123">123</a></span>
<span class="normal"><a href="#__codelineno-26-124">124</a></span>
<span class="normal"><a href="#__codelineno-26-125">125</a></span>
<span class="normal"><a href="#__codelineno-26-126">126</a></span>
<span class="normal"><a href="#__codelineno-26-127">127</a></span>
<span class="normal"><a href="#__codelineno-26-128">128</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-26-122" name="__codelineno-26-122"></a><span class="c1">// Combine GVCFs into a GenomicsDB datastore</span>
<a id="__codelineno-26-123" name="__codelineno-26-123"></a><span class="n">GATK_GENOMICSDB</span><span class="o">(</span>
<a id="__codelineno-26-124" name="__codelineno-26-124"></a><span class="w">    </span><span class="n">all_gvcfs_ch</span><span class="o">,</span>
<a id="__codelineno-26-125" name="__codelineno-26-125"></a><span class="w">    </span><span class="n">all_idxs_ch</span><span class="o">,</span>
<a id="__codelineno-26-126" name="__codelineno-26-126"></a><span class="w">    </span><span class="n">intervals_file</span><span class="o">,</span>
<a id="__codelineno-26-127" name="__codelineno-26-127"></a><span class="w">    </span><span class="n">params</span><span class="o">.</span><span class="na">cohort_name</span>
<a id="__codelineno-26-128" name="__codelineno-26-128"></a><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p>Ok, everything is wired up.</p>
<h3 id="25-run-the-workflow">2.5. Run the workflow<a class="headerlink" href="#25-run-the-workflow" title="Permanent link">&para;</a></h3>
<p>Let's see if this works.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-2.nf<span class="w"> </span>-resume
</code></pre></div>
<p>It run fairly quickly, since we're running with <code>-resume</code>, but...</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a>
<a id="__codelineno-28-3" name="__codelineno-28-3" href="#__codelineno-28-3"></a><span class="go"> ┃ Launching `genomics-2.nf` [mad_edison] DSL2 - revision: 6aea0cfded</span>
<a id="__codelineno-28-4" name="__codelineno-28-4" href="#__codelineno-28-4"></a>
<a id="__codelineno-28-5" name="__codelineno-28-5" href="#__codelineno-28-5"></a><span class="go">executor &gt;  local (1)</span>
<a id="__codelineno-28-6" name="__codelineno-28-6" href="#__codelineno-28-6"></a><span class="go">[a2/56a3a8] SAMTOOLS_INDEX (1)       | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-28-7" name="__codelineno-28-7" href="#__codelineno-28-7"></a><span class="go">[ce/096ac6] GATK_HAPLOTYPECALLER (1) | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-28-8" name="__codelineno-28-8" href="#__codelineno-28-8"></a><span class="go">[df/994a06] GATK_GENOMICSDB          | 0 of 1</span>
<a id="__codelineno-28-9" name="__codelineno-28-9" href="#__codelineno-28-9"></a><span class="go">ERROR ~ Error executing process &gt; &#39;GATK_GENOMICSDB&#39;</span>
<a id="__codelineno-28-10" name="__codelineno-28-10" href="#__codelineno-28-10"></a>
<a id="__codelineno-28-11" name="__codelineno-28-11" href="#__codelineno-28-11"></a><span class="go">Caused by:</span>
<a id="__codelineno-28-12" name="__codelineno-28-12" href="#__codelineno-28-12"></a><span class="go">  Process `GATK_GENOMICSDB` terminated with an error exit status (1)</span>
<a id="__codelineno-28-13" name="__codelineno-28-13" href="#__codelineno-28-13"></a>
<a id="__codelineno-28-14" name="__codelineno-28-14" href="#__codelineno-28-14"></a><span class="go">Command executed:</span>
<a id="__codelineno-28-15" name="__codelineno-28-15" href="#__codelineno-28-15"></a>
<a id="__codelineno-28-16" name="__codelineno-28-16" href="#__codelineno-28-16"></a><span class="go">  gatk GenomicsDBImport         -V reads_father.bam.g.vcf reads_son.bam.g.vcf reads_mother.bam.g.vcf         -L intervals.bed         --genomicsdb-workspace-path family_trio_gdb</span>
</code></pre></div>
<p>Ah. On the bright side, we see that Nextflow has picked up the <code>GATK_GENOMICSDB</code> process, and specifically called it just once.
That suggests that the <code>collect()</code> approach worked, to a point.
But, and it's a big one, the process call failed.</p>
<p>When we dig into the console output above, we can see the command executed isn't correct.</p>
<p>Can you spot the error?
Look at this bit: <code>-V reads_father.bam.g.vcf reads_son.bam.g.vcf reads_mother.bam.g.vcf</code></p>
<p>We gave <code>gatk GenomicsDBImport</code> multiple GVCF files for a single <code>-V</code> argument, but the tool expects a separate <code>-V</code> argument for each GVCF file.</p>
<p>As a reminder, this was the command we ran in the container:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>gatk<span class="w"> </span>GenomicsDBImport<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-2" name="__codelineno-29-2" href="#__codelineno-29-2"></a><span class="w">    </span>-V<span class="w"> </span>reads_mother.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-3" name="__codelineno-29-3" href="#__codelineno-29-3"></a><span class="w">    </span>-V<span class="w"> </span>reads_father.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-4" name="__codelineno-29-4" href="#__codelineno-29-4"></a><span class="w">    </span>-V<span class="w"> </span>reads_son.g.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-5" name="__codelineno-29-5" href="#__codelineno-29-5"></a><span class="w">    </span>-L<span class="w"> </span>/data/ref/intervals.bed<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-6" name="__codelineno-29-6" href="#__codelineno-29-6"></a><span class="w">    </span>--genomicsdb-workspace-path<span class="w"> </span>family_trio_gdb
</code></pre></div>
<p>So that means we need to somehow transform our bundle of GVCF files into a properly formatted command string.</p>
<h3 id="26-construct-a-command-line-with-a-separate-v-argument-for-each-input-gvcf">2.6. Construct a command line with a separate <code>-V</code> argument for each input GVCF<a class="headerlink" href="#26-construct-a-command-line-with-a-separate-v-argument-for-each-input-gvcf" title="Permanent link">&para;</a></h3>
<p>This is where Nextflow being based on Groovy comes in handy, because it's going to allow us to use some fairly straightforward string manipulations to construct the necessary command string.</p>
<p>Specifically, using this syntax: <code>all_gvcfs.collect { gvcf -&gt; "-V ${gvcf}" }.join(' ')</code></p>
<p>Once again, let's break it down into its components.</p>
<ol>
<li>First, we take the contents of the <code>all_gvcfs</code> input channel and apply <code>.collect()</code> on it (just like earlier).</li>
<li>That allows us to pass each individual GVCF file path in the bundle to the <strong>closure</strong>, <code>{ gvcf -&gt; "-V ${gvcf}" }</code>, where <code>gvcf</code> refers to that GVCF file path.
   The closure is a mini-function that we use to prepend <code>-V</code> to the file path, in the form of <code>"-V ${gvcf}"</code>.</li>
<li>Then we use <code>.join(' ')</code> to concatenate all three strings with a single space as separator.</li>
</ol>
<p>With a concrete example, it looks like this:</p>
<ol>
<li>We have three files:</li>
</ol>
<p><code>[A.ext, B.ext, C.ext]</code></p>
<ol>
<li>The closure modifies each one to create the strings:</li>
</ol>
<p><code>"-V A.ext", "-V B.ext", "-V C.ext"</code></p>
<ol>
<li>The <code>.join(' ')</code> operation generates the final string:</li>
</ol>
<p><code>"-V A.ext -V B.ext -V C.ext"</code></p>
<p>Once we have that string, we can assign it to a local variable, <code>gvcfs_line</code>, defined with the <code>def</code> keyword:</p>
<p><code>def gvcfs_line = all_gvcfs.collect { gvcf -&gt; "-V ${gvcf}" }.join(' ')</code></p>
<p>Ok, so we have our string manipulation thingy. Where do we put it?</p>
<p>We want this to go inside the process definition somewhere, because we want to do it <em>after</em> we've channeled the GVCF file paths into the process.
That is because Nextflow must see them as file paths in order to stage the files themselves correctly for execution.</p>
<p>But <em>where</em> in the process can we add this?</p>
<p>Fun fact: you can add arbitrary code after <code>script:</code> and before the <code>"""</code> !</p>
<p>Great, let's add our string manipulation line there then, and update the <code>gatk GenomicsDBImport</code> command to use the concatenated string it produces.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-30-87">87</a></span>
<span class="normal"><a href="#__codelineno-30-88">88</a></span>
<span class="normal"><a href="#__codelineno-30-89">89</a></span>
<span class="normal"><a href="#__codelineno-30-90">90</a></span>
<span class="normal"><a href="#__codelineno-30-91">91</a></span>
<span class="normal"><a href="#__codelineno-30-92">92</a></span>
<span class="normal"><a href="#__codelineno-30-93">93</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-30-87" name="__codelineno-30-87"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-30-88" name="__codelineno-30-88"></a><span class="hll"><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
</span><a id="__codelineno-30-89" name="__codelineno-30-89"></a><span class="s2">    gatk GenomicsDBImport \</span>
<a id="__codelineno-30-90" name="__codelineno-30-90"></a><span class="s2">        -V ${all_gvcfs} \</span>
<a id="__codelineno-30-91" name="__codelineno-30-91"></a><span class="s2">        -L ${interval_list} \</span>
<a id="__codelineno-30-92" name="__codelineno-30-92"></a><span class="s2">        --genomicsdb-workspace-path ${cohort_name}_gdb</span>
<a id="__codelineno-30-93" name="__codelineno-30-93"></a><span class="s2">    &quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-31-87">87</a></span>
<span class="normal"><a href="#__codelineno-31-88">88</a></span>
<span class="normal"><a href="#__codelineno-31-89">89</a></span>
<span class="normal"><a href="#__codelineno-31-90">90</a></span>
<span class="normal"><a href="#__codelineno-31-91">91</a></span>
<span class="normal"><a href="#__codelineno-31-92">92</a></span>
<span class="normal"><a href="#__codelineno-31-93">93</a></span>
<span class="normal"><a href="#__codelineno-31-94">94</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-31-87" name="__codelineno-31-87"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-31-88" name="__codelineno-31-88"></a><span class="hll"><span class="w">    </span><span class="kt">def</span><span class="w"> </span><span class="n">gvcfs_line</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">all_gvcfs</span><span class="o">.</span><span class="na">collect</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">gvcf</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="s2">&quot;-V ${gvcf}&quot;</span><span class="w"> </span><span class="o">}.</span><span class="na">join</span><span class="o">(</span><span class="s1">&#39; &#39;</span><span class="o">)</span>
</span><a id="__codelineno-31-89" name="__codelineno-31-89"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-31-90" name="__codelineno-31-90"></a><span class="s2">    gatk GenomicsDBImport \</span>
<a id="__codelineno-31-91" name="__codelineno-31-91"></a><span class="s2">        ${gvcfs_line} \</span>
<a id="__codelineno-31-92" name="__codelineno-31-92"></a><span class="s2">        -L ${interval_list} \</span>
<a id="__codelineno-31-93" name="__codelineno-31-93"></a><span class="s2">        --genomicsdb-workspace-path ${cohort_name}_gdb</span>
<a id="__codelineno-31-94" name="__codelineno-31-94"></a><span class="s2">    &quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p>That should be all that's needed to provide the inputs to <code>gatk GenomicsDBImport</code> correctly.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>When you update the <code>gatk GenomicsDBImport</code> command, make sure to remove the <code>-V</code> prefix when you swap in the <code>${gvcfs_line}</code> variable.</p>
</div>
<h3 id="27-run-the-workflow-to-verify-that-it-generates-the-genomicsdb-output-as-expected">2.7. Run the workflow to verify that it generates the GenomicsDB output as expected<a class="headerlink" href="#27-run-the-workflow-to-verify-that-it-generates-the-genomicsdb-output-as-expected" title="Permanent link">&para;</a></h3>
<p>Alright, let's see if that addressed the issue.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-2.nf<span class="w"> </span>-resume
</code></pre></div>
<p>Aha! It seems to be working now.</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-33-2" name="__codelineno-33-2" href="#__codelineno-33-2"></a>
<a id="__codelineno-33-3" name="__codelineno-33-3" href="#__codelineno-33-3"></a><span class="go"> ┃ Launching `genomics-2.nf` [special_noyce] DSL2 - revision: 11f7a51bbe</span>
<a id="__codelineno-33-4" name="__codelineno-33-4" href="#__codelineno-33-4"></a>
<a id="__codelineno-33-5" name="__codelineno-33-5" href="#__codelineno-33-5"></a><span class="go">executor &gt;  local (1)</span>
<a id="__codelineno-33-6" name="__codelineno-33-6" href="#__codelineno-33-6"></a><span class="go">[6a/5dcf6a] SAMTOOLS_INDEX (3)       | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-33-7" name="__codelineno-33-7" href="#__codelineno-33-7"></a><span class="go">[d6/d0d060] GATK_HAPLOTYPECALLER (3) | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-33-8" name="__codelineno-33-8" href="#__codelineno-33-8"></a><span class="go">[e8/749a05] GATK_GENOMICSDB          | 1 of 1 ✔</span>
</code></pre></div>
<p>The first two steps were successfully skipped, and the third step worked like a charm this time. We find our output data store in the results directory.</p>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a><span class="go">results_genomics/family_trio_gdb</span>
<a id="__codelineno-34-2" name="__codelineno-34-2" href="#__codelineno-34-2"></a><span class="go">├── 20_10037292_10066351$12912$14737</span>
<a id="__codelineno-34-3" name="__codelineno-34-3" href="#__codelineno-34-3"></a><span class="go">├── 20_10037292_10066351$3277$5495</span>
<a id="__codelineno-34-4" name="__codelineno-34-4" href="#__codelineno-34-4"></a><span class="go">├── 20_10037292_10066351$7536$9859</span>
<a id="__codelineno-34-5" name="__codelineno-34-5" href="#__codelineno-34-5"></a><span class="go">├── callset.json</span>
<a id="__codelineno-34-6" name="__codelineno-34-6" href="#__codelineno-34-6"></a><span class="go">├── __tiledb_workspace.tdb</span>
<a id="__codelineno-34-7" name="__codelineno-34-7" href="#__codelineno-34-7"></a><span class="go">├── vcfheader.vcf</span>
<a id="__codelineno-34-8" name="__codelineno-34-8" href="#__codelineno-34-8"></a><span class="go">└── vidmap.json</span>
</code></pre></div>
<p>By the way, we didn't have to do anything special to handle the output being a directory instead of a single file.
Isn't that nice?</p>
<h3 id="takeaway_2">Takeaway<a class="headerlink" href="#takeaway_2" title="Permanent link">&para;</a></h3>
<p>Now you know how to collect outputs from a channel and bundle them as a single input to another process.
You also know how to construct a command line to provide inputs to a given tool with the appropriate syntax.</p>
<h3 id="whats-next_2">What's next?<a class="headerlink" href="#whats-next_2" title="Permanent link">&para;</a></h3>
<p>Learn how to add a second command to the same process.</p>
<hr />
<h2 id="3-run-the-joint-genotyping-step-as-part-of-the-same-process">3. Run the joint genotyping step as part of the same process<a class="headerlink" href="#3-run-the-joint-genotyping-step-as-part-of-the-same-process" title="Permanent link">&para;</a></h2>
<p>Now that we have the combined genomic variant calls, we can run the joint genotyping tool, which will produce the final output that we actually care about: the VCF of cohort-level variant calls.</p>
<p>For logistical reasons, we decide to include the joint genotyping inside the same process.</p>
<h3 id="31-rename-the-process-from-gatk_genomicsdb-to-gatk_jointgenotyping">3.1. Rename the process from GATK_GENOMICSDB to GATK_JOINTGENOTYPING<a class="headerlink" href="#31-rename-the-process-from-gatk_genomicsdb-to-gatk_jointgenotyping" title="Permanent link">&para;</a></h3>
<p>Since the process will be running more than one tool, we change its name to refer to the overall operation rather than a single tool name.</p>
<p><em>Before:</em></p>
<div class="highlight"><span class="filename">genomics-2.nf</span><pre><span></span><code><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a><span class="cm">/*</span>
<a id="__codelineno-35-2" name="__codelineno-35-2" href="#__codelineno-35-2"></a><span class="cm"> * Combine GVCFs into GenomicsDB datastore</span>
<a id="__codelineno-35-3" name="__codelineno-35-3" href="#__codelineno-35-3"></a><span class="cm"> */</span>
<a id="__codelineno-35-4" name="__codelineno-35-4" href="#__codelineno-35-4"></a><span class="n">process</span><span class="w"> </span><span class="n">GATK_GENOMICSDB</span><span class="w"> </span><span class="o">{</span>
</code></pre></div>
<p><em>After:</em></p>
<div class="highlight"><span class="filename">genomics-2.nf</span><pre><span></span><code><a id="__codelineno-36-1" name="__codelineno-36-1" href="#__codelineno-36-1"></a><span class="cm">/*</span>
<a id="__codelineno-36-2" name="__codelineno-36-2" href="#__codelineno-36-2"></a><span class="cm"> * Combine GVCFs into GenomicsDB datastore and run joint genotyping to produce cohort-level calls</span>
<a id="__codelineno-36-3" name="__codelineno-36-3" href="#__codelineno-36-3"></a><span class="cm"> */</span>
<a id="__codelineno-36-4" name="__codelineno-36-4" href="#__codelineno-36-4"></a><span class="n">process</span><span class="w"> </span><span class="n">GATK_JOINTGENOTYPING</span><span class="w"> </span><span class="o">{</span>
</code></pre></div>
<p>Remember to keep your process names as descriptive as possible, to maximize readability for your colleagues —and your future self!</p>
<h3 id="32-add-the-joint-genotyping-command-to-the-gatk_jointgenotyping-process">3.2. Add the joint genotyping command to the GATK_JOINTGENOTYPING process<a class="headerlink" href="#32-add-the-joint-genotyping-command-to-the-gatk_jointgenotyping-process" title="Permanent link">&para;</a></h3>
<p>Simply add the second command after the first one inside the script section.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-37-89">89</a></span>
<span class="normal"><a href="#__codelineno-37-90">90</a></span>
<span class="normal"><a href="#__codelineno-37-91">91</a></span>
<span class="normal"><a href="#__codelineno-37-92">92</a></span>
<span class="normal"><a href="#__codelineno-37-93">93</a></span>
<span class="normal"><a href="#__codelineno-37-94">94</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-37-89" name="__codelineno-37-89"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-37-90" name="__codelineno-37-90"></a><span class="s2">    gatk GenomicsDBImport \</span>
<a id="__codelineno-37-91" name="__codelineno-37-91"></a><span class="s2">        ${gvcfs_line} \</span>
<a id="__codelineno-37-92" name="__codelineno-37-92"></a><span class="s2">        -L ${interval_list} \</span>
<a id="__codelineno-37-93" name="__codelineno-37-93"></a><span class="s2">        --genomicsdb-workspace-path ${cohort_name}_gdb</span>
<a id="__codelineno-37-94" name="__codelineno-37-94"></a><span class="s2">    &quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-38-89"> 89</a></span>
<span class="normal"><a href="#__codelineno-38-90"> 90</a></span>
<span class="normal"><a href="#__codelineno-38-91"> 91</a></span>
<span class="normal"><a href="#__codelineno-38-92"> 92</a></span>
<span class="normal"><a href="#__codelineno-38-93"> 93</a></span>
<span class="normal"><a href="#__codelineno-38-94"> 94</a></span>
<span class="normal"><a href="#__codelineno-38-95"> 95</a></span>
<span class="normal"><a href="#__codelineno-38-96"> 96</a></span>
<span class="normal"><a href="#__codelineno-38-97"> 97</a></span>
<span class="normal"><a href="#__codelineno-38-98"> 98</a></span>
<span class="normal"><a href="#__codelineno-38-99"> 99</a></span>
<span class="normal"><a href="#__codelineno-38-100">100</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-38-89" name="__codelineno-38-89"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-38-90" name="__codelineno-38-90"></a><span class="s2">    gatk GenomicsDBImport \</span>
<a id="__codelineno-38-91" name="__codelineno-38-91"></a><span class="s2">        ${gvcfs_line} \</span>
<a id="__codelineno-38-92" name="__codelineno-38-92"></a><span class="s2">        -L ${interval_list} \</span>
<a id="__codelineno-38-93" name="__codelineno-38-93"></a><span class="s2">        --genomicsdb-workspace-path ${cohort_name}_gdb</span>
<a id="__codelineno-38-94" name="__codelineno-38-94"></a><span class="hll">
</span><a id="__codelineno-38-95" name="__codelineno-38-95"></a><span class="hll"><span class="s2">    gatk GenotypeGVCFs \</span>
</span><a id="__codelineno-38-96" name="__codelineno-38-96"></a><span class="hll"><span class="s2">        -R ${ref_fasta} \</span>
</span><a id="__codelineno-38-97" name="__codelineno-38-97"></a><span class="hll"><span class="s2">        -V gendb://${cohort_name}_gdb \</span>
</span><a id="__codelineno-38-98" name="__codelineno-38-98"></a><span class="hll"><span class="s2">        -L ${interval_list} \</span>
</span><a id="__codelineno-38-99" name="__codelineno-38-99"></a><span class="s2">        -O ${cohort_name}.joint.vcf</span>
<a id="__codelineno-38-100" name="__codelineno-38-100"></a><span class="s2">    &quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p>The two commands will be run in serial, in the same way that they would if we were to run them manually in the terminal.</p>
<h3 id="33-add-the-reference-genome-files-to-the-gatk_jointgenotyping-process-input-definitions">3.3. Add the reference genome files to the GATK_JOINTGENOTYPING process input definitions<a class="headerlink" href="#33-add-the-reference-genome-files-to-the-gatk_jointgenotyping-process-input-definitions" title="Permanent link">&para;</a></h3>
<p>The second command requires the reference genome files, so we need to add those to the process inputs.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-39-78">78</a></span>
<span class="normal"><a href="#__codelineno-39-79">79</a></span>
<span class="normal"><a href="#__codelineno-39-80">80</a></span>
<span class="normal"><a href="#__codelineno-39-81">81</a></span>
<span class="normal"><a href="#__codelineno-39-82">82</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-39-78" name="__codelineno-39-78"></a><span class="nl">input:</span>
<a id="__codelineno-39-79" name="__codelineno-39-79"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">all_gvcfs</span>
<a id="__codelineno-39-80" name="__codelineno-39-80"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">all_idxs</span>
<a id="__codelineno-39-81" name="__codelineno-39-81"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">interval_list</span>
<a id="__codelineno-39-82" name="__codelineno-39-82"></a><span class="w">    </span><span class="n">val</span><span class="w"> </span><span class="n">cohort_name</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-40-78">78</a></span>
<span class="normal"><a href="#__codelineno-40-79">79</a></span>
<span class="normal"><a href="#__codelineno-40-80">80</a></span>
<span class="normal"><a href="#__codelineno-40-81">81</a></span>
<span class="normal"><a href="#__codelineno-40-82">82</a></span>
<span class="normal"><a href="#__codelineno-40-83">83</a></span>
<span class="normal"><a href="#__codelineno-40-84">84</a></span>
<span class="normal"><a href="#__codelineno-40-85">85</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-40-78" name="__codelineno-40-78"></a><span class="nl">input:</span>
<a id="__codelineno-40-79" name="__codelineno-40-79"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">all_gvcfs</span>
<a id="__codelineno-40-80" name="__codelineno-40-80"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">all_idxs</span>
<a id="__codelineno-40-81" name="__codelineno-40-81"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">interval_list</span>
<a id="__codelineno-40-82" name="__codelineno-40-82"></a><span class="hll"><span class="w">    </span><span class="n">val</span><span class="w"> </span><span class="n">cohort_name</span>
</span><a id="__codelineno-40-83" name="__codelineno-40-83"></a><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">ref_fasta</span>
</span><a id="__codelineno-40-84" name="__codelineno-40-84"></a><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">ref_index</span>
</span><a id="__codelineno-40-85" name="__codelineno-40-85"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">ref_dict</span>
</code></pre></div></td></tr></table></div>
<p>It may seem annoying to type these out, but remember, you only type them once, and then you can run the workflow a million times. Worth it?</p>
<h3 id="34-update-the-process-output-definition-to-emit-the-vcf-of-cohort-level-variant-calls">3.4. Update the process output definition to emit the VCF of cohort-level variant calls<a class="headerlink" href="#34-update-the-process-output-definition-to-emit-the-vcf-of-cohort-level-variant-calls" title="Permanent link">&para;</a></h3>
<p>We don't really care about saving the GenomicsDB datastore, which is just an intermediate format that only exists for logistical reasons, so we can just remove it from the output block if we want.</p>
<p>The output we're actually interested in is the VCF produced by the joint genotyping command.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-41-87">87</a></span>
<span class="normal"><a href="#__codelineno-41-88">88</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-41-87" name="__codelineno-41-87"></a><span class="nl">output:</span>
<a id="__codelineno-41-88" name="__codelineno-41-88"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${cohort_name}_gdb&quot;</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-42-87">87</a></span>
<span class="normal"><a href="#__codelineno-42-88">88</a></span>
<span class="normal"><a href="#__codelineno-42-89">89</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-42-87" name="__codelineno-42-87"></a><span class="nl">output:</span>
<a id="__codelineno-42-88" name="__codelineno-42-88"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${cohort_name}.joint.vcf&quot;</span><span class="w">     </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">vcf</span>
<a id="__codelineno-42-89" name="__codelineno-42-89"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${cohort_name}.joint.vcf.idx&quot;</span><span class="w"> </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">idx</span>
</code></pre></div></td></tr></table></div>
<p>We're almost done!</p>
<h3 id="35-update-the-process-call-from-gatk_genomicsdb-to-gatk_jointgenotyping">3.5. Update the process call from GATK_GENOMICSDB to GATK_JOINTGENOTYPING<a class="headerlink" href="#35-update-the-process-call-from-gatk_genomicsdb-to-gatk_jointgenotyping" title="Permanent link">&para;</a></h3>
<p>Let's not forget to rename the process call in the workflow body from GATK_GENOMICSDB to GATK_JOINTGENOTYPING. And while we're at it, we should also add the reference genome files as inputs, since we need to provide them to the joint genotyping tool.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-43-134">134</a></span>
<span class="normal"><a href="#__codelineno-43-135">135</a></span>
<span class="normal"><a href="#__codelineno-43-136">136</a></span>
<span class="normal"><a href="#__codelineno-43-137">137</a></span>
<span class="normal"><a href="#__codelineno-43-138">138</a></span>
<span class="normal"><a href="#__codelineno-43-139">139</a></span>
<span class="normal"><a href="#__codelineno-43-140">140</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-43-134" name="__codelineno-43-134"></a><span class="c1">// Combine GVCFs into a GenomicsDB data store</span>
<a id="__codelineno-43-135" name="__codelineno-43-135"></a><span class="n">GATK_GENOMICSDB</span><span class="o">(</span>
<a id="__codelineno-43-136" name="__codelineno-43-136"></a><span class="w">    </span><span class="n">all_gvcfs_ch</span><span class="o">,</span>
<a id="__codelineno-43-137" name="__codelineno-43-137"></a><span class="w">    </span><span class="n">all_idxs_ch</span><span class="o">,</span>
<a id="__codelineno-43-138" name="__codelineno-43-138"></a><span class="w">    </span><span class="n">intervals_file</span><span class="o">,</span>
<a id="__codelineno-43-139" name="__codelineno-43-139"></a><span class="w">    </span><span class="n">params</span><span class="o">.</span><span class="na">cohort_name</span>
<a id="__codelineno-43-140" name="__codelineno-43-140"></a><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-2.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-44-134">134</a></span>
<span class="normal"><a href="#__codelineno-44-135">135</a></span>
<span class="normal"><a href="#__codelineno-44-136">136</a></span>
<span class="normal"><a href="#__codelineno-44-137">137</a></span>
<span class="normal"><a href="#__codelineno-44-138">138</a></span>
<span class="normal"><a href="#__codelineno-44-139">139</a></span>
<span class="normal"><a href="#__codelineno-44-140">140</a></span>
<span class="normal"><a href="#__codelineno-44-141">141</a></span>
<span class="normal"><a href="#__codelineno-44-142">142</a></span>
<span class="normal"><a href="#__codelineno-44-143">143</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-44-134" name="__codelineno-44-134"></a><span class="c1">// Combine GVCFs into a GenomicsDB data store and apply joint genotyping</span>
<a id="__codelineno-44-135" name="__codelineno-44-135"></a><span class="n">GATK_JOINTGENOTYPING</span><span class="o">(</span>
<a id="__codelineno-44-136" name="__codelineno-44-136"></a><span class="w">    </span><span class="n">all_gvcfs_ch</span><span class="o">,</span>
<a id="__codelineno-44-137" name="__codelineno-44-137"></a><span class="w">    </span><span class="n">all_idxs_ch</span><span class="o">,</span>
<a id="__codelineno-44-138" name="__codelineno-44-138"></a><span class="w">    </span><span class="n">intervals_file</span><span class="o">,</span>
<a id="__codelineno-44-139" name="__codelineno-44-139"></a><span class="w">    </span><span class="n">params</span><span class="o">.</span><span class="na">cohort_name</span><span class="o">,</span>
<a id="__codelineno-44-140" name="__codelineno-44-140"></a><span class="w">    </span><span class="n">ref_file</span><span class="o">,</span>
<a id="__codelineno-44-141" name="__codelineno-44-141"></a><span class="w">    </span><span class="n">ref_index_file</span><span class="o">,</span>
<a id="__codelineno-44-142" name="__codelineno-44-142"></a><span class="w">    </span><span class="n">ref_dict_file</span>
<a id="__codelineno-44-143" name="__codelineno-44-143"></a><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p>Now everything should be completely wired up.</p>
<h3 id="36-run-the-workflow">3.6. Run the workflow<a class="headerlink" href="#36-run-the-workflow" title="Permanent link">&para;</a></h3>
<p>Finally, we can run the modified workflow...</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-45-1" name="__codelineno-45-1" href="#__codelineno-45-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-2.nf<span class="w"> </span>-resume
</code></pre></div>
<p>And it works!</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-46-1" name="__codelineno-46-1" href="#__codelineno-46-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-46-2" name="__codelineno-46-2" href="#__codelineno-46-2"></a>
<a id="__codelineno-46-3" name="__codelineno-46-3" href="#__codelineno-46-3"></a><span class="go"> ┃ Launching `genomics-2.nf` [modest_gilbert] DSL2 - revision: 4f49922223</span>
<a id="__codelineno-46-4" name="__codelineno-46-4" href="#__codelineno-46-4"></a>
<a id="__codelineno-46-5" name="__codelineno-46-5" href="#__codelineno-46-5"></a><span class="go">executor &gt;  local (1)</span>
<a id="__codelineno-46-6" name="__codelineno-46-6" href="#__codelineno-46-6"></a><span class="go">[6a/5dcf6a] SAMTOOLS_INDEX (3)       | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-46-7" name="__codelineno-46-7" href="#__codelineno-46-7"></a><span class="go">[fe/6c9ad4] GATK_HAPLOTYPECALLER (1) | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-46-8" name="__codelineno-46-8" href="#__codelineno-46-8"></a><span class="go">[e2/a8d95f] GATK_JOINTGENOTYPING     | 1 of 1 ✔</span>
</code></pre></div>
<p>You'll find the final output file, <code>family_trio.joint.vcf</code> (and its file index), in the results directory. If you're the skeptical type, you can click on it to open it and verify that the workflow has generated the same variant calls that you obtained by running the tools manually at the start of this section.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">family_trio.joint.vcf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-47-40">40</a></span>
<span class="normal"><a href="#__codelineno-47-41">41</a></span>
<span class="normal"><a href="#__codelineno-47-42">42</a></span>
<span class="normal"><a href="#__codelineno-47-43">43</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-47-40" name="__codelineno-47-40"></a><span class="gp">#</span>CHROM<span class="w">	</span>POS<span class="w">	</span>ID<span class="w">	</span>REF<span class="w">	</span>ALT<span class="w">	</span>QUAL<span class="w">	</span>FILTER<span class="w">	</span>INFO<span class="w">	</span>FORMAT<span class="w">	</span>reads_father<span class="w">	</span>reads_mother<span class="w">	</span>reads_son
<a id="__codelineno-47-41" name="__codelineno-47-41"></a><span class="go">20_10037292_10066351	3480	.	C	CT	1625.89	.	AC=5;AF=0.833;AN=6;BaseQRankSum=0.220;DP=85;ExcessHet=0.0000;FS=2.476;MLEAC=5;MLEAF=0.833;MQ=60.00;MQRankSum=0.00;QD=21.68;ReadPosRankSum=-1.147e+00;SOR=0.487	GT:AD:DP:GQ:PL	0/1:15,16:31:99:367,0,375	1/1:0,18:18:54:517,54,0	1/1:0,26:26:78:756,78,0</span>
<a id="__codelineno-47-42" name="__codelineno-47-42"></a><span class="go">20_10037292_10066351	3520	.	AT	A	1678.89	.	AC=5;AF=0.833;AN=6;BaseQRankSum=1.03;DP=80;ExcessHet=0.0000;FS=2.290;MLEAC=5;MLEAF=0.833;MQ=60.00;MQRankSum=0.00;QD=22.39;ReadPosRankSum=0.701;SOR=0.730	GT:AD:DP:GQ:PL	0/1:18,13:31:99:296,0,424	1/1:0,18:18:54:623,54,0	1/1:0,26:26:78:774,78,0</span>
<a id="__codelineno-47-43" name="__codelineno-47-43"></a><span class="go">20_10037292_10066351	3529	.	T	A	154.29	.	AC=1;AF=0.167;AN=6;BaseQRankSum=-5.440e-01;DP=104;ExcessHet=0.0000;FS=1.871;MLEAC=1;MLEAF=0.167;MQ=60.00;MQRankSum=0.00;QD=7.71;ReadPosRankSum=-1.158e+00;SOR=1.034	GT:AD:DP:GQ:PL	0/0:44,0:44:99:0,112,1347	0/1:12,8:20:99:163,0,328	0/0:39,0:39:99:0,105,1194</span>
</code></pre></div></td></tr></table></div>
<p>You now have an automated, fully reproducible joint variant calling workflow!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Keep in mind the data files we gave you cover only a tiny portion of chromosome 20.
The real size of a variant callset would be counted in millions of variants.
That's why we use only tiny subsets of data for training purposes!</p>
</div>
<h3 id="takeaway_3">Takeaway<a class="headerlink" href="#takeaway_3" title="Permanent link">&para;</a></h3>
<p>You know how to use some common operators as well as Groovy closures to control the flow of data in your workflow.</p>
<h3 id="whats-next_3">What's next?<a class="headerlink" href="#whats-next_3" title="Permanent link">&para;</a></h3>
<p>Celebrate your success and take an extra super mega long break! This was tough and you deserve it.</p>
<p>When you're ready to move on, have a look at our training portal to browse available training courses and select your next step.</p>
<p><strong>Good luck!</strong></p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Volver al principio
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Pie" >
        
          
          <a href="../01_per_sample_variant_calling/" class="md-footer__link md-footer__link--prev" aria-label="Anterior: Part 1: Per-sample variant calling">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Anterior
              </span>
              <div class="md-ellipsis">
                Part 1: Per-sample variant calling
              </div>
            </div>
          </a>
        
        
          
          <a href="../03_modules/" class="md-footer__link md-footer__link--next" aria-label="Siguiente: Part 3: Moving code into modules">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Siguiente
              </span>
              <div class="md-ellipsis">
                Part 3: Moving code into modules
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  <div>
    Copyright &copy; 2020-2025,
    <a href="https://seqera.io" target="_blank" rel="noopener">Seqera</a>.
  </div>
  <div class="footer_cc_link">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512">
      <!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.-->
      <path
        d="m245.83 214.87-33.22 17.28c-9.43-19.58-25.24-19.93-27.46-19.93-22.13 0-33.22 14.61-33.22 43.84 0 23.57 9.21 43.84 33.22 43.84 14.47 0 24.65-7.09 30.57-21.26l30.55 15.5c-6.17 11.51-25.69 38.98-65.1 38.98-22.6 0-73.96-10.32-73.96-77.05 0-58.69 43-77.06 72.63-77.06 30.72-.01 52.7 11.95 65.99 35.86zm143.05 0-32.78 17.28c-9.5-19.77-25.72-19.93-27.9-19.93-22.14 0-33.22 14.61-33.22 43.84 0 23.55 9.23 43.84 33.22 43.84 14.45 0 24.65-7.09 30.54-21.26l31 15.5c-2.1 3.75-21.39 38.98-65.09 38.98-22.69 0-73.96-9.87-73.96-77.05 0-58.67 42.97-77.06 72.63-77.06 30.71-.01 52.58 11.95 65.56 35.86zM247.56 8.05C104.74 8.05 0 123.11 0 256.05c0 138.49 113.6 248 247.56 248 129.93 0 248.44-100.87 248.44-248 0-137.87-106.62-248-248.44-248zm.87 450.81c-112.54 0-203.7-93.04-203.7-202.81 0-105.42 85.43-203.27 203.72-203.27 112.53 0 202.82 89.46 202.82 203.26-.01 121.69-99.68 202.82-202.84 202.82z"
      ></path>
    </svg>
    <a
      href="https://creativecommons.org/licenses/by-nc-nd/4.0/"
      target="_blank"
      rel="noopener"
      title="Creative Commons CC BY-NC-ND 4.0"
    >
      Creative Commons BY-NC-ND 4.0</a
    >.
  </div>
   Made with
  <a
    href="https://squidfunk.github.io/mkdocs-material/"
    target="_blank"
    rel="noopener"
  >
    Material for MkDocs
  </a>
  
</div>
      
        <div class="md-social">
  <a
    href="https://seqera.io"
    target="_blank"
    rel="noopener"
    class="footer_seqera_logo"
  >
    <img src="/assets/img/seqera_logo_white.png" />
  </a>
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
      <div class="md-consent" data-md-component="consent" id="__consent" hidden>
        <div class="md-consent__overlay"></div>
        <aside class="md-consent__inner">
          <form class="md-consent__form md-grid md-typeset" name="consent">
            


  


<h4>Cookie consent</h4>
<p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our training materials better.</p>
<input class="md-toggle" type="checkbox" id="__settings" >
<div class="md-consent__settings">
  <ul class="task-list">
    
    
      
        
  
  
    
    
  
  <li class="task-list-item">
    <label class="task-list-control">
      <input type="checkbox" name="analytics" checked>
      <span class="task-list-indicator"></span>
      Google Analytics
    </label>
  </li>

      
    
    
      
        
  
  
    
    
  
  <li class="task-list-item">
    <label class="task-list-control">
      <input type="checkbox" name="github" checked>
      <span class="task-list-indicator"></span>
      GitHub
    </label>
  </li>

      
    
    
  </ul>
</div>
<div class="md-consent__controls">
  
    
      <button class="md-button md-button--primary">Aceptar</button>
    
    
    
  
    
    
    
      <label class="md-button" for="__settings">Gestionar cookies</label>
    
  
</div>
          </form>
        </aside>
      </div>
      <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout((function(){document.querySelector("[data-md-component=consent]").hidden=!1}),250);var form=document.forms.consent;for(var action of["submit","reset"])form.addEventListener(action,(function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map((function(e){return[e,!0]})))),location.hash="",location.reload()}))</script>
    
    
      
      <script id="__config" type="application/json">{"base": "../../../..", "features": ["content.action.edit", "content.code.annotate", "content.code.copy", "navigation.footer", "navigation.instant", "navigation.top", "navigation.tracking", "search.share", "toc.follow"], "search": "../../../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copiado al portapapeles", "clipboard.copy": "Copiar al portapapeles", "search.result.more.one": "1 m\u00e1s en esta p\u00e1gina", "search.result.more.other": "# m\u00e1s en esta p\u00e1gina", "search.result.none": "No se encontraron documentos", "search.result.one": "1 documento encontrado", "search.result.other": "# documentos encontrados", "search.result.placeholder": "Teclee para comenzar b\u00fasqueda", "search.result.term.missing": "Falta", "select.version": "Seleccionar versi\u00f3n"}, "version": {"alias": true, "provider": "mike"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>