
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://training.nextflow.io/0.dev/other/hands_on/04_implementation/">
      
      
      
      
      <link rel="icon" href="../../../assets/img/nextflow-icon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Pipeline Implementation - training.nextflow.io</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/fonts.css">
    
      <link rel="stylesheet" href="../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-244N3GEN75"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-244N3GEN75",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-244N3GEN75",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>if("undefined"!=typeof __md_analytics){var consent=__md_get("__consent");consent&&consent.analytics&&__md_analytics()}</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#pipeline-implementation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
            
<div style="text-align: center">
  Nextflow Training Week (2025-Q1):
  <strong>March 10-14, 2025</strong>
  <a
    class="md-button"
    href="https://community.seqera.io/t/nextflow-training-week-2025-q1/1775/6"
    target="_blank"
  >
    More info
    <span class="twemoji"
      ><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path
          d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"
        ></path></svg
    ></span>
  </a>
</div>
<style type="text/css">
  .md-banner {
    display: none;
  }
</style>

<script>
  (function () {
    function hideAnnouncement() {
      var announcementDate = new Date(
        "2025-03-15T00:00:00Z",
      );
      var currentDate = new Date();
      var banner = document.querySelector(".md-banner");
      if (banner && currentDate <= announcementDate) {
        banner.style.display = "block";
      }
    }

    // Run on initial load
    hideAnnouncement();

    // Hook into XHR to run after each page update
    var originalOpen = XMLHttpRequest.prototype.open;
    XMLHttpRequest.prototype.open = function () {
      this.addEventListener("load", function () {
        setTimeout(hideAnnouncement, 0); // Ensure it runs after the request completes
      });
      return originalOpen.apply(this, arguments);
    };
  })();
</script>
 
          </div>
          
        </aside>
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="training.nextflow.io" class="md-header__button md-logo" aria-label="training.nextflow.io" data-md-component="logo">
      
  <img src="../../../assets/img/nextflow_logo_dark.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            training.nextflow.io
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Pipeline Implementation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
      <div class="md-header__option">
  <div class="md-select">
    
    <button class="md-header__button md-icon" aria-label="Select language">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12.87 15.07-2.54-2.51.03-.03A17.5 17.5 0 0 0 14.07 6H17V4h-7V2H8v2H1v2h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2zm-2.62 7 1.62-4.33L19.12 17z"/></svg>
    </button>
    <div class="md-select__inner">
      <ul class="md-select__list">
        
          <li class="md-select__item">
            <a href="./" hreflang="en" class="md-select__link">
              English
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../pt/other/hands_on/04_implementation/" hreflang="pt" class="md-select__link">
              Português
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../es/other/hands_on/04_implementation/" hreflang="es" class="md-select__link">
              Español
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../fr/other/hands_on/04_implementation/" hreflang="fr" class="md-select__link">
              Français
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../it/other/hands_on/04_implementation/" hreflang="it" class="md-select__link">
              Italiano
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../ko/other/hands_on/04_implementation/" hreflang="ko" class="md-select__link">
              Korean
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</div>
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/nextflow-io/training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    nextflow-io/training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="training.nextflow.io" class="md-nav__button md-logo" aria-label="training.nextflow.io" data-md-component="logo">
      
  <img src="../../../assets/img/nextflow_logo_dark.png" alt="logo">

    </a>
    training.nextflow.io
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/nextflow-io/training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    nextflow-io/training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Environment setup
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Environment setup
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Environment Setup
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/01_setup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    GitHub Codespaces
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/02_local/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/03_devcontainer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local installation using VSCode Devcontainers extension
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Hello Nextflow
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Hello Nextflow
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hello Nextflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/01_hello_world/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Hello World
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/02_hello_channels/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Hello Channels
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/03_hello_workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Hello Workflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/04_hello_modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 4: Hello Modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/05_hello_containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 5: Hello Containers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/06_hello_config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 6: Hello Config
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Hello nf-core
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Hello nf-core
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hello nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/01_run_demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Run a demo pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/02_rewrite_hello/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Rewrite Hello for nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Nextflow for Genomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Nextflow for Genomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow for Genomics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/01_per_sample_variant_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Per-sample variant calling
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/02_joint_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Joint calling on a cohort
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/03_modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Moving code into modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/04_testing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 4: Adding tests
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/genomics/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Nextflow for RNAseq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Nextflow for RNAseq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow for RNAseq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/01_method/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Method overview and manual testing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/02_single-sample/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Single-sample implementation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/03_multi-sample/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Multi-sample paired-end implementation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../nf4_science/rnaseq/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Side Quests
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Side Quests
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Side Quests
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/workflows_of_workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflows of Workflows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/splitting_and_grouping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Splitting and Grouping
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/nf-test/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Testing with nf-test
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/nf-core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Fundamentals Training
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Fundamentals Training
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Fundamentals Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basic concepts
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/rnaseq_pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Simple RNA-Seq workflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Dependencies and containers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/channels/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Channels
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/processes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Processes
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/operators/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Operators
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/groovy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Groovy introduction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Modularization
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/executors/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Deployment scenarios
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/seqera_platform/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Seqera Platform
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/cache_and_resume/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cache and resume
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/debugging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troubleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Advanced Training
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Advanced Training
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Advanced Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/operators/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Operator Tour
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Metadata Propagation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/grouping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Grouping and Splitting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/groovy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Groovy Imports
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/structure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow Structure
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/configuration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/summary/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Summary
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/support/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Support
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../help/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Help
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                      

<nav class="md-nav md-nav--secondary" aria-label="On this page">
       
  <label class="md-nav__title" for="__toc">
    <span class="md-nav__icon md-icon"></span>
    On this page
  </label>
  <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
     <li class="md-nav__item">
  <a href="#data-preparation" class="md-nav__link">
    <span class="md-ellipsis">
      Data preparation
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#input-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Input parameters
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-1a-create-a-fasta-genome-index" class="md-nav__link">
    <span class="md-ellipsis">
      Process 1A: Create a FASTA genome index
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-1b-create-a-fasta-genome-sequence-dictionary-with-picard-for-gatk" class="md-nav__link">
    <span class="md-ellipsis">
      Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-1c-create-star-genome-index-file" class="md-nav__link">
    <span class="md-ellipsis">
      Process 1C: Create STAR genome index file
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-1d-filtered-and-recoded-set-of-variants" class="md-nav__link">
    <span class="md-ellipsis">
      Process 1D: Filtered and recoded set of variants
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-2-star-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Process 2: STAR Mapping
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-3-gatk-split-on-n" class="md-nav__link">
    <span class="md-ellipsis">
      Process 3: GATK Split on N
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-4-gatk-recalibrate" class="md-nav__link">
    <span class="md-ellipsis">
      Process 4: GATK Recalibrate
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-5-gatk-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Process 5: GATK Variant Calling
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#processes-6a-and-6b-ase-rna-editing" class="md-nav__link">
    <span class="md-ellipsis">
      Processes 6A and 6B: ASE &amp; RNA Editing
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#process-7-allele-specific-expression-analysis-with-gatk-asereadcounter" class="md-nav__link">
    <span class="md-ellipsis">
      Process 7: Allele-Specific Expression analysis with GATK ASEReadCounter
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#results-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Results overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Results overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variant-calls" class="md-nav__link">
    <span class="md-ellipsis">
      Variant calls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#allele-specific-expression-quantification" class="md-nav__link">
    <span class="md-ellipsis">
      Allele specific expression quantification
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#allele-frequency-histogram" class="md-nav__link">
    <span class="md-ellipsis">
      Allele frequency histogram
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#bonus-step" class="md-nav__link">
    <span class="md-ellipsis">
      Bonus step
    </span>
  </a>
  
</li> 
  </ul>
  
</nav>
<nav class="md-nav md-nav--secondary sidebar_resources" aria-label="Resources">
  <label class="md-nav__title">Resources</label>
  <ul class="md-nav__list">
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/docs/latest/index.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M17.5 14.33c.79 0 1.63.08 2.5.24v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.33-4.5.99v-1.69c1.17-.53 2.67-.8 4.5-.8M13 12.46c1.29-.53 2.79-.79 4.5-.79.79 0 1.63.07 2.5.23v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.34-4.5.99m4.5-3.65c-1.9 0-3.39.32-4.5 1V9.84c1.23-.56 2.73-.84 4.5-.84.79 0 1.63.08 2.5.23v1.55c-.74-.19-1.59-.28-2.5-.28m3.5 8V7c-1.04-.33-2.21-.5-3.5-.5-2.05 0-3.88.5-5.5 1.5v11.5c1.62-1 3.45-1.5 5.5-1.5 1.19 0 2.36.16 3.5.5m-3.5-14c2.35 0 4.19.5 5.5 1.5v14.56c0 .12-.05.24-.16.35-.11.09-.23.17-.34.17-.11 0-.19-.02-.25-.05-1.28-.69-2.87-1.03-4.75-1.03-2.05 0-3.88.5-5.5 1.5-1.34-1-3.17-1.5-5.5-1.5-1.66 0-3.25.36-4.75 1.07-.03.01-.07.01-.12.03-.04.01-.08.02-.13.02-.11 0-.23-.04-.34-.12a.475.475 0 0 1-.16-.35V6c1.34-1 3.18-1.5 5.5-1.5 2.33 0 4.16.5 5.5 1.5 1.34-1 3.17-1.5 5.5-1.5Z"
          ></path>
        </svg>
        Nextflow Docs
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://www.nextflow.io/slack-invite.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M6 15a2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2h2v2m1 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-5m2-8a2 2 0 0 1-2-2 2 2 0 0 1 2-2 2 2 0 0 1 2 2v2H9m0 1a2 2 0 0 1 2 2 2 2 0 0 1-2 2H4a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5m8 2a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2h-2v-2m-1 0a2 2 0 0 1-2 2 2 2 0 0 1-2-2V5a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5m-2 8a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2v-2h2m0-1a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-5Z"
          ></path>
        </svg>
        Nextflow Slack
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="80 50 350 350"
        >
          <style type="text/css">
            .st0 {
              fill: "gray";
            }
          </style>
          <path
            class="st0"
            d="M124.5,171.5c39.4,2.6,71.1,34.3,73.5,73.8h46.1c-2.6-65-54.7-117.2-119.6-119.9V171.5z"
          />
          <path
            class="st0"
            d="M198.3,255.8c-2.6,39.5-34.3,71.1-73.8,73.5v46.1c65-2.6,117.2-54.7,119.9-119.6H198.3z"
          />
          <path
            class="st0"
            d="M300.7,245c2.6-39.5,34.3-71.1,73.8-73.5v-46.1c-65,2.6-117.2,54.7-119.9,119.6H300.7z"
          />
          <path
            class="st0"
            d="M374.5,329.3c-39.4-2.6-71.1-34.3-73.5-73.8h-46.1c2.6,65,54.7,117.2,119.6,119.9V329.3z"
          />
        </svg>
        Nextflow Homepage
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://seqera.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <?xml version="1.0" encoding="utf-8"?>
        <!-- Generator: Adobe Illustrator 28.3.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          x="0px"
          y="0px"
          viewBox="80 50 350 350"
          style="enable-background: new 0 0 500 500"
          xml:space="preserve"
        >
          <style type="text/css">
            .st0 {
              clip-path: url(#SVGID_00000006699037935518914080000000086571810102047141_);
            }
            .st1 {
              fill: "gray";
            }
          </style>
          <g>
            <defs>
              <rect id="SVGID_1_" x="125" y="125" width="250" height="250" />
            </defs>
            <clipPath
              id="SVGID_00000183214694142566821210000012389151436317298101_"
            >
              <use xlink:href="#SVGID_1_" style="overflow: visible" />
            </clipPath>
            <g
              style="
                clip-path: url(#SVGID_00000183214694142566821210000012389151436317298101_);
              "
            >
              <path
                class="st1"
                d="M375,333l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.2-0.9l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7L365,308l-0.6-0.6l-0.6-0.6l-0.6-0.6l-0.6-0.6L362,305
                      l-0.6-0.6l-0.7-0.6l-0.6-0.5l-1.2,0.7l-0.8,0.5l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.9,0.4l-0.9,0.4l-0.9,0.3l-0.9,0.3
                      l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0.1
                      l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0l-0.9,0h-57.5l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.3l0.1,0.3l0.1,0.4l0.1,0.4
                      l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4l-0.1,0.3l-0.2,0.3l-0.2,0.3
                      l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.1l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0H125V375h209.7l0.9,0l0.9,0l0.9,0l0.8-0.1
                      l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3
                      l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7
                      l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8
                      l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0-0.9l0-0.9l0-0.8L375,333z"
              />
              <path
                class="st1"
                d="M375,260.4l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.1-0.8l-0.1-0.8l-0.2-0.8l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.7
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.2-0.2h0h-89.8l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.4l0.1,0.3
                      l0.1,0.4l0.1,0.4l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4
                      l-0.2,0.4l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0h-82.5h-111H125v32.2
                      h122.8h86.9h0.9l0.9,0l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3
                      l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.4l0.8-0.4l0.2-0.1l0.6-0.3l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.5-0.7l0.5-0.7l0.5-0.7
                      l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7l0.4-0.7l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8
                      l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0-0.8l0-0.9l0-0.9
                      L375,260.4z"
              />
              <path
                class="st1"
                d="M165.3,197.6l-0.9,0l-0.8,0l-0.9,0l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.2,0.1l-0.6,0.3l-0.8,0.4
                      l-0.8,0.4l-0.8,0.4l-0.7,0.4l-0.7,0.5l-0.7,0.5L142,205l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.8l0,0.9l0.1,0.8l0.1,0.8
                      l0.1,0.9l0.1,0.9l0.1,0.8l0.2,0.9l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.2,0.2h89.8l-59.7-16.4l-0.4-0.1l-0.3-0.1
                      l-0.4-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3
                      l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4
                      l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3
                      l0.3-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.4-0.2l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0
                      h0.4h82.5h111H375v-32.3H252.2H165.3z"
              />
              <path
                class="st1"
                d="M165.3,125l-0.9,0l-0.8,0l-0.9,0.1l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4
                      l-0.8,0.4l-0.7,0.5l-0.7,0.4l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.9l0,0.8l0.1,0.9l0.1,0.9
                      l0.1,0.8l0.1,0.8l0.1,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.6,0.6l0.6,0.6
                      l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.7,0.6l0.6,0.5h0l1.1-0.7l0.8-0.5l0.8-0.5l0.8-0.4l0.8-0.4l0.9-0.4l0.9-0.4l0.9-0.4
                      l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.1l0.9-0.1l0.9-0.1l0.9-0.1
                      l0.9-0.1l0.9-0.1l0.9,0l0.9,0l0.9,0h57.5l-59.7-16.4l-0.4-0.1l-0.3-0.1l-0.4-0.1l-0.3-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2
                      l-0.3-0.2l-0.3-0.2l-0.3-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.3l-0.1-0.4
                      l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3
                      l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2
                      l0.3-0.2l0.4-0.1l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0l0.4,0H375V125H165.3z"
              />
            </g>
          </g>
        </svg>
        Seqera
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://github.com/nextflow-io/training"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"
          ></path>
        </svg>
        Training on GitHub
      </a>
    </li>
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/nextflow-io/training/edit/master/docs/other/hands_on/04_implementation.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<h1 id="pipeline-implementation">Pipeline Implementation<a class="headerlink" href="#pipeline-implementation" title="Permanent link">&para;</a></h1>
<h2 id="data-preparation">Data preparation<a class="headerlink" href="#data-preparation" title="Permanent link">&para;</a></h2>
<p>A first step in any pipeline is to prepare the input data. You will find all the data required to run the pipeline in the folder <code>data</code> within the <code>/workspaces/training/hands-on</code> repository directory.</p>
<p>There are four data inputs that we will use in this tutorial:</p>
<ol>
<li><strong>Genome File</strong> (<code>data/genome.fa</code>)</li>
<li>Human chromosome 22 in FASTA file format</li>
<li><strong>Read Files</strong> (<code>data/reads/</code>)</li>
<li>Sample ENCSR000COQ1: 76bp paired-end reads (<code>ENCSR000COQ1_1.fq.gz</code> and <code>ENCSR000COQ1_2.fq.gz</code>).</li>
<li><strong>Variants File</strong> (<code>data/known_variants.vcf.gz</code>)</li>
<li>Known variants, gzipped as a Variant Calling File (VCF) format.</li>
<li><strong>Blacklist File</strong> (<code>data/blacklist.bed</code>)</li>
<li>Genomic locations which are known to produce artifacts and spurious variants in Browser Extensible Data (BED) format.</li>
</ol>
<h2 id="input-parameters">Input parameters<a class="headerlink" href="#input-parameters" title="Permanent link">&para;</a></h2>
<p>We can begin writing the pipeline by creating and editing a text file called <code>main.nf</code> from the <code>/workspaces/training/hands-on</code> repository directory with your favourite text editor. In this example we are using <code>code</code>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="nb">cd</span><span class="w"> </span>/workspaces/training/hands-on
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>code<span class="w"> </span>main.nf
</code></pre></div>
<p>Edit this file to specify the input files as script parameters. Using this notation allows you to override them by specifying different values when launching the pipeline execution.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>Click the <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 13h-4v4h-2v-4H7v-2h4V7h2v4h4m-5-9A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10A10 10 0 0 0 12 2"/></svg></span> icons in the code for explanations.</p>
</div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-1-1">1</a></span>
<span class="normal"><a href="#__codelineno-1-2">2</a></span>
<span class="normal"><a href="#__codelineno-1-3">3</a></span>
<span class="normal"><a href="#__codelineno-1-4">4</a></span>
<span class="normal"><a href="#__codelineno-1-5">5</a></span>
<span class="normal"><a href="#__codelineno-1-6">6</a></span>
<span class="normal"><a href="#__codelineno-1-7">7</a></span>
<span class="normal"><a href="#__codelineno-1-8">8</a></span>
<span class="normal"><a href="#__codelineno-1-9">9</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1"></a><span class="cm">/*</span>
<a id="__codelineno-1-2" name="__codelineno-1-2"></a><span class="cm"> * Define the default parameters (1)</span>
<a id="__codelineno-1-3" name="__codelineno-1-3"></a><span class="cm"> */</span>
<a id="__codelineno-1-4" name="__codelineno-1-4"></a>
<a id="__codelineno-1-5" name="__codelineno-1-5"></a><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="w">     </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/genome.fa&quot;</span><span class="w"> </span><span class="c1">// (2)!</span>
<a id="__codelineno-1-6" name="__codelineno-1-6"></a><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/known_variants.vcf.gz&quot;</span>
<a id="__codelineno-1-7" name="__codelineno-1-7"></a><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/blacklist.bed&quot;</span>
<a id="__codelineno-1-8" name="__codelineno-1-8"></a><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/reads/ENCSR000COQ1_{1,2}.fastq.gz&quot;</span><span class="w"> </span><span class="c1">// (3)!</span>
<a id="__codelineno-1-9" name="__codelineno-1-9"></a><span class="n">params</span><span class="o">.</span><span class="na">results</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;results&quot;</span><span class="w"> </span><span class="c1">// (4)!</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The <code>/*</code>, <code>*</code> and <code>*/</code> specify comment lines which are ignored by Nextflow.</li>
<li>The <code>projectDir</code> variable represents the main script path location.</li>
<li>The <code>reads</code> parameter uses a glob pattern to specify the forward (<code>ENCSR000COQ1_1.fq.gz</code>) and reverse (<code>ENCSR000COQ1_2.fq.gz</code>) reads (paired-end) of a sample.</li>
<li>The <code>results</code> parameter is used to specify a directory called <code>results</code>.</li>
</ol>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>You can copy the above text (<span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 21H8V7h11m0-2H8a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h11a2 2 0 0 0 2-2V7a2 2 0 0 0-2-2m-3-4H4a2 2 0 0 0-2 2v14h2V3h12z"/></svg></span> top right or <span class="keys"><kbd class="key-command">Cmd</kbd><span>+</span><kbd class="key-c">C</kbd></span>), then move in the terminal window, open <code>code</code> and paste using the keyboard <span class="keys"><kbd class="key-command">Cmd</kbd><span>+</span><kbd class="key-v">V</kbd></span> shortcut.</p>
</div>
<p>Once you have the default parameters in the <code>main.nf</code> file, you can save and run the main script for the first time.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>With <code>code</code> you can save and close the file with <span class="keys"><kbd class="key-control">Ctrl</kbd><span>+</span><kbd class="key-o">O</kbd></span>, then <span class="keys"><kbd class="key-enter">Enter</kbd></span>, followed by <span class="keys"><kbd class="key-control">Ctrl</kbd><span>+</span><kbd class="key-x">X</kbd></span>.</p>
</div>
<p>To run the main script use the following command:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>main.nf
</code></pre></div>
<p>You should see the script execute, print Nextflow version and pipeline revision and then exit.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="go">N E X T F L O W  ~  version 23.04.1</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="go">Launching `main.nf` [elated_davinci] DSL2 - revision: 5187dd3166</span>
</code></pre></div>
<div class="admonition exercise">
<p class="admonition-title">Problem #1</p>
<p>Great, now we need to define a <a href="https://www.nextflow.io/docs/latest/channel.html">channel</a> variable to handle the read-pair files. To do that open the <code>main.nf</code> file and copy the lines below at the end of the file.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>In <code>code</code> you can move to the end of the file using <span class="keys"><kbd class="key-control">Ctrl</kbd><span>+</span><kbd class="key-w">W</kbd></span> and then <span class="keys"><kbd class="key-control">Ctrl</kbd><span>+</span><kbd class="key-v">V</kbd></span>.</p>
</div>
<p>This time you must fill the <code>BLANK</code> space with a channel factory that will create a channel out of the <code>params.reads</code> information.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-4-1">1</a></span>
<span class="normal"><a href="#__codelineno-4-2">2</a></span>
<span class="normal"><a href="#__codelineno-4-3">3</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-4-2" name="__codelineno-4-2"></a><span class="hll"><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">BLANK</span>
</span><a id="__codelineno-4-3" name="__codelineno-4-3"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Use the <a href="https://www.nextflow.io/docs/latest/channel.html#fromfilepairs">fromFilePairs</a> channel factory.</p>
</div>
<p>Once you think you have data organised, you can again run the pipeline. However this time, we can use the the <code>-resume</code> flag.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>main.nf<span class="w"> </span>-resume
</code></pre></div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>See <a href="https://www.nextflow.io/docs/latest/getstarted.html?highlight=resume#modify-and-resume">here</a> for more details about using the <code>resume</code> option.</p>
</div>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-6-1">1</a></span>
<span class="normal"><a href="#__codelineno-6-2">2</a></span>
<span class="normal"><a href="#__codelineno-6-3">3</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-6-2" name="__codelineno-6-2"></a><span class="hll"><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span><span class="w"> </span><span class="c1">// (1)!</span>
</span><a id="__codelineno-6-3" name="__codelineno-6-3"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>Creates a channel using the <a href="https://www.nextflow.io/docs/latest/channel.html#fromfilepairs">fromFilePairs()</a> channel factory.</li>
</ol>
</details>
</div>
<h2 id="process-1a-create-a-fasta-genome-index">Process 1A: Create a FASTA genome index<a class="headerlink" href="#process-1a-create-a-fasta-genome-index" title="Permanent link">&para;</a></h2>
<p>Now we have our inputs set up we can move onto the processes. In our first process we will create a genome index using <a href="http://www.htslib.org/">samtools</a>.</p>
<p>The first process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>prepare_genome_samtools</code></li>
<li><strong>Command</strong>: create a genome index for the genome fasta with samtools</li>
<li><strong>Input</strong>: the genome fasta file</li>
<li><strong>Output</strong>: the samtools genome index file</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #2</p>
<p>Copy the code below and paste it at the end of <code>main.nf</code>, removing the previous workflow block. Be careful not to accidentally have multiple workflow blocks.</p>
<p>Your aim is to replace the <code>BLANK</code> placeholder with the the correct process call.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-7-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-7-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-7-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-7-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-7-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-7-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-7-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-7-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-7-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-7-10">10</a></span>
<span class="normal"><a href="#__codelineno-7-11">11</a></span>
<span class="normal"><a href="#__codelineno-7-12">12</a></span>
<span class="normal"><a href="#__codelineno-7-13">13</a></span>
<span class="normal"><a href="#__codelineno-7-14">14</a></span>
<span class="normal"><a href="#__codelineno-7-15">15</a></span>
<span class="normal"><a href="#__codelineno-7-16">16</a></span>
<span class="normal"><a href="#__codelineno-7-17">17</a></span>
<span class="normal"><a href="#__codelineno-7-18">18</a></span>
<span class="normal"><a href="#__codelineno-7-19">19</a></span>
<span class="normal"><a href="#__codelineno-7-20">20</a></span>
<span class="normal"><a href="#__codelineno-7-21">21</a></span>
<span class="normal"><a href="#__codelineno-7-22">22</a></span>
<span class="normal"><a href="#__codelineno-7-23">23</a></span>
<span class="normal"><a href="#__codelineno-7-24">24</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1"></a><span class="cm">/*</span>
<a id="__codelineno-7-2" name="__codelineno-7-2"></a><span class="cm"> * Process 1A: Create a FASTA genome index with samtools</span>
<a id="__codelineno-7-3" name="__codelineno-7-3"></a><span class="cm"> */</span>
<a id="__codelineno-7-4" name="__codelineno-7-4"></a>
<a id="__codelineno-7-5" name="__codelineno-7-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_genome_samtools</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-7-6" name="__codelineno-7-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/samtools:1.3.1--h0cf4675_11&#39;</span>
<a id="__codelineno-7-7" name="__codelineno-7-7"></a>
<a id="__codelineno-7-8" name="__codelineno-7-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-7-9" name="__codelineno-7-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-7-10" name="__codelineno-7-10"></a>
<a id="__codelineno-7-11" name="__codelineno-7-11"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-7-12" name="__codelineno-7-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${genome}.fai&quot;</span><span class="w"> </span><span class="c1">// (2)!</span>
<a id="__codelineno-7-13" name="__codelineno-7-13"></a>
<a id="__codelineno-7-14" name="__codelineno-7-14"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-7-15" name="__codelineno-7-15"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-7-16" name="__codelineno-7-16"></a><span class="s2">    samtools faidx ${genome}</span>
<a id="__codelineno-7-17" name="__codelineno-7-17"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-7-18" name="__codelineno-7-18"></a><span class="o">}</span>
<a id="__codelineno-7-19" name="__codelineno-7-19"></a>
<a id="__codelineno-7-20" name="__codelineno-7-20"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-7-21" name="__codelineno-7-21"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-7-22" name="__codelineno-7-22"></a>
<a id="__codelineno-7-23" name="__codelineno-7-23"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-7-24" name="__codelineno-7-24"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>Take as input the genome file from the <code>params.genome</code> parameter and refer to it within the process as <code>genome</code></li>
<li>The output is a path named just like the input file but adding <code>.fai</code> to the end, e.g. <code>ref_genome.fa.fai</code></li>
</ol>
<p>In plain english, the process could be written as:</p>
<ul>
<li>A <strong>process</strong> called <code>prepare_genome_samtools</code></li>
<li>takes as <strong>input</strong> the genome file</li>
<li>and creates as <strong>output</strong> a genome index file</li>
<li><strong>script</strong>: using samtools create the genome index from the genome file</li>
</ul>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-8-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-8-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-8-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-8-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-8-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-8-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-8-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-8-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-8-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-8-10">10</a></span>
<span class="normal"><a href="#__codelineno-8-11">11</a></span>
<span class="normal"><a href="#__codelineno-8-12">12</a></span>
<span class="normal"><a href="#__codelineno-8-13">13</a></span>
<span class="normal"><a href="#__codelineno-8-14">14</a></span>
<span class="normal"><a href="#__codelineno-8-15">15</a></span>
<span class="normal"><a href="#__codelineno-8-16">16</a></span>
<span class="normal"><a href="#__codelineno-8-17">17</a></span>
<span class="normal"><a href="#__codelineno-8-18">18</a></span>
<span class="normal"><a href="#__codelineno-8-19">19</a></span>
<span class="normal"><a href="#__codelineno-8-20">20</a></span>
<span class="normal"><a href="#__codelineno-8-21">21</a></span>
<span class="normal"><a href="#__codelineno-8-22">22</a></span>
<span class="normal"><a href="#__codelineno-8-23">23</a></span>
<span class="normal"><a href="#__codelineno-8-24">24</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1"></a><span class="cm">/*</span>
<a id="__codelineno-8-2" name="__codelineno-8-2"></a><span class="cm"> * Process 1A: Create a FASTA genome index with samtools</span>
<a id="__codelineno-8-3" name="__codelineno-8-3"></a><span class="cm"> */</span>
<a id="__codelineno-8-4" name="__codelineno-8-4"></a>
<a id="__codelineno-8-5" name="__codelineno-8-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_genome_samtools</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-8-6" name="__codelineno-8-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/samtools:1.3.1--h0cf4675_11&#39;</span>
<a id="__codelineno-8-7" name="__codelineno-8-7"></a>
<a id="__codelineno-8-8" name="__codelineno-8-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-8-9" name="__codelineno-8-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-8-10" name="__codelineno-8-10"></a>
<a id="__codelineno-8-11" name="__codelineno-8-11"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-8-12" name="__codelineno-8-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${genome}.fai&quot;</span>
<a id="__codelineno-8-13" name="__codelineno-8-13"></a>
<a id="__codelineno-8-14" name="__codelineno-8-14"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-8-15" name="__codelineno-8-15"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-8-16" name="__codelineno-8-16"></a><span class="s2">    samtools faidx ${genome}</span>
<a id="__codelineno-8-17" name="__codelineno-8-17"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-8-18" name="__codelineno-8-18"></a><span class="o">}</span>
<a id="__codelineno-8-19" name="__codelineno-8-19"></a>
<a id="__codelineno-8-20" name="__codelineno-8-20"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-8-21" name="__codelineno-8-21"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-8-22" name="__codelineno-8-22"></a>
<a id="__codelineno-8-23" name="__codelineno-8-23"></a><span class="hll"><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span><span class="w"> </span><span class="c1">// (1)!</span>
</span><a id="__codelineno-8-24" name="__codelineno-8-24"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The solution is to provide <strong><code>params.genome</code></strong> as input to the <code>prepare_genome_samtools</code> process.</li>
</ol>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><code>params.genome</code> is just a regular variable, not a channel, but when passed as input to a process, it's automatically converted into a value channel.</p>
</div>
</details>
<p>Now when we run the pipeline, we see that the process <code>prepare_genome_samtools</code> is submitted:</p>
<p><div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>main.nf<span class="w"> </span>-resume
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="go">N E X T F L O W  ~  version 23.04.1</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="go">Launching `main.nf` [cranky_bose] - revision: d1df5b7267</span>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="go">executor &gt;  local (1)</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="go">[cd/47f882] process &gt; prepare_genome_samtools [100%] 1 of 1 ✔</span>
</code></pre></div></p>
</div>
<h2 id="process-1b-create-a-fasta-genome-sequence-dictionary-with-picard-for-gatk">Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK<a class="headerlink" href="#process-1b-create-a-fasta-genome-sequence-dictionary-with-picard-for-gatk" title="Permanent link">&para;</a></h2>
<p>Our first process created the genome index for GATK using samtools. For the next process we must do something very similar, this time creating a genome sequence dictionary using <a href="https://broadinstitute.github.io/picard/">Picard</a>.</p>
<p>The next process should have the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>prepare_genome_picard</code></li>
<li><strong>Command</strong>: create a genome dictionary for the genome fasta with Picard tools</li>
<li><strong>Input</strong>: the genome fasta file</li>
<li><strong>Output</strong>: the genome dictionary file</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #3</p>
<p>Your aim is to replace the <code>BLANK</code> placeholder with the the correct process call.</p>
<p>Copy the code below and paste it at the end of <code>main.nf</code>, removing the previous workflow block.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>You can choose any channel output name that makes sense to you.</p>
</div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-11-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-11-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-11-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-11-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-11-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-11-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-11-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-11-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-11-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-11-10">10</a></span>
<span class="normal"><a href="#__codelineno-11-11">11</a></span>
<span class="normal"><a href="#__codelineno-11-12">12</a></span>
<span class="normal"><a href="#__codelineno-11-13">13</a></span>
<span class="normal"><a href="#__codelineno-11-14">14</a></span>
<span class="normal"><a href="#__codelineno-11-15">15</a></span>
<span class="normal"><a href="#__codelineno-11-16">16</a></span>
<span class="normal"><a href="#__codelineno-11-17">17</a></span>
<span class="normal"><a href="#__codelineno-11-18">18</a></span>
<span class="normal"><a href="#__codelineno-11-19">19</a></span>
<span class="normal"><a href="#__codelineno-11-20">20</a></span>
<span class="normal"><a href="#__codelineno-11-21">21</a></span>
<span class="normal"><a href="#__codelineno-11-22">22</a></span>
<span class="normal"><a href="#__codelineno-11-23">23</a></span>
<span class="normal"><a href="#__codelineno-11-24">24</a></span>
<span class="normal"><a href="#__codelineno-11-25">25</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1"></a><span class="cm">/*</span>
<a id="__codelineno-11-2" name="__codelineno-11-2"></a><span class="cm"> * Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK</span>
<a id="__codelineno-11-3" name="__codelineno-11-3"></a><span class="cm"> */</span>
<a id="__codelineno-11-4" name="__codelineno-11-4"></a>
<a id="__codelineno-11-5" name="__codelineno-11-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_genome_picard</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-11-6" name="__codelineno-11-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/picard:1.141--hdfd78af_6&#39;</span>
<a id="__codelineno-11-7" name="__codelineno-11-7"></a>
<a id="__codelineno-11-8" name="__codelineno-11-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-11-9" name="__codelineno-11-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-11-10" name="__codelineno-11-10"></a>
<a id="__codelineno-11-11" name="__codelineno-11-11"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-11-12" name="__codelineno-11-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${genome.baseName}.dict&quot;</span>
<a id="__codelineno-11-13" name="__codelineno-11-13"></a>
<a id="__codelineno-11-14" name="__codelineno-11-14"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-11-15" name="__codelineno-11-15"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-11-16" name="__codelineno-11-16"></a><span class="s2">    picard CreateSequenceDictionary R= ${genome} O= ${genome.baseName}.dict</span>
<a id="__codelineno-11-17" name="__codelineno-11-17"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-11-18" name="__codelineno-11-18"></a><span class="o">}</span>
<a id="__codelineno-11-19" name="__codelineno-11-19"></a>
<a id="__codelineno-11-20" name="__codelineno-11-20"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-11-21" name="__codelineno-11-21"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-11-22" name="__codelineno-11-22"></a>
<a id="__codelineno-11-23" name="__codelineno-11-23"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-11-24" name="__codelineno-11-24"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-11-25" name="__codelineno-11-25"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code>.baseName</code> returns the filename without the file suffix. If <code>"${genome}"</code> is <code>human.fa</code>, then <code>"${genome.baseName}.dict"</code> would be <code>human.dict</code>.</p>
</div>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-12-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-12-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-12-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-12-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-12-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-12-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-12-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-12-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-12-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-12-10">10</a></span>
<span class="normal"><a href="#__codelineno-12-11">11</a></span>
<span class="normal"><a href="#__codelineno-12-12">12</a></span>
<span class="normal"><a href="#__codelineno-12-13">13</a></span>
<span class="normal"><a href="#__codelineno-12-14">14</a></span>
<span class="normal"><a href="#__codelineno-12-15">15</a></span>
<span class="normal"><a href="#__codelineno-12-16">16</a></span>
<span class="normal"><a href="#__codelineno-12-17">17</a></span>
<span class="normal"><a href="#__codelineno-12-18">18</a></span>
<span class="normal"><a href="#__codelineno-12-19">19</a></span>
<span class="normal"><a href="#__codelineno-12-20">20</a></span>
<span class="normal"><a href="#__codelineno-12-21">21</a></span>
<span class="normal"><a href="#__codelineno-12-22">22</a></span>
<span class="normal"><a href="#__codelineno-12-23">23</a></span>
<span class="normal"><a href="#__codelineno-12-24">24</a></span>
<span class="normal"><a href="#__codelineno-12-25">25</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1"></a><span class="cm">/*</span>
<a id="__codelineno-12-2" name="__codelineno-12-2"></a><span class="cm"> * Process 1B: Create a FASTA genome sequence dictionary with Picard for GATK</span>
<a id="__codelineno-12-3" name="__codelineno-12-3"></a><span class="cm"> */</span>
<a id="__codelineno-12-4" name="__codelineno-12-4"></a>
<a id="__codelineno-12-5" name="__codelineno-12-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_genome_picard</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-12-6" name="__codelineno-12-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/picard:1.141--hdfd78af_6&#39;</span>
<a id="__codelineno-12-7" name="__codelineno-12-7"></a>
<a id="__codelineno-12-8" name="__codelineno-12-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-12-9" name="__codelineno-12-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-12-10" name="__codelineno-12-10"></a>
<a id="__codelineno-12-11" name="__codelineno-12-11"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-12-12" name="__codelineno-12-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${genome.baseName}.dict&quot;</span>
<a id="__codelineno-12-13" name="__codelineno-12-13"></a>
<a id="__codelineno-12-14" name="__codelineno-12-14"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-12-15" name="__codelineno-12-15"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-12-16" name="__codelineno-12-16"></a><span class="s2">    picard CreateSequenceDictionary R= ${genome} O= ${genome.baseName}.dict</span>
<a id="__codelineno-12-17" name="__codelineno-12-17"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-12-18" name="__codelineno-12-18"></a><span class="o">}</span>
<a id="__codelineno-12-19" name="__codelineno-12-19"></a>
<a id="__codelineno-12-20" name="__codelineno-12-20"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-12-21" name="__codelineno-12-21"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-12-22" name="__codelineno-12-22"></a>
<a id="__codelineno-12-23" name="__codelineno-12-23"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-12-24" name="__codelineno-12-24"></a><span class="hll"><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span><span class="w"> </span><span class="c1">// (1)!</span>
</span><a id="__codelineno-12-25" name="__codelineno-12-25"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The solution is to provide <strong><code>params.genome</code></strong> as input to the <code>prepare_genome_picard</code> process.</li>
</ol>
</details>
</div>
<h2 id="process-1c-create-star-genome-index-file">Process 1C: Create STAR genome index file<a class="headerlink" href="#process-1c-create-star-genome-index-file" title="Permanent link">&para;</a></h2>
<p>Next we must create a genome index for the <a href="https://github.com/alexdobin/STAR">STAR</a> mapping software.</p>
<p>The next process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>prepare_star_genome_index</code></li>
<li><strong>Command</strong>: create a STAR genome index for the genome fasta</li>
<li><strong>Input</strong>: the genome fasta file</li>
<li><strong>Output</strong>: a directory containing the STAR genome index</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #4</p>
<p>This is a similar exercise as problem 3.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-13-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-13-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-13-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-13-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-13-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-13-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-13-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-13-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-13-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-13-10">10</a></span>
<span class="normal"><a href="#__codelineno-13-11">11</a></span>
<span class="normal"><a href="#__codelineno-13-12">12</a></span>
<span class="normal"><a href="#__codelineno-13-13">13</a></span>
<span class="normal"><a href="#__codelineno-13-14">14</a></span>
<span class="normal"><a href="#__codelineno-13-15">15</a></span>
<span class="normal"><a href="#__codelineno-13-16">16</a></span>
<span class="normal"><a href="#__codelineno-13-17">17</a></span>
<span class="normal"><a href="#__codelineno-13-18">18</a></span>
<span class="normal"><a href="#__codelineno-13-19">19</a></span>
<span class="normal"><a href="#__codelineno-13-20">20</a></span>
<span class="normal"><a href="#__codelineno-13-21">21</a></span>
<span class="normal"><a href="#__codelineno-13-22">22</a></span>
<span class="normal"><a href="#__codelineno-13-23">23</a></span>
<span class="normal"><a href="#__codelineno-13-24">24</a></span>
<span class="normal"><a href="#__codelineno-13-25">25</a></span>
<span class="normal"><a href="#__codelineno-13-26">26</a></span>
<span class="normal"><a href="#__codelineno-13-27">27</a></span>
<span class="normal"><a href="#__codelineno-13-28">28</a></span>
<span class="normal"><a href="#__codelineno-13-29">29</a></span>
<span class="normal"><a href="#__codelineno-13-30">30</a></span>
<span class="normal"><a href="#__codelineno-13-31">31</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1"></a><span class="cm">/*</span>
<a id="__codelineno-13-2" name="__codelineno-13-2"></a><span class="cm"> * Process 1C: Create the genome index file for STAR</span>
<a id="__codelineno-13-3" name="__codelineno-13-3"></a><span class="cm"> */</span>
<a id="__codelineno-13-4" name="__codelineno-13-4"></a>
<a id="__codelineno-13-5" name="__codelineno-13-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_star_genome_index</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-13-6" name="__codelineno-13-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/star:2.7.10b--h6b7c446_1&#39;</span>
<a id="__codelineno-13-7" name="__codelineno-13-7"></a>
<a id="__codelineno-13-8" name="__codelineno-13-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-13-9" name="__codelineno-13-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-13-10" name="__codelineno-13-10"></a>
<a id="__codelineno-13-11" name="__codelineno-13-11"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-13-12" name="__codelineno-13-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s1">&#39;genome_dir&#39;</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-13-13" name="__codelineno-13-13"></a>
<a id="__codelineno-13-14" name="__codelineno-13-14"></a><span class="w">    </span><span class="nl">script:</span><span class="w"> </span><span class="c1">// (2)!</span>
<a id="__codelineno-13-15" name="__codelineno-13-15"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-13-16" name="__codelineno-13-16"></a><span class="s2">    mkdir -p genome_dir</span>
<a id="__codelineno-13-17" name="__codelineno-13-17"></a>
<a id="__codelineno-13-18" name="__codelineno-13-18"></a><span class="s2">    STAR --runMode genomeGenerate \</span>
<a id="__codelineno-13-19" name="__codelineno-13-19"></a><span class="s2">         --genomeDir genome_dir \</span>
<a id="__codelineno-13-20" name="__codelineno-13-20"></a><span class="s2">         --genomeFastaFiles ${genome} \</span>
<a id="__codelineno-13-21" name="__codelineno-13-21"></a><span class="s2">         --runThreadN ${task.cpus}</span>
<a id="__codelineno-13-22" name="__codelineno-13-22"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-13-23" name="__codelineno-13-23"></a><span class="o">}</span>
<a id="__codelineno-13-24" name="__codelineno-13-24"></a>
<a id="__codelineno-13-25" name="__codelineno-13-25"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-13-26" name="__codelineno-13-26"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-13-27" name="__codelineno-13-27"></a>
<a id="__codelineno-13-28" name="__codelineno-13-28"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-13-29" name="__codelineno-13-29"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-13-30" name="__codelineno-13-30"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-13-31" name="__codelineno-13-31"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The <code>output</code> is a <code>path</code> called <code>genome_dir</code></li>
<li>Before running <code>STAR</code>, it creates the output directory that will contain the resulting STAR genome index.</li>
</ol>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The output of the STAR genomeGenerate command is specified here as <code>genome_dir</code>.</p>
</div>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-14-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-14-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-14-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-14-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-14-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-14-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-14-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-14-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-14-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-14-10">10</a></span>
<span class="normal"><a href="#__codelineno-14-11">11</a></span>
<span class="normal"><a href="#__codelineno-14-12">12</a></span>
<span class="normal"><a href="#__codelineno-14-13">13</a></span>
<span class="normal"><a href="#__codelineno-14-14">14</a></span>
<span class="normal"><a href="#__codelineno-14-15">15</a></span>
<span class="normal"><a href="#__codelineno-14-16">16</a></span>
<span class="normal"><a href="#__codelineno-14-17">17</a></span>
<span class="normal"><a href="#__codelineno-14-18">18</a></span>
<span class="normal"><a href="#__codelineno-14-19">19</a></span>
<span class="normal"><a href="#__codelineno-14-20">20</a></span>
<span class="normal"><a href="#__codelineno-14-21">21</a></span>
<span class="normal"><a href="#__codelineno-14-22">22</a></span>
<span class="normal"><a href="#__codelineno-14-23">23</a></span>
<span class="normal"><a href="#__codelineno-14-24">24</a></span>
<span class="normal"><a href="#__codelineno-14-25">25</a></span>
<span class="normal"><a href="#__codelineno-14-26">26</a></span>
<span class="normal"><a href="#__codelineno-14-27">27</a></span>
<span class="normal"><a href="#__codelineno-14-28">28</a></span>
<span class="normal"><a href="#__codelineno-14-29">29</a></span>
<span class="normal"><a href="#__codelineno-14-30">30</a></span>
<span class="normal"><a href="#__codelineno-14-31">31</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1"></a><span class="cm">/*</span>
<a id="__codelineno-14-2" name="__codelineno-14-2"></a><span class="cm">* Process 1C: Create the genome index file for STAR</span>
<a id="__codelineno-14-3" name="__codelineno-14-3"></a><span class="cm">*/</span>
<a id="__codelineno-14-4" name="__codelineno-14-4"></a>
<a id="__codelineno-14-5" name="__codelineno-14-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_star_genome_index</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-14-6" name="__codelineno-14-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/star:2.7.10b--h6b7c446_1&#39;</span>
<a id="__codelineno-14-7" name="__codelineno-14-7"></a>
<a id="__codelineno-14-8" name="__codelineno-14-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-14-9" name="__codelineno-14-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-14-10" name="__codelineno-14-10"></a>
<a id="__codelineno-14-11" name="__codelineno-14-11"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-14-12" name="__codelineno-14-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s1">&#39;genome_dir&#39;</span>
<a id="__codelineno-14-13" name="__codelineno-14-13"></a>
<a id="__codelineno-14-14" name="__codelineno-14-14"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-14-15" name="__codelineno-14-15"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-14-16" name="__codelineno-14-16"></a><span class="s2">    mkdir -p genome_dir</span>
<a id="__codelineno-14-17" name="__codelineno-14-17"></a>
<a id="__codelineno-14-18" name="__codelineno-14-18"></a><span class="s2">    STAR --runMode genomeGenerate \</span>
<a id="__codelineno-14-19" name="__codelineno-14-19"></a><span class="s2">         --genomeDir genome_dir \</span>
<a id="__codelineno-14-20" name="__codelineno-14-20"></a><span class="s2">         --genomeFastaFiles ${genome} \</span>
<a id="__codelineno-14-21" name="__codelineno-14-21"></a><span class="s2">         --runThreadN ${task.cpus}</span>
<a id="__codelineno-14-22" name="__codelineno-14-22"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-14-23" name="__codelineno-14-23"></a><span class="o">}</span>
<a id="__codelineno-14-24" name="__codelineno-14-24"></a>
<a id="__codelineno-14-25" name="__codelineno-14-25"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-14-26" name="__codelineno-14-26"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-14-27" name="__codelineno-14-27"></a>
<a id="__codelineno-14-28" name="__codelineno-14-28"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-14-29" name="__codelineno-14-29"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-14-30" name="__codelineno-14-30"></a><span class="hll"><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
</span><a id="__codelineno-14-31" name="__codelineno-14-31"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The path in this case is a directory however it makes no difference, it could be a text file, for example.</p>
</div>
</details>
</div>
<h2 id="process-1d-filtered-and-recoded-set-of-variants">Process 1D: Filtered and recoded set of variants<a class="headerlink" href="#process-1d-filtered-and-recoded-set-of-variants" title="Permanent link">&para;</a></h2>
<p>Next on to something a little more tricky. The next process takes two inputs: the variants file and the blacklist file.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>In Nextflow, tuples can be defined in the input or output using the <a href="https://www.nextflow.io/docs/latest/process.html#input-type-tuple"><code>tuple</code></a> qualifier.</p>
</div>
<p>The next process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>prepare_vcf_file</code></li>
<li><strong>Command</strong>: create a filtered and recoded set of variants</li>
<li><strong>Input</strong>:</li>
<li>the variants file</li>
<li>the blacklisted regions file</li>
<li><strong>Output</strong>: a tuple containing the filtered/recoded VCF file and the tab index (TBI) file.</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #5</p>
<p>You must fill in the <code>BLANK</code>.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-15-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-15-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-15-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-15-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-15-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-15-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-15-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-15-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-15-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-15-10">10</a></span>
<span class="normal"><a href="#__codelineno-15-11">11</a></span>
<span class="normal"><a href="#__codelineno-15-12">12</a></span>
<span class="normal"><a href="#__codelineno-15-13">13</a></span>
<span class="normal"><a href="#__codelineno-15-14">14</a></span>
<span class="normal"><a href="#__codelineno-15-15">15</a></span>
<span class="normal"><a href="#__codelineno-15-16">16</a></span>
<span class="normal"><a href="#__codelineno-15-17">17</a></span>
<span class="normal"><a href="#__codelineno-15-18">18</a></span>
<span class="normal"><a href="#__codelineno-15-19">19</a></span>
<span class="normal"><a href="#__codelineno-15-20">20</a></span>
<span class="normal"><a href="#__codelineno-15-21">21</a></span>
<span class="normal"><a href="#__codelineno-15-22">22</a></span>
<span class="normal"><a href="#__codelineno-15-23">23</a></span>
<span class="normal"><a href="#__codelineno-15-24">24</a></span>
<span class="normal"><a href="#__codelineno-15-25">25</a></span>
<span class="normal"><a href="#__codelineno-15-26">26</a></span>
<span class="normal"><a href="#__codelineno-15-27">27</a></span>
<span class="normal"><a href="#__codelineno-15-28">28</a></span>
<span class="normal"><a href="#__codelineno-15-29">29</a></span>
<span class="normal"><a href="#__codelineno-15-30">30</a></span>
<span class="normal"><a href="#__codelineno-15-31">31</a></span>
<span class="normal"><a href="#__codelineno-15-32">32</a></span>
<span class="normal"><a href="#__codelineno-15-33">33</a></span>
<span class="normal"><a href="#__codelineno-15-34">34</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1"></a><span class="cm">/*</span>
<a id="__codelineno-15-2" name="__codelineno-15-2"></a><span class="cm"> * Process 1D: Create a file containing the filtered and recoded set of variants</span>
<a id="__codelineno-15-3" name="__codelineno-15-3"></a><span class="cm"> */</span>
<a id="__codelineno-15-4" name="__codelineno-15-4"></a>
<a id="__codelineno-15-5" name="__codelineno-15-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_vcf_file</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-15-6" name="__codelineno-15-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-b9358559e3ae3b9d7d8dbf1f401ae1fcaf757de3:ac05763cf181a5070c2fdb9bb5461f8d08f7b93b-0&#39;</span>
<a id="__codelineno-15-7" name="__codelineno-15-7"></a>
<a id="__codelineno-15-8" name="__codelineno-15-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-15-9" name="__codelineno-15-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">variantsFile</span>
<a id="__codelineno-15-10" name="__codelineno-15-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">blacklisted</span>
<a id="__codelineno-15-11" name="__codelineno-15-11"></a>
<a id="__codelineno-15-12" name="__codelineno-15-12"></a><span class="w">    </span><span class="nl">output:</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-15-13" name="__codelineno-15-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${variantsFile.baseName}.filtered.recode.vcf.gz&quot;</span><span class="o">),</span>
<a id="__codelineno-15-14" name="__codelineno-15-14"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${variantsFile.baseName}.filtered.recode.vcf.gz.tbi&quot;</span><span class="o">)</span>
<a id="__codelineno-15-15" name="__codelineno-15-15"></a>
<a id="__codelineno-15-16" name="__codelineno-15-16"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-15-17" name="__codelineno-15-17"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-15-18" name="__codelineno-15-18"></a><span class="s2">    vcftools --gzvcf ${variantsFile} -c \</span>
<a id="__codelineno-15-19" name="__codelineno-15-19"></a><span class="s2">             --exclude-bed ${blacklisted} \</span>
<a id="__codelineno-15-20" name="__codelineno-15-20"></a><span class="s2">             --recode | bgzip -c \</span>
<a id="__codelineno-15-21" name="__codelineno-15-21"></a><span class="s2">             &gt; ${variantsFile.baseName}.filtered.recode.vcf.gz</span>
<a id="__codelineno-15-22" name="__codelineno-15-22"></a>
<a id="__codelineno-15-23" name="__codelineno-15-23"></a><span class="s2">    tabix ${variantsFile.baseName}.filtered.recode.vcf.gz</span>
<a id="__codelineno-15-24" name="__codelineno-15-24"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-15-25" name="__codelineno-15-25"></a><span class="o">}</span>
<a id="__codelineno-15-26" name="__codelineno-15-26"></a>
<a id="__codelineno-15-27" name="__codelineno-15-27"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-15-28" name="__codelineno-15-28"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-15-29" name="__codelineno-15-29"></a>
<a id="__codelineno-15-30" name="__codelineno-15-30"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-15-31" name="__codelineno-15-31"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-15-32" name="__codelineno-15-32"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-15-33" name="__codelineno-15-33"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-15-34" name="__codelineno-15-34"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The output is a tuple with two paths that here are files</li>
</ol>
<p>Broken down, here is what the script is doing:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>vcftools<span class="w"> </span>--gzvcf<span class="w"> </span><span class="si">${</span><span class="nv">variantsFile</span><span class="si">}</span><span class="w"> </span>-c<span class="w"> </span><span class="se">\ </span><span class="c1"># (1)!</span>
<a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="w">         </span>--exclude-bed<span class="w"> </span><span class="si">${</span><span class="nv">blacklisted</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="c1"># (2)!</span>
<a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a><span class="w">         </span>--recode<span class="w"> </span><span class="p">|</span><span class="w"> </span>bgzip<span class="w"> </span>-c<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a><span class="w">         </span>&gt;<span class="w"> </span><span class="si">${</span><span class="nv">variantsFile</span><span class="p">.baseName</span><span class="si">}</span>.filtered.recode.vcf.gz<span class="w"> </span><span class="c1"># (3)!</span>
<a id="__codelineno-16-5" name="__codelineno-16-5" href="#__codelineno-16-5"></a>
<a id="__codelineno-16-6" name="__codelineno-16-6" href="#__codelineno-16-6"></a>tabix<span class="w"> </span><span class="si">${</span><span class="nv">variantsFile</span><span class="p">.baseName</span><span class="si">}</span>.filtered.recode.vcf.gz<span class="w"> </span><span class="c1"># (4)!</span>
</code></pre></div>
<ol>
<li>The <code>variantsFile</code> variable contains the path to the file with the known variants</li>
<li>The <code>blacklisted</code> variable contains the path to the file with the genomic locations which are known to produce artifacts and spurious variants</li>
<li>The <code>&gt;</code> symbol is used to redirect the output to the file specified after it</li>
<li><code>tabix</code> is used here to create the second output that we want to consider from this process</li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-17-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-17-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-17-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-17-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-17-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-17-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-17-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-17-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-17-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-17-10">10</a></span>
<span class="normal"><a href="#__codelineno-17-11">11</a></span>
<span class="normal"><a href="#__codelineno-17-12">12</a></span>
<span class="normal"><a href="#__codelineno-17-13">13</a></span>
<span class="normal"><a href="#__codelineno-17-14">14</a></span>
<span class="normal"><a href="#__codelineno-17-15">15</a></span>
<span class="normal"><a href="#__codelineno-17-16">16</a></span>
<span class="normal"><a href="#__codelineno-17-17">17</a></span>
<span class="normal"><a href="#__codelineno-17-18">18</a></span>
<span class="normal"><a href="#__codelineno-17-19">19</a></span>
<span class="normal"><a href="#__codelineno-17-20">20</a></span>
<span class="normal"><a href="#__codelineno-17-21">21</a></span>
<span class="normal"><a href="#__codelineno-17-22">22</a></span>
<span class="normal"><a href="#__codelineno-17-23">23</a></span>
<span class="normal"><a href="#__codelineno-17-24">24</a></span>
<span class="normal"><a href="#__codelineno-17-25">25</a></span>
<span class="normal"><a href="#__codelineno-17-26">26</a></span>
<span class="normal"><a href="#__codelineno-17-27">27</a></span>
<span class="normal"><a href="#__codelineno-17-28">28</a></span>
<span class="normal"><a href="#__codelineno-17-29">29</a></span>
<span class="normal"><a href="#__codelineno-17-30">30</a></span>
<span class="normal"><a href="#__codelineno-17-31">31</a></span>
<span class="normal"><a href="#__codelineno-17-32">32</a></span>
<span class="normal"><a href="#__codelineno-17-33">33</a></span>
<span class="normal"><a href="#__codelineno-17-34">34</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1"></a><span class="cm">/*</span>
<a id="__codelineno-17-2" name="__codelineno-17-2"></a><span class="cm"> * Process 1D: Create a file containing the filtered and recoded set of variants</span>
<a id="__codelineno-17-3" name="__codelineno-17-3"></a><span class="cm"> */</span>
<a id="__codelineno-17-4" name="__codelineno-17-4"></a>
<a id="__codelineno-17-5" name="__codelineno-17-5"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_vcf_file</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-17-6" name="__codelineno-17-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-b9358559e3ae3b9d7d8dbf1f401ae1fcaf757de3:ac05763cf181a5070c2fdb9bb5461f8d08f7b93b-0&#39;</span>
<a id="__codelineno-17-7" name="__codelineno-17-7"></a>
<a id="__codelineno-17-8" name="__codelineno-17-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-17-9" name="__codelineno-17-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">variantsFile</span>
<a id="__codelineno-17-10" name="__codelineno-17-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">blacklisted</span>
<a id="__codelineno-17-11" name="__codelineno-17-11"></a>
<a id="__codelineno-17-12" name="__codelineno-17-12"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-17-13" name="__codelineno-17-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${variantsFile.baseName}.filtered.recode.vcf.gz&quot;</span><span class="o">),</span>
<a id="__codelineno-17-14" name="__codelineno-17-14"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${variantsFile.baseName}.filtered.recode.vcf.gz.tbi&quot;</span><span class="o">)</span>
<a id="__codelineno-17-15" name="__codelineno-17-15"></a>
<a id="__codelineno-17-16" name="__codelineno-17-16"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-17-17" name="__codelineno-17-17"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-17-18" name="__codelineno-17-18"></a><span class="s2">    vcftools --gzvcf ${variantsFile} -c \</span>
<a id="__codelineno-17-19" name="__codelineno-17-19"></a><span class="s2">             --exclude-bed ${blacklisted} \</span>
<a id="__codelineno-17-20" name="__codelineno-17-20"></a><span class="s2">             --recode | bgzip -c \</span>
<a id="__codelineno-17-21" name="__codelineno-17-21"></a><span class="s2">             &gt; ${variantsFile.baseName}.filtered.recode.vcf.gz</span>
<a id="__codelineno-17-22" name="__codelineno-17-22"></a>
<a id="__codelineno-17-23" name="__codelineno-17-23"></a><span class="s2">    tabix ${variantsFile.baseName}.filtered.recode.vcf.gz</span>
<a id="__codelineno-17-24" name="__codelineno-17-24"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-17-25" name="__codelineno-17-25"></a><span class="o">}</span>
<a id="__codelineno-17-26" name="__codelineno-17-26"></a>
<a id="__codelineno-17-27" name="__codelineno-17-27"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-17-28" name="__codelineno-17-28"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-17-29" name="__codelineno-17-29"></a>
<a id="__codelineno-17-30" name="__codelineno-17-30"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-17-31" name="__codelineno-17-31"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-17-32" name="__codelineno-17-32"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-17-33" name="__codelineno-17-33"></a><span class="hll"><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
</span><a id="__codelineno-17-34" name="__codelineno-17-34"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
<p>Try running the pipeline from the project directory with:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>main.nf<span class="w"> </span>-resume
</code></pre></div>
</div>
<p>Congratulations! Part 1 is now complete.</p>
<p>We have all the data prepared and into channels ready for the more serious steps.</p>
<h2 id="process-2-star-mapping">Process 2: STAR Mapping<a class="headerlink" href="#process-2-star-mapping" title="Permanent link">&para;</a></h2>
<p>In this process, for each sample, we align the reads to our genome using the STAR index we created previously.</p>
<p>The process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>rnaseq_mapping_star</code></li>
<li><strong>Command</strong>: mapping of the RNA-Seq reads using STAR</li>
<li><strong>Input</strong>:</li>
<li>the genome fasta file</li>
<li>the STAR genome index</li>
<li>a tuple containing the replicate id and paired read files</li>
<li><strong>Output</strong>: a tuple containing replicate id, aligned bam file &amp; aligned bam file index</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #6</p>
<p>Copy the code below and paste it at the end of <code>main.nf</code>, removing the previous workflow block.</p>
<p>You must fill the <code>BLANK</code> space with the correct process call and inputs.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-19-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-19-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-19-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-19-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-19-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-19-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-19-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-19-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-19-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-19-10">10</a></span>
<span class="normal"><a href="#__codelineno-19-11">11</a></span>
<span class="normal"><a href="#__codelineno-19-12">12</a></span>
<span class="normal"><a href="#__codelineno-19-13">13</a></span>
<span class="normal"><a href="#__codelineno-19-14">14</a></span>
<span class="normal"><a href="#__codelineno-19-15">15</a></span>
<span class="normal"><a href="#__codelineno-19-16">16</a></span>
<span class="normal"><a href="#__codelineno-19-17">17</a></span>
<span class="normal"><a href="#__codelineno-19-18">18</a></span>
<span class="normal"><a href="#__codelineno-19-19">19</a></span>
<span class="normal"><a href="#__codelineno-19-20">20</a></span>
<span class="normal"><a href="#__codelineno-19-21">21</a></span>
<span class="normal"><a href="#__codelineno-19-22">22</a></span>
<span class="normal"><a href="#__codelineno-19-23">23</a></span>
<span class="normal"><a href="#__codelineno-19-24">24</a></span>
<span class="normal"><a href="#__codelineno-19-25">25</a></span>
<span class="normal"><a href="#__codelineno-19-26">26</a></span>
<span class="normal"><a href="#__codelineno-19-27">27</a></span>
<span class="normal"><a href="#__codelineno-19-28">28</a></span>
<span class="normal"><a href="#__codelineno-19-29">29</a></span>
<span class="normal"><a href="#__codelineno-19-30">30</a></span>
<span class="normal"><a href="#__codelineno-19-31">31</a></span>
<span class="normal"><a href="#__codelineno-19-32">32</a></span>
<span class="normal"><a href="#__codelineno-19-33">33</a></span>
<span class="normal"><a href="#__codelineno-19-34">34</a></span>
<span class="normal"><a href="#__codelineno-19-35">35</a></span>
<span class="normal"><a href="#__codelineno-19-36">36</a></span>
<span class="normal"><a href="#__codelineno-19-37">37</a></span>
<span class="normal"><a href="#__codelineno-19-38">38</a></span>
<span class="normal"><a href="#__codelineno-19-39">39</a></span>
<span class="normal"><a href="#__codelineno-19-40">40</a></span>
<span class="normal"><a href="#__codelineno-19-41">41</a></span>
<span class="normal"><a href="#__codelineno-19-42">42</a></span>
<span class="normal"><a href="#__codelineno-19-43">43</a></span>
<span class="normal"><a href="#__codelineno-19-44">44</a></span>
<span class="normal"><a href="#__codelineno-19-45">45</a></span>
<span class="normal"><a href="#__codelineno-19-46">46</a></span>
<span class="normal"><a href="#__codelineno-19-47">47</a></span>
<span class="normal"><a href="#__codelineno-19-48">48</a></span>
<span class="normal"><a href="#__codelineno-19-49">49</a></span>
<span class="normal"><a href="#__codelineno-19-50">50</a></span>
<span class="normal"><a href="#__codelineno-19-51">51</a></span>
<span class="normal"><a href="#__codelineno-19-52">52</a></span>
<span class="normal"><a href="#__codelineno-19-53">53</a></span>
<span class="normal"><a href="#__codelineno-19-54">54</a></span>
<span class="normal"><a href="#__codelineno-19-55">55</a></span>
<span class="normal"><a href="#__codelineno-19-56">56</a></span>
<span class="normal"><a href="#__codelineno-19-57">57</a></span>
<span class="normal"><a href="#__codelineno-19-58">58</a></span>
<span class="normal"><a href="#__codelineno-19-59">59</a></span>
<span class="normal"><a href="#__codelineno-19-60">60</a></span>
<span class="normal"><a href="#__codelineno-19-61">61</a></span>
<span class="normal"><a href="#__codelineno-19-62">62</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1"></a><span class="cm">/*</span>
<a id="__codelineno-19-2" name="__codelineno-19-2"></a><span class="cm"> * Process 2: Align RNA-Seq reads to the genome with STAR</span>
<a id="__codelineno-19-3" name="__codelineno-19-3"></a><span class="cm"> */</span>
<a id="__codelineno-19-4" name="__codelineno-19-4"></a>
<a id="__codelineno-19-5" name="__codelineno-19-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_mapping_star</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-19-6" name="__codelineno-19-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-52f8f283e3c401243cee4ee45f80122fbf6df3bb:e3bc54570927dc255f0e580cba1789b64690d611-0&#39;</span>
<a id="__codelineno-19-7" name="__codelineno-19-7"></a>
<a id="__codelineno-19-8" name="__codelineno-19-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-19-9" name="__codelineno-19-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-19-10" name="__codelineno-19-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genomeDir</span>
<a id="__codelineno-19-11" name="__codelineno-19-11"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">reads</span><span class="o">)</span>
<a id="__codelineno-19-12" name="__codelineno-19-12"></a>
<a id="__codelineno-19-13" name="__codelineno-19-13"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-19-14" name="__codelineno-19-14"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span>
<a id="__codelineno-19-15" name="__codelineno-19-15"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;Aligned.sortedByCoord.out.bam&#39;</span><span class="o">),</span>
<a id="__codelineno-19-16" name="__codelineno-19-16"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;Aligned.sortedByCoord.out.bam.bai&#39;</span><span class="o">)</span>
<a id="__codelineno-19-17" name="__codelineno-19-17"></a>
<a id="__codelineno-19-18" name="__codelineno-19-18"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-19-19" name="__codelineno-19-19"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-19-20" name="__codelineno-19-20"></a><span class="s2">    STAR --genomeDir ${genomeDir} \</span>
<a id="__codelineno-19-21" name="__codelineno-19-21"></a><span class="s2">         --readFilesIn ${reads} \</span>
<a id="__codelineno-19-22" name="__codelineno-19-22"></a><span class="s2">         --runThreadN ${task.cpus} \</span>
<a id="__codelineno-19-23" name="__codelineno-19-23"></a><span class="s2">         --readFilesCommand zcat \</span>
<a id="__codelineno-19-24" name="__codelineno-19-24"></a><span class="s2">         --outFilterType BySJout \</span>
<a id="__codelineno-19-25" name="__codelineno-19-25"></a><span class="s2">         --alignSJoverhangMin 8 \</span>
<a id="__codelineno-19-26" name="__codelineno-19-26"></a><span class="s2">         --alignSJDBoverhangMin 1 \</span>
<a id="__codelineno-19-27" name="__codelineno-19-27"></a><span class="s2">         --outFilterMismatchNmax 999</span>
<a id="__codelineno-19-28" name="__codelineno-19-28"></a>
<a id="__codelineno-19-29" name="__codelineno-19-29"></a><span class="s2">    mkdir -p genomeDir</span>
<a id="__codelineno-19-30" name="__codelineno-19-30"></a><span class="s2">    STAR --runMode genomeGenerate \</span>
<a id="__codelineno-19-31" name="__codelineno-19-31"></a><span class="s2">         --genomeDir genomeDir \</span>
<a id="__codelineno-19-32" name="__codelineno-19-32"></a><span class="s2">         --genomeFastaFiles ${genome} \</span>
<a id="__codelineno-19-33" name="__codelineno-19-33"></a><span class="s2">         --sjdbFileChrStartEnd SJ.out.tab \</span>
<a id="__codelineno-19-34" name="__codelineno-19-34"></a><span class="s2">         --sjdbOverhang 75 \</span>
<a id="__codelineno-19-35" name="__codelineno-19-35"></a><span class="s2">         --runThreadN ${task.cpus}</span>
<a id="__codelineno-19-36" name="__codelineno-19-36"></a>
<a id="__codelineno-19-37" name="__codelineno-19-37"></a><span class="s2">    STAR --genomeDir genomeDir \</span>
<a id="__codelineno-19-38" name="__codelineno-19-38"></a><span class="s2">         --readFilesIn ${reads} \</span>
<a id="__codelineno-19-39" name="__codelineno-19-39"></a><span class="s2">         --runThreadN ${task.cpus} \</span>
<a id="__codelineno-19-40" name="__codelineno-19-40"></a><span class="s2">         --readFilesCommand zcat \</span>
<a id="__codelineno-19-41" name="__codelineno-19-41"></a><span class="s2">         --outFilterType BySJout \</span>
<a id="__codelineno-19-42" name="__codelineno-19-42"></a><span class="s2">         --alignSJoverhangMin 8 \</span>
<a id="__codelineno-19-43" name="__codelineno-19-43"></a><span class="s2">         --alignSJDBoverhangMin 1 \</span>
<a id="__codelineno-19-44" name="__codelineno-19-44"></a><span class="s2">         --outFilterMismatchNmax 999 \</span>
<a id="__codelineno-19-45" name="__codelineno-19-45"></a><span class="s2">         --outSAMtype BAM SortedByCoordinate \</span>
<a id="__codelineno-19-46" name="__codelineno-19-46"></a><span class="s2">         --outSAMattrRGline ID:${replicateId} LB:library PL:illumina \</span>
<a id="__codelineno-19-47" name="__codelineno-19-47"></a><span class="s2">                            PU:machine SM:GM12878</span>
<a id="__codelineno-19-48" name="__codelineno-19-48"></a>
<a id="__codelineno-19-49" name="__codelineno-19-49"></a><span class="s2">    samtools index Aligned.sortedByCoord.out.bam</span>
<a id="__codelineno-19-50" name="__codelineno-19-50"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-19-51" name="__codelineno-19-51"></a><span class="o">}</span>
<a id="__codelineno-19-52" name="__codelineno-19-52"></a>
<a id="__codelineno-19-53" name="__codelineno-19-53"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-19-54" name="__codelineno-19-54"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-19-55" name="__codelineno-19-55"></a>
<a id="__codelineno-19-56" name="__codelineno-19-56"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-19-57" name="__codelineno-19-57"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-19-58" name="__codelineno-19-58"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-19-59" name="__codelineno-19-59"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-19-60" name="__codelineno-19-60"></a>
<a id="__codelineno-19-61" name="__codelineno-19-61"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-19-62" name="__codelineno-19-62"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The final command produces a bam index which is the full filename with an additional <code>.bai</code> suffix.</p>
</div>
<p>Broken down, here is what the script is doing:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>STAR<span class="w"> </span>--genomeDir<span class="w"> </span><span class="si">${</span><span class="nv">genomeDir</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="c1"># (1)!</span>
<a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a><span class="w">     </span>--readFilesIn<span class="w"> </span><span class="si">${</span><span class="nv">reads</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a><span class="w">     </span>--runThreadN<span class="w"> </span><span class="si">${</span><span class="nv">task</span><span class="p">.cpus</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-4" name="__codelineno-20-4" href="#__codelineno-20-4"></a><span class="w">     </span>--readFilesCommand<span class="w"> </span>zcat<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-5" name="__codelineno-20-5" href="#__codelineno-20-5"></a><span class="w">     </span>--outFilterType<span class="w"> </span>BySJout<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-6" name="__codelineno-20-6" href="#__codelineno-20-6"></a><span class="w">     </span>--alignSJoverhangMin<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-7" name="__codelineno-20-7" href="#__codelineno-20-7"></a><span class="w">     </span>--alignSJDBoverhangMin<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-8" name="__codelineno-20-8" href="#__codelineno-20-8"></a><span class="w">     </span>--outFilterMismatchNmax<span class="w"> </span><span class="m">999</span>
<a id="__codelineno-20-9" name="__codelineno-20-9" href="#__codelineno-20-9"></a>
<a id="__codelineno-20-10" name="__codelineno-20-10" href="#__codelineno-20-10"></a>mkdir<span class="w"> </span>-p<span class="w"> </span>genomeDir<span class="w"> </span><span class="c1"># (2)!</span>
<a id="__codelineno-20-11" name="__codelineno-20-11" href="#__codelineno-20-11"></a>STAR<span class="w"> </span>--runMode<span class="w"> </span>genomeGenerate<span class="w"> </span><span class="se">\ </span><span class="c1"># (3)!</span>
<a id="__codelineno-20-12" name="__codelineno-20-12" href="#__codelineno-20-12"></a><span class="w">     </span>--genomeDir<span class="w"> </span>genomeDir<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-13" name="__codelineno-20-13" href="#__codelineno-20-13"></a><span class="w">     </span>--genomeFastaFiles<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-14" name="__codelineno-20-14" href="#__codelineno-20-14"></a><span class="w">     </span>--sjdbFileChrStartEnd<span class="w"> </span>SJ.out.tab<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-15" name="__codelineno-20-15" href="#__codelineno-20-15"></a><span class="w">     </span>--sjdbOverhang<span class="w"> </span><span class="m">75</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-16" name="__codelineno-20-16" href="#__codelineno-20-16"></a><span class="w">     </span>--runThreadN<span class="w"> </span><span class="si">${</span><span class="nv">task</span><span class="p">.cpus</span><span class="si">}</span>
<a id="__codelineno-20-17" name="__codelineno-20-17" href="#__codelineno-20-17"></a>
<a id="__codelineno-20-18" name="__codelineno-20-18" href="#__codelineno-20-18"></a>STAR<span class="w"> </span>--genomeDir<span class="w"> </span>genomeDir<span class="w"> </span><span class="se">\ </span><span class="c1"># (4)!</span>
<a id="__codelineno-20-19" name="__codelineno-20-19" href="#__codelineno-20-19"></a><span class="w">     </span>--readFilesIn<span class="w"> </span><span class="si">${</span><span class="nv">reads</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-20" name="__codelineno-20-20" href="#__codelineno-20-20"></a><span class="w">     </span>--runThreadN<span class="w"> </span><span class="si">${</span><span class="nv">task</span><span class="p">.cpus</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-21" name="__codelineno-20-21" href="#__codelineno-20-21"></a><span class="w">     </span>--readFilesCommand<span class="w"> </span>zcat<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-22" name="__codelineno-20-22" href="#__codelineno-20-22"></a><span class="w">     </span>--outFilterType<span class="w"> </span>BySJout<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-23" name="__codelineno-20-23" href="#__codelineno-20-23"></a><span class="w">     </span>--alignSJoverhangMin<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-24" name="__codelineno-20-24" href="#__codelineno-20-24"></a><span class="w">     </span>--alignSJDBoverhangMin<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-25" name="__codelineno-20-25" href="#__codelineno-20-25"></a><span class="w">     </span>--outFilterMismatchNmax<span class="w"> </span><span class="m">999</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-26" name="__codelineno-20-26" href="#__codelineno-20-26"></a><span class="w">     </span>--outSAMtype<span class="w"> </span>BAM<span class="w"> </span>SortedByCoordinate<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-27" name="__codelineno-20-27" href="#__codelineno-20-27"></a><span class="w">     </span>--outSAMattrRGline<span class="w"> </span>ID:<span class="si">${</span><span class="nv">replicateId</span><span class="si">}</span><span class="w"> </span>LB:library<span class="w"> </span>PL:illumina<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-20-28" name="__codelineno-20-28" href="#__codelineno-20-28"></a><span class="w">                        </span>PU:machine<span class="w"> </span>SM:GM12878
<a id="__codelineno-20-29" name="__codelineno-20-29" href="#__codelineno-20-29"></a>
<a id="__codelineno-20-30" name="__codelineno-20-30" href="#__codelineno-20-30"></a>samtools<span class="w"> </span>index<span class="w"> </span>Aligned.sortedByCoord.out.bam<span class="w"> </span><span class="c1"># (5)!</span>
</code></pre></div>
<ol>
<li>Align reads to the reference genome</li>
<li>Create output directory <code>genomeDir</code> for next STAR calls within the same task</li>
<li>2nd pass (improve alignments using table of splice junctions and create a new index)</li>
<li>Final read alignments</li>
<li>Index the BAM file</li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-21-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-21-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-21-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-21-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-21-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-21-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-21-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-21-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-21-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-21-10">10</a></span>
<span class="normal"><a href="#__codelineno-21-11">11</a></span>
<span class="normal"><a href="#__codelineno-21-12">12</a></span>
<span class="normal"><a href="#__codelineno-21-13">13</a></span>
<span class="normal"><a href="#__codelineno-21-14">14</a></span>
<span class="normal"><a href="#__codelineno-21-15">15</a></span>
<span class="normal"><a href="#__codelineno-21-16">16</a></span>
<span class="normal"><a href="#__codelineno-21-17">17</a></span>
<span class="normal"><a href="#__codelineno-21-18">18</a></span>
<span class="normal"><a href="#__codelineno-21-19">19</a></span>
<span class="normal"><a href="#__codelineno-21-20">20</a></span>
<span class="normal"><a href="#__codelineno-21-21">21</a></span>
<span class="normal"><a href="#__codelineno-21-22">22</a></span>
<span class="normal"><a href="#__codelineno-21-23">23</a></span>
<span class="normal"><a href="#__codelineno-21-24">24</a></span>
<span class="normal"><a href="#__codelineno-21-25">25</a></span>
<span class="normal"><a href="#__codelineno-21-26">26</a></span>
<span class="normal"><a href="#__codelineno-21-27">27</a></span>
<span class="normal"><a href="#__codelineno-21-28">28</a></span>
<span class="normal"><a href="#__codelineno-21-29">29</a></span>
<span class="normal"><a href="#__codelineno-21-30">30</a></span>
<span class="normal"><a href="#__codelineno-21-31">31</a></span>
<span class="normal"><a href="#__codelineno-21-32">32</a></span>
<span class="normal"><a href="#__codelineno-21-33">33</a></span>
<span class="normal"><a href="#__codelineno-21-34">34</a></span>
<span class="normal"><a href="#__codelineno-21-35">35</a></span>
<span class="normal"><a href="#__codelineno-21-36">36</a></span>
<span class="normal"><a href="#__codelineno-21-37">37</a></span>
<span class="normal"><a href="#__codelineno-21-38">38</a></span>
<span class="normal"><a href="#__codelineno-21-39">39</a></span>
<span class="normal"><a href="#__codelineno-21-40">40</a></span>
<span class="normal"><a href="#__codelineno-21-41">41</a></span>
<span class="normal"><a href="#__codelineno-21-42">42</a></span>
<span class="normal"><a href="#__codelineno-21-43">43</a></span>
<span class="normal"><a href="#__codelineno-21-44">44</a></span>
<span class="normal"><a href="#__codelineno-21-45">45</a></span>
<span class="normal"><a href="#__codelineno-21-46">46</a></span>
<span class="normal"><a href="#__codelineno-21-47">47</a></span>
<span class="normal"><a href="#__codelineno-21-48">48</a></span>
<span class="normal"><a href="#__codelineno-21-49">49</a></span>
<span class="normal"><a href="#__codelineno-21-50">50</a></span>
<span class="normal"><a href="#__codelineno-21-51">51</a></span>
<span class="normal"><a href="#__codelineno-21-52">52</a></span>
<span class="normal"><a href="#__codelineno-21-53">53</a></span>
<span class="normal"><a href="#__codelineno-21-54">54</a></span>
<span class="normal"><a href="#__codelineno-21-55">55</a></span>
<span class="normal"><a href="#__codelineno-21-56">56</a></span>
<span class="normal"><a href="#__codelineno-21-57">57</a></span>
<span class="normal"><a href="#__codelineno-21-58">58</a></span>
<span class="normal"><a href="#__codelineno-21-59">59</a></span>
<span class="normal"><a href="#__codelineno-21-60">60</a></span>
<span class="normal"><a href="#__codelineno-21-61">61</a></span>
<span class="normal"><a href="#__codelineno-21-62">62</a></span>
<span class="normal"><a href="#__codelineno-21-63">63</a></span>
<span class="normal"><a href="#__codelineno-21-64">64</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1"></a><span class="cm">/*</span>
<a id="__codelineno-21-2" name="__codelineno-21-2"></a><span class="cm"> * Process 2: Align RNA-Seq reads to the genome with STAR</span>
<a id="__codelineno-21-3" name="__codelineno-21-3"></a><span class="cm"> */</span>
<a id="__codelineno-21-4" name="__codelineno-21-4"></a>
<a id="__codelineno-21-5" name="__codelineno-21-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_mapping_star</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-21-6" name="__codelineno-21-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-52f8f283e3c401243cee4ee45f80122fbf6df3bb:e3bc54570927dc255f0e580cba1789b64690d611-0&#39;</span>
<a id="__codelineno-21-7" name="__codelineno-21-7"></a>
<a id="__codelineno-21-8" name="__codelineno-21-8"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-21-9" name="__codelineno-21-9"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-21-10" name="__codelineno-21-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genomeDir</span>
<a id="__codelineno-21-11" name="__codelineno-21-11"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">reads</span><span class="o">)</span>
<a id="__codelineno-21-12" name="__codelineno-21-12"></a>
<a id="__codelineno-21-13" name="__codelineno-21-13"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-21-14" name="__codelineno-21-14"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span>
<a id="__codelineno-21-15" name="__codelineno-21-15"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;Aligned.sortedByCoord.out.bam&#39;</span><span class="o">),</span>
<a id="__codelineno-21-16" name="__codelineno-21-16"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;Aligned.sortedByCoord.out.bam.bai&#39;</span><span class="o">)</span>
<a id="__codelineno-21-17" name="__codelineno-21-17"></a>
<a id="__codelineno-21-18" name="__codelineno-21-18"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-21-19" name="__codelineno-21-19"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-21-20" name="__codelineno-21-20"></a><span class="s2">    STAR --genomeDir ${genomeDir} \</span>
<a id="__codelineno-21-21" name="__codelineno-21-21"></a><span class="s2">         --readFilesIn ${reads} \</span>
<a id="__codelineno-21-22" name="__codelineno-21-22"></a><span class="s2">         --runThreadN ${task.cpus} \</span>
<a id="__codelineno-21-23" name="__codelineno-21-23"></a><span class="s2">         --readFilesCommand zcat \</span>
<a id="__codelineno-21-24" name="__codelineno-21-24"></a><span class="s2">         --outFilterType BySJout \</span>
<a id="__codelineno-21-25" name="__codelineno-21-25"></a><span class="s2">         --alignSJoverhangMin 8 \</span>
<a id="__codelineno-21-26" name="__codelineno-21-26"></a><span class="s2">         --alignSJDBoverhangMin 1 \</span>
<a id="__codelineno-21-27" name="__codelineno-21-27"></a><span class="s2">         --outFilterMismatchNmax 999</span>
<a id="__codelineno-21-28" name="__codelineno-21-28"></a>
<a id="__codelineno-21-29" name="__codelineno-21-29"></a><span class="s2">    mkdir -p genomeDir</span>
<a id="__codelineno-21-30" name="__codelineno-21-30"></a><span class="s2">    STAR --runMode genomeGenerate \</span>
<a id="__codelineno-21-31" name="__codelineno-21-31"></a><span class="s2">         --genomeDir genomeDir \</span>
<a id="__codelineno-21-32" name="__codelineno-21-32"></a><span class="s2">         --genomeFastaFiles ${genome} \</span>
<a id="__codelineno-21-33" name="__codelineno-21-33"></a><span class="s2">         --sjdbFileChrStartEnd SJ.out.tab \</span>
<a id="__codelineno-21-34" name="__codelineno-21-34"></a><span class="s2">         --sjdbOverhang 75 \</span>
<a id="__codelineno-21-35" name="__codelineno-21-35"></a><span class="s2">         --runThreadN ${task.cpus}</span>
<a id="__codelineno-21-36" name="__codelineno-21-36"></a>
<a id="__codelineno-21-37" name="__codelineno-21-37"></a><span class="s2">    STAR --genomeDir genomeDir \</span>
<a id="__codelineno-21-38" name="__codelineno-21-38"></a><span class="s2">         --readFilesIn ${reads} \</span>
<a id="__codelineno-21-39" name="__codelineno-21-39"></a><span class="s2">         --runThreadN ${task.cpus} \</span>
<a id="__codelineno-21-40" name="__codelineno-21-40"></a><span class="s2">         --readFilesCommand zcat \</span>
<a id="__codelineno-21-41" name="__codelineno-21-41"></a><span class="s2">         --outFilterType BySJout \</span>
<a id="__codelineno-21-42" name="__codelineno-21-42"></a><span class="s2">         --alignSJoverhangMin 8 \</span>
<a id="__codelineno-21-43" name="__codelineno-21-43"></a><span class="s2">         --alignSJDBoverhangMin 1 \</span>
<a id="__codelineno-21-44" name="__codelineno-21-44"></a><span class="s2">         --outFilterMismatchNmax 999 \</span>
<a id="__codelineno-21-45" name="__codelineno-21-45"></a><span class="s2">         --outSAMtype BAM SortedByCoordinate \</span>
<a id="__codelineno-21-46" name="__codelineno-21-46"></a><span class="s2">         --outSAMattrRGline ID:${replicateId} LB:library PL:illumina \</span>
<a id="__codelineno-21-47" name="__codelineno-21-47"></a><span class="s2">                            PU:machine SM:GM12878</span>
<a id="__codelineno-21-48" name="__codelineno-21-48"></a>
<a id="__codelineno-21-49" name="__codelineno-21-49"></a><span class="s2">    samtools index Aligned.sortedByCoord.out.bam</span>
<a id="__codelineno-21-50" name="__codelineno-21-50"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-21-51" name="__codelineno-21-51"></a><span class="o">}</span>
<a id="__codelineno-21-52" name="__codelineno-21-52"></a>
<a id="__codelineno-21-53" name="__codelineno-21-53"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-21-54" name="__codelineno-21-54"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-21-55" name="__codelineno-21-55"></a>
<a id="__codelineno-21-56" name="__codelineno-21-56"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-21-57" name="__codelineno-21-57"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-21-58" name="__codelineno-21-58"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-21-59" name="__codelineno-21-59"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-21-60" name="__codelineno-21-60"></a>
<a id="__codelineno-21-61" name="__codelineno-21-61"></a><span class="hll"><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
</span><a id="__codelineno-21-62" name="__codelineno-21-62"></a><span class="hll"><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-21-63" name="__codelineno-21-63"></a><span class="hll"><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
</span><a id="__codelineno-21-64" name="__codelineno-21-64"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
</div>
<p>The next step is a filtering step using GATK. For each sample, we split all the reads that contain N characters in their <a href="http://genome.sph.umich.edu/wiki/SAM#What_is_a_CIGAR.3F">CIGAR</a> string.</p>
<h2 id="process-3-gatk-split-on-n">Process 3: GATK Split on N<a class="headerlink" href="#process-3-gatk-split-on-n" title="Permanent link">&para;</a></h2>
<p>The process creates <code>k+1</code> new reads (where <code>k</code> is the number of <code>N</code> cigar elements) that correspond to the segments of the original read beside/between the splicing events represented by the <code>N</code>s in the original CIGAR.</p>
<p>The next process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>rnaseq_gatk_splitNcigar</code></li>
<li><strong>Command</strong>: split reads on Ns in CIGAR string using GATK</li>
<li><strong>Input</strong>:</li>
<li>the genome fasta file</li>
<li>the genome index made with samtools</li>
<li>the genome dictionary made with picard</li>
<li>a tuple containing replicate id, aligned bam file and aligned bam file index from the STAR mapping</li>
<li><strong>Output</strong>: a tuple containing the replicate id, the split bam file and the split bam index file</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #7</p>
<p>Copy the code below and paste it at the end of <code>main.nf</code>, removing the previous workflow block.</p>
<p>You must fill the <code>BLANK</code> space with the correct process call and inputs.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>There is an optional <a href="https://www.nextflow.io/docs/latest/process.html#tag"><code>tag</code></a> line added to the start of this process. The <a href="https://www.nextflow.io/docs/latest/process.html#tag"><code>tag</code></a> line allows you to assign a name to a specific task (single instance of a process). This is particularly useful when there are many samples/replicates which pass through the same process.</p>
</div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-22-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-22-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-22-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-22-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-22-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-22-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-22-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-22-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-22-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-22-10">10</a></span>
<span class="normal"><a href="#__codelineno-22-11">11</a></span>
<span class="normal"><a href="#__codelineno-22-12">12</a></span>
<span class="normal"><a href="#__codelineno-22-13">13</a></span>
<span class="normal"><a href="#__codelineno-22-14">14</a></span>
<span class="normal"><a href="#__codelineno-22-15">15</a></span>
<span class="normal"><a href="#__codelineno-22-16">16</a></span>
<span class="normal"><a href="#__codelineno-22-17">17</a></span>
<span class="normal"><a href="#__codelineno-22-18">18</a></span>
<span class="normal"><a href="#__codelineno-22-19">19</a></span>
<span class="normal"><a href="#__codelineno-22-20">20</a></span>
<span class="normal"><a href="#__codelineno-22-21">21</a></span>
<span class="normal"><a href="#__codelineno-22-22">22</a></span>
<span class="normal"><a href="#__codelineno-22-23">23</a></span>
<span class="normal"><a href="#__codelineno-22-24">24</a></span>
<span class="normal"><a href="#__codelineno-22-25">25</a></span>
<span class="normal"><a href="#__codelineno-22-26">26</a></span>
<span class="normal"><a href="#__codelineno-22-27">27</a></span>
<span class="normal"><a href="#__codelineno-22-28">28</a></span>
<span class="normal"><a href="#__codelineno-22-29">29</a></span>
<span class="normal"><a href="#__codelineno-22-30">30</a></span>
<span class="normal"><a href="#__codelineno-22-31">31</a></span>
<span class="normal"><a href="#__codelineno-22-32">32</a></span>
<span class="normal"><a href="#__codelineno-22-33">33</a></span>
<span class="normal"><a href="#__codelineno-22-34">34</a></span>
<span class="normal"><a href="#__codelineno-22-35">35</a></span>
<span class="normal"><a href="#__codelineno-22-36">36</a></span>
<span class="normal"><a href="#__codelineno-22-37">37</a></span>
<span class="normal"><a href="#__codelineno-22-38">38</a></span>
<span class="normal"><a href="#__codelineno-22-39">39</a></span>
<span class="normal"><a href="#__codelineno-22-40">40</a></span>
<span class="normal"><a href="#__codelineno-22-41">41</a></span>
<span class="normal"><a href="#__codelineno-22-42">42</a></span>
<span class="normal"><a href="#__codelineno-22-43">43</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1"></a><span class="cm">/*</span>
<a id="__codelineno-22-2" name="__codelineno-22-2"></a><span class="cm"> * Process 3: GATK Split on N</span>
<a id="__codelineno-22-3" name="__codelineno-22-3"></a><span class="cm"> */</span>
<a id="__codelineno-22-4" name="__codelineno-22-4"></a>
<a id="__codelineno-22-5" name="__codelineno-22-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-22-6" name="__codelineno-22-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/broadinstitute/gotc-prod-gatk:1.0.0-4.1.8.0-1626439571&#39;</span>
<a id="__codelineno-22-7" name="__codelineno-22-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${replicateId}&quot;</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-22-8" name="__codelineno-22-8"></a>
<a id="__codelineno-22-9" name="__codelineno-22-9"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-22-10" name="__codelineno-22-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span><span class="w"> </span><span class="c1">// (2)!</span>
<a id="__codelineno-22-11" name="__codelineno-22-11"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="c1">// (3)!</span>
<a id="__codelineno-22-12" name="__codelineno-22-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome_dict</span><span class="w"> </span><span class="c1">// (4)!</span>
<a id="__codelineno-22-13" name="__codelineno-22-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span><span class="w"> </span><span class="c1">// (5)!</span>
<a id="__codelineno-22-14" name="__codelineno-22-14"></a>
<a id="__codelineno-22-15" name="__codelineno-22-15"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-22-16" name="__codelineno-22-16"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;split.bam&#39;</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;split.bai&#39;</span><span class="o">)</span><span class="w"> </span><span class="c1">// (6)!</span>
<a id="__codelineno-22-17" name="__codelineno-22-17"></a>
<a id="__codelineno-22-18" name="__codelineno-22-18"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-22-19" name="__codelineno-22-19"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-22-20" name="__codelineno-22-20"></a><span class="s2">    java -jar /usr/gitc/GATK35.jar -T SplitNCigarReads \</span>
<a id="__codelineno-22-21" name="__codelineno-22-21"></a><span class="s2">                                   -R ${genome} -I ${bam} \</span>
<a id="__codelineno-22-22" name="__codelineno-22-22"></a><span class="s2">                                   -o split.bam \</span>
<a id="__codelineno-22-23" name="__codelineno-22-23"></a><span class="s2">                                   -rf ReassignOneMappingQuality \</span>
<a id="__codelineno-22-24" name="__codelineno-22-24"></a><span class="s2">                                   -RMQF 255 -RMQT 60 \</span>
<a id="__codelineno-22-25" name="__codelineno-22-25"></a><span class="s2">                                   -U ALLOW_N_CIGAR_READS \</span>
<a id="__codelineno-22-26" name="__codelineno-22-26"></a><span class="s2">                                   --fix_misencoded_quality_scores</span>
<a id="__codelineno-22-27" name="__codelineno-22-27"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-22-28" name="__codelineno-22-28"></a><span class="o">}</span>
<a id="__codelineno-22-29" name="__codelineno-22-29"></a>
<a id="__codelineno-22-30" name="__codelineno-22-30"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-22-31" name="__codelineno-22-31"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-22-32" name="__codelineno-22-32"></a>
<a id="__codelineno-22-33" name="__codelineno-22-33"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-22-34" name="__codelineno-22-34"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-22-35" name="__codelineno-22-35"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-22-36" name="__codelineno-22-36"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-22-37" name="__codelineno-22-37"></a>
<a id="__codelineno-22-38" name="__codelineno-22-38"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-22-39" name="__codelineno-22-39"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-22-40" name="__codelineno-22-40"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-22-41" name="__codelineno-22-41"></a>
<a id="__codelineno-22-42" name="__codelineno-22-42"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-22-43" name="__codelineno-22-43"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li><a href="https://www.nextflow.io/docs/latest/process.html#tag"><code>tag</code></a> line using the replicate id as the tag.</li>
<li>The genome fasta file</li>
<li>The genome index in the output channel from the <code>prepare_genome_samtools</code> process</li>
<li>The genome dictionary in the output channel from the <code>prepare_genome_picard</code> process</li>
<li>The tuple containing the aligned reads in the output channel from the <code>rnaseq_mapping_star</code> process</li>
<li>A tuple containing the replicate id, the split bam file and the split bam index</li>
</ol>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The GATK command above automatically creates a bam index (<code>.bai</code>) of the <code>split.bam</code> output file</p>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>A <code>tag</code> line would also be useful in <a href="#process-2-star-mapping">Process 2</a></p>
</div>
<p>Broken down, here is what the script is doing:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>java<span class="w"> </span>-jar<span class="w"> </span>/usr/gitc/GATK35.jar<span class="w"> </span>-T<span class="w"> </span>SplitNCigarReads<span class="w"> </span><span class="se">\ </span><span class="c1"># (1)!</span>
<a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a><span class="w">                               </span>-R<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span>-I<span class="w"> </span><span class="si">${</span><span class="nv">bam</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="c1"># (2)!</span>
<a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a><span class="w">                               </span>-o<span class="w"> </span>split.bam<span class="w"> </span><span class="se">\ </span><span class="c1"># (3)!</span>
<a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a><span class="w">                               </span>-rf<span class="w"> </span>ReassignOneMappingQuality<span class="w"> </span><span class="se">\ </span><span class="c1"># (4)!</span>
<a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a><span class="w">                               </span>-RMQF<span class="w"> </span><span class="m">255</span><span class="w"> </span>-RMQT<span class="w"> </span><span class="m">60</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-23-6" name="__codelineno-23-6" href="#__codelineno-23-6"></a><span class="w">                               </span>-U<span class="w"> </span>ALLOW_N_CIGAR_READS<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-23-7" name="__codelineno-23-7" href="#__codelineno-23-7"></a><span class="w">                               </span>--fix_misencoded_quality_scores
</code></pre></div>
<ol>
<li>Use the <code>SplitNCigarReads</code> tool from GATK</li>
<li>Set the reference sequence file (<code>-R</code>) and the input files containing reads (<code>-I</code>)</li>
<li>Write the output BAM file to <code>split.bam</code> (<code>-o</code>)</li>
<li>Reassign mapping qualities too</li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-24-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-24-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-24-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-24-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-24-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-24-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-24-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-24-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-24-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-24-10">10</a></span>
<span class="normal"><a href="#__codelineno-24-11">11</a></span>
<span class="normal"><a href="#__codelineno-24-12">12</a></span>
<span class="normal"><a href="#__codelineno-24-13">13</a></span>
<span class="normal"><a href="#__codelineno-24-14">14</a></span>
<span class="normal"><a href="#__codelineno-24-15">15</a></span>
<span class="normal"><a href="#__codelineno-24-16">16</a></span>
<span class="normal"><a href="#__codelineno-24-17">17</a></span>
<span class="normal"><a href="#__codelineno-24-18">18</a></span>
<span class="normal"><a href="#__codelineno-24-19">19</a></span>
<span class="normal"><a href="#__codelineno-24-20">20</a></span>
<span class="normal"><a href="#__codelineno-24-21">21</a></span>
<span class="normal"><a href="#__codelineno-24-22">22</a></span>
<span class="normal"><a href="#__codelineno-24-23">23</a></span>
<span class="normal"><a href="#__codelineno-24-24">24</a></span>
<span class="normal"><a href="#__codelineno-24-25">25</a></span>
<span class="normal"><a href="#__codelineno-24-26">26</a></span>
<span class="normal"><a href="#__codelineno-24-27">27</a></span>
<span class="normal"><a href="#__codelineno-24-28">28</a></span>
<span class="normal"><a href="#__codelineno-24-29">29</a></span>
<span class="normal"><a href="#__codelineno-24-30">30</a></span>
<span class="normal"><a href="#__codelineno-24-31">31</a></span>
<span class="normal"><a href="#__codelineno-24-32">32</a></span>
<span class="normal"><a href="#__codelineno-24-33">33</a></span>
<span class="normal"><a href="#__codelineno-24-34">34</a></span>
<span class="normal"><a href="#__codelineno-24-35">35</a></span>
<span class="normal"><a href="#__codelineno-24-36">36</a></span>
<span class="normal"><a href="#__codelineno-24-37">37</a></span>
<span class="normal"><a href="#__codelineno-24-38">38</a></span>
<span class="normal"><a href="#__codelineno-24-39">39</a></span>
<span class="normal"><a href="#__codelineno-24-40">40</a></span>
<span class="normal"><a href="#__codelineno-24-41">41</a></span>
<span class="normal"><a href="#__codelineno-24-42">42</a></span>
<span class="normal"><a href="#__codelineno-24-43">43</a></span>
<span class="normal"><a href="#__codelineno-24-44">44</a></span>
<span class="normal"><a href="#__codelineno-24-45">45</a></span>
<span class="normal"><a href="#__codelineno-24-46">46</a></span>
<span class="normal"><a href="#__codelineno-24-47">47</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-24-1" name="__codelineno-24-1"></a><span class="cm">/*</span>
<a id="__codelineno-24-2" name="__codelineno-24-2"></a><span class="cm"> * Process 3: GATK Split on N</span>
<a id="__codelineno-24-3" name="__codelineno-24-3"></a><span class="cm"> */</span>
<a id="__codelineno-24-4" name="__codelineno-24-4"></a>
<a id="__codelineno-24-5" name="__codelineno-24-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-24-6" name="__codelineno-24-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/broadinstitute/gotc-prod-gatk:1.0.0-4.1.8.0-1626439571&#39;</span>
<a id="__codelineno-24-7" name="__codelineno-24-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${replicateId}&quot;</span>
<a id="__codelineno-24-8" name="__codelineno-24-8"></a>
<a id="__codelineno-24-9" name="__codelineno-24-9"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-24-10" name="__codelineno-24-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-24-11" name="__codelineno-24-11"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span>
<a id="__codelineno-24-12" name="__codelineno-24-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome_dict</span>
<a id="__codelineno-24-13" name="__codelineno-24-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span>
<a id="__codelineno-24-14" name="__codelineno-24-14"></a>
<a id="__codelineno-24-15" name="__codelineno-24-15"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-24-16" name="__codelineno-24-16"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;split.bam&#39;</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;split.bai&#39;</span><span class="o">)</span>
<a id="__codelineno-24-17" name="__codelineno-24-17"></a>
<a id="__codelineno-24-18" name="__codelineno-24-18"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-24-19" name="__codelineno-24-19"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-24-20" name="__codelineno-24-20"></a><span class="s2">    java -jar /usr/gitc/GATK35.jar -T SplitNCigarReads \</span>
<a id="__codelineno-24-21" name="__codelineno-24-21"></a><span class="s2">                                   -R ${genome} -I ${bam} \</span>
<a id="__codelineno-24-22" name="__codelineno-24-22"></a><span class="s2">                                   -o split.bam \</span>
<a id="__codelineno-24-23" name="__codelineno-24-23"></a><span class="s2">                                   -rf ReassignOneMappingQuality \</span>
<a id="__codelineno-24-24" name="__codelineno-24-24"></a><span class="s2">                                   -RMQF 255 -RMQT 60 \</span>
<a id="__codelineno-24-25" name="__codelineno-24-25"></a><span class="s2">                                   -U ALLOW_N_CIGAR_READS \</span>
<a id="__codelineno-24-26" name="__codelineno-24-26"></a><span class="s2">                                   --fix_misencoded_quality_scores</span>
<a id="__codelineno-24-27" name="__codelineno-24-27"></a>
<a id="__codelineno-24-28" name="__codelineno-24-28"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-24-29" name="__codelineno-24-29"></a><span class="o">}</span>
<a id="__codelineno-24-30" name="__codelineno-24-30"></a>
<a id="__codelineno-24-31" name="__codelineno-24-31"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-24-32" name="__codelineno-24-32"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-24-33" name="__codelineno-24-33"></a>
<a id="__codelineno-24-34" name="__codelineno-24-34"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-24-35" name="__codelineno-24-35"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-24-36" name="__codelineno-24-36"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-24-37" name="__codelineno-24-37"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-24-38" name="__codelineno-24-38"></a>
<a id="__codelineno-24-39" name="__codelineno-24-39"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-24-40" name="__codelineno-24-40"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-24-41" name="__codelineno-24-41"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-24-42" name="__codelineno-24-42"></a>
<a id="__codelineno-24-43" name="__codelineno-24-43"></a><span class="hll"><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
</span><a id="__codelineno-24-44" name="__codelineno-24-44"></a><span class="hll"><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-24-45" name="__codelineno-24-45"></a><span class="hll"><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-24-46" name="__codelineno-24-46"></a><span class="hll"><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
</span><a id="__codelineno-24-47" name="__codelineno-24-47"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
</div>
<p>Next we perform a Base Quality Score Recalibration step using GATK.</p>
<h2 id="process-4-gatk-recalibrate">Process 4: GATK Recalibrate<a class="headerlink" href="#process-4-gatk-recalibrate" title="Permanent link">&para;</a></h2>
<p>This step uses GATK to detect systematic errors in the base quality scores, select unique alignments and then index the resulting bam file with samtools. You can find details of the specific GATK BaseRecalibrator parameters <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/3.8-0/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php">here</a>.</p>
<p>The next process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>rnaseq_gatk_recalibrate</code></li>
<li><strong>Command</strong>: recalibrate reads from each replicate using GATK</li>
<li><strong>Input</strong></li>
<li>the genome fasta file</li>
<li>the genome index made with samtools</li>
<li>the genome dictionary made with picard</li>
<li>a tuple containing replicate id, aligned bam file and aligned bam file index from process 3</li>
<li>a tuple containing the filtered/recoded VCF file and the tab index (TBI) file from process 1D</li>
<li><strong>Output</strong>: a tuple containing the sample id, the unique bam file and the unique bam index file</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #8</p>
<p>Copy the code below and paste it at the end of <code>main.nf</code>, removing the previous workflow block.</p>
<p>Your aim is to replace the <code>BLANK</code> placeholder with the the correct process call.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-25-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-25-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-25-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-25-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-25-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-25-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-25-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-25-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-25-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-25-10">10</a></span>
<span class="normal"><a href="#__codelineno-25-11">11</a></span>
<span class="normal"><a href="#__codelineno-25-12">12</a></span>
<span class="normal"><a href="#__codelineno-25-13">13</a></span>
<span class="normal"><a href="#__codelineno-25-14">14</a></span>
<span class="normal"><a href="#__codelineno-25-15">15</a></span>
<span class="normal"><a href="#__codelineno-25-16">16</a></span>
<span class="normal"><a href="#__codelineno-25-17">17</a></span>
<span class="normal"><a href="#__codelineno-25-18">18</a></span>
<span class="normal"><a href="#__codelineno-25-19">19</a></span>
<span class="normal"><a href="#__codelineno-25-20">20</a></span>
<span class="normal"><a href="#__codelineno-25-21">21</a></span>
<span class="normal"><a href="#__codelineno-25-22">22</a></span>
<span class="normal"><a href="#__codelineno-25-23">23</a></span>
<span class="normal"><a href="#__codelineno-25-24">24</a></span>
<span class="normal"><a href="#__codelineno-25-25">25</a></span>
<span class="normal"><a href="#__codelineno-25-26">26</a></span>
<span class="normal"><a href="#__codelineno-25-27">27</a></span>
<span class="normal"><a href="#__codelineno-25-28">28</a></span>
<span class="normal"><a href="#__codelineno-25-29">29</a></span>
<span class="normal"><a href="#__codelineno-25-30">30</a></span>
<span class="normal"><a href="#__codelineno-25-31">31</a></span>
<span class="normal"><a href="#__codelineno-25-32">32</a></span>
<span class="normal"><a href="#__codelineno-25-33">33</a></span>
<span class="normal"><a href="#__codelineno-25-34">34</a></span>
<span class="normal"><a href="#__codelineno-25-35">35</a></span>
<span class="normal"><a href="#__codelineno-25-36">36</a></span>
<span class="normal"><a href="#__codelineno-25-37">37</a></span>
<span class="normal"><a href="#__codelineno-25-38">38</a></span>
<span class="normal"><a href="#__codelineno-25-39">39</a></span>
<span class="normal"><a href="#__codelineno-25-40">40</a></span>
<span class="normal"><a href="#__codelineno-25-41">41</a></span>
<span class="normal"><a href="#__codelineno-25-42">42</a></span>
<span class="normal"><a href="#__codelineno-25-43">43</a></span>
<span class="normal"><a href="#__codelineno-25-44">44</a></span>
<span class="normal"><a href="#__codelineno-25-45">45</a></span>
<span class="normal"><a href="#__codelineno-25-46">46</a></span>
<span class="normal"><a href="#__codelineno-25-47">47</a></span>
<span class="normal"><a href="#__codelineno-25-48">48</a></span>
<span class="normal"><a href="#__codelineno-25-49">49</a></span>
<span class="normal"><a href="#__codelineno-25-50">50</a></span>
<span class="normal"><a href="#__codelineno-25-51">51</a></span>
<span class="normal"><a href="#__codelineno-25-52">52</a></span>
<span class="normal"><a href="#__codelineno-25-53">53</a></span>
<span class="normal"><a href="#__codelineno-25-54">54</a></span>
<span class="normal"><a href="#__codelineno-25-55">55</a></span>
<span class="normal"><a href="#__codelineno-25-56">56</a></span>
<span class="normal"><a href="#__codelineno-25-57">57</a></span>
<span class="normal"><a href="#__codelineno-25-58">58</a></span>
<span class="normal"><a href="#__codelineno-25-59">59</a></span>
<span class="normal"><a href="#__codelineno-25-60">60</a></span>
<span class="normal"><a href="#__codelineno-25-61">61</a></span>
<span class="normal"><a href="#__codelineno-25-62">62</a></span>
<span class="normal"><a href="#__codelineno-25-63">63</a></span>
<span class="normal"><a href="#__codelineno-25-64">64</a></span>
<span class="normal"><a href="#__codelineno-25-65">65</a></span>
<span class="normal"><a href="#__codelineno-25-66">66</a></span>
<span class="normal"><a href="#__codelineno-25-67">67</a></span>
<span class="normal"><a href="#__codelineno-25-68">68</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-25-1" name="__codelineno-25-1"></a><span class="cm">/*</span>
<a id="__codelineno-25-2" name="__codelineno-25-2"></a><span class="cm"> * Process 4: GATK Recalibrate</span>
<a id="__codelineno-25-3" name="__codelineno-25-3"></a><span class="cm"> */</span>
<a id="__codelineno-25-4" name="__codelineno-25-4"></a>
<a id="__codelineno-25-5" name="__codelineno-25-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_gatk_recalibrate</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-25-6" name="__codelineno-25-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-aa1d7bddaee5eb6c4cbab18f8a072e3ea7ec3969:f963c36fd770e89d267eeaa27cad95c1c3dbe660-0&#39;</span>
<a id="__codelineno-25-7" name="__codelineno-25-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${replicateId}&quot;</span>
<a id="__codelineno-25-8" name="__codelineno-25-8"></a>
<a id="__codelineno-25-9" name="__codelineno-25-9"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-25-10" name="__codelineno-25-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-25-11" name="__codelineno-25-11"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span>
<a id="__codelineno-25-12" name="__codelineno-25-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">dict</span>
<a id="__codelineno-25-13" name="__codelineno-25-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-25-14" name="__codelineno-25-14"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">prepared_variants_file</span><span class="o">),</span>
<a id="__codelineno-25-15" name="__codelineno-25-15"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="n">prepared_variants_file_index</span><span class="o">)</span><span class="w"> </span><span class="c1">// (2)!</span>
<a id="__codelineno-25-16" name="__codelineno-25-16"></a>
<a id="__codelineno-25-17" name="__codelineno-25-17"></a><span class="w">    </span><span class="nl">output:</span><span class="w"> </span><span class="c1">// (3)!</span>
<a id="__codelineno-25-18" name="__codelineno-25-18"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span>
<a id="__codelineno-25-19" name="__codelineno-25-19"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${replicateId}.final.uniq.bam&quot;</span><span class="o">),</span>
<a id="__codelineno-25-20" name="__codelineno-25-20"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${replicateId}.final.uniq.bam.bai&quot;</span><span class="o">)</span>
<a id="__codelineno-25-21" name="__codelineno-25-21"></a>
<a id="__codelineno-25-22" name="__codelineno-25-22"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-25-23" name="__codelineno-25-23"></a><span class="w">    </span><span class="n">sampleId</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">replicateId</span><span class="o">.</span><span class="na">replaceAll</span><span class="o">(</span><span class="s">/[12]$/</span><span class="o">,</span><span class="s1">&#39;&#39;</span><span class="o">)</span><span class="w"> </span><span class="c1">// (4)!</span>
<a id="__codelineno-25-24" name="__codelineno-25-24"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-25-25" name="__codelineno-25-25"></a><span class="s2">    gatk3 -T BaseRecalibrator \</span>
<a id="__codelineno-25-26" name="__codelineno-25-26"></a><span class="s2">          --default_platform illumina \</span>
<a id="__codelineno-25-27" name="__codelineno-25-27"></a><span class="s2">          -cov ReadGroupCovariate \</span>
<a id="__codelineno-25-28" name="__codelineno-25-28"></a><span class="s2">          -cov QualityScoreCovariate \</span>
<a id="__codelineno-25-29" name="__codelineno-25-29"></a><span class="s2">          -cov CycleCovariate \</span>
<a id="__codelineno-25-30" name="__codelineno-25-30"></a><span class="s2">          -knownSites ${prepared_variants_file} \</span>
<a id="__codelineno-25-31" name="__codelineno-25-31"></a><span class="s2">          -cov ContextCovariate \</span>
<a id="__codelineno-25-32" name="__codelineno-25-32"></a><span class="s2">          -R ${genome} -I ${bam} \</span>
<a id="__codelineno-25-33" name="__codelineno-25-33"></a><span class="s2">          --downsampling_type NONE \</span>
<a id="__codelineno-25-34" name="__codelineno-25-34"></a><span class="s2">          -nct ${task.cpus} \</span>
<a id="__codelineno-25-35" name="__codelineno-25-35"></a><span class="s2">          -o final.rnaseq.grp</span>
<a id="__codelineno-25-36" name="__codelineno-25-36"></a>
<a id="__codelineno-25-37" name="__codelineno-25-37"></a><span class="s2">    gatk3 -T PrintReads \</span>
<a id="__codelineno-25-38" name="__codelineno-25-38"></a><span class="s2">          -R ${genome} -I ${bam} \</span>
<a id="__codelineno-25-39" name="__codelineno-25-39"></a><span class="s2">          -BQSR final.rnaseq.grp \</span>
<a id="__codelineno-25-40" name="__codelineno-25-40"></a><span class="s2">          -nct ${task.cpus} \</span>
<a id="__codelineno-25-41" name="__codelineno-25-41"></a><span class="s2">          -o final.bam</span>
<a id="__codelineno-25-42" name="__codelineno-25-42"></a>
<a id="__codelineno-25-43" name="__codelineno-25-43"></a><span class="s2">    (samtools view -H final.bam; samtools view final.bam | \</span>
<a id="__codelineno-25-44" name="__codelineno-25-44"></a><span class="s2">    grep -w &#39;NH:i:1&#39;) | samtools view -Sb -  &gt; ${replicateId}.final.uniq.bam</span>
<a id="__codelineno-25-45" name="__codelineno-25-45"></a>
<a id="__codelineno-25-46" name="__codelineno-25-46"></a><span class="s2">    samtools index ${replicateId}.final.uniq.bam</span>
<a id="__codelineno-25-47" name="__codelineno-25-47"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-25-48" name="__codelineno-25-48"></a><span class="o">}</span>
<a id="__codelineno-25-49" name="__codelineno-25-49"></a>
<a id="__codelineno-25-50" name="__codelineno-25-50"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-25-51" name="__codelineno-25-51"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-25-52" name="__codelineno-25-52"></a>
<a id="__codelineno-25-53" name="__codelineno-25-53"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-25-54" name="__codelineno-25-54"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-25-55" name="__codelineno-25-55"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-25-56" name="__codelineno-25-56"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-25-57" name="__codelineno-25-57"></a>
<a id="__codelineno-25-58" name="__codelineno-25-58"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-25-59" name="__codelineno-25-59"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-25-60" name="__codelineno-25-60"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-25-61" name="__codelineno-25-61"></a>
<a id="__codelineno-25-62" name="__codelineno-25-62"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-25-63" name="__codelineno-25-63"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-25-64" name="__codelineno-25-64"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-25-65" name="__codelineno-25-65"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-25-66" name="__codelineno-25-66"></a>
<a id="__codelineno-25-67" name="__codelineno-25-67"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-25-68" name="__codelineno-25-68"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The tuple containing the split reads in the output channel from the <code>rnaseq_gatk_splitNcigar</code> process.</li>
<li>The tuple containing the filtered/recoded VCF file and the tab index (TBI) file in the output channel from the <code>prepare_vcf_file</code> process.</li>
<li>The tuple containing the replicate id, the unique bam file and the unique bam index file which goes into two channels.</li>
<li>Line specifying the filename of the output bam file</li>
</ol>
<p>Broken down, here is what the script is doing:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>gatk3<span class="w"> </span>-T<span class="w"> </span>BaseRecalibrator<span class="w"> </span><span class="se">\ </span><span class="c1"># (1)!</span>
<a id="__codelineno-26-2" name="__codelineno-26-2" href="#__codelineno-26-2"></a><span class="w">      </span>--default_platform<span class="w"> </span>illumina<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-3" name="__codelineno-26-3" href="#__codelineno-26-3"></a><span class="w">      </span>-cov<span class="w"> </span>ReadGroupCovariate<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-4" name="__codelineno-26-4" href="#__codelineno-26-4"></a><span class="w">      </span>-cov<span class="w"> </span>QualityScoreCovariate<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-5" name="__codelineno-26-5" href="#__codelineno-26-5"></a><span class="w">      </span>-cov<span class="w"> </span>CycleCovariate<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-6" name="__codelineno-26-6" href="#__codelineno-26-6"></a><span class="w">      </span>-knownSites<span class="w"> </span><span class="si">${</span><span class="nv">prepared_variants_file</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-7" name="__codelineno-26-7" href="#__codelineno-26-7"></a><span class="w">      </span>-cov<span class="w"> </span>ContextCovariate<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-8" name="__codelineno-26-8" href="#__codelineno-26-8"></a><span class="w">      </span>-R<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span>-I<span class="w"> </span><span class="si">${</span><span class="nv">bam</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-9" name="__codelineno-26-9" href="#__codelineno-26-9"></a><span class="w">      </span>--downsampling_type<span class="w"> </span>NONE<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-10" name="__codelineno-26-10" href="#__codelineno-26-10"></a><span class="w">      </span>-nct<span class="w"> </span><span class="si">${</span><span class="nv">task</span><span class="p">.cpus</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-11" name="__codelineno-26-11" href="#__codelineno-26-11"></a><span class="w">      </span>-o<span class="w"> </span>final.rnaseq.grp
<a id="__codelineno-26-12" name="__codelineno-26-12" href="#__codelineno-26-12"></a>
<a id="__codelineno-26-13" name="__codelineno-26-13" href="#__codelineno-26-13"></a>gatk3<span class="w"> </span>-T<span class="w"> </span>PrintReads<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-14" name="__codelineno-26-14" href="#__codelineno-26-14"></a><span class="w">      </span>-R<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span>-I<span class="w"> </span><span class="si">${</span><span class="nv">bam</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-15" name="__codelineno-26-15" href="#__codelineno-26-15"></a><span class="w">      </span>-BQSR<span class="w"> </span>final.rnaseq.grp<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-16" name="__codelineno-26-16" href="#__codelineno-26-16"></a><span class="w">      </span>-nct<span class="w"> </span><span class="si">${</span><span class="nv">task</span><span class="p">.cpus</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-17" name="__codelineno-26-17" href="#__codelineno-26-17"></a><span class="w">      </span>-o<span class="w"> </span>final.bam
<a id="__codelineno-26-18" name="__codelineno-26-18" href="#__codelineno-26-18"></a>
<a id="__codelineno-26-19" name="__codelineno-26-19" href="#__codelineno-26-19"></a><span class="o">(</span>samtools<span class="w"> </span>view<span class="w"> </span>-H<span class="w"> </span>final.bam<span class="p">;</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>final.bam<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-26-20" name="__codelineno-26-20" href="#__codelineno-26-20"></a>grep<span class="w"> </span>-w<span class="w"> </span><span class="s1">&#39;NH:i:1&#39;</span><span class="o">)</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-Sb<span class="w"> </span>-<span class="w">  </span>&gt;<span class="w"> </span><span class="si">${</span><span class="nv">replicateId</span><span class="si">}</span>.final.uniq.bam<span class="w"> </span><span class="c1"># (2)!</span>
<a id="__codelineno-26-21" name="__codelineno-26-21" href="#__codelineno-26-21"></a>
<a id="__codelineno-26-22" name="__codelineno-26-22" href="#__codelineno-26-22"></a>samtools<span class="w"> </span>index<span class="w"> </span><span class="si">${</span><span class="nv">replicateId</span><span class="si">}</span>.final.uniq.bam<span class="w"> </span><span class="c1"># (3)!</span>
</code></pre></div>
<ol>
<li>Generates recalibration table for Base Quality Score Recalibration (BQSR)</li>
<li>Select only unique alignments, no multimaps</li>
<li>Index BAM file</li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-27-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-27-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-27-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-27-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-27-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-27-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-27-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-27-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-27-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-27-10">10</a></span>
<span class="normal"><a href="#__codelineno-27-11">11</a></span>
<span class="normal"><a href="#__codelineno-27-12">12</a></span>
<span class="normal"><a href="#__codelineno-27-13">13</a></span>
<span class="normal"><a href="#__codelineno-27-14">14</a></span>
<span class="normal"><a href="#__codelineno-27-15">15</a></span>
<span class="normal"><a href="#__codelineno-27-16">16</a></span>
<span class="normal"><a href="#__codelineno-27-17">17</a></span>
<span class="normal"><a href="#__codelineno-27-18">18</a></span>
<span class="normal"><a href="#__codelineno-27-19">19</a></span>
<span class="normal"><a href="#__codelineno-27-20">20</a></span>
<span class="normal"><a href="#__codelineno-27-21">21</a></span>
<span class="normal"><a href="#__codelineno-27-22">22</a></span>
<span class="normal"><a href="#__codelineno-27-23">23</a></span>
<span class="normal"><a href="#__codelineno-27-24">24</a></span>
<span class="normal"><a href="#__codelineno-27-25">25</a></span>
<span class="normal"><a href="#__codelineno-27-26">26</a></span>
<span class="normal"><a href="#__codelineno-27-27">27</a></span>
<span class="normal"><a href="#__codelineno-27-28">28</a></span>
<span class="normal"><a href="#__codelineno-27-29">29</a></span>
<span class="normal"><a href="#__codelineno-27-30">30</a></span>
<span class="normal"><a href="#__codelineno-27-31">31</a></span>
<span class="normal"><a href="#__codelineno-27-32">32</a></span>
<span class="normal"><a href="#__codelineno-27-33">33</a></span>
<span class="normal"><a href="#__codelineno-27-34">34</a></span>
<span class="normal"><a href="#__codelineno-27-35">35</a></span>
<span class="normal"><a href="#__codelineno-27-36">36</a></span>
<span class="normal"><a href="#__codelineno-27-37">37</a></span>
<span class="normal"><a href="#__codelineno-27-38">38</a></span>
<span class="normal"><a href="#__codelineno-27-39">39</a></span>
<span class="normal"><a href="#__codelineno-27-40">40</a></span>
<span class="normal"><a href="#__codelineno-27-41">41</a></span>
<span class="normal"><a href="#__codelineno-27-42">42</a></span>
<span class="normal"><a href="#__codelineno-27-43">43</a></span>
<span class="normal"><a href="#__codelineno-27-44">44</a></span>
<span class="normal"><a href="#__codelineno-27-45">45</a></span>
<span class="normal"><a href="#__codelineno-27-46">46</a></span>
<span class="normal"><a href="#__codelineno-27-47">47</a></span>
<span class="normal"><a href="#__codelineno-27-48">48</a></span>
<span class="normal"><a href="#__codelineno-27-49">49</a></span>
<span class="normal"><a href="#__codelineno-27-50">50</a></span>
<span class="normal"><a href="#__codelineno-27-51">51</a></span>
<span class="normal"><a href="#__codelineno-27-52">52</a></span>
<span class="normal"><a href="#__codelineno-27-53">53</a></span>
<span class="normal"><a href="#__codelineno-27-54">54</a></span>
<span class="normal"><a href="#__codelineno-27-55">55</a></span>
<span class="normal"><a href="#__codelineno-27-56">56</a></span>
<span class="normal"><a href="#__codelineno-27-57">57</a></span>
<span class="normal"><a href="#__codelineno-27-58">58</a></span>
<span class="normal"><a href="#__codelineno-27-59">59</a></span>
<span class="normal"><a href="#__codelineno-27-60">60</a></span>
<span class="normal"><a href="#__codelineno-27-61">61</a></span>
<span class="normal"><a href="#__codelineno-27-62">62</a></span>
<span class="normal"><a href="#__codelineno-27-63">63</a></span>
<span class="normal"><a href="#__codelineno-27-64">64</a></span>
<span class="normal"><a href="#__codelineno-27-65">65</a></span>
<span class="normal"><a href="#__codelineno-27-66">66</a></span>
<span class="normal"><a href="#__codelineno-27-67">67</a></span>
<span class="normal"><a href="#__codelineno-27-68">68</a></span>
<span class="normal"><a href="#__codelineno-27-69">69</a></span>
<span class="normal"><a href="#__codelineno-27-70">70</a></span>
<span class="normal"><a href="#__codelineno-27-71">71</a></span>
<span class="normal"><a href="#__codelineno-27-72">72</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1"></a><span class="cm">/*</span>
<a id="__codelineno-27-2" name="__codelineno-27-2"></a><span class="cm"> * Process 4: GATK Recalibrate</span>
<a id="__codelineno-27-3" name="__codelineno-27-3"></a><span class="cm"> */</span>
<a id="__codelineno-27-4" name="__codelineno-27-4"></a>
<a id="__codelineno-27-5" name="__codelineno-27-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_gatk_recalibrate</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-27-6" name="__codelineno-27-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-aa1d7bddaee5eb6c4cbab18f8a072e3ea7ec3969:f963c36fd770e89d267eeaa27cad95c1c3dbe660-0&#39;</span>
<a id="__codelineno-27-7" name="__codelineno-27-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${replicateId}&quot;</span>
<a id="__codelineno-27-8" name="__codelineno-27-8"></a>
<a id="__codelineno-27-9" name="__codelineno-27-9"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-27-10" name="__codelineno-27-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-27-11" name="__codelineno-27-11"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span>
<a id="__codelineno-27-12" name="__codelineno-27-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">dict</span>
<a id="__codelineno-27-13" name="__codelineno-27-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">replicateId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span>
<a id="__codelineno-27-14" name="__codelineno-27-14"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">prepared_variants_file</span><span class="o">),</span>
<a id="__codelineno-27-15" name="__codelineno-27-15"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="n">prepared_variants_file_index</span><span class="o">)</span>
<a id="__codelineno-27-16" name="__codelineno-27-16"></a>
<a id="__codelineno-27-17" name="__codelineno-27-17"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-27-18" name="__codelineno-27-18"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span>
<a id="__codelineno-27-19" name="__codelineno-27-19"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${replicateId}.final.uniq.bam&quot;</span><span class="o">),</span>
<a id="__codelineno-27-20" name="__codelineno-27-20"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${replicateId}.final.uniq.bam.bai&quot;</span><span class="o">)</span>
<a id="__codelineno-27-21" name="__codelineno-27-21"></a>
<a id="__codelineno-27-22" name="__codelineno-27-22"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-27-23" name="__codelineno-27-23"></a><span class="w">    </span><span class="n">sampleId</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">replicateId</span><span class="o">.</span><span class="na">replaceAll</span><span class="o">(</span><span class="s">/[12]$/</span><span class="o">,</span><span class="s1">&#39;&#39;</span><span class="o">)</span>
<a id="__codelineno-27-24" name="__codelineno-27-24"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-27-25" name="__codelineno-27-25"></a><span class="s2">    gatk3 -T BaseRecalibrator \</span>
<a id="__codelineno-27-26" name="__codelineno-27-26"></a><span class="s2">          --default_platform illumina \</span>
<a id="__codelineno-27-27" name="__codelineno-27-27"></a><span class="s2">          -cov ReadGroupCovariate \</span>
<a id="__codelineno-27-28" name="__codelineno-27-28"></a><span class="s2">          -cov QualityScoreCovariate \</span>
<a id="__codelineno-27-29" name="__codelineno-27-29"></a><span class="s2">          -cov CycleCovariate \</span>
<a id="__codelineno-27-30" name="__codelineno-27-30"></a><span class="s2">          -knownSites ${prepared_variants_file} \</span>
<a id="__codelineno-27-31" name="__codelineno-27-31"></a><span class="s2">          -cov ContextCovariate \</span>
<a id="__codelineno-27-32" name="__codelineno-27-32"></a><span class="s2">          -R ${genome} -I ${bam} \</span>
<a id="__codelineno-27-33" name="__codelineno-27-33"></a><span class="s2">          --downsampling_type NONE \</span>
<a id="__codelineno-27-34" name="__codelineno-27-34"></a><span class="s2">          -nct ${task.cpus} \</span>
<a id="__codelineno-27-35" name="__codelineno-27-35"></a><span class="s2">          -o final.rnaseq.grp</span>
<a id="__codelineno-27-36" name="__codelineno-27-36"></a>
<a id="__codelineno-27-37" name="__codelineno-27-37"></a><span class="s2">    gatk3 -T PrintReads \</span>
<a id="__codelineno-27-38" name="__codelineno-27-38"></a><span class="s2">          -R ${genome} -I ${bam} \</span>
<a id="__codelineno-27-39" name="__codelineno-27-39"></a><span class="s2">          -BQSR final.rnaseq.grp \</span>
<a id="__codelineno-27-40" name="__codelineno-27-40"></a><span class="s2">          -nct ${task.cpus} \</span>
<a id="__codelineno-27-41" name="__codelineno-27-41"></a><span class="s2">          -o final.bam</span>
<a id="__codelineno-27-42" name="__codelineno-27-42"></a>
<a id="__codelineno-27-43" name="__codelineno-27-43"></a><span class="s2">    (samtools view -H final.bam; samtools view final.bam | \</span>
<a id="__codelineno-27-44" name="__codelineno-27-44"></a><span class="s2">    grep -w &#39;NH:i:1&#39;) | samtools view -Sb -  &gt; ${replicateId}.final.uniq.bam</span>
<a id="__codelineno-27-45" name="__codelineno-27-45"></a>
<a id="__codelineno-27-46" name="__codelineno-27-46"></a><span class="s2">    samtools index ${replicateId}.final.uniq.bam</span>
<a id="__codelineno-27-47" name="__codelineno-27-47"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-27-48" name="__codelineno-27-48"></a><span class="o">}</span>
<a id="__codelineno-27-49" name="__codelineno-27-49"></a>
<a id="__codelineno-27-50" name="__codelineno-27-50"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-27-51" name="__codelineno-27-51"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-27-52" name="__codelineno-27-52"></a>
<a id="__codelineno-27-53" name="__codelineno-27-53"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-27-54" name="__codelineno-27-54"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-27-55" name="__codelineno-27-55"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-27-56" name="__codelineno-27-56"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-27-57" name="__codelineno-27-57"></a>
<a id="__codelineno-27-58" name="__codelineno-27-58"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-27-59" name="__codelineno-27-59"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-27-60" name="__codelineno-27-60"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-27-61" name="__codelineno-27-61"></a>
<a id="__codelineno-27-62" name="__codelineno-27-62"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-27-63" name="__codelineno-27-63"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-27-64" name="__codelineno-27-64"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-27-65" name="__codelineno-27-65"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-27-66" name="__codelineno-27-66"></a>
<a id="__codelineno-27-67" name="__codelineno-27-67"></a><span class="hll"><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
</span><a id="__codelineno-27-68" name="__codelineno-27-68"></a><span class="hll"><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-27-69" name="__codelineno-27-69"></a><span class="hll"><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-27-70" name="__codelineno-27-70"></a><span class="hll"><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-27-71" name="__codelineno-27-71"></a><span class="hll"><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
</span><a id="__codelineno-27-72" name="__codelineno-27-72"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
</div>
<p>Now we are ready to perform the variant calling with GATK.</p>
<h2 id="process-5-gatk-variant-calling">Process 5: GATK Variant Calling<a class="headerlink" href="#process-5-gatk-variant-calling" title="Permanent link">&para;</a></h2>
<p>This steps call variants with GATK HaplotypeCaller. You can find details of the specific GATK HaplotypeCaller parameters <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">here</a>.</p>
<p>The next process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>rnaseq_call_variants</code></li>
<li><strong>Command</strong>: variant calling of each sample using GATK</li>
<li><strong>Input</strong>:</li>
<li>the genome fasta file</li>
<li>the genome index made with samtools</li>
<li>the genome dictionary made with picard</li>
<li>a tuple containing replicate id, aligned bam file and aligned bam file index from process 4</li>
<li><strong>Output</strong>: a tuple containing the sample id the resulting variant calling file (vcf)</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #9</p>
<p>Copy the code below and paste it at the end of <code>main.nf</code>, removing the previous workflow block. Be careful not to accidentally have multiple workflow blocks.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Note that in process 4, we used the sampleID (not replicateID) as the first element of the tuple in the output. Now we combine the replicates by grouping them on the sample ID. It follows from this that process 4 is run one time per replicate and process 5 is run one time per sample.</p>
</div>
<p>Your aim is to replace the <code>BLANK</code> placeholder with the the correct process call.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-28-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-28-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-28-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-28-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-28-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-28-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-28-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-28-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-28-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-28-10">10</a></span>
<span class="normal"><a href="#__codelineno-28-11">11</a></span>
<span class="normal"><a href="#__codelineno-28-12">12</a></span>
<span class="normal"><a href="#__codelineno-28-13">13</a></span>
<span class="normal"><a href="#__codelineno-28-14">14</a></span>
<span class="normal"><a href="#__codelineno-28-15">15</a></span>
<span class="normal"><a href="#__codelineno-28-16">16</a></span>
<span class="normal"><a href="#__codelineno-28-17">17</a></span>
<span class="normal"><a href="#__codelineno-28-18">18</a></span>
<span class="normal"><a href="#__codelineno-28-19">19</a></span>
<span class="normal"><a href="#__codelineno-28-20">20</a></span>
<span class="normal"><a href="#__codelineno-28-21">21</a></span>
<span class="normal"><a href="#__codelineno-28-22">22</a></span>
<span class="normal"><a href="#__codelineno-28-23">23</a></span>
<span class="normal"><a href="#__codelineno-28-24">24</a></span>
<span class="normal"><a href="#__codelineno-28-25">25</a></span>
<span class="normal"><a href="#__codelineno-28-26">26</a></span>
<span class="normal"><a href="#__codelineno-28-27">27</a></span>
<span class="normal"><a href="#__codelineno-28-28">28</a></span>
<span class="normal"><a href="#__codelineno-28-29">29</a></span>
<span class="normal"><a href="#__codelineno-28-30">30</a></span>
<span class="normal"><a href="#__codelineno-28-31">31</a></span>
<span class="normal"><a href="#__codelineno-28-32">32</a></span>
<span class="normal"><a href="#__codelineno-28-33">33</a></span>
<span class="normal"><a href="#__codelineno-28-34">34</a></span>
<span class="normal"><a href="#__codelineno-28-35">35</a></span>
<span class="normal"><a href="#__codelineno-28-36">36</a></span>
<span class="normal"><a href="#__codelineno-28-37">37</a></span>
<span class="normal"><a href="#__codelineno-28-38">38</a></span>
<span class="normal"><a href="#__codelineno-28-39">39</a></span>
<span class="normal"><a href="#__codelineno-28-40">40</a></span>
<span class="normal"><a href="#__codelineno-28-41">41</a></span>
<span class="normal"><a href="#__codelineno-28-42">42</a></span>
<span class="normal"><a href="#__codelineno-28-43">43</a></span>
<span class="normal"><a href="#__codelineno-28-44">44</a></span>
<span class="normal"><a href="#__codelineno-28-45">45</a></span>
<span class="normal"><a href="#__codelineno-28-46">46</a></span>
<span class="normal"><a href="#__codelineno-28-47">47</a></span>
<span class="normal"><a href="#__codelineno-28-48">48</a></span>
<span class="normal"><a href="#__codelineno-28-49">49</a></span>
<span class="normal"><a href="#__codelineno-28-50">50</a></span>
<span class="normal"><a href="#__codelineno-28-51">51</a></span>
<span class="normal"><a href="#__codelineno-28-52">52</a></span>
<span class="normal"><a href="#__codelineno-28-53">53</a></span>
<span class="normal"><a href="#__codelineno-28-54">54</a></span>
<span class="normal"><a href="#__codelineno-28-55">55</a></span>
<span class="normal"><a href="#__codelineno-28-56">56</a></span>
<span class="normal"><a href="#__codelineno-28-57">57</a></span>
<span class="normal"><a href="#__codelineno-28-58">58</a></span>
<span class="normal"><a href="#__codelineno-28-59">59</a></span>
<span class="normal"><a href="#__codelineno-28-60">60</a></span>
<span class="normal"><a href="#__codelineno-28-61">61</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1"></a><span class="cm">/*</span>
<a id="__codelineno-28-2" name="__codelineno-28-2"></a><span class="cm"> * Process 5: GATK Variant Calling</span>
<a id="__codelineno-28-3" name="__codelineno-28-3"></a><span class="cm"> */</span>
<a id="__codelineno-28-4" name="__codelineno-28-4"></a>
<a id="__codelineno-28-5" name="__codelineno-28-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_call_variants</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-28-6" name="__codelineno-28-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/broadinstitute/gotc-prod-gatk:1.0.0-4.1.8.0-1626439571&#39;</span>
<a id="__codelineno-28-7" name="__codelineno-28-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-28-8" name="__codelineno-28-8"></a>
<a id="__codelineno-28-9" name="__codelineno-28-9"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-28-10" name="__codelineno-28-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span><span class="w"> </span><span class="c1">// (2)!</span>
<a id="__codelineno-28-11" name="__codelineno-28-11"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span><span class="w"> </span><span class="c1">// (3)!</span>
<a id="__codelineno-28-12" name="__codelineno-28-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">dict</span><span class="w">  </span><span class="c1">// (4)!</span>
<a id="__codelineno-28-13" name="__codelineno-28-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span><span class="w"> </span><span class="c1">// (5)!</span>
<a id="__codelineno-28-14" name="__codelineno-28-14"></a>
<a id="__codelineno-28-15" name="__codelineno-28-15"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-28-16" name="__codelineno-28-16"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">)</span><span class="w"> </span><span class="c1">// (6)!</span>
<a id="__codelineno-28-17" name="__codelineno-28-17"></a>
<a id="__codelineno-28-18" name="__codelineno-28-18"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-28-19" name="__codelineno-28-19"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-28-20" name="__codelineno-28-20"></a><span class="s2">    echo &quot;${bam.join(&#39;\n&#39;)}&quot; &gt; bam.list</span>
<a id="__codelineno-28-21" name="__codelineno-28-21"></a>
<a id="__codelineno-28-22" name="__codelineno-28-22"></a><span class="s2">    java -jar /usr/gitc/GATK35.jar -T HaplotypeCaller \</span>
<a id="__codelineno-28-23" name="__codelineno-28-23"></a><span class="s2">                                   -R ${genome} -I bam.list \</span>
<a id="__codelineno-28-24" name="__codelineno-28-24"></a><span class="s2">                                   -dontUseSoftClippedBases \</span>
<a id="__codelineno-28-25" name="__codelineno-28-25"></a><span class="s2">                                   -stand_call_conf 20.0 \</span>
<a id="__codelineno-28-26" name="__codelineno-28-26"></a><span class="s2">                                   -o output.gatk.vcf.gz</span>
<a id="__codelineno-28-27" name="__codelineno-28-27"></a>
<a id="__codelineno-28-28" name="__codelineno-28-28"></a><span class="s2">    java -jar /usr/gitc/GATK35.jar -T VariantFiltration \</span>
<a id="__codelineno-28-29" name="__codelineno-28-29"></a><span class="s2">                                   -R ${genome} -V output.gatk.vcf.gz \</span>
<a id="__codelineno-28-30" name="__codelineno-28-30"></a><span class="s2">                                   -window 35 -cluster 3 \</span>
<a id="__codelineno-28-31" name="__codelineno-28-31"></a><span class="s2">                                   -filterName FS -filter &quot;FS &gt; 30.0&quot; \</span>
<a id="__codelineno-28-32" name="__codelineno-28-32"></a><span class="s2">                                   -filterName QD -filter &quot;QD &lt; 2.0&quot; \</span>
<a id="__codelineno-28-33" name="__codelineno-28-33"></a><span class="s2">                                   -o final.vcf</span>
<a id="__codelineno-28-34" name="__codelineno-28-34"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-28-35" name="__codelineno-28-35"></a><span class="o">}</span>
<a id="__codelineno-28-36" name="__codelineno-28-36"></a>
<a id="__codelineno-28-37" name="__codelineno-28-37"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-28-38" name="__codelineno-28-38"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-28-39" name="__codelineno-28-39"></a>
<a id="__codelineno-28-40" name="__codelineno-28-40"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-28-41" name="__codelineno-28-41"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-28-42" name="__codelineno-28-42"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-28-43" name="__codelineno-28-43"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-28-44" name="__codelineno-28-44"></a>
<a id="__codelineno-28-45" name="__codelineno-28-45"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-28-46" name="__codelineno-28-46"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-28-47" name="__codelineno-28-47"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-28-48" name="__codelineno-28-48"></a>
<a id="__codelineno-28-49" name="__codelineno-28-49"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-28-50" name="__codelineno-28-50"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-28-51" name="__codelineno-28-51"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-28-52" name="__codelineno-28-52"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-28-53" name="__codelineno-28-53"></a>
<a id="__codelineno-28-54" name="__codelineno-28-54"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-28-55" name="__codelineno-28-55"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-28-56" name="__codelineno-28-56"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-28-57" name="__codelineno-28-57"></a><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-28-58" name="__codelineno-28-58"></a><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-28-59" name="__codelineno-28-59"></a>
<a id="__codelineno-28-60" name="__codelineno-28-60"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-28-61" name="__codelineno-28-61"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li><a href="https://www.nextflow.io/docs/latest/process.html#tag"><code>tag</code></a> line with the using the sample id as the tag.</li>
<li>The genome fasta file.</li>
<li>The genome index in the output channel from the <code>prepare_genome_samtools</code> process.</li>
<li>The genome dictionary in the output channel from the <code>prepare_genome_picard</code> process.</li>
<li>The tuples grouped by sampleID in the output channel from the <code>rnaseq_gatk_recalibrate</code> process.</li>
<li>The tuple containing the sample ID and final VCF file.</li>
</ol>
<p>Broken down, here is what the script is doing:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">bam</span><span class="p">.join(</span><span class="s1">&#39;\n&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>bam.list<span class="w"> </span><span class="c1"># (1)!</span>
<a id="__codelineno-29-2" name="__codelineno-29-2" href="#__codelineno-29-2"></a>
<a id="__codelineno-29-3" name="__codelineno-29-3" href="#__codelineno-29-3"></a>java<span class="w"> </span>-jar<span class="w"> </span>/usr/gitc/GATK35.jar<span class="w"> </span>-T<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\ </span><span class="c1"># (2)!</span>
<a id="__codelineno-29-4" name="__codelineno-29-4" href="#__codelineno-29-4"></a><span class="w">                               </span>-R<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span>-I<span class="w"> </span>bam.list<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-5" name="__codelineno-29-5" href="#__codelineno-29-5"></a><span class="w">                               </span>-dontUseSoftClippedBases<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-6" name="__codelineno-29-6" href="#__codelineno-29-6"></a><span class="w">                               </span>-stand_call_conf<span class="w"> </span><span class="m">20</span>.0<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-7" name="__codelineno-29-7" href="#__codelineno-29-7"></a><span class="w">                               </span>-o<span class="w"> </span>output.gatk.vcf.gz
<a id="__codelineno-29-8" name="__codelineno-29-8" href="#__codelineno-29-8"></a>
<a id="__codelineno-29-9" name="__codelineno-29-9" href="#__codelineno-29-9"></a>java<span class="w"> </span>-jar<span class="w"> </span>/usr/gitc/GATK35.jar<span class="w"> </span>-T<span class="w"> </span>VariantFiltration<span class="w"> </span><span class="se">\ </span><span class="c1"># (3)!</span>
<a id="__codelineno-29-10" name="__codelineno-29-10" href="#__codelineno-29-10"></a><span class="w">                               </span>-R<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span>-V<span class="w"> </span>output.gatk.vcf.gz<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-11" name="__codelineno-29-11" href="#__codelineno-29-11"></a><span class="w">                               </span>-window<span class="w"> </span><span class="m">35</span><span class="w"> </span>-cluster<span class="w"> </span><span class="m">3</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-12" name="__codelineno-29-12" href="#__codelineno-29-12"></a><span class="w">                               </span>-filterName<span class="w"> </span>FS<span class="w"> </span>-filter<span class="w"> </span><span class="s2">&quot;FS &gt; 30.0&quot;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-13" name="__codelineno-29-13" href="#__codelineno-29-13"></a><span class="w">                               </span>-filterName<span class="w"> </span>QD<span class="w"> </span>-filter<span class="w"> </span><span class="s2">&quot;QD &lt; 2.0&quot;</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-29-14" name="__codelineno-29-14" href="#__codelineno-29-14"></a><span class="w">                               </span>-o<span class="w"> </span>final.vcf
</code></pre></div>
<ol>
<li>Create a file containing a list of BAM files</li>
<li>Variant calling step</li>
<li>Variant filtering step</li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-30-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-30-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-30-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-30-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-30-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-30-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-30-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-30-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-30-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-30-10">10</a></span>
<span class="normal"><a href="#__codelineno-30-11">11</a></span>
<span class="normal"><a href="#__codelineno-30-12">12</a></span>
<span class="normal"><a href="#__codelineno-30-13">13</a></span>
<span class="normal"><a href="#__codelineno-30-14">14</a></span>
<span class="normal"><a href="#__codelineno-30-15">15</a></span>
<span class="normal"><a href="#__codelineno-30-16">16</a></span>
<span class="normal"><a href="#__codelineno-30-17">17</a></span>
<span class="normal"><a href="#__codelineno-30-18">18</a></span>
<span class="normal"><a href="#__codelineno-30-19">19</a></span>
<span class="normal"><a href="#__codelineno-30-20">20</a></span>
<span class="normal"><a href="#__codelineno-30-21">21</a></span>
<span class="normal"><a href="#__codelineno-30-22">22</a></span>
<span class="normal"><a href="#__codelineno-30-23">23</a></span>
<span class="normal"><a href="#__codelineno-30-24">24</a></span>
<span class="normal"><a href="#__codelineno-30-25">25</a></span>
<span class="normal"><a href="#__codelineno-30-26">26</a></span>
<span class="normal"><a href="#__codelineno-30-27">27</a></span>
<span class="normal"><a href="#__codelineno-30-28">28</a></span>
<span class="normal"><a href="#__codelineno-30-29">29</a></span>
<span class="normal"><a href="#__codelineno-30-30">30</a></span>
<span class="normal"><a href="#__codelineno-30-31">31</a></span>
<span class="normal"><a href="#__codelineno-30-32">32</a></span>
<span class="normal"><a href="#__codelineno-30-33">33</a></span>
<span class="normal"><a href="#__codelineno-30-34">34</a></span>
<span class="normal"><a href="#__codelineno-30-35">35</a></span>
<span class="normal"><a href="#__codelineno-30-36">36</a></span>
<span class="normal"><a href="#__codelineno-30-37">37</a></span>
<span class="normal"><a href="#__codelineno-30-38">38</a></span>
<span class="normal"><a href="#__codelineno-30-39">39</a></span>
<span class="normal"><a href="#__codelineno-30-40">40</a></span>
<span class="normal"><a href="#__codelineno-30-41">41</a></span>
<span class="normal"><a href="#__codelineno-30-42">42</a></span>
<span class="normal"><a href="#__codelineno-30-43">43</a></span>
<span class="normal"><a href="#__codelineno-30-44">44</a></span>
<span class="normal"><a href="#__codelineno-30-45">45</a></span>
<span class="normal"><a href="#__codelineno-30-46">46</a></span>
<span class="normal"><a href="#__codelineno-30-47">47</a></span>
<span class="normal"><a href="#__codelineno-30-48">48</a></span>
<span class="normal"><a href="#__codelineno-30-49">49</a></span>
<span class="normal"><a href="#__codelineno-30-50">50</a></span>
<span class="normal"><a href="#__codelineno-30-51">51</a></span>
<span class="normal"><a href="#__codelineno-30-52">52</a></span>
<span class="normal"><a href="#__codelineno-30-53">53</a></span>
<span class="normal"><a href="#__codelineno-30-54">54</a></span>
<span class="normal"><a href="#__codelineno-30-55">55</a></span>
<span class="normal"><a href="#__codelineno-30-56">56</a></span>
<span class="normal"><a href="#__codelineno-30-57">57</a></span>
<span class="normal"><a href="#__codelineno-30-58">58</a></span>
<span class="normal"><a href="#__codelineno-30-59">59</a></span>
<span class="normal"><a href="#__codelineno-30-60">60</a></span>
<span class="normal"><a href="#__codelineno-30-61">61</a></span>
<span class="normal"><a href="#__codelineno-30-62">62</a></span>
<span class="normal"><a href="#__codelineno-30-63">63</a></span>
<span class="normal"><a href="#__codelineno-30-64">64</a></span>
<span class="normal"><a href="#__codelineno-30-65">65</a></span>
<span class="normal"><a href="#__codelineno-30-66">66</a></span>
<span class="normal"><a href="#__codelineno-30-67">67</a></span>
<span class="normal"><a href="#__codelineno-30-68">68</a></span>
<span class="normal"><a href="#__codelineno-30-69">69</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-30-1" name="__codelineno-30-1"></a><span class="cm">/*</span>
<a id="__codelineno-30-2" name="__codelineno-30-2"></a><span class="cm"> * Process 5: GATK Variant Calling</span>
<a id="__codelineno-30-3" name="__codelineno-30-3"></a><span class="cm"> */</span>
<a id="__codelineno-30-4" name="__codelineno-30-4"></a>
<a id="__codelineno-30-5" name="__codelineno-30-5"></a><span class="n">process</span><span class="w"> </span><span class="n">rnaseq_call_variants</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-30-6" name="__codelineno-30-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/broadinstitute/gotc-prod-gatk:1.0.0-4.1.8.0-1626439571&#39;</span>
<a id="__codelineno-30-7" name="__codelineno-30-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span>
<a id="__codelineno-30-8" name="__codelineno-30-8"></a>
<a id="__codelineno-30-9" name="__codelineno-30-9"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-30-10" name="__codelineno-30-10"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
<a id="__codelineno-30-11" name="__codelineno-30-11"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span>
<a id="__codelineno-30-12" name="__codelineno-30-12"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">dict</span>
<a id="__codelineno-30-13" name="__codelineno-30-13"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span>
<a id="__codelineno-30-14" name="__codelineno-30-14"></a>
<a id="__codelineno-30-15" name="__codelineno-30-15"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-30-16" name="__codelineno-30-16"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">)</span>
<a id="__codelineno-30-17" name="__codelineno-30-17"></a>
<a id="__codelineno-30-18" name="__codelineno-30-18"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-30-19" name="__codelineno-30-19"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-30-20" name="__codelineno-30-20"></a><span class="s2">    echo &quot;${bam.join(&#39;\n&#39;)}&quot; &gt; bam.list</span>
<a id="__codelineno-30-21" name="__codelineno-30-21"></a>
<a id="__codelineno-30-22" name="__codelineno-30-22"></a><span class="s2">    java -jar /usr/gitc/GATK35.jar -T HaplotypeCaller \</span>
<a id="__codelineno-30-23" name="__codelineno-30-23"></a><span class="s2">                                   -R ${genome} -I bam.list \</span>
<a id="__codelineno-30-24" name="__codelineno-30-24"></a><span class="s2">                                   -dontUseSoftClippedBases \</span>
<a id="__codelineno-30-25" name="__codelineno-30-25"></a><span class="s2">                                   -stand_call_conf 20.0 \</span>
<a id="__codelineno-30-26" name="__codelineno-30-26"></a><span class="s2">                                   -o output.gatk.vcf.gz</span>
<a id="__codelineno-30-27" name="__codelineno-30-27"></a>
<a id="__codelineno-30-28" name="__codelineno-30-28"></a><span class="s2">    java -jar /usr/gitc/GATK35.jar -T VariantFiltration \</span>
<a id="__codelineno-30-29" name="__codelineno-30-29"></a><span class="s2">                                   -R ${genome} -V output.gatk.vcf.gz \</span>
<a id="__codelineno-30-30" name="__codelineno-30-30"></a><span class="s2">                                   -window 35 -cluster 3 \</span>
<a id="__codelineno-30-31" name="__codelineno-30-31"></a><span class="s2">                                   -filterName FS -filter &quot;FS &gt; 30.0&quot; \</span>
<a id="__codelineno-30-32" name="__codelineno-30-32"></a><span class="s2">                                   -filterName QD -filter &quot;QD &lt; 2.0&quot; \</span>
<a id="__codelineno-30-33" name="__codelineno-30-33"></a><span class="s2">                                   -o final.vcf</span>
<a id="__codelineno-30-34" name="__codelineno-30-34"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-30-35" name="__codelineno-30-35"></a><span class="o">}</span>
<a id="__codelineno-30-36" name="__codelineno-30-36"></a>
<a id="__codelineno-30-37" name="__codelineno-30-37"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-30-38" name="__codelineno-30-38"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-30-39" name="__codelineno-30-39"></a>
<a id="__codelineno-30-40" name="__codelineno-30-40"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-30-41" name="__codelineno-30-41"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-30-42" name="__codelineno-30-42"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-30-43" name="__codelineno-30-43"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-30-44" name="__codelineno-30-44"></a>
<a id="__codelineno-30-45" name="__codelineno-30-45"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-30-46" name="__codelineno-30-46"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-30-47" name="__codelineno-30-47"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-30-48" name="__codelineno-30-48"></a>
<a id="__codelineno-30-49" name="__codelineno-30-49"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-30-50" name="__codelineno-30-50"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-30-51" name="__codelineno-30-51"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-30-52" name="__codelineno-30-52"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-30-53" name="__codelineno-30-53"></a>
<a id="__codelineno-30-54" name="__codelineno-30-54"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-30-55" name="__codelineno-30-55"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-30-56" name="__codelineno-30-56"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-30-57" name="__codelineno-30-57"></a><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-30-58" name="__codelineno-30-58"></a><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-30-59" name="__codelineno-30-59"></a>
<a id="__codelineno-30-60" name="__codelineno-30-60"></a><span class="hll"><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span>
</span><a id="__codelineno-30-61" name="__codelineno-30-61"></a><span class="hll"><span class="w">        </span><span class="o">.</span><span class="na">out</span>
</span><a id="__codelineno-30-62" name="__codelineno-30-62"></a><span class="hll"><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">groupTuple</span>
</span><a id="__codelineno-30-63" name="__codelineno-30-63"></a><span class="hll"><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">recalibrated_samples</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// (1)!</span>
</span><a id="__codelineno-30-64" name="__codelineno-30-64"></a><span class="hll">
</span><a id="__codelineno-30-65" name="__codelineno-30-65"></a><span class="hll"><span class="w">    </span><span class="n">rnaseq_call_variants</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
</span><a id="__codelineno-30-66" name="__codelineno-30-66"></a><span class="hll"><span class="w">                         </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-30-67" name="__codelineno-30-67"></a><span class="hll"><span class="w">                         </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-30-68" name="__codelineno-30-68"></a><span class="hll"><span class="w">                         </span><span class="n">recalibrated_samples</span><span class="o">)</span>
</span><a id="__codelineno-30-69" name="__codelineno-30-69"></a><span class="hll"><span class="o">}</span>
</span></code></pre></div></td></tr></table></div>
<ol>
<li>New channel to aggregate the <code>bam</code> files from different replicates into sample level.</li>
</ol>
</details>
</div>
<h2 id="processes-6a-and-6b-ase-rna-editing">Processes 6A and 6B: ASE &amp; RNA Editing<a class="headerlink" href="#processes-6a-and-6b-ase-rna-editing" title="Permanent link">&para;</a></h2>
<p>In the final steps we will create processes for Allele-Specific Expression and RNA Editing Analysis.</p>
<p>We must process the VCF result to prepare variants file for allele specific expression (ASE) analysis. We will implement both processes together.</p>
<p>You should implement two processes having the following structure:</p>
<ul>
<li><em>1st process</em></li>
<li><strong>Name</strong>: <code>post_process_vcf</code></li>
<li><strong>Command</strong>: post-process the variant calling file (vcf) of each sample</li>
<li><strong>Input</strong>:<ul>
<li>tuple containing the sample ID and vcf file</li>
<li>a tuple containing the filtered/recoded VCF file and the tab index (TBI) file from process 1D</li>
</ul>
</li>
<li><strong>Output</strong>: a tuple containing the sample id, the variant calling file (vcf) and a file containing common SNPs</li>
<li><em>2nd process</em></li>
<li><strong>Name</strong>: <code>prepare_vcf_for_ase</code></li>
<li><strong>Command</strong>: prepare the VCF for allele specific expression (ASE) and generate a figure in R.</li>
<li><strong>Input</strong>: a tuple containing the sample id, the variant calling file (vcf) and a file containing common SNPs</li>
<li><strong>Output</strong>:<ul>
<li>a tuple containing the sample ID and known SNPs in the sample for ASE</li>
<li>a figure of the SNPs generated in R as a PDF file</li>
</ul>
</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #10</p>
<p>Here we introduce the <code>publishDir</code> directive. This allows us to specify a location for the outputs of the process. See <a href="https://www.nextflow.io/docs/latest/process.html#publishdir">here</a> for more details.</p>
<p>You must have the output of process 6A become the input of process 6B.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-31-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-31-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-31-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-31-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-31-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-31-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-31-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-31-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-31-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-31-10">10</a></span>
<span class="normal"><a href="#__codelineno-31-11">11</a></span>
<span class="normal"><a href="#__codelineno-31-12">12</a></span>
<span class="normal"><a href="#__codelineno-31-13">13</a></span>
<span class="normal"><a href="#__codelineno-31-14">14</a></span>
<span class="normal"><a href="#__codelineno-31-15">15</a></span>
<span class="normal"><a href="#__codelineno-31-16">16</a></span>
<span class="normal"><a href="#__codelineno-31-17">17</a></span>
<span class="normal"><a href="#__codelineno-31-18">18</a></span>
<span class="normal"><a href="#__codelineno-31-19">19</a></span>
<span class="normal"><a href="#__codelineno-31-20">20</a></span>
<span class="normal"><a href="#__codelineno-31-21">21</a></span>
<span class="normal"><a href="#__codelineno-31-22">22</a></span>
<span class="normal"><a href="#__codelineno-31-23">23</a></span>
<span class="normal"><a href="#__codelineno-31-24">24</a></span>
<span class="normal"><a href="#__codelineno-31-25">25</a></span>
<span class="normal"><a href="#__codelineno-31-26">26</a></span>
<span class="normal"><a href="#__codelineno-31-27">27</a></span>
<span class="normal"><a href="#__codelineno-31-28">28</a></span>
<span class="normal"><a href="#__codelineno-31-29">29</a></span>
<span class="normal"><a href="#__codelineno-31-30">30</a></span>
<span class="normal"><a href="#__codelineno-31-31">31</a></span>
<span class="normal"><a href="#__codelineno-31-32">32</a></span>
<span class="normal"><a href="#__codelineno-31-33">33</a></span>
<span class="normal"><a href="#__codelineno-31-34">34</a></span>
<span class="normal"><a href="#__codelineno-31-35">35</a></span>
<span class="normal"><a href="#__codelineno-31-36">36</a></span>
<span class="normal"><a href="#__codelineno-31-37">37</a></span>
<span class="normal"><a href="#__codelineno-31-38">38</a></span>
<span class="normal"><a href="#__codelineno-31-39">39</a></span>
<span class="normal"><a href="#__codelineno-31-40">40</a></span>
<span class="normal"><a href="#__codelineno-31-41">41</a></span>
<span class="normal"><a href="#__codelineno-31-42">42</a></span>
<span class="normal"><a href="#__codelineno-31-43">43</a></span>
<span class="normal"><a href="#__codelineno-31-44">44</a></span>
<span class="normal"><a href="#__codelineno-31-45">45</a></span>
<span class="normal"><a href="#__codelineno-31-46">46</a></span>
<span class="normal"><a href="#__codelineno-31-47">47</a></span>
<span class="normal"><a href="#__codelineno-31-48">48</a></span>
<span class="normal"><a href="#__codelineno-31-49">49</a></span>
<span class="normal"><a href="#__codelineno-31-50">50</a></span>
<span class="normal"><a href="#__codelineno-31-51">51</a></span>
<span class="normal"><a href="#__codelineno-31-52">52</a></span>
<span class="normal"><a href="#__codelineno-31-53">53</a></span>
<span class="normal"><a href="#__codelineno-31-54">54</a></span>
<span class="normal"><a href="#__codelineno-31-55">55</a></span>
<span class="normal"><a href="#__codelineno-31-56">56</a></span>
<span class="normal"><a href="#__codelineno-31-57">57</a></span>
<span class="normal"><a href="#__codelineno-31-58">58</a></span>
<span class="normal"><a href="#__codelineno-31-59">59</a></span>
<span class="normal"><a href="#__codelineno-31-60">60</a></span>
<span class="normal"><a href="#__codelineno-31-61">61</a></span>
<span class="normal"><a href="#__codelineno-31-62">62</a></span>
<span class="normal"><a href="#__codelineno-31-63">63</a></span>
<span class="normal"><a href="#__codelineno-31-64">64</a></span>
<span class="normal"><a href="#__codelineno-31-65">65</a></span>
<span class="normal"><a href="#__codelineno-31-66">66</a></span>
<span class="normal"><a href="#__codelineno-31-67">67</a></span>
<span class="normal"><a href="#__codelineno-31-68">68</a></span>
<span class="normal"><a href="#__codelineno-31-69">69</a></span>
<span class="normal"><a href="#__codelineno-31-70">70</a></span>
<span class="normal"><a href="#__codelineno-31-71">71</a></span>
<span class="normal"><a href="#__codelineno-31-72">72</a></span>
<span class="normal"><a href="#__codelineno-31-73">73</a></span>
<span class="normal"><a href="#__codelineno-31-74">74</a></span>
<span class="normal"><a href="#__codelineno-31-75">75</a></span>
<span class="normal"><a href="#__codelineno-31-76">76</a></span>
<span class="normal"><a href="#__codelineno-31-77">77</a></span>
<span class="normal"><a href="#__codelineno-31-78">78</a></span>
<span class="normal"><a href="#__codelineno-31-79">79</a></span>
<span class="normal"><a href="#__codelineno-31-80">80</a></span>
<span class="normal"><a href="#__codelineno-31-81">81</a></span>
<span class="normal"><a href="#__codelineno-31-82">82</a></span>
<span class="normal"><a href="#__codelineno-31-83">83</a></span>
<span class="normal"><a href="#__codelineno-31-84">84</a></span>
<span class="normal"><a href="#__codelineno-31-85">85</a></span>
<span class="normal"><a href="#__codelineno-31-86">86</a></span>
<span class="normal"><a href="#__codelineno-31-87">87</a></span>
<span class="normal"><a href="#__codelineno-31-88">88</a></span>
<span class="normal"><a href="#__codelineno-31-89">89</a></span>
<span class="normal"><a href="#__codelineno-31-90">90</a></span>
<span class="normal"><a href="#__codelineno-31-91">91</a></span>
<span class="normal"><a href="#__codelineno-31-92">92</a></span>
<span class="normal"><a href="#__codelineno-31-93">93</a></span>
<span class="normal"><a href="#__codelineno-31-94">94</a></span>
<span class="normal"><a href="#__codelineno-31-95">95</a></span>
<span class="normal"><a href="#__codelineno-31-96">96</a></span>
<span class="normal"><a href="#__codelineno-31-97">97</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-31-1" name="__codelineno-31-1"></a><span class="cm">/*</span>
<a id="__codelineno-31-2" name="__codelineno-31-2"></a><span class="cm"> * Processes 6: ASE &amp; RNA Editing</span>
<a id="__codelineno-31-3" name="__codelineno-31-3"></a><span class="cm"> */</span>
<a id="__codelineno-31-4" name="__codelineno-31-4"></a>
<a id="__codelineno-31-5" name="__codelineno-31-5"></a><span class="n">process</span><span class="w"> </span><span class="n">post_process_vcf</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-31-6" name="__codelineno-31-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-b9358559e3ae3b9d7d8dbf1f401ae1fcaf757de3:ac05763cf181a5070c2fdb9bb5461f8d08f7b93b-0&#39;</span>
<a id="__codelineno-31-7" name="__codelineno-31-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span>
<a id="__codelineno-31-8" name="__codelineno-31-8"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="s2">&quot;${params.results}/${sampleId}&quot;</span><span class="w"> </span><span class="c1">// (1)!</span>
<a id="__codelineno-31-9" name="__codelineno-31-9"></a>
<a id="__codelineno-31-10" name="__codelineno-31-10"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-31-11" name="__codelineno-31-11"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">)</span>
<a id="__codelineno-31-12" name="__codelineno-31-12"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;filtered.recode.vcf.gz&#39;</span><span class="o">),</span>
<a id="__codelineno-31-13" name="__codelineno-31-13"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;filtered.recode.vcf.gz.tbi&#39;</span><span class="o">)</span>
<a id="__codelineno-31-14" name="__codelineno-31-14"></a>
<a id="__codelineno-31-15" name="__codelineno-31-15"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-31-16" name="__codelineno-31-16"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span>
<a id="__codelineno-31-17" name="__codelineno-31-17"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">),</span>
<a id="__codelineno-31-18" name="__codelineno-31-18"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;commonSNPs.diff.sites_in_files&#39;</span><span class="o">)</span>
<a id="__codelineno-31-19" name="__codelineno-31-19"></a>
<a id="__codelineno-31-20" name="__codelineno-31-20"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-31-21" name="__codelineno-31-21"></a><span class="w">    </span><span class="s1">&#39;&#39;&#39;</span>
<a id="__codelineno-31-22" name="__codelineno-31-22"></a><span class="s1">    grep -v &#39;#&#39; final.vcf | awk &#39;$7~/PASS/&#39; | perl -ne &#39;chomp($_); \</span>
<a id="__codelineno-31-23" name="__codelineno-31-23"></a><span class="s1">         ($dp)=$_=~/DP\\=(\\d+)\\;/; if($dp&gt;=8){print $_.&quot;\\n&quot;};&#39; \</span>
<a id="__codelineno-31-24" name="__codelineno-31-24"></a><span class="s1">         &gt; result.DP8.vcf</span>
<a id="__codelineno-31-25" name="__codelineno-31-25"></a>
<a id="__codelineno-31-26" name="__codelineno-31-26"></a><span class="s1">    vcftools --vcf result.DP8.vcf --gzdiff filtered.recode.vcf.gz \</span>
<a id="__codelineno-31-27" name="__codelineno-31-27"></a><span class="s1">             --diff-site --out commonSNPs</span>
<a id="__codelineno-31-28" name="__codelineno-31-28"></a><span class="s1">    &#39;&#39;&#39;</span>
<a id="__codelineno-31-29" name="__codelineno-31-29"></a><span class="o">}</span>
<a id="__codelineno-31-30" name="__codelineno-31-30"></a>
<a id="__codelineno-31-31" name="__codelineno-31-31"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_vcf_for_ase</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-31-32" name="__codelineno-31-32"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;cbcrg/callings-with-gatk:latest&#39;</span>
<a id="__codelineno-31-33" name="__codelineno-31-33"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span>
<a id="__codelineno-31-34" name="__codelineno-31-34"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="s2">&quot;${params.results}/${sampleId}&quot;</span>
<a id="__codelineno-31-35" name="__codelineno-31-35"></a>
<a id="__codelineno-31-36" name="__codelineno-31-36"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-31-37" name="__codelineno-31-37"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span>
<a id="__codelineno-31-38" name="__codelineno-31-38"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">),</span>
<a id="__codelineno-31-39" name="__codelineno-31-39"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;commonSNPs.diff.sites_in_files&#39;</span><span class="o">)</span>
<a id="__codelineno-31-40" name="__codelineno-31-40"></a>
<a id="__codelineno-31-41" name="__codelineno-31-41"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-31-42" name="__codelineno-31-42"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;known_snps.vcf&#39;</span><span class="o">),</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">vcf_for_ASE</span>
<a id="__codelineno-31-43" name="__codelineno-31-43"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s1">&#39;AF.histogram.pdf&#39;</span><span class="w">                    </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">gghist_pdfs</span>
<a id="__codelineno-31-44" name="__codelineno-31-44"></a>
<a id="__codelineno-31-45" name="__codelineno-31-45"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-31-46" name="__codelineno-31-46"></a><span class="w">    </span><span class="s1">&#39;&#39;&#39;</span>
<a id="__codelineno-31-47" name="__codelineno-31-47"></a><span class="s1">    awk &#39;BEGIN{OFS=&quot;\t&quot;} $4~/B/{print $1,$2,$3}&#39; \</span>
<a id="__codelineno-31-48" name="__codelineno-31-48"></a><span class="s1">        commonSNPs.diff.sites_in_files  &gt; test.bed</span>
<a id="__codelineno-31-49" name="__codelineno-31-49"></a>
<a id="__codelineno-31-50" name="__codelineno-31-50"></a><span class="s1">    vcftools --vcf final.vcf --bed test.bed --recode --keep-INFO-all \</span>
<a id="__codelineno-31-51" name="__codelineno-31-51"></a><span class="s1">             --stdout &gt; known_snps.vcf</span>
<a id="__codelineno-31-52" name="__codelineno-31-52"></a>
<a id="__codelineno-31-53" name="__codelineno-31-53"></a><span class="s1">    grep -v &#39;#&#39;  known_snps.vcf | awk -F &#39;\\t&#39; &#39;{print $10}&#39; \</span>
<a id="__codelineno-31-54" name="__codelineno-31-54"></a><span class="s1">                | awk -F &#39;:&#39; &#39;{print $2}&#39; | perl -ne &#39;chomp($_); \</span>
<a id="__codelineno-31-55" name="__codelineno-31-55"></a><span class="s1">                @v=split(/\\,/,$_); if($v[0]!=0 ||$v[1] !=0)\</span>
<a id="__codelineno-31-56" name="__codelineno-31-56"></a><span class="s1">                {print  $v[1]/($v[1]+$v[0]).&quot;\\n&quot;; }&#39; | awk &#39;$1!=1&#39; \</span>
<a id="__codelineno-31-57" name="__codelineno-31-57"></a><span class="s1">                &gt;AF.4R</span>
<a id="__codelineno-31-58" name="__codelineno-31-58"></a>
<a id="__codelineno-31-59" name="__codelineno-31-59"></a><span class="s1">    gghist.R -i AF.4R -o AF.histogram.pdf</span>
<a id="__codelineno-31-60" name="__codelineno-31-60"></a><span class="s1">    &#39;&#39;&#39;</span>
<a id="__codelineno-31-61" name="__codelineno-31-61"></a><span class="o">}</span>
<a id="__codelineno-31-62" name="__codelineno-31-62"></a>
<a id="__codelineno-31-63" name="__codelineno-31-63"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-31-64" name="__codelineno-31-64"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-31-65" name="__codelineno-31-65"></a>
<a id="__codelineno-31-66" name="__codelineno-31-66"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-31-67" name="__codelineno-31-67"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-31-68" name="__codelineno-31-68"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-31-69" name="__codelineno-31-69"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-31-70" name="__codelineno-31-70"></a>
<a id="__codelineno-31-71" name="__codelineno-31-71"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-31-72" name="__codelineno-31-72"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-73" name="__codelineno-31-73"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-31-74" name="__codelineno-31-74"></a>
<a id="__codelineno-31-75" name="__codelineno-31-75"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-31-76" name="__codelineno-31-76"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-77" name="__codelineno-31-77"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-78" name="__codelineno-31-78"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-31-79" name="__codelineno-31-79"></a>
<a id="__codelineno-31-80" name="__codelineno-31-80"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-31-81" name="__codelineno-31-81"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-82" name="__codelineno-31-82"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-83" name="__codelineno-31-83"></a><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-84" name="__codelineno-31-84"></a><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-31-85" name="__codelineno-31-85"></a>
<a id="__codelineno-31-86" name="__codelineno-31-86"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span>
<a id="__codelineno-31-87" name="__codelineno-31-87"></a><span class="w">        </span><span class="o">.</span><span class="na">out</span>
<a id="__codelineno-31-88" name="__codelineno-31-88"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">groupTuple</span>
<a id="__codelineno-31-89" name="__codelineno-31-89"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">recalibrated_samples</span><span class="w"> </span><span class="o">}</span>
<a id="__codelineno-31-90" name="__codelineno-31-90"></a>
<a id="__codelineno-31-91" name="__codelineno-31-91"></a><span class="w">    </span><span class="n">rnaseq_call_variants</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-31-92" name="__codelineno-31-92"></a><span class="w">                         </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-93" name="__codelineno-31-93"></a><span class="w">                         </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-31-94" name="__codelineno-31-94"></a><span class="w">                         </span><span class="n">recalibrated_samples</span><span class="o">)</span>
<a id="__codelineno-31-95" name="__codelineno-31-95"></a>
<a id="__codelineno-31-96" name="__codelineno-31-96"></a><span class="hll"><span class="w">    </span><span class="n">BLANK</span>
</span><a id="__codelineno-31-97" name="__codelineno-31-97"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<ol>
<li>The path to the <code>publishDir</code> process directive consists of variables that will be evaluated before saving the files over there</li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-32-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-32-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-32-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-32-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-32-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-32-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-32-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-32-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-32-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-32-10">10</a></span>
<span class="normal"><a href="#__codelineno-32-11">11</a></span>
<span class="normal"><a href="#__codelineno-32-12">12</a></span>
<span class="normal"><a href="#__codelineno-32-13">13</a></span>
<span class="normal"><a href="#__codelineno-32-14">14</a></span>
<span class="normal"><a href="#__codelineno-32-15">15</a></span>
<span class="normal"><a href="#__codelineno-32-16">16</a></span>
<span class="normal"><a href="#__codelineno-32-17">17</a></span>
<span class="normal"><a href="#__codelineno-32-18">18</a></span>
<span class="normal"><a href="#__codelineno-32-19">19</a></span>
<span class="normal"><a href="#__codelineno-32-20">20</a></span>
<span class="normal"><a href="#__codelineno-32-21">21</a></span>
<span class="normal"><a href="#__codelineno-32-22">22</a></span>
<span class="normal"><a href="#__codelineno-32-23">23</a></span>
<span class="normal"><a href="#__codelineno-32-24">24</a></span>
<span class="normal"><a href="#__codelineno-32-25">25</a></span>
<span class="normal"><a href="#__codelineno-32-26">26</a></span>
<span class="normal"><a href="#__codelineno-32-27">27</a></span>
<span class="normal"><a href="#__codelineno-32-28">28</a></span>
<span class="normal"><a href="#__codelineno-32-29">29</a></span>
<span class="normal"><a href="#__codelineno-32-30">30</a></span>
<span class="normal"><a href="#__codelineno-32-31">31</a></span>
<span class="normal"><a href="#__codelineno-32-32">32</a></span>
<span class="normal"><a href="#__codelineno-32-33">33</a></span>
<span class="normal"><a href="#__codelineno-32-34">34</a></span>
<span class="normal"><a href="#__codelineno-32-35">35</a></span>
<span class="normal"><a href="#__codelineno-32-36">36</a></span>
<span class="normal"><a href="#__codelineno-32-37">37</a></span>
<span class="normal"><a href="#__codelineno-32-38">38</a></span>
<span class="normal"><a href="#__codelineno-32-39">39</a></span>
<span class="normal"><a href="#__codelineno-32-40">40</a></span>
<span class="normal"><a href="#__codelineno-32-41">41</a></span>
<span class="normal"><a href="#__codelineno-32-42">42</a></span>
<span class="normal"><a href="#__codelineno-32-43">43</a></span>
<span class="normal"><a href="#__codelineno-32-44">44</a></span>
<span class="normal"><a href="#__codelineno-32-45">45</a></span>
<span class="normal"><a href="#__codelineno-32-46">46</a></span>
<span class="normal"><a href="#__codelineno-32-47">47</a></span>
<span class="normal"><a href="#__codelineno-32-48">48</a></span>
<span class="normal"><a href="#__codelineno-32-49">49</a></span>
<span class="normal"><a href="#__codelineno-32-50">50</a></span>
<span class="normal"><a href="#__codelineno-32-51">51</a></span>
<span class="normal"><a href="#__codelineno-32-52">52</a></span>
<span class="normal"><a href="#__codelineno-32-53">53</a></span>
<span class="normal"><a href="#__codelineno-32-54">54</a></span>
<span class="normal"><a href="#__codelineno-32-55">55</a></span>
<span class="normal"><a href="#__codelineno-32-56">56</a></span>
<span class="normal"><a href="#__codelineno-32-57">57</a></span>
<span class="normal"><a href="#__codelineno-32-58">58</a></span>
<span class="normal"><a href="#__codelineno-32-59">59</a></span>
<span class="normal"><a href="#__codelineno-32-60">60</a></span>
<span class="normal"><a href="#__codelineno-32-61">61</a></span>
<span class="normal"><a href="#__codelineno-32-62">62</a></span>
<span class="normal"><a href="#__codelineno-32-63">63</a></span>
<span class="normal"><a href="#__codelineno-32-64">64</a></span>
<span class="normal"><a href="#__codelineno-32-65">65</a></span>
<span class="normal"><a href="#__codelineno-32-66">66</a></span>
<span class="normal"><a href="#__codelineno-32-67">67</a></span>
<span class="normal"><a href="#__codelineno-32-68">68</a></span>
<span class="normal"><a href="#__codelineno-32-69">69</a></span>
<span class="normal"><a href="#__codelineno-32-70">70</a></span>
<span class="normal"><a href="#__codelineno-32-71">71</a></span>
<span class="normal"><a href="#__codelineno-32-72">72</a></span>
<span class="normal"><a href="#__codelineno-32-73">73</a></span>
<span class="normal"><a href="#__codelineno-32-74">74</a></span>
<span class="normal"><a href="#__codelineno-32-75">75</a></span>
<span class="normal"><a href="#__codelineno-32-76">76</a></span>
<span class="normal"><a href="#__codelineno-32-77">77</a></span>
<span class="normal"><a href="#__codelineno-32-78">78</a></span>
<span class="normal"><a href="#__codelineno-32-79">79</a></span>
<span class="normal"><a href="#__codelineno-32-80">80</a></span>
<span class="normal"><a href="#__codelineno-32-81">81</a></span>
<span class="normal"><a href="#__codelineno-32-82">82</a></span>
<span class="normal"><a href="#__codelineno-32-83">83</a></span>
<span class="normal"><a href="#__codelineno-32-84">84</a></span>
<span class="normal"><a href="#__codelineno-32-85">85</a></span>
<span class="normal"><a href="#__codelineno-32-86">86</a></span>
<span class="normal"><a href="#__codelineno-32-87">87</a></span>
<span class="normal"><a href="#__codelineno-32-88">88</a></span>
<span class="normal"><a href="#__codelineno-32-89">89</a></span>
<span class="normal"><a href="#__codelineno-32-90">90</a></span>
<span class="normal"><a href="#__codelineno-32-91">91</a></span>
<span class="normal"><a href="#__codelineno-32-92">92</a></span>
<span class="normal"><a href="#__codelineno-32-93">93</a></span>
<span class="normal"><a href="#__codelineno-32-94">94</a></span>
<span class="normal"><a href="#__codelineno-32-95">95</a></span>
<span class="normal"><a href="#__codelineno-32-96">96</a></span>
<span class="normal"><a href="#__codelineno-32-97">97</a></span>
<span class="normal"><a href="#__codelineno-32-98">98</a></span>
<span class="normal"><a href="#__codelineno-32-99">99</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-32-1" name="__codelineno-32-1"></a><span class="cm">/*</span>
<a id="__codelineno-32-2" name="__codelineno-32-2"></a><span class="cm"> * Processes 6: ASE &amp; RNA Editing</span>
<a id="__codelineno-32-3" name="__codelineno-32-3"></a><span class="cm"> */</span>
<a id="__codelineno-32-4" name="__codelineno-32-4"></a>
<a id="__codelineno-32-5" name="__codelineno-32-5"></a><span class="n">process</span><span class="w"> </span><span class="n">post_process_vcf</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-32-6" name="__codelineno-32-6"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/biocontainers/mulled-v2-b9358559e3ae3b9d7d8dbf1f401ae1fcaf757de3:ac05763cf181a5070c2fdb9bb5461f8d08f7b93b-0&#39;</span>
<a id="__codelineno-32-7" name="__codelineno-32-7"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span>
<a id="__codelineno-32-8" name="__codelineno-32-8"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="s2">&quot;${params.results}/${sampleId}&quot;</span>
<a id="__codelineno-32-9" name="__codelineno-32-9"></a>
<a id="__codelineno-32-10" name="__codelineno-32-10"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-32-11" name="__codelineno-32-11"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">)</span>
<a id="__codelineno-32-12" name="__codelineno-32-12"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;filtered.recode.vcf.gz&#39;</span><span class="o">),</span>
<a id="__codelineno-32-13" name="__codelineno-32-13"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;filtered.recode.vcf.gz.tbi&#39;</span><span class="o">)</span>
<a id="__codelineno-32-14" name="__codelineno-32-14"></a>
<a id="__codelineno-32-15" name="__codelineno-32-15"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-32-16" name="__codelineno-32-16"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span>
<a id="__codelineno-32-17" name="__codelineno-32-17"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">),</span>
<a id="__codelineno-32-18" name="__codelineno-32-18"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;commonSNPs.diff.sites_in_files&#39;</span><span class="o">)</span>
<a id="__codelineno-32-19" name="__codelineno-32-19"></a>
<a id="__codelineno-32-20" name="__codelineno-32-20"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-32-21" name="__codelineno-32-21"></a><span class="w">    </span><span class="s1">&#39;&#39;&#39;</span>
<a id="__codelineno-32-22" name="__codelineno-32-22"></a><span class="s1">    grep -v &#39;#&#39; final.vcf | awk &#39;$7~/PASS/&#39; | perl -ne &#39;chomp($_); \</span>
<a id="__codelineno-32-23" name="__codelineno-32-23"></a><span class="s1">            ($dp)=$_=~/DP\\=(\\d+)\\;/; if($dp&gt;=8){print $_.&quot;\\n&quot;};&#39; \</span>
<a id="__codelineno-32-24" name="__codelineno-32-24"></a><span class="s1">            &gt; result.DP8.vcf</span>
<a id="__codelineno-32-25" name="__codelineno-32-25"></a>
<a id="__codelineno-32-26" name="__codelineno-32-26"></a><span class="s1">    vcftools --vcf result.DP8.vcf --gzdiff filtered.recode.vcf.gz \</span>
<a id="__codelineno-32-27" name="__codelineno-32-27"></a><span class="s1">             --diff-site --out commonSNPs</span>
<a id="__codelineno-32-28" name="__codelineno-32-28"></a><span class="s1">    &#39;&#39;&#39;</span>
<a id="__codelineno-32-29" name="__codelineno-32-29"></a><span class="o">}</span>
<a id="__codelineno-32-30" name="__codelineno-32-30"></a>
<a id="__codelineno-32-31" name="__codelineno-32-31"></a><span class="n">process</span><span class="w"> </span><span class="n">prepare_vcf_for_ase</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-32-32" name="__codelineno-32-32"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;cbcrg/callings-with-gatk:latest&#39;</span>
<a id="__codelineno-32-33" name="__codelineno-32-33"></a><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span>
<a id="__codelineno-32-34" name="__codelineno-32-34"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="s2">&quot;${params.results}/${sampleId}&quot;</span>
<a id="__codelineno-32-35" name="__codelineno-32-35"></a>
<a id="__codelineno-32-36" name="__codelineno-32-36"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-32-37" name="__codelineno-32-37"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span>
<a id="__codelineno-32-38" name="__codelineno-32-38"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;final.vcf&#39;</span><span class="o">),</span>
<a id="__codelineno-32-39" name="__codelineno-32-39"></a><span class="w">          </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;commonSNPs.diff.sites_in_files&#39;</span><span class="o">)</span>
<a id="__codelineno-32-40" name="__codelineno-32-40"></a>
<a id="__codelineno-32-41" name="__codelineno-32-41"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-32-42" name="__codelineno-32-42"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s1">&#39;known_snps.vcf&#39;</span><span class="o">),</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">vcf_for_ASE</span>
<a id="__codelineno-32-43" name="__codelineno-32-43"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s1">&#39;AF.histogram.pdf&#39;</span><span class="w">                    </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">gghist_pdfs</span>
<a id="__codelineno-32-44" name="__codelineno-32-44"></a>
<a id="__codelineno-32-45" name="__codelineno-32-45"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-32-46" name="__codelineno-32-46"></a><span class="w">    </span><span class="s1">&#39;&#39;&#39;</span>
<a id="__codelineno-32-47" name="__codelineno-32-47"></a><span class="s1">    awk &#39;BEGIN{OFS=&quot;\t&quot;} $4~/B/{print $1,$2,$3}&#39; \</span>
<a id="__codelineno-32-48" name="__codelineno-32-48"></a><span class="s1">        commonSNPs.diff.sites_in_files  &gt; test.bed</span>
<a id="__codelineno-32-49" name="__codelineno-32-49"></a>
<a id="__codelineno-32-50" name="__codelineno-32-50"></a><span class="s1">    vcftools --vcf final.vcf --bed test.bed --recode --keep-INFO-all \</span>
<a id="__codelineno-32-51" name="__codelineno-32-51"></a><span class="s1">             --stdout &gt; known_snps.vcf</span>
<a id="__codelineno-32-52" name="__codelineno-32-52"></a>
<a id="__codelineno-32-53" name="__codelineno-32-53"></a><span class="s1">    grep -v &#39;#&#39;  known_snps.vcf | awk -F &#39;\\t&#39; &#39;{print $10}&#39; \</span>
<a id="__codelineno-32-54" name="__codelineno-32-54"></a><span class="s1">                | awk -F &#39;:&#39; &#39;{print $2}&#39; | perl -ne &#39;chomp($_); \</span>
<a id="__codelineno-32-55" name="__codelineno-32-55"></a><span class="s1">                @v=split(/\\,/,$_); if($v[0]!=0 ||$v[1] !=0)\</span>
<a id="__codelineno-32-56" name="__codelineno-32-56"></a><span class="s1">                {print  $v[1]/($v[1]+$v[0]).&quot;\\n&quot;; }&#39; | awk &#39;$1!=1&#39; \</span>
<a id="__codelineno-32-57" name="__codelineno-32-57"></a><span class="s1">                &gt;AF.4R</span>
<a id="__codelineno-32-58" name="__codelineno-32-58"></a>
<a id="__codelineno-32-59" name="__codelineno-32-59"></a><span class="s1">    gghist.R -i AF.4R -o AF.histogram.pdf</span>
<a id="__codelineno-32-60" name="__codelineno-32-60"></a><span class="s1">    &#39;&#39;&#39;</span>
<a id="__codelineno-32-61" name="__codelineno-32-61"></a><span class="o">}</span>
<a id="__codelineno-32-62" name="__codelineno-32-62"></a>
<a id="__codelineno-32-63" name="__codelineno-32-63"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-32-64" name="__codelineno-32-64"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-32-65" name="__codelineno-32-65"></a>
<a id="__codelineno-32-66" name="__codelineno-32-66"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-32-67" name="__codelineno-32-67"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-32-68" name="__codelineno-32-68"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-32-69" name="__codelineno-32-69"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-32-70" name="__codelineno-32-70"></a>
<a id="__codelineno-32-71" name="__codelineno-32-71"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-32-72" name="__codelineno-32-72"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-73" name="__codelineno-32-73"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-32-74" name="__codelineno-32-74"></a>
<a id="__codelineno-32-75" name="__codelineno-32-75"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-32-76" name="__codelineno-32-76"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-77" name="__codelineno-32-77"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-78" name="__codelineno-32-78"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-32-79" name="__codelineno-32-79"></a>
<a id="__codelineno-32-80" name="__codelineno-32-80"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-32-81" name="__codelineno-32-81"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-82" name="__codelineno-32-82"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-83" name="__codelineno-32-83"></a><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-84" name="__codelineno-32-84"></a><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-32-85" name="__codelineno-32-85"></a>
<a id="__codelineno-32-86" name="__codelineno-32-86"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span>
<a id="__codelineno-32-87" name="__codelineno-32-87"></a><span class="w">        </span><span class="o">.</span><span class="na">out</span>
<a id="__codelineno-32-88" name="__codelineno-32-88"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">groupTuple</span>
<a id="__codelineno-32-89" name="__codelineno-32-89"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">recalibrated_samples</span><span class="w"> </span><span class="o">}</span>
<a id="__codelineno-32-90" name="__codelineno-32-90"></a>
<a id="__codelineno-32-91" name="__codelineno-32-91"></a><span class="w">    </span><span class="n">rnaseq_call_variants</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-32-92" name="__codelineno-32-92"></a><span class="w">                         </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-93" name="__codelineno-32-93"></a><span class="w">                         </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-32-94" name="__codelineno-32-94"></a><span class="w">                         </span><span class="n">recalibrated_samples</span><span class="o">)</span>
<a id="__codelineno-32-95" name="__codelineno-32-95"></a>
<a id="__codelineno-32-96" name="__codelineno-32-96"></a><span class="hll"><span class="w">    </span><span class="n">post_process_vcf</span><span class="o">(</span><span class="n">rnaseq_call_variants</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-32-97" name="__codelineno-32-97"></a><span class="hll"><span class="w">                     </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
</span><a id="__codelineno-32-98" name="__codelineno-32-98"></a><span class="hll"><span class="w">    </span><span class="n">prepare_vcf_for_ase</span><span class="o">(</span><span class="n">post_process_vcf</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
</span><a id="__codelineno-32-99" name="__codelineno-32-99"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
</div>
<p>The final step is the GATK ASEReadCounter.</p>
<div class="admonition exercise">
<p class="admonition-title">Problem #11</p>
<p>We have seen the basics of using processes in Nextflow. Yet one of the features of Nextflow is the operations that can be performed on channels outside of processes. See <a href="https://www.nextflow.io/docs/latest/operator.html">here</a> for details on the specific operators.</p>
<p>Before we perform the GATK ASEReadCounter process, we must group the data for allele-specific expression. To do this we must combine channels.</p>
<p>The output channel of the <code>rnaseq_gatk_recalibrate</code> process (<code>rnaseq_gatk_recalibrate.out</code>) emites tuples having the following structure, holding the final BAM/BAI files:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>&lt;<span class="w"> </span>sample_id,<span class="w"> </span>file_bam,<span class="w"> </span>file_bai<span class="w"> </span>&gt;
</code></pre></div>
<p>The <code>vcf_for_ASE</code> channel emits tuples having the following structure:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-34-1" name="__codelineno-34-1" href="#__codelineno-34-1"></a>&lt;<span class="w"> </span>sample_id,<span class="w"> </span>output.vcf<span class="w"> </span>&gt;
</code></pre></div>
<p>In the first operation, the BAM files are grouped together by sample id.</p>
<p>Next, this resulting channel is merged with the VCFs having the same sample id.</p>
<p>We must take the merged channel and creates a channel named <code>grouped_vcf_bam_bai_ch</code> emitting the following tuples:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-35-1" name="__codelineno-35-1" href="#__codelineno-35-1"></a>&lt;<span class="w"> </span>sample_id,<span class="w"> </span>file_vcf,<span class="w"> </span>List<span class="o">[</span>file_bam<span class="o">]</span>,<span class="w"> </span>List<span class="o">[</span>file_bai<span class="o">]</span><span class="w"> </span>&gt;
</code></pre></div>
<p>Your aim is to fill in the <code>BLANKS</code> below.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-36-1">1</a></span>
<span class="normal"><a href="#__codelineno-36-2">2</a></span>
<span class="normal"><a href="#__codelineno-36-3">3</a></span>
<span class="normal"><a href="#__codelineno-36-4">4</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-36-1" name="__codelineno-36-1"></a><span class="n">recalibrated_samples</span>
<a id="__codelineno-36-2" name="__codelineno-36-2"></a><span class="hll"><span class="w">    </span><span class="o">.</span><span class="na">BLANK</span><span class="w"> </span><span class="c1">// (1)!</span>
</span><a id="__codelineno-36-3" name="__codelineno-36-3"></a><span class="hll"><span class="w">    </span><span class="o">.</span><span class="na">map</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">BLANK</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// (2)!</span>
</span><a id="__codelineno-36-4" name="__codelineno-36-4"></a><span class="hll"><span class="w">    </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">BLANK</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// (3)!</span>
</span></code></pre></div></td></tr></table></div>
<ol>
<li>An operator that joins two channels taking a key into consideration. See <a href="https://www.nextflow.io/docs/latest/operator.html?join#join">here</a> for more details</li>
<li>The map operator can apply any function to every item on a channel. In this case we take our tuple from the previous step, define the separate elements and create a new tuple.</li>
<li>Rename the resulting as <code>grouped_vcf_bam_bai_ch</code></li>
</ol>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-37-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-37-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-37-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-37-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-37-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-37-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-37-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-37-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-37-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-37-10">10</a></span>
<span class="normal"><a href="#__codelineno-37-11">11</a></span>
<span class="normal"><a href="#__codelineno-37-12">12</a></span>
<span class="normal"><a href="#__codelineno-37-13">13</a></span>
<span class="normal"><a href="#__codelineno-37-14">14</a></span>
<span class="normal"><a href="#__codelineno-37-15">15</a></span>
<span class="normal"><a href="#__codelineno-37-16">16</a></span>
<span class="normal"><a href="#__codelineno-37-17">17</a></span>
<span class="normal"><a href="#__codelineno-37-18">18</a></span>
<span class="normal"><a href="#__codelineno-37-19">19</a></span>
<span class="normal"><a href="#__codelineno-37-20">20</a></span>
<span class="normal"><a href="#__codelineno-37-21">21</a></span>
<span class="normal"><a href="#__codelineno-37-22">22</a></span>
<span class="normal"><a href="#__codelineno-37-23">23</a></span>
<span class="normal"><a href="#__codelineno-37-24">24</a></span>
<span class="normal"><a href="#__codelineno-37-25">25</a></span>
<span class="normal"><a href="#__codelineno-37-26">26</a></span>
<span class="normal"><a href="#__codelineno-37-27">27</a></span>
<span class="normal"><a href="#__codelineno-37-28">28</a></span>
<span class="normal"><a href="#__codelineno-37-29">29</a></span>
<span class="normal"><a href="#__codelineno-37-30">30</a></span>
<span class="normal"><a href="#__codelineno-37-31">31</a></span>
<span class="normal"><a href="#__codelineno-37-32">32</a></span>
<span class="normal"><a href="#__codelineno-37-33">33</a></span>
<span class="normal"><a href="#__codelineno-37-34">34</a></span>
<span class="normal"><a href="#__codelineno-37-35">35</a></span>
<span class="normal"><a href="#__codelineno-37-36">36</a></span>
<span class="normal"><a href="#__codelineno-37-37">37</a></span>
<span class="normal"><a href="#__codelineno-37-38">38</a></span>
<span class="normal"><a href="#__codelineno-37-39">39</a></span>
<span class="normal"><a href="#__codelineno-37-40">40</a></span>
<span class="normal"><a href="#__codelineno-37-41">41</a></span>
<span class="normal"><a href="#__codelineno-37-42">42</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-37-1" name="__codelineno-37-1"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-37-2" name="__codelineno-37-2"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-37-3" name="__codelineno-37-3"></a>
<a id="__codelineno-37-4" name="__codelineno-37-4"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-37-5" name="__codelineno-37-5"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-37-6" name="__codelineno-37-6"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-37-7" name="__codelineno-37-7"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-37-8" name="__codelineno-37-8"></a>
<a id="__codelineno-37-9" name="__codelineno-37-9"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-37-10" name="__codelineno-37-10"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-11" name="__codelineno-37-11"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-37-12" name="__codelineno-37-12"></a>
<a id="__codelineno-37-13" name="__codelineno-37-13"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-37-14" name="__codelineno-37-14"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-15" name="__codelineno-37-15"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-16" name="__codelineno-37-16"></a><span class="w">                            </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-37-17" name="__codelineno-37-17"></a>
<a id="__codelineno-37-18" name="__codelineno-37-18"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-37-19" name="__codelineno-37-19"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-20" name="__codelineno-37-20"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-21" name="__codelineno-37-21"></a><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-22" name="__codelineno-37-22"></a><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-37-23" name="__codelineno-37-23"></a>
<a id="__codelineno-37-24" name="__codelineno-37-24"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span>
<a id="__codelineno-37-25" name="__codelineno-37-25"></a><span class="w">        </span><span class="o">.</span><span class="na">out</span>
<a id="__codelineno-37-26" name="__codelineno-37-26"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">groupTuple</span>
<a id="__codelineno-37-27" name="__codelineno-37-27"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">recalibrated_samples</span><span class="w"> </span><span class="o">}</span>
<a id="__codelineno-37-28" name="__codelineno-37-28"></a>
<a id="__codelineno-37-29" name="__codelineno-37-29"></a><span class="w">    </span><span class="n">rnaseq_call_variants</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-37-30" name="__codelineno-37-30"></a><span class="w">                         </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-31" name="__codelineno-37-31"></a><span class="w">                         </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-32" name="__codelineno-37-32"></a><span class="w">                         </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-37-33" name="__codelineno-37-33"></a>
<a id="__codelineno-37-34" name="__codelineno-37-34"></a><span class="w">    </span><span class="n">post_process_vcf</span><span class="o">(</span><span class="n">rnaseq_call_variants</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-37-35" name="__codelineno-37-35"></a><span class="w">                     </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-37-36" name="__codelineno-37-36"></a><span class="w">    </span><span class="n">prepare_vcf_for_ase</span><span class="o">(</span><span class="n">post_process_vcf</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-37-37" name="__codelineno-37-37"></a>
<a id="__codelineno-37-38" name="__codelineno-37-38"></a><span class="hll"><span class="w">    </span><span class="n">recalibrated_samples</span>
</span><a id="__codelineno-37-39" name="__codelineno-37-39"></a><span class="hll"><span class="w">        </span><span class="o">.</span><span class="na">join</span><span class="o">(</span><span class="n">prepare_vcf_for_ase</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">vcf_for_ASE</span><span class="o">)</span>
</span><a id="__codelineno-37-40" name="__codelineno-37-40"></a><span class="hll"><span class="w">        </span><span class="o">.</span><span class="na">map</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">meta</span><span class="o">,</span><span class="w"> </span><span class="n">bams</span><span class="o">,</span><span class="w"> </span><span class="n">bais</span><span class="o">,</span><span class="w"> </span><span class="n">vcf</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="o">[</span><span class="n">meta</span><span class="o">,</span><span class="w"> </span><span class="n">vcf</span><span class="o">,</span><span class="w"> </span><span class="n">bams</span><span class="o">,</span><span class="w"> </span><span class="n">bais</span><span class="o">]</span><span class="w"> </span><span class="o">}</span>
</span><a id="__codelineno-37-41" name="__codelineno-37-41"></a><span class="hll"><span class="w">        </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">grouped_vcf_bam_bai_ch</span><span class="w"> </span><span class="o">}</span>
</span><a id="__codelineno-37-42" name="__codelineno-37-42"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
</div>
<h2 id="process-7-allele-specific-expression-analysis-with-gatk-asereadcounter">Process 7: Allele-Specific Expression analysis with GATK ASEReadCounter<a class="headerlink" href="#process-7-allele-specific-expression-analysis-with-gatk-asereadcounter" title="Permanent link">&para;</a></h2>
<p>Now we are ready for the final process.</p>
<p>The next process has the following structure:</p>
<ul>
<li><strong>Name</strong>: <code>ASE_knownSNPs</code></li>
<li><strong>Command</strong>: calculate allele counts at a set of positions with GATK tools</li>
<li><strong>Input</strong>:</li>
<li>genome fasta file</li>
<li>genome index file from samtools</li>
<li>genome dictionary file</li>
<li>the <code>grouped_vcf_bam_bai_ch</code> channel</li>
<li><strong>Output</strong>: the allele specific expression file (<code>ASE.tsv</code>)</li>
</ul>
<div class="admonition exercise">
<p class="admonition-title">Problem #12</p>
<p>You should construct the process from scratch, add the process call and inputs to the workflow block and run the pipeline in its entirety.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-38-1">1</a></span>
<span class="normal"><a href="#__codelineno-38-2">2</a></span>
<span class="normal"><a href="#__codelineno-38-3">3</a></span>
<span class="normal"><a href="#__codelineno-38-4">4</a></span>
<span class="normal"><a href="#__codelineno-38-5">5</a></span>
<span class="normal"><a href="#__codelineno-38-6">6</a></span>
<span class="normal"><a href="#__codelineno-38-7">7</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-38-1" name="__codelineno-38-1"></a><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">bam</span><span class="p">.join(</span><span class="s1">&#39;\n&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>bam.list
<a id="__codelineno-38-2" name="__codelineno-38-2"></a>
<a id="__codelineno-38-3" name="__codelineno-38-3"></a>java<span class="w"> </span>-jar<span class="w"> </span>/usr/gitc/GATK35.jar<span class="w"> </span>-R<span class="w"> </span><span class="si">${</span><span class="nv">genome</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-38-4" name="__codelineno-38-4"></a><span class="w">                               </span>-T<span class="w"> </span>ASEReadCounter<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-38-5" name="__codelineno-38-5"></a><span class="w">                               </span>-o<span class="w"> </span>ASE.tsv<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-38-6" name="__codelineno-38-6"></a><span class="w">                               </span>-I<span class="w"> </span>bam.list<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-38-7" name="__codelineno-38-7"></a><span class="w">                               </span>-sites<span class="w"> </span><span class="si">${</span><span class="nv">vcf</span><span class="si">}</span>
</code></pre></div></td></tr></table></div>
<details class="solution">
<summary>Solution</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-39-1"> 1</a></span>
<span class="normal"><a href="#__codelineno-39-2"> 2</a></span>
<span class="normal"><a href="#__codelineno-39-3"> 3</a></span>
<span class="normal"><a href="#__codelineno-39-4"> 4</a></span>
<span class="normal"><a href="#__codelineno-39-5"> 5</a></span>
<span class="normal"><a href="#__codelineno-39-6"> 6</a></span>
<span class="normal"><a href="#__codelineno-39-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-39-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-39-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-39-10">10</a></span>
<span class="normal"><a href="#__codelineno-39-11">11</a></span>
<span class="normal"><a href="#__codelineno-39-12">12</a></span>
<span class="normal"><a href="#__codelineno-39-13">13</a></span>
<span class="normal"><a href="#__codelineno-39-14">14</a></span>
<span class="normal"><a href="#__codelineno-39-15">15</a></span>
<span class="normal"><a href="#__codelineno-39-16">16</a></span>
<span class="normal"><a href="#__codelineno-39-17">17</a></span>
<span class="normal"><a href="#__codelineno-39-18">18</a></span>
<span class="normal"><a href="#__codelineno-39-19">19</a></span>
<span class="normal"><a href="#__codelineno-39-20">20</a></span>
<span class="normal"><a href="#__codelineno-39-21">21</a></span>
<span class="normal"><a href="#__codelineno-39-22">22</a></span>
<span class="normal"><a href="#__codelineno-39-23">23</a></span>
<span class="normal"><a href="#__codelineno-39-24">24</a></span>
<span class="normal"><a href="#__codelineno-39-25">25</a></span>
<span class="normal"><a href="#__codelineno-39-26">26</a></span>
<span class="normal"><a href="#__codelineno-39-27">27</a></span>
<span class="normal"><a href="#__codelineno-39-28">28</a></span>
<span class="normal"><a href="#__codelineno-39-29">29</a></span>
<span class="normal"><a href="#__codelineno-39-30">30</a></span>
<span class="normal"><a href="#__codelineno-39-31">31</a></span>
<span class="normal"><a href="#__codelineno-39-32">32</a></span>
<span class="normal"><a href="#__codelineno-39-33">33</a></span>
<span class="normal"><a href="#__codelineno-39-34">34</a></span>
<span class="normal"><a href="#__codelineno-39-35">35</a></span>
<span class="normal"><a href="#__codelineno-39-36">36</a></span>
<span class="normal"><a href="#__codelineno-39-37">37</a></span>
<span class="normal"><a href="#__codelineno-39-38">38</a></span>
<span class="normal"><a href="#__codelineno-39-39">39</a></span>
<span class="normal"><a href="#__codelineno-39-40">40</a></span>
<span class="normal"><a href="#__codelineno-39-41">41</a></span>
<span class="normal"><a href="#__codelineno-39-42">42</a></span>
<span class="normal"><a href="#__codelineno-39-43">43</a></span>
<span class="normal"><a href="#__codelineno-39-44">44</a></span>
<span class="normal"><a href="#__codelineno-39-45">45</a></span>
<span class="normal"><a href="#__codelineno-39-46">46</a></span>
<span class="normal"><a href="#__codelineno-39-47">47</a></span>
<span class="normal"><a href="#__codelineno-39-48">48</a></span>
<span class="normal"><a href="#__codelineno-39-49">49</a></span>
<span class="normal"><a href="#__codelineno-39-50">50</a></span>
<span class="normal"><a href="#__codelineno-39-51">51</a></span>
<span class="normal"><a href="#__codelineno-39-52">52</a></span>
<span class="normal"><a href="#__codelineno-39-53">53</a></span>
<span class="normal"><a href="#__codelineno-39-54">54</a></span>
<span class="normal"><a href="#__codelineno-39-55">55</a></span>
<span class="normal"><a href="#__codelineno-39-56">56</a></span>
<span class="normal"><a href="#__codelineno-39-57">57</a></span>
<span class="normal"><a href="#__codelineno-39-58">58</a></span>
<span class="normal"><a href="#__codelineno-39-59">59</a></span>
<span class="normal"><a href="#__codelineno-39-60">60</a></span>
<span class="normal"><a href="#__codelineno-39-61">61</a></span>
<span class="normal"><a href="#__codelineno-39-62">62</a></span>
<span class="normal"><a href="#__codelineno-39-63">63</a></span>
<span class="normal"><a href="#__codelineno-39-64">64</a></span>
<span class="normal"><a href="#__codelineno-39-65">65</a></span>
<span class="normal"><a href="#__codelineno-39-66">66</a></span>
<span class="normal"><a href="#__codelineno-39-67">67</a></span>
<span class="normal"><a href="#__codelineno-39-68">68</a></span>
<span class="normal"><a href="#__codelineno-39-69">69</a></span>
<span class="normal"><a href="#__codelineno-39-70">70</a></span>
<span class="normal"><a href="#__codelineno-39-71">71</a></span>
<span class="normal"><a href="#__codelineno-39-72">72</a></span>
<span class="normal"><a href="#__codelineno-39-73">73</a></span>
<span class="normal"><a href="#__codelineno-39-74">74</a></span>
<span class="normal"><a href="#__codelineno-39-75">75</a></span>
<span class="normal"><a href="#__codelineno-39-76">76</a></span>
<span class="normal"><a href="#__codelineno-39-77">77</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-39-1" name="__codelineno-39-1"></a><span class="cm">/*</span>
<a id="__codelineno-39-2" name="__codelineno-39-2"></a><span class="cm"> * Processes 7: Allele-Specific Expression analysis with GATK ASEReadCounter</span>
<a id="__codelineno-39-3" name="__codelineno-39-3"></a><span class="cm"> */</span>
<a id="__codelineno-39-4" name="__codelineno-39-4"></a>
<a id="__codelineno-39-5" name="__codelineno-39-5"></a><span class="hll"><span class="n">process</span><span class="w"> </span><span class="n">ASE_knownSNPs</span><span class="w"> </span><span class="o">{</span>
</span><a id="__codelineno-39-6" name="__codelineno-39-6"></a><span class="hll"><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;quay.io/broadinstitute/gotc-prod-gatk:1.0.0-4.1.8.0-1626439571&#39;</span>
</span><a id="__codelineno-39-7" name="__codelineno-39-7"></a><span class="hll"><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="s2">&quot;${sampleId}&quot;</span>
</span><a id="__codelineno-39-8" name="__codelineno-39-8"></a><span class="hll"><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="s2">&quot;${params.results}/${sampleId}&quot;</span>
</span><a id="__codelineno-39-9" name="__codelineno-39-9"></a><span class="hll">
</span><a id="__codelineno-39-10" name="__codelineno-39-10"></a><span class="hll"><span class="w">    </span><span class="nl">input:</span>
</span><a id="__codelineno-39-11" name="__codelineno-39-11"></a><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">genome</span>
</span><a id="__codelineno-39-12" name="__codelineno-39-12"></a><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">index</span>
</span><a id="__codelineno-39-13" name="__codelineno-39-13"></a><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">dict</span>
</span><a id="__codelineno-39-14" name="__codelineno-39-14"></a><span class="hll"><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">val</span><span class="o">(</span><span class="n">sampleId</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">vcf</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">bai</span><span class="o">)</span>
</span><a id="__codelineno-39-15" name="__codelineno-39-15"></a><span class="hll">
</span><a id="__codelineno-39-16" name="__codelineno-39-16"></a><span class="hll"><span class="w">    </span><span class="nl">output:</span>
</span><a id="__codelineno-39-17" name="__codelineno-39-17"></a><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;ASE.tsv&quot;</span>
</span><a id="__codelineno-39-18" name="__codelineno-39-18"></a><span class="hll">
</span><a id="__codelineno-39-19" name="__codelineno-39-19"></a><span class="hll"><span class="w">    </span><span class="nl">script:</span>
</span><a id="__codelineno-39-20" name="__codelineno-39-20"></a><span class="hll"><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
</span><a id="__codelineno-39-21" name="__codelineno-39-21"></a><span class="hll"><span class="s2">    echo &quot;${bam.join(&#39;\n&#39;)}&quot; &gt; bam.list</span>
</span><a id="__codelineno-39-22" name="__codelineno-39-22"></a><span class="hll">
</span><a id="__codelineno-39-23" name="__codelineno-39-23"></a><span class="hll"><span class="s2">    java -jar /usr/gitc/GATK35.jar -R ${genome} \</span>
</span><a id="__codelineno-39-24" name="__codelineno-39-24"></a><span class="hll"><span class="s2">                                   -T ASEReadCounter \</span>
</span><a id="__codelineno-39-25" name="__codelineno-39-25"></a><span class="hll"><span class="s2">                                   -o ASE.tsv \</span>
</span><a id="__codelineno-39-26" name="__codelineno-39-26"></a><span class="hll"><span class="s2">                                   -I bam.list \</span>
</span><a id="__codelineno-39-27" name="__codelineno-39-27"></a><span class="hll"><span class="s2">                                   -sites ${vcf}</span>
</span><a id="__codelineno-39-28" name="__codelineno-39-28"></a><span class="hll"><span class="s2">    &quot;&quot;&quot;</span>
</span><a id="__codelineno-39-29" name="__codelineno-39-29"></a><span class="hll"><span class="o">}</span>
</span><a id="__codelineno-39-30" name="__codelineno-39-30"></a>
<a id="__codelineno-39-31" name="__codelineno-39-31"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-39-32" name="__codelineno-39-32"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromFilePairs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="o">)</span>
<a id="__codelineno-39-33" name="__codelineno-39-33"></a>
<a id="__codelineno-39-34" name="__codelineno-39-34"></a><span class="w">    </span><span class="n">prepare_genome_samtools</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-39-35" name="__codelineno-39-35"></a><span class="w">    </span><span class="n">prepare_genome_picard</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-39-36" name="__codelineno-39-36"></a><span class="w">    </span><span class="n">prepare_star_genome_index</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">)</span>
<a id="__codelineno-39-37" name="__codelineno-39-37"></a><span class="w">    </span><span class="n">prepare_vcf_file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">variants</span><span class="o">,</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">blacklist</span><span class="o">)</span>
<a id="__codelineno-39-38" name="__codelineno-39-38"></a>
<a id="__codelineno-39-39" name="__codelineno-39-39"></a><span class="w">    </span><span class="n">rnaseq_mapping_star</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-39-40" name="__codelineno-39-40"></a><span class="w">                        </span><span class="n">prepare_star_genome_index</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-41" name="__codelineno-39-41"></a><span class="w">                        </span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-39-42" name="__codelineno-39-42"></a>
<a id="__codelineno-39-43" name="__codelineno-39-43"></a><span class="w">    </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-39-44" name="__codelineno-39-44"></a><span class="w">                        </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-45" name="__codelineno-39-45"></a><span class="w">                        </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-46" name="__codelineno-39-46"></a><span class="w">                        </span><span class="n">rnaseq_mapping_star</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-39-47" name="__codelineno-39-47"></a>
<a id="__codelineno-39-48" name="__codelineno-39-48"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-39-49" name="__codelineno-39-49"></a><span class="w">                            </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-50" name="__codelineno-39-50"></a><span class="w">                            </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-51" name="__codelineno-39-51"></a><span class="w">                            </span><span class="n">rnaseq_gatk_splitNcigar</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-52" name="__codelineno-39-52"></a><span class="w">                            </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-39-53" name="__codelineno-39-53"></a>
<a id="__codelineno-39-54" name="__codelineno-39-54"></a><span class="w">    </span><span class="n">rnaseq_gatk_recalibrate</span>
<a id="__codelineno-39-55" name="__codelineno-39-55"></a><span class="w">        </span><span class="o">.</span><span class="na">out</span>
<a id="__codelineno-39-56" name="__codelineno-39-56"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">groupTuple</span>
<a id="__codelineno-39-57" name="__codelineno-39-57"></a><span class="w">        </span><span class="o">|</span><span class="w"> </span><span class="n">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">recalibrated_samples</span><span class="w"> </span><span class="o">}</span>
<a id="__codelineno-39-58" name="__codelineno-39-58"></a>
<a id="__codelineno-39-59" name="__codelineno-39-59"></a><span class="w">    </span><span class="n">rnaseq_call_variants</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
<a id="__codelineno-39-60" name="__codelineno-39-60"></a><span class="w">                         </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-61" name="__codelineno-39-61"></a><span class="w">                         </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-62" name="__codelineno-39-62"></a><span class="w">                         </span><span class="n">recalibrated_samples</span><span class="o">)</span>
<a id="__codelineno-39-63" name="__codelineno-39-63"></a>
<a id="__codelineno-39-64" name="__codelineno-39-64"></a><span class="w">    </span><span class="n">post_process_vcf</span><span class="o">(</span><span class="n">rnaseq_call_variants</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-39-65" name="__codelineno-39-65"></a><span class="w">                     </span><span class="n">prepare_vcf_file</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-39-66" name="__codelineno-39-66"></a><span class="w">    </span><span class="n">prepare_vcf_for_ase</span><span class="o">(</span><span class="n">post_process_vcf</span><span class="o">.</span><span class="na">out</span><span class="o">)</span>
<a id="__codelineno-39-67" name="__codelineno-39-67"></a>
<a id="__codelineno-39-68" name="__codelineno-39-68"></a><span class="w">    </span><span class="n">recalibrated_samples</span>
<a id="__codelineno-39-69" name="__codelineno-39-69"></a><span class="w">        </span><span class="o">.</span><span class="na">join</span><span class="o">(</span><span class="n">prepare_vcf_for_ase</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">vcf_for_ASE</span><span class="o">)</span>
<a id="__codelineno-39-70" name="__codelineno-39-70"></a><span class="w">        </span><span class="o">.</span><span class="na">map</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">meta</span><span class="o">,</span><span class="w"> </span><span class="n">bams</span><span class="o">,</span><span class="w"> </span><span class="n">bais</span><span class="o">,</span><span class="w"> </span><span class="n">vcf</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="o">[</span><span class="n">meta</span><span class="o">,</span><span class="w"> </span><span class="n">vcf</span><span class="o">,</span><span class="w"> </span><span class="n">bams</span><span class="o">,</span><span class="w"> </span><span class="n">bais</span><span class="o">]</span><span class="w"> </span><span class="o">}</span>
<a id="__codelineno-39-71" name="__codelineno-39-71"></a><span class="w">        </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">grouped_vcf_bam_bai_ch</span><span class="w"> </span><span class="o">}</span>
<a id="__codelineno-39-72" name="__codelineno-39-72"></a>
<a id="__codelineno-39-73" name="__codelineno-39-73"></a><span class="hll"><span class="w">    </span><span class="n">ASE_knownSNPs</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">genome</span><span class="o">,</span>
</span><a id="__codelineno-39-74" name="__codelineno-39-74"></a><span class="hll"><span class="w">                  </span><span class="n">prepare_genome_samtools</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-39-75" name="__codelineno-39-75"></a><span class="hll"><span class="w">                  </span><span class="n">prepare_genome_picard</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</span><a id="__codelineno-39-76" name="__codelineno-39-76"></a><span class="hll"><span class="w">                  </span><span class="n">grouped_vcf_bam_bai_ch</span><span class="o">)</span>
</span><a id="__codelineno-39-77" name="__codelineno-39-77"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
</details>
</div>
<p>Congratulations! If you made it this far you now have all the basics to create your own Nextflow workflows.</p>
<h2 id="results-overview">Results overview<a class="headerlink" href="#results-overview" title="Permanent link">&para;</a></h2>
<p>For each processed sample the pipeline stores results into a folder named after the sample identifier. These folders are created in the directory specified as a parameter in <code>params.results</code>.</p>
<p>Result files for this workshop can be found in the folder <code>results</code> within the current folder. There you should see a directory called <code>ENCSR000COQ/</code> containing the following files:</p>
<h3 id="variant-calls">Variant calls<a class="headerlink" href="#variant-calls" title="Permanent link">&para;</a></h3>
<p><code>final.vcf</code></p>
<p>This file contains all somatic variants (SNVs) called from RNAseq data. You will see variants that pass all filters, with the <code>PASS</code> keyword in the <span class="red">7th</span> field of the vcf file (<code>filter status</code>), and also those that did not pass one or more filters.</p>
<p><code>commonSNPs.diff.sites_in_files</code></p>
<p>Tab-separated file with comparison between variants obtained from RNAseq and "known" variants from DNA.</p>
<p>The file is sorted by genomic position and contains 8 fields:</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td><code>CHROM</code></td>
<td>chromosome name;</td>
</tr>
<tr>
<td>2</td>
<td><code>POS1</code></td>
<td>position of the SNV in file #1 (RNAseq data);</td>
</tr>
<tr>
<td>3</td>
<td><code>POS2</code></td>
<td>position of SNV in file #2 (DNA "known" variants);</td>
</tr>
<tr>
<td>4</td>
<td><code>IN_FILE</code></td>
<td>flag whether SNV is present in the file #1 <em>1</em>, in the file #2 <em>2</em>, or in both files <em>B</em>;</td>
</tr>
<tr>
<td>5</td>
<td><code>REF1</code></td>
<td>reference sequence in the file 1;</td>
</tr>
<tr>
<td>6</td>
<td><code>REF2</code></td>
<td>reference sequence in the file 2;</td>
</tr>
<tr>
<td>7</td>
<td><code>ALT1</code></td>
<td>alternative sequence in the file 1;</td>
</tr>
<tr>
<td>8</td>
<td><code>ALT2</code></td>
<td>alternative sequence in the file 2</td>
</tr>
</tbody>
</table>
<p><code>known_snps.vcf</code></p>
<p>Variants that are common to RNAseq and "known" variants from DNA.</p>
<h3 id="allele-specific-expression-quantification">Allele specific expression quantification<a class="headerlink" href="#allele-specific-expression-quantification" title="Permanent link">&para;</a></h3>
<p><code>ASE.tsv</code></p>
<p>Tab-separated file with allele counts at common SNVs positions (only SNVs from the file <code>known_snps.vcf</code>)</p>
<p>The file is sorted by coordinates and contains 13 fields:</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td><code>contig</code></td>
<td>contig, scaffold or chromosome name of the variant</td>
</tr>
<tr>
<td>2</td>
<td><code>position</code></td>
<td>position of the variant</td>
</tr>
<tr>
<td>3</td>
<td><code>variant ID</code></td>
<td>variant ID in the dbSNP</td>
</tr>
<tr>
<td>4</td>
<td><code>refAllele</code></td>
<td>reference allele sequence</td>
</tr>
<tr>
<td>5</td>
<td><code>altAllele</code></td>
<td>alternate allele sequence</td>
</tr>
<tr>
<td>6</td>
<td><code>refCount</code></td>
<td>number of reads that support the reference allele</td>
</tr>
<tr>
<td>7</td>
<td><code>altCount</code></td>
<td>number of reads that support the alternate allele</td>
</tr>
<tr>
<td>8</td>
<td><code>totalCount</code></td>
<td>total number of reads at the site that support both reference and alternate allele and any other alleles present at the site</td>
</tr>
<tr>
<td>9</td>
<td><code>lowMAPQDepth</code></td>
<td>number of reads that have low mapping quality</td>
</tr>
<tr>
<td>10</td>
<td><code>lowBaseQDepth</code></td>
<td>number of reads that have low base quality</td>
</tr>
<tr>
<td>11</td>
<td><code>rawDepth</code></td>
<td>total number of reads at the site that support both reference and alternate allele and any other alleles present at the site</td>
</tr>
<tr>
<td>12</td>
<td><code>otherBases</code></td>
<td>number of reads that support bases other than reference and alternate bases</td>
</tr>
<tr>
<td>13</td>
<td><code>improperPairs</code></td>
<td>number of reads that have malformed pairs</td>
</tr>
</tbody>
</table>
<h3 id="allele-frequency-histogram">Allele frequency histogram<a class="headerlink" href="#allele-frequency-histogram" title="Permanent link">&para;</a></h3>
<p><code>AF.histogram.pdf</code></p>
<p>This file contains a histogram plot of allele frequency for SNVs common to RNA-seq and "known" variants from DNA.</p>
<h2 id="bonus-step">Bonus step<a class="headerlink" href="#bonus-step" title="Permanent link">&para;</a></h2>
<p>Until now the pipeline has been executed using just a single sample (<code>ENCSR000COQ1</code>).</p>
<p>Now we can re-execute the pipeline specifying a large set of samples by using the command shown below:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-40-1" name="__codelineno-40-1" href="#__codelineno-40-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>main.nf<span class="w"> </span>-resume<span class="w"> </span>--reads<span class="w"> </span><span class="s1">&#39;data/reads/ENCSR000C*_{1,2}.fastq.gz&#39;</span>
</code></pre></div>
<p>Or run the final version of the Nextflow pipeline that is already prepared for you:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-41-1" name="__codelineno-41-1" href="#__codelineno-41-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>final_main.nf<span class="w"> </span>-resume
</code></pre></div>
<p>It will immediately print an output similar to the one below:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-42-1" name="__codelineno-42-1" href="#__codelineno-42-1"></a><span class="go">Launching `final_main.nf` [evil_almeida] DSL2 - revision: 09802c02ae</span>
<a id="__codelineno-42-2" name="__codelineno-42-2" href="#__codelineno-42-2"></a><span class="go">executor &gt;  local (5)</span>
<a id="__codelineno-42-3" name="__codelineno-42-3" href="#__codelineno-42-3"></a><span class="go">[6d/fafc39] process &gt; prepare_genome_samtools                [100%] 1 of 1, cached: 1 ✔</span>
<a id="__codelineno-42-4" name="__codelineno-42-4" href="#__codelineno-42-4"></a><span class="go">[38/5e729e] process &gt; prepare_genome_picard                  [100%] 1 of 1, cached: 1 ✔</span>
<a id="__codelineno-42-5" name="__codelineno-42-5" href="#__codelineno-42-5"></a><span class="go">[cb/7d9286] process &gt; prepare_star_genome_index              [100%] 1 of 1, cached: 1 ✔</span>
<a id="__codelineno-42-6" name="__codelineno-42-6" href="#__codelineno-42-6"></a><span class="go">[f7/3b5953] process &gt; prepare_vcf_file                       [100%] 1 of 1, cached: 1 ✔</span>
<a id="__codelineno-42-7" name="__codelineno-42-7" href="#__codelineno-42-7"></a><span class="go">[29/83ed80] process &gt; rnaseq_mapping_star (3)                [ 16%] 1 of 6, cached: 1</span>
<a id="__codelineno-42-8" name="__codelineno-42-8" href="#__codelineno-42-8"></a><span class="go">[e6/332dcc] process &gt; rnaseq_gatk_splitNcigar (ENCSR000COQ1) [100%] 1 of 1, cached: 1</span>
<a id="__codelineno-42-9" name="__codelineno-42-9" href="#__codelineno-42-9"></a><span class="go">[9d/359eac] process &gt; rnaseq_gatk_recalibrate (ENCSR000COQ1) [100%] 1 of 1, cached: 1</span>
<a id="__codelineno-42-10" name="__codelineno-42-10" href="#__codelineno-42-10"></a><span class="go">[-        ] process &gt; rnaseq_call_variants                   -</span>
<a id="__codelineno-42-11" name="__codelineno-42-11" href="#__codelineno-42-11"></a><span class="go">[-        ] process &gt; post_process_vcf                       -</span>
<a id="__codelineno-42-12" name="__codelineno-42-12" href="#__codelineno-42-12"></a><span class="go">[-        ] process &gt; prepare_vcf_for_ase                    -</span>
<a id="__codelineno-42-13" name="__codelineno-42-13" href="#__codelineno-42-13"></a><span class="go">[-        ] process &gt; ASE_knownSNPs                          -</span>
</code></pre></div>
<p>As you can see, it won't process everything from scratch. Even though we're running a different Nextflwo script, cache from previous runs will be taken into consideration.</p>
<p>At the end, you should see something like the output below:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-43-1" name="__codelineno-43-1" href="#__codelineno-43-1"></a><span class="go">N E X T F L O W  ~  version 23.04.3</span>
<a id="__codelineno-43-2" name="__codelineno-43-2" href="#__codelineno-43-2"></a><span class="go">Launching `final_main.nf` [cranky_rosalind] DSL2 - revision: 09802c02ae</span>
<a id="__codelineno-43-3" name="__codelineno-43-3" href="#__codelineno-43-3"></a><span class="go">executor &gt;  local (34)</span>
<a id="__codelineno-43-4" name="__codelineno-43-4" href="#__codelineno-43-4"></a><span class="go">[0a/e7bf16] process &gt; prepare_genome_samtools                [100%] 1 of 1 ✔</span>
<a id="__codelineno-43-5" name="__codelineno-43-5" href="#__codelineno-43-5"></a><span class="go">[87/233b95] process &gt; prepare_genome_picard                  [100%] 1 of 1 ✔</span>
<a id="__codelineno-43-6" name="__codelineno-43-6" href="#__codelineno-43-6"></a><span class="go">[25/228128] process &gt; prepare_star_genome_index              [100%] 1 of 1 ✔</span>
<a id="__codelineno-43-7" name="__codelineno-43-7" href="#__codelineno-43-7"></a><span class="go">[e3/ef7c40] process &gt; prepare_vcf_file                       [100%] 1 of 1 ✔</span>
<a id="__codelineno-43-8" name="__codelineno-43-8" href="#__codelineno-43-8"></a><span class="go">[14/3eccab] process &gt; rnaseq_mapping_star (6)                [100%] 6 of 6 ✔</span>
<a id="__codelineno-43-9" name="__codelineno-43-9" href="#__codelineno-43-9"></a><span class="go">[71/ac2c3e] process &gt; rnaseq_gatk_splitNcigar (ENCSR000CPO2) [100%] 6 of 6 ✔</span>
<a id="__codelineno-43-10" name="__codelineno-43-10" href="#__codelineno-43-10"></a><span class="go">[79/47a057] process &gt; rnaseq_gatk_recalibrate (ENCSR000CPO2) [100%] 6 of 6 ✔</span>
<a id="__codelineno-43-11" name="__codelineno-43-11" href="#__codelineno-43-11"></a><span class="go">[36/75c49b] process &gt; rnaseq_call_variants (ENCSR000CPO)     [100%] 3 of 3 ✔</span>
<a id="__codelineno-43-12" name="__codelineno-43-12" href="#__codelineno-43-12"></a><span class="go">[22/c86980] process &gt; post_process_vcf (ENCSR000CPO)         [100%] 3 of 3 ✔</span>
<a id="__codelineno-43-13" name="__codelineno-43-13" href="#__codelineno-43-13"></a><span class="go">[11/e66735] process &gt; prepare_vcf_for_ase (ENCSR000CPO)      [100%] 3 of 3 ✔</span>
<a id="__codelineno-43-14" name="__codelineno-43-14" href="#__codelineno-43-14"></a><span class="go">[7d/b7376e] process &gt; ASE_knownSNPs (ENCSR000CPO)            [100%] 3 of 3 ✔</span>
</code></pre></div>
<p>You can notice that this time the pipeline spawns the execution of more tasks because three samples have been provided instead of one.</p>
<p>This shows the ability of Nextflow to implicitly handle multiple parallel task executions depending on the specified pipeline input dataset.</p>
<p>A fully functional version of this pipeline is available at the following GitHub repository: <a href="https://github.com/CRG-CNAG/CalliNGS-NF">CalliNGS-NF</a>.</p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  <div>
    Copyright &copy; 2020-2025,
    <a href="https://seqera.io" target="_blank" rel="noopener">Seqera</a>.
  </div>
  <div class="footer_cc_link">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512">
      <!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.-->
      <path
        d="m245.83 214.87-33.22 17.28c-9.43-19.58-25.24-19.93-27.46-19.93-22.13 0-33.22 14.61-33.22 43.84 0 23.57 9.21 43.84 33.22 43.84 14.47 0 24.65-7.09 30.57-21.26l30.55 15.5c-6.17 11.51-25.69 38.98-65.1 38.98-22.6 0-73.96-10.32-73.96-77.05 0-58.69 43-77.06 72.63-77.06 30.72-.01 52.7 11.95 65.99 35.86zm143.05 0-32.78 17.28c-9.5-19.77-25.72-19.93-27.9-19.93-22.14 0-33.22 14.61-33.22 43.84 0 23.55 9.23 43.84 33.22 43.84 14.45 0 24.65-7.09 30.54-21.26l31 15.5c-2.1 3.75-21.39 38.98-65.09 38.98-22.69 0-73.96-9.87-73.96-77.05 0-58.67 42.97-77.06 72.63-77.06 30.71-.01 52.58 11.95 65.56 35.86zM247.56 8.05C104.74 8.05 0 123.11 0 256.05c0 138.49 113.6 248 247.56 248 129.93 0 248.44-100.87 248.44-248 0-137.87-106.62-248-248.44-248zm.87 450.81c-112.54 0-203.7-93.04-203.7-202.81 0-105.42 85.43-203.27 203.72-203.27 112.53 0 202.82 89.46 202.82 203.26-.01 121.69-99.68 202.82-202.84 202.82z"
      ></path>
    </svg>
    <a
      href="https://creativecommons.org/licenses/by-nc-nd/4.0/"
      target="_blank"
      rel="noopener"
      title="Creative Commons CC BY-NC-ND 4.0"
    >
      Creative Commons BY-NC-ND 4.0</a
    >.
  </div>
   Made with
  <a
    href="https://squidfunk.github.io/mkdocs-material/"
    target="_blank"
    rel="noopener"
  >
    Material for MkDocs
  </a>
  
</div>
      
        <div class="md-social">
  <a
    href="https://seqera.io"
    target="_blank"
    rel="noopener"
    class="footer_seqera_logo"
  >
    <img src="/assets/img/seqera_logo_white.png" />
  </a>
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
      <div class="md-consent" data-md-component="consent" id="__consent" hidden>
        <div class="md-consent__overlay"></div>
        <aside class="md-consent__inner">
          <form class="md-consent__form md-grid md-typeset" name="consent">
            


  


<h4>Cookie consent</h4>
<p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our training materials better.</p>
<input class="md-toggle" type="checkbox" id="__settings" >
<div class="md-consent__settings">
  <ul class="task-list">
    
    
      
        
  
  
    
    
  
  <li class="task-list-item">
    <label class="task-list-control">
      <input type="checkbox" name="analytics" checked>
      <span class="task-list-indicator"></span>
      Google Analytics
    </label>
  </li>

      
    
    
      
        
  
  
    
    
  
  <li class="task-list-item">
    <label class="task-list-control">
      <input type="checkbox" name="github" checked>
      <span class="task-list-indicator"></span>
      GitHub
    </label>
  </li>

      
    
    
  </ul>
</div>
<div class="md-consent__controls">
  
    
      <button class="md-button md-button--primary">Accept</button>
    
    
    
  
    
    
    
      <label class="md-button" for="__settings">Manage settings</label>
    
  
</div>
          </form>
        </aside>
      </div>
      <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout((function(){document.querySelector("[data-md-component=consent]").hidden=!1}),250);var form=document.forms.consent;for(var action of["submit","reset"])form.addEventListener(action,(function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map((function(e){return[e,!0]})))),location.hash="",location.reload()}))</script>
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["content.action.edit", "content.code.annotate", "content.code.copy", "navigation.footer", "navigation.instant", "navigation.top", "navigation.tracking", "search.share", "toc.follow"], "search": "../../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"alias": true, "provider": "mike"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>