
<!doctype html>
<html lang="pt" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://training.nextflow.io/0.dev/pt/nf4_science/genomics/01_per_sample_variant_calling/">
      
      
        <link rel="prev" href="../00_orientation/">
      
      
        <link rel="next" href="../02_joint_calling/">
      
      
      <link rel="icon" href="../../../../assets/img/nextflow-icon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Part 1: Per-sample variant calling - training.nextflow.io</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/fonts.css">
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-244N3GEN75"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-244N3GEN75",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-244N3GEN75",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>if("undefined"!=typeof __md_analytics){var consent=__md_get("__consent");consent&&consent.analytics&&__md_analytics()}</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#part-1-per-sample-variant-calling" class="md-skip">
          Ir para o conteúdo
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
            
<div style="text-align: center">
  Nextflow Training Week (2025-Q1):
  <strong>March 10-14, 2025</strong>
  <a
    class="md-button"
    href="https://community.seqera.io/t/nextflow-training-week-2025-q1/1775/6"
    target="_blank"
  >
    More info
    <span class="twemoji"
      ><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
        <path
          d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"
        ></path></svg
    ></span>
  </a>
</div>
<style type="text/css">
  .md-banner {
    display: none;
  }
</style>

<script>
  (function () {
    function hideAnnouncement() {
      var announcementDate = new Date(
        "2025-03-15T00:00:00Z",
      );
      var currentDate = new Date();
      var banner = document.querySelector(".md-banner");
      if (banner && currentDate <= announcementDate) {
        banner.style.display = "block";
      }
    }

    // Run on initial load
    hideAnnouncement();

    // Hook into XHR to run after each page update
    var originalOpen = XMLHttpRequest.prototype.open;
    XMLHttpRequest.prototype.open = function () {
      this.addEventListener("load", function () {
        setTimeout(hideAnnouncement, 0); // Ensure it runs after the request completes
      });
      return originalOpen.apply(this, arguments);
    };
  })();
</script>
 
          </div>
          
        </aside>
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Cabeçalho">
    <a href="../../../" title="training.nextflow.io" class="md-header__button md-logo" aria-label="training.nextflow.io" data-md-component="logo">
      
  <img src="../../../../assets/img/nextflow_logo_dark.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            training.nextflow.io
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Part 1: Per-sample variant calling
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
      <div class="md-header__option">
  <div class="md-select">
    
    <button class="md-header__button md-icon" aria-label="Selecione o idioma">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12.87 15.07-2.54-2.51.03-.03A17.5 17.5 0 0 0 14.07 6H17V4h-7V2H8v2H1v2h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2zm-2.62 7 1.62-4.33L19.12 17z"/></svg>
    </button>
    <div class="md-select__inner">
      <ul class="md-select__list">
        
          <li class="md-select__item">
            <a href="../../../../nf4_science/genomics/01_per_sample_variant_calling/" hreflang="en" class="md-select__link">
              English
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="./" hreflang="pt" class="md-select__link">
              Português
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../es/nf4_science/genomics/01_per_sample_variant_calling/" hreflang="es" class="md-select__link">
              Español
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../fr/nf4_science/genomics/01_per_sample_variant_calling/" hreflang="fr" class="md-select__link">
              Français
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../it/nf4_science/genomics/01_per_sample_variant_calling/" hreflang="it" class="md-select__link">
              Italiano
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../../../ko/nf4_science/genomics/01_per_sample_variant_calling/" hreflang="ko" class="md-select__link">
              Korean
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</div>
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Buscar" placeholder="Buscar" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Pesquisar">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Compartilhar" aria-label="Compartilhar" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Limpar" aria-label="Limpar" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Inicializando a pesquisa
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/nextflow-io/training" title="Ir ao repositório" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    nextflow-io/training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navegação" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../" title="training.nextflow.io" class="md-nav__button md-logo" aria-label="training.nextflow.io" data-md-component="logo">
      
  <img src="../../../../assets/img/nextflow_logo_dark.png" alt="logo">

    </a>
    training.nextflow.io
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/nextflow-io/training" title="Ir ao repositório" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    nextflow-io/training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Environment setup
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Environment setup
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuração do ambiente
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/01_setup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Gitpod
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/02_local/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Instalação local
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../envsetup/03_devcontainer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local installation using VSCode Devcontainers extension
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Hello Nextflow
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Hello Nextflow
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hello Nextflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientação
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/01_hello_world/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Hello World
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/02_hello_channels/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Hello Channels
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/03_hello_workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Hello Workflow
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/04_hello_modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 4: Hello Modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/05_hello_containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 5: Hello Containers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/06_hello_config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 6: Hello Config
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nextflow/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Hello nf-core
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Hello nf-core
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hello nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/01_run_demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Run a demo pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/02_rewrite_hello/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Rewrite Hello for nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../hello_nf-core/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Nextflow for Genomics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Nextflow for Genomics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow for Genomics
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Part 1: Per-sample variant calling
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Part 1: Per-sample variant calling
    
  </span>
  

      </a>
      
          

<nav class="md-nav md-nav--secondary" aria-label="On this page">
       
  <label class="md-nav__title" for="__toc">
    <span class="md-nav__icon md-icon"></span>
    On this page
  </label>
  <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
     <li class="md-nav__item">
  <a href="#method-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Method overview
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Dataset
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#0-warmup-test-the-samtools-and-gatk-commands-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0. Warmup: Test the Samtools and GATK commands interactively
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0. Warmup: Test the Samtools and GATK commands interactively">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#01-index-a-bam-input-file-with-samtools" class="md-nav__link">
    <span class="md-ellipsis">
      0.1. Index a BAM input file with Samtools
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.1. Index a BAM input file with Samtools">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#011-pull-the-samtools-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.1. Pull the Samtools container
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#012-spin-up-the-samtools-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.2. Spin up the Samtools container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#013-run-the-indexing-command" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.3. Run the indexing command
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#014-exit-the-samtools-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.4. Exit the Samtools container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#02-call-variants-with-gatk-haplotypecaller" class="md-nav__link">
    <span class="md-ellipsis">
      0.2. Call variants with GATK HaplotypeCaller
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.2. Call variants with GATK HaplotypeCaller">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#021-pull-the-gatk-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.1. Pull the GATK container
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#022-spin-up-the-gatk-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.2. Spin up the GATK container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#023-run-the-variant-calling-command" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.3. Run the variant calling command
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#024-exit-the-gatk-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.4. Exit the GATK container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#1-write-a-single-stage-workflow-that-runs-samtools-index-on-a-bam-file" class="md-nav__link">
    <span class="md-ellipsis">
      1. Write a single-stage workflow that runs Samtools index on a BAM file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Write a single-stage workflow that runs Samtools index on a BAM file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-define-the-indexing-process" class="md-nav__link">
    <span class="md-ellipsis">
      1.1. Define the indexing process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-add-an-input-and-output-parameter-declaration" class="md-nav__link">
    <span class="md-ellipsis">
      1.2. Add an input and output parameter declaration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-add-workflow-block-to-run-samtools_index" class="md-nav__link">
    <span class="md-ellipsis">
      1.3. Add workflow block to run SAMTOOLS_INDEX
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-run-the-workflow-to-verify-that-the-indexing-step-works" class="md-nav__link">
    <span class="md-ellipsis">
      1.4. Run the workflow to verify that the indexing step works
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_1" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_1" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#2-add-a-second-process-to-run-gatk-haplotypecaller-on-the-indexed-bam-file" class="md-nav__link">
    <span class="md-ellipsis">
      2. Add a second process to run GATK HaplotypeCaller on the indexed BAM file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Add a second process to run GATK HaplotypeCaller on the indexed BAM file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-define-the-variant-calling-process" class="md-nav__link">
    <span class="md-ellipsis">
      2.1. Define the variant calling process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-add-definitions-for-accessory-inputs" class="md-nav__link">
    <span class="md-ellipsis">
      2.2. Add definitions for accessory inputs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-create-variables-to-hold-the-accessory-file-paths" class="md-nav__link">
    <span class="md-ellipsis">
      2.3. Create variables to hold the accessory file paths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-add-a-call-to-the-workflow-block-to-run-gatk_haplotypecaller" class="md-nav__link">
    <span class="md-ellipsis">
      2.4. Add a call to the workflow block to run GATK_HAPLOTYPECALLER
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#25-run-the-workflow-to-verify-that-the-variant-calling-step-works" class="md-nav__link">
    <span class="md-ellipsis">
      2.5. Run the workflow to verify that the variant calling step works
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_2" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_2" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#3-adapt-the-workflow-to-run-on-a-batch-of-samples" class="md-nav__link">
    <span class="md-ellipsis">
      3. Adapt the workflow to run on a batch of samples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Adapt the workflow to run on a batch of samples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31-turn-the-input-parameter-declaration-into-an-array-listing-the-three-samples" class="md-nav__link">
    <span class="md-ellipsis">
      3.1. Turn the input parameter declaration into an array listing the three samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-run-the-workflow-to-verify-that-it-runs-on-all-three-samples" class="md-nav__link">
    <span class="md-ellipsis">
      3.2. Run the workflow to verify that it runs on all three samples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3.2. Run the workflow to verify that it runs on all three samples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#321-check-the-work-directories-for-the-relevant-calls" class="md-nav__link">
    <span class="md-ellipsis">
      3.2.1. Check the work directories for the relevant calls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#321-use-the-view-operator-to-inspect-channel-contents" class="md-nav__link">
    <span class="md-ellipsis">
      3.2.1. Use the view() operator to inspect channel contents
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33-change-the-output-of-the-samtools_index-process-into-a-tuple-that-keeps-the-input-file-and-its-index-together" class="md-nav__link">
    <span class="md-ellipsis">
      3.3. Change the output of the SAMTOOLS_INDEX process into a tuple that keeps the input file and its index together
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34-change-the-input-to-the-gatk_haplotypecaller-process-to-be-a-tuple" class="md-nav__link">
    <span class="md-ellipsis">
      3.4. Change the input to the GATK_HAPLOTYPECALLER process to be a tuple
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#35-update-the-call-to-gatk_haplotypecaller-in-the-workflow-block" class="md-nav__link">
    <span class="md-ellipsis">
      3.5. Update the call to GATK_HAPLOTYPECALLER in the workflow block
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#36-run-the-workflow-to-verify-it-works-correctly-on-all-three-samples-every-time" class="md-nav__link">
    <span class="md-ellipsis">
      3.6. Run the workflow to verify it works correctly on all three samples every time
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_3" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_3" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#4-make-the-workflow-accept-a-text-file-containing-a-batch-of-input-files" class="md-nav__link">
    <span class="md-ellipsis">
      4. Make the workflow accept a text file containing a batch of input files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Make the workflow accept a text file containing a batch of input files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41-examine-the-provided-text-file-listing-the-input-file-paths" class="md-nav__link">
    <span class="md-ellipsis">
      4.1. Examine the provided text file listing the input file paths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-update-the-parameter-default" class="md-nav__link">
    <span class="md-ellipsis">
      4.2. Update the parameter default
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#43-update-the-channel-factory-to-read-lines-from-a-file" class="md-nav__link">
    <span class="md-ellipsis">
      4.3. Update the channel factory to read lines from a file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#44-run-the-workflow-to-verify-that-it-works-correctly" class="md-nav__link">
    <span class="md-ellipsis">
      4.4. Run the workflow to verify that it works correctly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_4" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_4" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li> 
  </ul>
  
</nav>
<nav class="md-nav md-nav--secondary sidebar_resources" aria-label="Resources">
  <label class="md-nav__title">Resources</label>
  <ul class="md-nav__list">
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/docs/latest/index.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M17.5 14.33c.79 0 1.63.08 2.5.24v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.33-4.5.99v-1.69c1.17-.53 2.67-.8 4.5-.8M13 12.46c1.29-.53 2.79-.79 4.5-.79.79 0 1.63.07 2.5.23v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.34-4.5.99m4.5-3.65c-1.9 0-3.39.32-4.5 1V9.84c1.23-.56 2.73-.84 4.5-.84.79 0 1.63.08 2.5.23v1.55c-.74-.19-1.59-.28-2.5-.28m3.5 8V7c-1.04-.33-2.21-.5-3.5-.5-2.05 0-3.88.5-5.5 1.5v11.5c1.62-1 3.45-1.5 5.5-1.5 1.19 0 2.36.16 3.5.5m-3.5-14c2.35 0 4.19.5 5.5 1.5v14.56c0 .12-.05.24-.16.35-.11.09-.23.17-.34.17-.11 0-.19-.02-.25-.05-1.28-.69-2.87-1.03-4.75-1.03-2.05 0-3.88.5-5.5 1.5-1.34-1-3.17-1.5-5.5-1.5-1.66 0-3.25.36-4.75 1.07-.03.01-.07.01-.12.03-.04.01-.08.02-.13.02-.11 0-.23-.04-.34-.12a.475.475 0 0 1-.16-.35V6c1.34-1 3.18-1.5 5.5-1.5 2.33 0 4.16.5 5.5 1.5 1.34-1 3.17-1.5 5.5-1.5Z"
          ></path>
        </svg>
        Nextflow Docs
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://www.nextflow.io/slack-invite.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M6 15a2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2h2v2m1 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-5m2-8a2 2 0 0 1-2-2 2 2 0 0 1 2-2 2 2 0 0 1 2 2v2H9m0 1a2 2 0 0 1 2 2 2 2 0 0 1-2 2H4a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5m8 2a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2h-2v-2m-1 0a2 2 0 0 1-2 2 2 2 0 0 1-2-2V5a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5m-2 8a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2v-2h2m0-1a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-5Z"
          ></path>
        </svg>
        Nextflow Slack
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="80 50 350 350"
        >
          <style type="text/css">
            .st0 {
              fill: "gray";
            }
          </style>
          <path
            class="st0"
            d="M124.5,171.5c39.4,2.6,71.1,34.3,73.5,73.8h46.1c-2.6-65-54.7-117.2-119.6-119.9V171.5z"
          />
          <path
            class="st0"
            d="M198.3,255.8c-2.6,39.5-34.3,71.1-73.8,73.5v46.1c65-2.6,117.2-54.7,119.9-119.6H198.3z"
          />
          <path
            class="st0"
            d="M300.7,245c2.6-39.5,34.3-71.1,73.8-73.5v-46.1c-65,2.6-117.2,54.7-119.9,119.6H300.7z"
          />
          <path
            class="st0"
            d="M374.5,329.3c-39.4-2.6-71.1-34.3-73.5-73.8h-46.1c2.6,65,54.7,117.2,119.6,119.9V329.3z"
          />
        </svg>
        Nextflow Homepage
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://seqera.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <?xml version="1.0" encoding="utf-8"?>
        <!-- Generator: Adobe Illustrator 28.3.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          x="0px"
          y="0px"
          viewBox="80 50 350 350"
          style="enable-background: new 0 0 500 500"
          xml:space="preserve"
        >
          <style type="text/css">
            .st0 {
              clip-path: url(#SVGID_00000006699037935518914080000000086571810102047141_);
            }
            .st1 {
              fill: "gray";
            }
          </style>
          <g>
            <defs>
              <rect id="SVGID_1_" x="125" y="125" width="250" height="250" />
            </defs>
            <clipPath
              id="SVGID_00000183214694142566821210000012389151436317298101_"
            >
              <use xlink:href="#SVGID_1_" style="overflow: visible" />
            </clipPath>
            <g
              style="
                clip-path: url(#SVGID_00000183214694142566821210000012389151436317298101_);
              "
            >
              <path
                class="st1"
                d="M375,333l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.2-0.9l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7L365,308l-0.6-0.6l-0.6-0.6l-0.6-0.6l-0.6-0.6L362,305
                      l-0.6-0.6l-0.7-0.6l-0.6-0.5l-1.2,0.7l-0.8,0.5l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.9,0.4l-0.9,0.4l-0.9,0.3l-0.9,0.3
                      l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0.1
                      l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0l-0.9,0h-57.5l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.3l0.1,0.3l0.1,0.4l0.1,0.4
                      l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4l-0.1,0.3l-0.2,0.3l-0.2,0.3
                      l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.1l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0H125V375h209.7l0.9,0l0.9,0l0.9,0l0.8-0.1
                      l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3
                      l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7
                      l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8
                      l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0-0.9l0-0.9l0-0.8L375,333z"
              />
              <path
                class="st1"
                d="M375,260.4l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.1-0.8l-0.1-0.8l-0.2-0.8l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.7
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.2-0.2h0h-89.8l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.4l0.1,0.3
                      l0.1,0.4l0.1,0.4l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4
                      l-0.2,0.4l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0h-82.5h-111H125v32.2
                      h122.8h86.9h0.9l0.9,0l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3
                      l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.4l0.8-0.4l0.2-0.1l0.6-0.3l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.5-0.7l0.5-0.7l0.5-0.7
                      l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7l0.4-0.7l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8
                      l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0-0.8l0-0.9l0-0.9
                      L375,260.4z"
              />
              <path
                class="st1"
                d="M165.3,197.6l-0.9,0l-0.8,0l-0.9,0l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.2,0.1l-0.6,0.3l-0.8,0.4
                      l-0.8,0.4l-0.8,0.4l-0.7,0.4l-0.7,0.5l-0.7,0.5L142,205l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.8l0,0.9l0.1,0.8l0.1,0.8
                      l0.1,0.9l0.1,0.9l0.1,0.8l0.2,0.9l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.2,0.2h89.8l-59.7-16.4l-0.4-0.1l-0.3-0.1
                      l-0.4-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3
                      l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4
                      l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3
                      l0.3-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.4-0.2l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0
                      h0.4h82.5h111H375v-32.3H252.2H165.3z"
              />
              <path
                class="st1"
                d="M165.3,125l-0.9,0l-0.8,0l-0.9,0.1l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4
                      l-0.8,0.4l-0.7,0.5l-0.7,0.4l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.9l0,0.8l0.1,0.9l0.1,0.9
                      l0.1,0.8l0.1,0.8l0.1,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.6,0.6l0.6,0.6
                      l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.7,0.6l0.6,0.5h0l1.1-0.7l0.8-0.5l0.8-0.5l0.8-0.4l0.8-0.4l0.9-0.4l0.9-0.4l0.9-0.4
                      l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.1l0.9-0.1l0.9-0.1l0.9-0.1
                      l0.9-0.1l0.9-0.1l0.9,0l0.9,0l0.9,0h57.5l-59.7-16.4l-0.4-0.1l-0.3-0.1l-0.4-0.1l-0.3-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2
                      l-0.3-0.2l-0.3-0.2l-0.3-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.3l-0.1-0.4
                      l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3
                      l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2
                      l0.3-0.2l0.4-0.1l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0l0.4,0H375V125H165.3z"
              />
            </g>
          </g>
        </svg>
        Seqera
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://github.com/nextflow-io/training"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"
          ></path>
        </svg>
        Training on GitHub
      </a>
    </li>
  </ul>
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../02_joint_calling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Joint calling on a cohort
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../03_modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Moving code into modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../04_testing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 4: Adding tests
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Nextflow for RNAseq
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Nextflow for RNAseq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Nextflow for RNAseq
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/00_orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/01_method/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 1: Method overview and manual testing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/02_single-sample/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 2: Single-sample implementation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/03_multi-sample/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Part 3: Multi-sample paired-end implementation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/survey/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback survey
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../rnaseq/next_steps/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Next Steps
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Side Quests
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Side Quests
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Side Quests
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/workflows_of_workflows/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflows of Workflows
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/splitting_and_grouping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Splitting and Grouping
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/nf-test/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Testing with nf-test
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../side_quests/nf-core/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to nf-core
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Fundamentals Training
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Fundamentals Training
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Index
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientação
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Intro
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/rnaseq_pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Rnaseq pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/containers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Containers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/channels/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Channels
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/processes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Processes
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/operators/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Operators
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/groovy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introdução ao Groovy
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/modules/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Modules
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/config/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuração
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/executors/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Executors
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/seqera_platform/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Seqera Platform
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/cache_and_resume/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cache e reentrância
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../basic_training/debugging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Resolução de problemas
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Advanced Training
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            Advanced Training
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Advanced Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/orientation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Orientation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/operators/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Operator Tour
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/metadata/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Metadata Propagation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/grouping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Grouping and Splitting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/groovy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Groovy Imports
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/structure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow Structure
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/configuration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/summary/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Summary
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../advanced/support/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Support
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../help/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Help
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                      

<nav class="md-nav md-nav--secondary" aria-label="On this page">
       
  <label class="md-nav__title" for="__toc">
    <span class="md-nav__icon md-icon"></span>
    On this page
  </label>
  <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
     <li class="md-nav__item">
  <a href="#method-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Method overview
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Dataset
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
</li>  <li class="md-nav__item">
  <a href="#0-warmup-test-the-samtools-and-gatk-commands-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0. Warmup: Test the Samtools and GATK commands interactively
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0. Warmup: Test the Samtools and GATK commands interactively">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#01-index-a-bam-input-file-with-samtools" class="md-nav__link">
    <span class="md-ellipsis">
      0.1. Index a BAM input file with Samtools
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.1. Index a BAM input file with Samtools">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#011-pull-the-samtools-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.1. Pull the Samtools container
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#012-spin-up-the-samtools-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.2. Spin up the Samtools container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#013-run-the-indexing-command" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.3. Run the indexing command
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#014-exit-the-samtools-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.1.4. Exit the Samtools container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#02-call-variants-with-gatk-haplotypecaller" class="md-nav__link">
    <span class="md-ellipsis">
      0.2. Call variants with GATK HaplotypeCaller
    </span>
  </a>
  
    <nav class="md-nav" aria-label="0.2. Call variants with GATK HaplotypeCaller">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#021-pull-the-gatk-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.1. Pull the GATK container
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#022-spin-up-the-gatk-container-interactively" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.2. Spin up the GATK container interactively
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#023-run-the-variant-calling-command" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.3. Run the variant calling command
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#024-exit-the-gatk-container" class="md-nav__link">
    <span class="md-ellipsis">
      0.2.4. Exit the GATK container
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#1-write-a-single-stage-workflow-that-runs-samtools-index-on-a-bam-file" class="md-nav__link">
    <span class="md-ellipsis">
      1. Write a single-stage workflow that runs Samtools index on a BAM file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Write a single-stage workflow that runs Samtools index on a BAM file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-define-the-indexing-process" class="md-nav__link">
    <span class="md-ellipsis">
      1.1. Define the indexing process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-add-an-input-and-output-parameter-declaration" class="md-nav__link">
    <span class="md-ellipsis">
      1.2. Add an input and output parameter declaration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-add-workflow-block-to-run-samtools_index" class="md-nav__link">
    <span class="md-ellipsis">
      1.3. Add workflow block to run SAMTOOLS_INDEX
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-run-the-workflow-to-verify-that-the-indexing-step-works" class="md-nav__link">
    <span class="md-ellipsis">
      1.4. Run the workflow to verify that the indexing step works
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_1" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_1" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#2-add-a-second-process-to-run-gatk-haplotypecaller-on-the-indexed-bam-file" class="md-nav__link">
    <span class="md-ellipsis">
      2. Add a second process to run GATK HaplotypeCaller on the indexed BAM file
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Add a second process to run GATK HaplotypeCaller on the indexed BAM file">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-define-the-variant-calling-process" class="md-nav__link">
    <span class="md-ellipsis">
      2.1. Define the variant calling process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-add-definitions-for-accessory-inputs" class="md-nav__link">
    <span class="md-ellipsis">
      2.2. Add definitions for accessory inputs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23-create-variables-to-hold-the-accessory-file-paths" class="md-nav__link">
    <span class="md-ellipsis">
      2.3. Create variables to hold the accessory file paths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24-add-a-call-to-the-workflow-block-to-run-gatk_haplotypecaller" class="md-nav__link">
    <span class="md-ellipsis">
      2.4. Add a call to the workflow block to run GATK_HAPLOTYPECALLER
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#25-run-the-workflow-to-verify-that-the-variant-calling-step-works" class="md-nav__link">
    <span class="md-ellipsis">
      2.5. Run the workflow to verify that the variant calling step works
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_2" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_2" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#3-adapt-the-workflow-to-run-on-a-batch-of-samples" class="md-nav__link">
    <span class="md-ellipsis">
      3. Adapt the workflow to run on a batch of samples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Adapt the workflow to run on a batch of samples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#31-turn-the-input-parameter-declaration-into-an-array-listing-the-three-samples" class="md-nav__link">
    <span class="md-ellipsis">
      3.1. Turn the input parameter declaration into an array listing the three samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#32-run-the-workflow-to-verify-that-it-runs-on-all-three-samples" class="md-nav__link">
    <span class="md-ellipsis">
      3.2. Run the workflow to verify that it runs on all three samples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3.2. Run the workflow to verify that it runs on all three samples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#321-check-the-work-directories-for-the-relevant-calls" class="md-nav__link">
    <span class="md-ellipsis">
      3.2.1. Check the work directories for the relevant calls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#321-use-the-view-operator-to-inspect-channel-contents" class="md-nav__link">
    <span class="md-ellipsis">
      3.2.1. Use the view() operator to inspect channel contents
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#33-change-the-output-of-the-samtools_index-process-into-a-tuple-that-keeps-the-input-file-and-its-index-together" class="md-nav__link">
    <span class="md-ellipsis">
      3.3. Change the output of the SAMTOOLS_INDEX process into a tuple that keeps the input file and its index together
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#34-change-the-input-to-the-gatk_haplotypecaller-process-to-be-a-tuple" class="md-nav__link">
    <span class="md-ellipsis">
      3.4. Change the input to the GATK_HAPLOTYPECALLER process to be a tuple
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#35-update-the-call-to-gatk_haplotypecaller-in-the-workflow-block" class="md-nav__link">
    <span class="md-ellipsis">
      3.5. Update the call to GATK_HAPLOTYPECALLER in the workflow block
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#36-run-the-workflow-to-verify-it-works-correctly-on-all-three-samples-every-time" class="md-nav__link">
    <span class="md-ellipsis">
      3.6. Run the workflow to verify it works correctly on all three samples every time
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_3" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_3" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>  <li class="md-nav__item">
  <a href="#4-make-the-workflow-accept-a-text-file-containing-a-batch-of-input-files" class="md-nav__link">
    <span class="md-ellipsis">
      4. Make the workflow accept a text file containing a batch of input files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Make the workflow accept a text file containing a batch of input files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#41-examine-the-provided-text-file-listing-the-input-file-paths" class="md-nav__link">
    <span class="md-ellipsis">
      4.1. Examine the provided text file listing the input file paths
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#42-update-the-parameter-default" class="md-nav__link">
    <span class="md-ellipsis">
      4.2. Update the parameter default
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#43-update-the-channel-factory-to-read-lines-from-a-file" class="md-nav__link">
    <span class="md-ellipsis">
      4.3. Update the channel factory to read lines from a file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#44-run-the-workflow-to-verify-that-it-works-correctly" class="md-nav__link">
    <span class="md-ellipsis">
      4.4. Run the workflow to verify that it works correctly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#takeaway_4" class="md-nav__link">
    <span class="md-ellipsis">
      Takeaway
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#whats-next_4" class="md-nav__link">
    <span class="md-ellipsis">
      What's next?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li> 
  </ul>
  
</nav>
<nav class="md-nav md-nav--secondary sidebar_resources" aria-label="Resources">
  <label class="md-nav__title">Resources</label>
  <ul class="md-nav__list">
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/docs/latest/index.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M17.5 14.33c.79 0 1.63.08 2.5.24v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.33-4.5.99v-1.69c1.17-.53 2.67-.8 4.5-.8M13 12.46c1.29-.53 2.79-.79 4.5-.79.79 0 1.63.07 2.5.23v1.5c-.62-.16-1.46-.24-2.5-.24-1.9 0-3.39.34-4.5.99m4.5-3.65c-1.9 0-3.39.32-4.5 1V9.84c1.23-.56 2.73-.84 4.5-.84.79 0 1.63.08 2.5.23v1.55c-.74-.19-1.59-.28-2.5-.28m3.5 8V7c-1.04-.33-2.21-.5-3.5-.5-2.05 0-3.88.5-5.5 1.5v11.5c1.62-1 3.45-1.5 5.5-1.5 1.19 0 2.36.16 3.5.5m-3.5-14c2.35 0 4.19.5 5.5 1.5v14.56c0 .12-.05.24-.16.35-.11.09-.23.17-.34.17-.11 0-.19-.02-.25-.05-1.28-.69-2.87-1.03-4.75-1.03-2.05 0-3.88.5-5.5 1.5-1.34-1-3.17-1.5-5.5-1.5-1.66 0-3.25.36-4.75 1.07-.03.01-.07.01-.12.03-.04.01-.08.02-.13.02-.11 0-.23-.04-.34-.12a.475.475 0 0 1-.16-.35V6c1.34-1 3.18-1.5 5.5-1.5 2.33 0 4.16.5 5.5 1.5 1.34-1 3.17-1.5 5.5-1.5Z"
          ></path>
        </svg>
        Nextflow Docs
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://www.nextflow.io/slack-invite.html"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M6 15a2 2 0 0 1-2 2 2 2 0 0 1-2-2 2 2 0 0 1 2-2h2v2m1 0a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-5m2-8a2 2 0 0 1-2-2 2 2 0 0 1 2-2 2 2 0 0 1 2 2v2H9m0 1a2 2 0 0 1 2 2 2 2 0 0 1-2 2H4a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5m8 2a2 2 0 0 1 2-2 2 2 0 0 1 2 2 2 2 0 0 1-2 2h-2v-2m-1 0a2 2 0 0 1-2 2 2 2 0 0 1-2-2V5a2 2 0 0 1 2-2 2 2 0 0 1 2 2v5m-2 8a2 2 0 0 1 2 2 2 2 0 0 1-2 2 2 2 0 0 1-2-2v-2h2m0-1a2 2 0 0 1-2-2 2 2 0 0 1 2-2h5a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-5Z"
          ></path>
        </svg>
        Nextflow Slack
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://nextflow.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="80 50 350 350"
        >
          <style type="text/css">
            .st0 {
              fill: "gray";
            }
          </style>
          <path
            class="st0"
            d="M124.5,171.5c39.4,2.6,71.1,34.3,73.5,73.8h46.1c-2.6-65-54.7-117.2-119.6-119.9V171.5z"
          />
          <path
            class="st0"
            d="M198.3,255.8c-2.6,39.5-34.3,71.1-73.8,73.5v46.1c65-2.6,117.2-54.7,119.9-119.6H198.3z"
          />
          <path
            class="st0"
            d="M300.7,245c2.6-39.5,34.3-71.1,73.8-73.5v-46.1c-65,2.6-117.2,54.7-119.9,119.6H300.7z"
          />
          <path
            class="st0"
            d="M374.5,329.3c-39.4-2.6-71.1-34.3-73.5-73.8h-46.1c2.6,65,54.7,117.2,119.6,119.9V329.3z"
          />
        </svg>
        Nextflow Homepage
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://seqera.io/"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <?xml version="1.0" encoding="utf-8"?>
        <!-- Generator: Adobe Illustrator 28.3.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
        <svg
          version="1.1"
          id="Layer_1"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          x="0px"
          y="0px"
          viewBox="80 50 350 350"
          style="enable-background: new 0 0 500 500"
          xml:space="preserve"
        >
          <style type="text/css">
            .st0 {
              clip-path: url(#SVGID_00000006699037935518914080000000086571810102047141_);
            }
            .st1 {
              fill: "gray";
            }
          </style>
          <g>
            <defs>
              <rect id="SVGID_1_" x="125" y="125" width="250" height="250" />
            </defs>
            <clipPath
              id="SVGID_00000183214694142566821210000012389151436317298101_"
            >
              <use xlink:href="#SVGID_1_" style="overflow: visible" />
            </clipPath>
            <g
              style="
                clip-path: url(#SVGID_00000183214694142566821210000012389151436317298101_);
              "
            >
              <path
                class="st1"
                d="M375,333l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.2-0.9l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7L365,308l-0.6-0.6l-0.6-0.6l-0.6-0.6l-0.6-0.6L362,305
                      l-0.6-0.6l-0.7-0.6l-0.6-0.5l-1.2,0.7l-0.8,0.5l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.9,0.4l-0.9,0.4l-0.9,0.3l-0.9,0.3
                      l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.3l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.2l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0.1
                      l-0.9,0.1l-0.9,0.1l-0.9,0.1l-0.9,0l-0.9,0h-57.5l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.3l0.1,0.3l0.1,0.4l0.1,0.4
                      l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4l-0.1,0.3l-0.2,0.3l-0.2,0.3
                      l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.1l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0H125V375h209.7l0.9,0l0.9,0l0.9,0l0.8-0.1
                      l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3
                      l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7
                      l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8
                      l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0.1-0.8l0-0.9l0-0.9l0-0.8L375,333z"
              />
              <path
                class="st1"
                d="M375,260.4l-0.1-0.9l-0.1-0.9l-0.1-0.8l-0.1-0.9l-0.1-0.8l-0.1-0.8l-0.1-0.8l-0.2-0.8l-0.2-0.8l-0.2-0.8
                      l-0.2-0.8l-0.2-0.8l-0.2-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.3-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.8l-0.4-0.7
                      l-0.4-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.5-0.7l-0.2-0.2h0h-89.8l59.7,16.4l0.4,0.1l0.4,0.1l0.4,0.1l0.3,0.2l0.3,0.2l0.3,0.2
                      l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.2l0.3,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.2,0.3l0.1,0.4l0.1,0.3
                      l0.1,0.4l0.1,0.4l0.1,0.4l0.1,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l0,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.4l-0.1,0.3l-0.1,0.4
                      l-0.2,0.4l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.2,0.3l-0.3,0.3l-0.3,0.3l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.3,0.2
                      l-0.3,0.2l-0.3,0.2l-0.3,0.2l-0.4,0.2l-0.4,0.1l-0.3,0.1l-0.4,0.1l-0.4,0.1l-0.4,0.1l-0.4,0l-0.4,0l-0.4,0h-82.5h-111H125v32.2
                      h122.8h86.9h0.9l0.9,0l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.9-0.1l0.8-0.1l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.2l0.8-0.3
                      l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.3l0.8-0.4l0.8-0.4l0.2-0.1l0.6-0.3l0.8-0.4l0.8-0.4l0.7-0.4l0.7-0.4l0.7-0.5l0.7-0.5l0.7-0.5
                      l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.5l0.7-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.6-0.6l0.5-0.7l0.5-0.7l0.5-0.7
                      l0.5-0.7l0.5-0.7l0.5-0.7l0.5-0.7l0.4-0.7l0.4-0.7l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.4-0.8l0.3-0.8l0.3-0.8l0.3-0.8l0.3-0.8
                      l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.2-0.8l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.8l0.1-0.9l0.1-0.9l0-0.8l0-0.9l0-0.9
                      L375,260.4z"
              />
              <path
                class="st1"
                d="M165.3,197.6l-0.9,0l-0.8,0l-0.9,0l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.2,0.1l-0.6,0.3l-0.8,0.4
                      l-0.8,0.4l-0.8,0.4l-0.7,0.4l-0.7,0.5l-0.7,0.5L142,205l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.8l0,0.9l0.1,0.8l0.1,0.8
                      l0.1,0.9l0.1,0.9l0.1,0.8l0.2,0.9l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.2,0.2h89.8l-59.7-16.4l-0.4-0.1l-0.3-0.1
                      l-0.4-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3
                      l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4
                      l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3
                      l0.3-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.4-0.2l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0
                      h0.4h82.5h111H375v-32.3H252.2H165.3z"
              />
              <path
                class="st1"
                d="M165.3,125l-0.9,0l-0.8,0l-0.9,0.1l-0.9,0.1l-0.8,0.1l-0.9,0.1l-0.8,0.1l-0.8,0.1l-0.8,0.2l-0.8,0.2l-0.8,0.2
                      l-0.8,0.2l-0.8,0.2l-0.8,0.2l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.3l-0.8,0.4l-0.8,0.4l-0.8,0.4l-0.8,0.4
                      l-0.8,0.4l-0.7,0.5l-0.7,0.4l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.5l-0.7,0.6l-0.7,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6
                      l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.6,0.6l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.5,0.7l-0.4,0.7l-0.5,0.7l-0.4,0.7
                      l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.4,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.3,0.8l-0.2,0.8l-0.2,0.8l-0.2,0.8
                      l-0.2,0.8l-0.2,0.8l-0.2,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l-0.1,0.8l-0.1,0.9l0,0.9l0,0.9v0.9v0.9l0,0.9l0,0.8l0.1,0.9l0.1,0.9
                      l0.1,0.8l0.1,0.8l0.1,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.2,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8l0.3,0.8
                      l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.8l0.4,0.7l0.5,0.7l0.4,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.5,0.7l0.6,0.6l0.6,0.6
                      l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.6,0.6l0.7,0.6l0.6,0.5h0l1.1-0.7l0.8-0.5l0.8-0.5l0.8-0.4l0.8-0.4l0.9-0.4l0.9-0.4l0.9-0.4
                      l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.3l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.2l0.9-0.1l0.9-0.1l0.9-0.1l0.9-0.1
                      l0.9-0.1l0.9-0.1l0.9,0l0.9,0l0.9,0h57.5l-59.7-16.4l-0.4-0.1l-0.3-0.1l-0.4-0.1l-0.3-0.1l-0.3-0.2l-0.3-0.2l-0.3-0.2l-0.3-0.2
                      l-0.3-0.2l-0.3-0.2l-0.3-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.2-0.3l-0.1-0.3l-0.1-0.4
                      l-0.1-0.3l-0.1-0.4l-0.1-0.4l-0.1-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.4l0.1-0.3
                      l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.2-0.3l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2l0.3-0.2
                      l0.3-0.2l0.4-0.1l0.3-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4-0.1l0.4,0l0.4,0l0.4,0H375V125H165.3z"
              />
            </g>
          </g>
        </svg>
        Seqera
      </a>
    </li>
    <li class="md-nav__item">
      <a
        href="https://github.com/nextflow-io/training"
        target="_blank"
        rel="noopener"
        class="md-nav__link"
      >
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path
            d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"
          ></path>
        </svg>
        Training on GitHub
      </a>
    </li>
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/nextflow-io/training/edit/master/docs/nf4_science/genomics/01_per_sample_variant_calling.md" title="Editar esta página" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<h1 id="part-1-per-sample-variant-calling">Part 1: Per-sample variant calling<a class="headerlink" href="#part-1-per-sample-variant-calling" title="Permanent link">&para;</a></h1>
<p>In the first part of this course, we show you how to build a simple variant calling pipeline that applies GATK variant calling to individual sequencing samples.</p>
<h3 id="method-overview">Method overview<a class="headerlink" href="#method-overview" title="Permanent link">&para;</a></h3>
<p>Variant calling is a genomic analysis method that aims to identify variations in a genome sequence relative to a reference genome.
Here we are going to use tools and methods designed for calling short variants, <em>i.e.</em> SNPs and indels.</p>
<p><img alt="GATK pipeline" src="../../../../nf4_science/genomics/img/gatk-pipeline.png" /></p>
<p>A full variant calling pipeline typically involves a lot of steps, including mapping to the reference (sometime referred to as genome alignment) and variant filtering and prioritization.
For simplicity, in this part of the course we are going to focus on just the variant calling part.</p>
<h3 id="dataset">Dataset<a class="headerlink" href="#dataset" title="Permanent link">&para;</a></h3>
<p>We provide the following data and related resources:</p>
<ul>
<li><strong>A reference genome</strong> consisting of a small region of the human chromosome 20 (from hg19/b37) and its accessory files (index and sequence dictionary).</li>
<li><strong>Three whole genome sequencing samples</strong> corresponding to a family trio (mother, father and son), which have been subset to a small slice of data on chromosome 20 to keep the file sizes small.
  This is Illumina short-read sequencing data that have already been mapped to the reference genome, provided in <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a> format (Binary Alignment Map, a compressed version of SAM, Sequence Alignment Map).</li>
<li><strong>A list of genomic intervals</strong>, i.e. coordinates on the genome where our samples have data suitable for calling variants, provided in BED format.</li>
</ul>
<h3 id="workflow">Workflow<a class="headerlink" href="#workflow" title="Permanent link">&para;</a></h3>
<p>In this part of the course, we're going to develop a workflow that does the following:</p>
<ol>
<li>Generate an index file for each BAM input file using <a href="https://www.htslib.org/">Samtools</a></li>
<li>Run the GATK HaplotypeCaller on each BAM input file to generate per-sample variant calls in VCF (Variant Call Format)</li>
</ol>
<figure class="excalidraw">
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 902.9047234696894 562.94140625" width="1805.8094469393789" height="1125.8828125">
  <!-- svg-source:excalidraw -->

  <defs>
    <style class="style-fonts">
      @font-face {
        font-family: Virgil;
        src: url(https://unpkg.com/@excalidraw/excalidraw@undefined/dist/prod/Virgil-Regular-YHAB2VGJ.woff2);
          }
    </style>

  </defs>
  <rect x="0" y="0" width="902.9047234696894" height="562.94140625" fill="#ffffff"></rect><g stroke-linecap="round" transform="translate(10 11.091796875) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C52.17 2.44, 83.37 4.1, 153.5 0 C166.04 -0.79, 170.84 3.38, 174.42 20.92 C175.18 37.07, 170.33 48.16, 174.42 62.75 C173.91 77.37, 167.35 86.91, 153.5 83.66 C107.36 78.9, 60.95 79.43, 20.92 83.66 C5.96 83.73, -0.91 78.57, 0 62.75 C-0.87 50.02, -3.47 38.79, 0 20.92 C0 8.7, 7.26 0.43, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C48 -2.37, 76.33 -1.5, 153.5 0 M20.92 0 C54.16 -1.23, 85.53 -0.82, 153.5 0 M153.5 0 C167.21 -1.87, 173.37 6.53, 174.42 20.92 M153.5 0 C167.01 -1.06, 173.07 7.8, 174.42 20.92 M174.42 20.92 C176.23 29.92, 176.26 39.63, 174.42 62.75 M174.42 20.92 C174.44 33.17, 174.53 43.24, 174.42 62.75 M174.42 62.75 C172.69 78.28, 167.49 82.74, 153.5 83.66 M174.42 62.75 C172.54 76.73, 167.23 84.43, 153.5 83.66 M153.5 83.66 C117.09 83.91, 80.29 84, 20.92 83.66 M153.5 83.66 C103.23 84.89, 53.56 84.61, 20.92 83.66 M20.92 83.66 C6.7 81.9, -0.87 77.17, 0 62.75 M20.92 83.66 C4.84 83.81, 0.87 76.34, 0 62.75 M0 62.75 C-0.13 51.17, 1.55 41.75, 0 20.92 M0 62.75 C0.16 49.93, 0.66 37.7, 0 20.92 M0 20.92 C1.17 8.03, 5.37 -0.24, 20.92 0 M0 20.92 C1.09 8.99, 4.73 2.24, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(75.71902970189899 40.421875) rotate(0 21.489990234375 12.5)"><text x="21.489990234375" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">BAM</text></g><g stroke-linecap="round" transform="translate(269.54925471968943 10) rotate(0 109.333984375 41.830078125)"><path d="M4.91 5.59 C4.91 5.59, 4.91 5.59, 4.91 5.59 M4.91 5.59 C4.91 5.59, 4.91 5.59, 4.91 5.59 M1.11 22.16 C6.94 13.81, 13.66 5.33, 20.13 0.28 M1.11 22.16 C8.48 15.51, 14.46 7.62, 20.13 0.28 M0.58 34.96 C9.49 24.06, 19.08 15.64, 30.76 0.24 M0.58 34.96 C9.33 25.13, 16.39 16.31, 30.76 0.24 M0.72 47 C14.11 31.26, 30.55 14.48, 41.39 0.21 M0.72 47 C13.09 32.34, 25.74 20.16, 41.39 0.21 M0.85 59.04 C9.94 48.58, 21.24 35.85, 52.02 0.17 M0.85 59.04 C19.43 39.07, 35.95 17.48, 52.02 0.17 M1.64 70.33 C24.63 44.06, 42.97 20.29, 62.65 0.14 M1.64 70.33 C18.62 49.89, 36.98 27.58, 62.65 0.14 M5.71 77.84 C27.26 55.42, 46.82 30.2, 73.28 0.1 M5.71 77.84 C28.81 52.45, 51.5 26.26, 73.28 0.1 M12.4 82.33 C35.07 56.81, 57.38 30.69, 83.91 0.07 M12.4 82.33 C37.15 52.76, 62.87 23.57, 83.91 0.07 M21.72 83.81 C37.76 64.37, 55.38 47.34, 94.54 0.04 M21.72 83.81 C49.14 53.43, 75.3 21.3, 94.54 0.04 M32.35 83.77 C58.24 52.53, 86.93 19.66, 104.52 0.76 M32.35 83.77 C52.27 60.56, 72.26 38.04, 104.52 0.76 M42.33 84.49 C67.16 51.85, 94.14 21.89, 115.15 0.72 M42.33 84.49 C64.65 56.98, 87.21 30.81, 115.15 0.72 M52.96 84.46 C78.5 57.54, 102.11 29.18, 125.78 0.69 M52.96 84.46 C77.55 54.81, 102.31 27.56, 125.78 0.69 M63.59 84.42 C81.73 65.24, 100.2 44.45, 136.41 0.65 M63.59 84.42 C90 54.29, 117.36 23.93, 136.41 0.65 M74.22 84.39 C101.53 51.44, 132.65 17.55, 147.04 0.62 M74.22 84.39 C90.28 65.7, 105.96 49.96, 147.04 0.62 M84.85 84.36 C108.82 56.24, 134.95 27.65, 157.67 0.58 M84.85 84.36 C102.24 65.9, 118.66 46.24, 157.67 0.58 M95.48 84.32 C110.95 67.07, 126.41 49.21, 168.3 0.55 M95.48 84.32 C124.39 51.75, 151.93 21.28, 168.3 0.55 M106.11 84.29 C134.46 51.38, 160.28 21.63, 178.93 0.51 M106.11 84.29 C130.34 56.18, 153.53 27.69, 178.93 0.51 M116.74 84.25 C131.23 68.36, 146.91 48.98, 189.56 0.48 M116.74 84.25 C141.95 54.25, 168.66 23.45, 189.56 0.48 M127.37 84.22 C149.24 56.22, 173.25 30.45, 200.19 0.45 M127.37 84.22 C144.06 65.03, 162.22 44.34, 200.19 0.45 M138 84.18 C162.55 56.82, 188.29 25.58, 208.85 2.67 M138 84.18 C159.93 58.41, 181.18 31.95, 208.85 2.67 M148.63 84.15 C172.14 55.7, 199.45 27.27, 214.89 7.92 M148.63 84.15 C163.58 67.68, 179.51 49.75, 214.89 7.92 M159.26 84.11 C176.03 64.73, 196.42 40.93, 218.3 16.19 M159.26 84.11 C183.4 58.5, 204.38 31.72, 218.3 16.19 M169.89 84.08 C185.71 66.1, 200.84 47.58, 218.44 28.23 M169.89 84.08 C183.13 69.12, 196.09 52.61, 218.44 28.23 M180.52 84.05 C188.69 70.87, 202.77 61.45, 218.57 40.27 M180.52 84.05 C193.57 69.19, 207.78 53.01, 218.57 40.27 M191.15 84.01 C196.8 75.68, 204.83 70.22, 218.7 52.31 M191.15 84.01 C199.73 73.38, 208.23 64.12, 218.7 52.31 M203.09 82.47 C209.17 76.55, 216.67 67.67, 218.83 64.35 M203.09 82.47 C206.51 77.5, 209.58 73.12, 218.83 64.35 M4.93 78.17 C4.93 78.17, 4.93 78.17, 4.93 78.17 M4.93 78.17 C4.93 78.17, 4.93 78.17, 4.93 78.17 M22.25 82.63 C15.78 78.25, 12.46 73.02, 2.63 65.57 M22.25 82.63 C17.06 77.54, 11.42 71.81, 2.63 65.57 M34.29 82.49 C24.32 73.38, 13.85 65.26, 2.59 54.94 M34.29 82.49 C23.11 73.76, 11.54 62.56, 2.59 54.94 M46.33 82.36 C36.59 72.79, 28.31 65.17, 2.56 44.31 M46.33 82.36 C30.35 67.76, 11.69 53.44, 2.56 44.31 M58.37 82.23 C38.45 63.16, 16.54 46.34, 2.52 33.68 M58.37 82.23 C44.73 70.63, 31.41 57.26, 2.52 33.68 M70.41 82.09 C47.33 63.62, 23.36 43.21, 2.49 23.05 M70.41 82.09 C54.97 66.57, 37.54 53.61, 2.49 23.05 M83.21 82.62 C60.45 61, 36.07 40.04, 3.21 13.08 M83.21 82.62 C66.95 69.53, 51.02 55.36, 3.21 13.08 M95.25 82.48 C75.51 65.79, 54.1 47.49, 5.44 4.41 M95.25 82.48 C75.35 65.59, 55.59 47.67, 5.44 4.41 M107.29 82.35 C86.05 62.67, 67.52 47.03, 12.95 0.34 M107.29 82.35 C78.56 55.93, 50.37 31.55, 12.95 0.34 M119.33 82.22 C98.59 66.22, 80.59 45.62, 22.73 -1.76 M119.33 82.22 C85.01 51.39, 48.74 20.61, 22.73 -1.76 M131.37 82.08 C107.78 62.25, 83.44 40.63, 34.77 -1.89 M131.37 82.08 C107.54 62.05, 83.87 41.68, 34.77 -1.89 M144.17 82.61 C110.86 51.83, 74.16 22.55, 46.81 -2.02 M144.17 82.61 C115.22 58.21, 88.96 34.29, 46.81 -2.02 M156.21 82.47 C127.28 55.38, 94.19 30.56, 58.85 -2.16 M156.21 82.47 C120.69 52, 85.32 22.01, 58.85 -2.16 M168.25 82.34 C147.36 63.13, 123.8 43.05, 71.65 -1.63 M168.25 82.34 C139.96 57.93, 111.52 31.8, 71.65 -1.63 M180.29 82.21 C144.48 51.26, 104.3 19.68, 83.69 -1.77 M180.29 82.21 C157.23 63.62, 136.27 45.87, 83.69 -1.77 M193.09 82.73 C158.08 50.29, 121.98 21.84, 95.73 -1.9 M193.09 82.73 C169.3 62.08, 144.43 40.52, 95.73 -1.9 M201.35 79.32 C175.76 58.83, 153.79 35.47, 107.77 -2.03 M201.35 79.32 C181.73 61.84, 162.26 46.79, 107.77 -2.03 M212.64 78.53 C183.39 50.27, 149.89 23.16, 119.81 -2.17 M212.64 78.53 C183.44 52.26, 153.91 27.8, 119.81 -2.17 M217.13 71.83 C189.47 46.04, 162.45 23.76, 132.61 -1.64 M217.13 71.83 C187.71 48.51, 159.35 23.81, 132.61 -1.64 M218.61 62.52 C196.06 39.55, 169.02 21.26, 144.65 -1.78 M218.61 62.52 C202.55 49.91, 187.36 36.26, 144.65 -1.78 M218.57 51.89 C203.39 38.96, 189.57 25.17, 156.69 -1.91 M218.57 51.89 C203.11 37.7, 187.02 24.96, 156.69 -1.91 M218.54 41.26 C197.63 25.9, 180.36 7.76, 168.73 -2.04 M218.54 41.26 C199.81 25.92, 181.88 9.01, 168.73 -2.04 M218.5 30.63 C209.22 22.81, 199.59 14.41, 180.77 -2.18 M218.5 30.63 C209.32 23.88, 200.94 15.31, 180.77 -2.18 M218.47 20 C210.17 13.69, 202.98 6.43, 193.56 -1.65 M218.47 20 C209.47 12.93, 200.67 5.63, 193.56 -1.65" stroke="#ffec99" stroke-width="1" fill="none"></path><path d="M20.92 0 C61.76 1.98, 104.49 1.96, 197.75 0 M20.92 0 C74.85 0.46, 130.88 0.24, 197.75 0 M197.75 0 C211.01 -1.84, 218.24 6.31, 218.67 20.92 M197.75 0 C212.37 0.03, 220.66 6.14, 218.67 20.92 M218.67 20.92 C219.23 31.58, 218.21 40.1, 218.67 62.75 M218.67 20.92 C219.24 35.25, 219.46 50.93, 218.67 62.75 M218.67 62.75 C217.25 75.4, 210.64 84.12, 197.75 83.66 M218.67 62.75 C217.58 74.53, 210.08 85.67, 197.75 83.66 M197.75 83.66 C144.25 82.53, 90.8 83.1, 20.92 83.66 M197.75 83.66 C159.19 84.1, 122.5 83.58, 20.92 83.66 M20.92 83.66 C8.19 84.31, -0.92 75.81, 0 62.75 M20.92 83.66 C8.83 81.83, -1.21 76.05, 0 62.75 M0 62.75 C1.42 54.53, -0.32 45, 0 20.92 M0 62.75 C-0.83 50.56, 0.46 37.91, 0 20.92 M0 20.92 C-0.4 7.67, 5.48 -1.29, 20.92 0 M0 20.92 C0.37 8.77, 6.66 1.11, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(330.0132821244746 26.830078125) rotate(0 48.869956970214844 25)"><text x="48.869956970214844" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Samtools </text><text x="48.869956970214844" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">index</text></g><g stroke-linecap="round" transform="translate(293.69732569464554 165.921875) rotate(0 87.20901993627393 41.830078125)"><path d="M20.92 0 C53.87 -1.29, 87.07 -1.25, 153.5 0 C170.57 -0.1, 176.03 7.28, 174.42 20.92 C175.59 35.38, 174.51 45.05, 174.42 62.75 C172.85 77.99, 169.11 82.53, 153.5 83.66 C101.5 85.16, 49.57 87.82, 20.92 83.66 C10.19 84.75, 1.76 78.17, 0 62.75 C0.15 46.65, -1.52 39.09, 0 20.92 C2.33 6.44, 3.67 -1.91, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C50.72 0.76, 82.73 1.08, 153.5 0 M20.92 0 C69.33 -0.53, 118.43 0.3, 153.5 0 M153.5 0 C167.96 1.56, 174.97 6.67, 174.42 20.92 M153.5 0 C167.11 -1.36, 175.45 6.39, 174.42 20.92 M174.42 20.92 C172.26 31.78, 174.09 39.8, 174.42 62.75 M174.42 20.92 C174.97 34.67, 174.11 50.03, 174.42 62.75 M174.42 62.75 C173.95 77.75, 165.95 83.29, 153.5 83.66 M174.42 62.75 C175.11 76.99, 168.22 84.71, 153.5 83.66 M153.5 83.66 C117.37 85.39, 84.53 85.74, 20.92 83.66 M153.5 83.66 C103.9 83.43, 53.42 82.49, 20.92 83.66 M20.92 83.66 C5.85 85.13, 0.69 75.37, 0 62.75 M20.92 83.66 C6.33 85.57, 0.05 78.55, 0 62.75 M0 62.75 C-0.5 52.46, -0.72 41.51, 0 20.92 M0 62.75 C1.09 48.7, 0.42 33.7, 0 20.92 M0 20.92 C-0.41 6.01, 7.89 -0.86, 20.92 0 M0 20.92 C-2.09 9.23, 7.52 -0.15, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(329.90637507561405 195.251953125) rotate(0 50.99997055530548 12.5)"><text x="50.99997055530548" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">BAM index</text></g><g stroke-linecap="round" transform="translate(506.0624047713194 165.8125) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C53.87 -4.1, 88.73 -0.01, 153.5 0 C165.12 -3.58, 174.02 4.98, 174.42 20.92 C172.46 31.72, 175.29 49.65, 174.42 62.75 C174.9 73.7, 165.85 82.36, 153.5 83.66 C126.52 84.35, 93.77 81.79, 20.92 83.66 C8.43 81.03, -2.87 78.51, 0 62.75 C0.3 49.29, -1.28 29.75, 0 20.92 C3.45 5.73, 10.55 0.5, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C57.78 0.13, 99.78 0.7, 153.5 0 M20.92 0 C66.25 0.36, 113.73 2, 153.5 0 M153.5 0 C168.57 -1.73, 174.09 6.2, 174.42 20.92 M153.5 0 C166.69 -0.96, 176.55 6.13, 174.42 20.92 M174.42 20.92 C172.63 32.24, 172.78 43.79, 174.42 62.75 M174.42 20.92 C174.18 31.35, 174.1 40.61, 174.42 62.75 M174.42 62.75 C175.43 78.07, 168.97 85.23, 153.5 83.66 M174.42 62.75 C174.93 78.63, 166.45 84.28, 153.5 83.66 M153.5 83.66 C116.31 84.17, 74.05 84.59, 20.92 83.66 M153.5 83.66 C110.88 84.55, 69.39 83.64, 20.92 83.66 M20.92 83.66 C5.55 81.81, 1.05 76.76, 0 62.75 M20.92 83.66 C4.97 83.94, 1.58 75.51, 0 62.75 M0 62.75 C-0.6 50.2, 0.81 34.76, 0 20.92 M0 62.75 C-0.05 47.24, -0.91 30.91, 0 20.92 M0 20.92 C0.57 6.23, 6.48 1.7, 20.92 0 M0 20.92 C-0.05 7.79, 8.1 -2.29, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(547.6514553843159 195.142578125) rotate(0 45.61996932327747 12.5)"><text x="45.61996932327747" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Intervals</text></g><g stroke-linecap="round" transform="translate(718.4866835971415 166.6953125) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C72.18 -4.21, 119.69 -0.07, 153.5 0 C170.32 2.79, 175.29 10.2, 174.42 20.92 C174.38 37.48, 173.22 48.85, 174.42 62.75 C177.98 73.82, 169.88 84.14, 153.5 83.66 C116.91 82.41, 82.49 82.73, 20.92 83.66 C7.18 83.98, 0.15 74.37, 0 62.75 C-2.67 52.18, -2.98 43.81, 0 20.92 C-2.31 6.43, 8.25 -0.96, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C72.56 -1.41, 123.27 0.25, 153.5 0 M20.92 0 C70.1 0.94, 120.16 0.75, 153.5 0 M153.5 0 C165.88 1.31, 174.4 7.8, 174.42 20.92 M153.5 0 C168.31 -1.39, 173.35 5.3, 174.42 20.92 M174.42 20.92 C173.37 31.08, 175.62 36.87, 174.42 62.75 M174.42 20.92 C174.66 30.39, 175.11 40.87, 174.42 62.75 M174.42 62.75 C175.15 75.3, 167.89 82.38, 153.5 83.66 M174.42 62.75 C176 76.6, 169.73 84.94, 153.5 83.66 M153.5 83.66 C106.16 84.42, 57.4 86.15, 20.92 83.66 M153.5 83.66 C116.59 84.14, 81.78 84.04, 20.92 83.66 M20.92 83.66 C6.04 85.29, 0.47 75.3, 0 62.75 M20.92 83.66 C9.21 84.85, 2.28 78.34, 0 62.75 M0 62.75 C0.23 53.63, -1.41 40.04, 0 20.92 M0 62.75 C0.48 47.62, -0.12 32.09, 0 20.92 M0 20.92 C-0.22 7.69, 8.15 -1.74, 20.92 0 M0 20.92 C1.17 5.38, 8.89 -2.12, 20.92 0" stroke="#1e1e1e" stroke-width="4" fill="none"></path></g><g transform="translate(734.4957431824251 183.525390625) rotate(0 71.1999603509903 25)"><text x="71.1999603509903" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">Reference</text><text x="71.1999603509903" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">+ index &amp; dict</text></g><g stroke-linecap="round" transform="translate(271.13519221968943 320.8828125) rotate(0 109.333984375 41.830078125)"><path d="M4.91 5.59 C4.91 5.59, 4.91 5.59, 4.91 5.59 M4.91 5.59 C4.91 5.59, 4.91 5.59, 4.91 5.59 M3.73 19.14 C6.63 13.72, 11.47 8.83, 16.2 4.8 M3.73 19.14 C7.28 14.69, 9.86 12.15, 16.2 4.8 M3.87 31.19 C11.11 24.59, 15.37 18.48, 26.83 4.77 M3.87 31.19 C12.09 20.72, 21.31 11.87, 26.83 4.77 M4 43.23 C14.86 31.7, 25.6 18.94, 37.46 4.74 M4 43.23 C13.97 33.64, 21.21 22.23, 37.46 4.74 M4.13 55.27 C17.63 42.9, 29.22 30.49, 48.09 4.7 M4.13 55.27 C21.54 34.82, 38.47 14.52, 48.09 4.7 M2.3 69.57 C14.11 53.59, 27.44 41.88, 58.72 4.67 M2.3 69.57 C13.93 55.52, 26.41 40.41, 58.72 4.67 M6.37 77.08 C28.86 49.07, 56.85 19.3, 69.35 4.63 M6.37 77.08 C21.5 59.16, 37.72 41.82, 69.35 4.63 M12.4 82.33 C25.26 64.88, 38.98 49.85, 79.32 5.35 M12.4 82.33 C38.32 51.08, 66.46 21.53, 79.32 5.35 M21.72 83.81 C45.1 55.64, 71.17 28.52, 89.95 5.32 M21.72 83.81 C37.25 66.39, 53.26 47.46, 89.95 5.32 M32.35 83.77 C58.73 52.95, 84.47 23.23, 100.58 5.28 M32.35 83.77 C51.63 62.18, 70.01 39.89, 100.58 5.28 M42.33 84.49 C60.86 61.16, 84.35 37.98, 111.21 5.25 M42.33 84.49 C66.27 57.53, 89.83 29.83, 111.21 5.25 M52.96 84.46 C78.41 52.99, 105.24 23.72, 121.84 5.21 M52.96 84.46 C77.23 55.83, 101.69 25.61, 121.84 5.21 M63.59 84.42 C90.32 55.15, 117.46 24.83, 132.47 5.18 M63.59 84.42 C87.74 57.47, 110.82 31.88, 132.47 5.18 M74.22 84.39 C100.99 55.92, 123.14 27.99, 143.1 5.15 M74.22 84.39 C100.27 52.36, 128.57 23.62, 143.1 5.15 M84.85 84.36 C101.12 67.8, 114.24 47.77, 153.73 5.11 M84.85 84.36 C102.76 61.84, 121.83 39.86, 153.73 5.11 M95.48 84.32 C109.51 68.52, 122.27 51.4, 164.36 5.08 M95.48 84.32 C116.13 61.35, 136.21 37.6, 164.36 5.08 M106.11 84.29 C120.64 65.65, 136.99 50.49, 174.99 5.04 M106.11 84.29 C130 56.52, 155.26 29.71, 174.99 5.04 M116.74 84.25 C137.64 59.01, 163.43 31.2, 185.62 5.01 M116.74 84.25 C134.81 64.99, 152.87 43.45, 185.62 5.01 M127.37 84.22 C146.43 61.84, 165.65 41.65, 199.53 1.2 M127.37 84.22 C144.72 63.19, 165.01 41.15, 199.53 1.2 M138 84.18 C158.98 58.39, 177.91 37.86, 208.85 2.67 M138 84.18 C158.6 62.4, 178.38 40.49, 208.85 2.67 M148.63 84.15 C161.87 67.52, 178.07 52.05, 215.54 7.17 M148.63 84.15 C163.35 65.87, 179.27 48.26, 215.54 7.17 M159.26 84.11 C179.06 58.37, 199.42 37.38, 218.3 16.19 M159.26 84.11 C180.11 60.07, 201.77 37.13, 218.3 16.19 M169.89 84.08 C185.72 61.75, 206.2 41.44, 218.44 28.23 M169.89 84.08 C179.89 70.85, 191.96 59.75, 218.44 28.23 M180.52 84.05 C190.19 72.91, 205.57 56.56, 218.57 40.27 M180.52 84.05 C193.05 68.89, 204.86 55.36, 218.57 40.27 M191.15 84.01 C199.92 74.49, 206.01 65.76, 218.7 52.31 M191.15 84.01 C199.99 73.7, 207.59 65.52, 218.7 52.31 M200.47 85.49 C204.87 79.5, 211.11 73.42, 219.49 63.6 M200.47 85.49 C205.38 80.36, 210.39 74.05, 219.49 63.6 M4.93 78.17 C4.93 78.17, 4.93 78.17, 4.93 78.17 M4.93 78.17 C4.93 78.17, 4.93 78.17, 4.93 78.17 M22.25 82.63 C17.18 76.03, 12.04 71.68, 2.63 65.57 M22.25 82.63 C17.84 77.5, 11.61 73.52, 2.63 65.57 M34.29 82.49 C24.03 71, 11.9 60.64, 2.59 54.94 M34.29 82.49 C25.55 75.76, 18.06 68.57, 2.59 54.94 M46.33 82.36 C27.42 65.89, 13.57 53.59, 2.56 44.31 M46.33 82.36 C32.48 69.33, 16.63 55.77, 2.56 44.31 M58.37 82.23 C38.75 65.94, 16.44 45.13, 2.52 33.68 M58.37 82.23 C46.37 72.42, 34.99 61.08, 2.52 33.68 M70.41 82.09 C53.46 64.95, 33.04 51.49, 2.49 23.05 M70.41 82.09 C49.74 64.44, 28.19 46.96, 2.49 23.05 M83.21 82.62 C61.73 61.15, 34.75 40.12, 3.21 13.08 M83.21 82.62 C61.03 63.78, 40.84 44.57, 3.21 13.08 M95.25 82.48 C72.69 63.59, 50.48 44, 5.44 4.41 M95.25 82.48 C66.18 57.96, 37.54 35.19, 5.44 4.41 M107.29 82.35 C76.61 55.52, 47.47 32.1, 12.95 0.34 M107.29 82.35 C85.08 62.77, 64.39 43.61, 12.95 0.34 M119.33 82.22 C96.13 63.17, 72.25 42.37, 22.73 -1.76 M119.33 82.22 C97.79 63.96, 76.38 45.98, 22.73 -1.76 M131.37 82.08 C92.12 48.35, 54.48 17.05, 34.77 -1.89 M131.37 82.08 C104.65 58.37, 76.68 34.9, 34.77 -1.89 M144.17 82.61 C118.78 61.97, 95.67 39.27, 46.81 -2.02 M144.17 82.61 C123.41 63.66, 101.82 45.26, 46.81 -2.02 M156.21 82.47 C124.31 54.94, 93.46 27.27, 58.85 -2.16 M156.21 82.47 C131.81 62.26, 108.86 40.18, 58.85 -2.16 M168.25 82.34 C149.03 66.08, 126.56 49.07, 71.65 -1.63 M168.25 82.34 C146.37 64.98, 125.64 46.31, 71.65 -1.63 M180.29 82.21 C150.47 53.53, 115.63 27.26, 83.69 -1.77 M180.29 82.21 C143.81 48.63, 104.06 15.48, 83.69 -1.77 M193.09 82.73 C155.44 52.27, 122.08 20, 95.73 -1.9 M193.09 82.73 C166.79 58.24, 138 35.91, 95.73 -1.9 M201.35 79.32 C175.74 59.18, 154.6 40.08, 107.77 -2.03 M201.35 79.32 C182.43 61.41, 163.55 46.45, 107.77 -2.03 M212.64 78.53 C180.28 54.92, 151.71 27.58, 119.81 -2.17 M212.64 78.53 C177.98 48.98, 143.07 19.56, 119.81 -2.17 M217.13 71.83 C188.78 47.09, 158.88 21.38, 132.61 -1.64 M217.13 71.83 C182.61 43.41, 149.18 14.12, 132.61 -1.64 M218.61 62.52 C194.17 41.33, 166.6 18.51, 144.65 -1.78 M218.61 62.52 C188.65 36.39, 160.26 11.12, 144.65 -1.78 M218.57 51.89 C200.27 34.95, 178.92 18.14, 156.69 -1.91 M218.57 51.89 C199.72 34.61, 179.84 18.65, 156.69 -1.91 M218.54 41.26 C204.43 28.7, 192.34 18.49, 168.73 -2.04 M218.54 41.26 C198.49 24.77, 180.12 8.43, 168.73 -2.04 M218.5 30.63 C208.97 25.28, 199.68 14.83, 180.77 -2.18 M218.5 30.63 C206.5 20.21, 195.22 9.6, 180.77 -2.18 M218.47 20 C214.59 16.89, 205.76 9.92, 193.56 -1.65 M218.47 20 C211.09 13.03, 204.3 6.6, 193.56 -1.65" stroke="#ffec99" stroke-width="1" fill="none"></path><path d="M20.92 0 C73.62 0, 127.82 -0.58, 197.75 0 M20.92 0 C60.11 0.7, 96.59 1.87, 197.75 0 M197.75 0 C213.1 -0.47, 218.81 7.46, 218.67 20.92 M197.75 0 C210.44 -1.65, 220.84 6.82, 218.67 20.92 M218.67 20.92 C219.06 33.82, 218.05 43.85, 218.67 62.75 M218.67 20.92 C217.92 31.49, 219.09 43.46, 218.67 62.75 M218.67 62.75 C217.87 74.83, 213.23 82.64, 197.75 83.66 M218.67 62.75 C218.09 75.93, 209.76 83.23, 197.75 83.66 M197.75 83.66 C141.23 81.99, 87.62 84.1, 20.92 83.66 M197.75 83.66 C126.73 81.07, 58.51 81.68, 20.92 83.66 M20.92 83.66 C6.63 85.61, -1.14 76.49, 0 62.75 M20.92 83.66 C6.36 83.51, 2.18 77.46, 0 62.75 M0 62.75 C-1.74 44.15, 0.12 28.31, 0 20.92 M0 62.75 C0.02 52.7, -0.78 42.65, 0 20.92 M0 20.92 C-0.55 7.97, 8.65 0.95, 20.92 0 M0 20.92 C0.75 7.74, 4.77 0.79, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(305.689241920188 337.712890625) rotate(0 74.77993467450142 25)"><text x="74.77993467450142" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">GATK </text><text x="74.77993467450142" y="42.62" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">HaplotypeCaller</text></g><g stroke-linecap="round" transform="translate(293.750293891252 469.28125) rotate(0 87.20901993627399 41.830078125)"><path d="M20.92 0 C68.74 0.97, 113.08 2.59, 153.5 0 C166.35 -1.31, 174.25 6.32, 174.42 20.92 C172.52 34.83, 171.41 57.67, 174.42 62.75 C171.7 74.25, 167.3 82.67, 153.5 83.66 C126.4 84.38, 97.04 82.15, 20.92 83.66 C6.59 83.14, 0.94 78.84, 0 62.75 C-1.18 45.69, -2.83 34.39, 0 20.92 C2 3.42, 3.62 0.99, 20.92 0" stroke="none" stroke-width="0" fill="#a5d8ff"></path><path d="M20.92 0 C66.29 -0.96, 114.98 -1.31, 153.5 0 M20.92 0 C52.15 -0.3, 82.47 -0.88, 153.5 0 M153.5 0 C166.82 0.22, 176 8.68, 174.42 20.92 M153.5 0 C169.12 -0.36, 175.64 8.74, 174.42 20.92 M174.42 20.92 C172.87 37.16, 172.23 49.78, 174.42 62.75 M174.42 20.92 C175.08 32.33, 174.77 41.95, 174.42 62.75 M174.42 62.75 C176.27 74.88, 167.49 81.99, 153.5 83.66 M174.42 62.75 C172.65 77.61, 169.15 85.95, 153.5 83.66 M153.5 83.66 C126.01 85.16, 95.39 84.74, 20.92 83.66 M153.5 83.66 C111.67 83.47, 69.78 82.69, 20.92 83.66 M20.92 83.66 C7.37 85.15, 1.17 74.94, 0 62.75 M20.92 83.66 C7.73 85.59, -0.39 77.89, 0 62.75 M0 62.75 C1.36 53.92, -1.65 44.26, 0 20.92 M0 62.75 C-0.14 48.01, 0.16 32.6, 0 20.92 M0 20.92 C-1.53 7.76, 7.52 0.79, 20.92 0 M0 20.92 C-0.98 8.04, 8.83 -0.88, 20.92 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(322.8893426609749 498.611328125) rotate(0 58.06997116655111 12.5)"><text x="58.06997116655111" y="17.619999999999997" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="middle" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">VCF + index</text></g><g stroke-linecap="round"><g transform="translate(96.94140625 107.619140625) rotate(0 76.6328929632758 129.97283362320871)"><path d="M-1.11 -0.86 C0.32 35.35, -17.94 174.22, 8.77 217.61 C35.48 261, 133.83 252.3, 159.15 259.45 M0.51 1.3 C1.77 37.69, -18.36 175.91, 7.92 219.17 C34.2 262.42, 133.2 254.24, 158.2 260.81" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(96.94140625 107.619140625) rotate(0 76.6328929632758 129.97283362320871)"><path d="M134 267.07 C140.24 265.61, 149.44 262.73, 158.2 260.81 M134 267.07 C141.11 264.64, 148.62 263.29, 158.2 260.81" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(96.94140625 107.619140625) rotate(0 76.6328929632758 129.97283362320871)"><path d="M135.64 250.04 C141.4 253.99, 150.08 256.53, 158.2 260.81 M135.64 250.04 C142.42 252.86, 149.42 256.72, 158.2 260.81" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(380.2293586378545 105.09804847491444) rotate(0 0.1301044309677195 24.439764024615755)"><path d="M0.11 0.11 C0.21 8.11, 0.77 40.46, 0.76 48.48 M-0.5 -0.31 C-0.46 8.02, 0.21 41.03, 0.49 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(380.2293586378545 105.09804847491444) rotate(0 0.1301044309677195 24.439764024615755)"><path d="M-8.38 26.65 C-5.19 34.57, -2.43 40.67, 0.49 49.19 M-8.38 26.65 C-5.76 32.15, -4.23 38.16, 0.49 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(380.2293586378545 105.09804847491444) rotate(0 0.1301044309677195 24.439764024615755)"><path d="M8.18 26.22 C5.97 34.3, 3.32 40.54, 0.49 49.19 M8.18 26.22 C6.55 31.88, 3.83 38.01, 0.49 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(380.7613654270848 261.1536390604771) rotate(0 0.22226737518252548 24.26726710375675)"><path d="M-0.53 -0.42 C-0.38 7.69, 0.8 40.76, 0.98 48.95 M0.19 0.55 C0.32 8.5, 0.82 40.37, 0.82 48.25" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(380.7613654270848 261.1536390604771) rotate(0 0.22226737518252548 24.26726710375675)"><path d="M-7.67 25.57 C-5.74 33.61, -1.41 39.8, 0.82 48.25 M-7.67 25.57 C-5.05 33.11, -1.41 41.5, 0.82 48.25" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(380.7613654270848 261.1536390604771) rotate(0 0.22226737518252548 24.26726710375675)"><path d="M8.89 25.41 C5.69 33.42, 4.88 39.67, 0.82 48.25 M8.89 25.41 C5.63 33.07, 3.38 41.52, 0.82 48.25" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(380.8983448637614 411.92357307814893) rotate(0 0.3785265712478463 24.535335349170907)"><path d="M-0.13 0.24 C0.05 8.23, 0.78 40.43, 0.89 48.47 M0.8 -0.11 C0.96 8.21, 0.78 40.99, 0.69 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(380.8983448637614 411.92357307814893) rotate(0 0.3785265712478463 24.535335349170907)"><path d="M-7.42 26.36 C-4.73 33.24, -2.96 38.64, 0.69 49.19 M-7.42 26.36 C-5.36 33.54, -2.44 40.5, 0.69 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(380.8983448637614 411.92357307814893) rotate(0 0.3785265712478463 24.535335349170907)"><path d="M9.15 26.49 C7.44 33.31, 4.81 38.67, 0.69 49.19 M9.15 26.49 C6.19 33.56, 4.1 40.48, 0.69 49.19" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(595.6613677465953 258.79565813882994) rotate(0 -47.35407096480634 46.09440729512727)"><path d="M-0.48 -0.87 C-3.12 10.62, 1.2 54.63, -14.7 70.07 C-30.61 85.5, -82.38 88.18, -95.9 91.74 M1.47 1.29 C-1.45 12.93, 0.46 52.93, -15.82 68.22 C-32.09 83.52, -82.97 89.03, -96.17 93.06" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(595.6613677465953 258.79565813882994) rotate(0 -47.35407096480634 46.09440729512727)"><path d="M-74.96 79.83 C-84.62 84.43, -90.64 91.72, -96.17 93.06 M-74.96 79.83 C-80.83 82.94, -88.03 86.74, -96.17 93.06" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(595.6613677465953 258.79565813882994) rotate(0 -47.35407096480634 46.09440729512727)"><path d="M-71.42 96.56 C-82.64 94.56, -90.05 95.27, -96.17 93.06 M-71.42 96.56 C-78.5 94.66, -86.76 93.45, -96.17 93.06" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(801.1498254589234 263.1575657921736) rotate(0 -148.39641986341223 55.6850621379798)"><path d="M0.06 -0.75 C-7.34 15.13, 4.02 77.56, -45.38 96.21 C-94.79 114.86, -254.59 108.73, -296.37 111.15 M-1.36 1.47 C-8.28 17.41, 6.24 78.92, -43.01 97.36 C-92.26 115.8, -254.8 109.8, -296.86 112.12" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(801.1498254589234 263.1575657921736) rotate(0 -148.39641986341223 55.6850621379798)"><path d="M-273.46 103.3 C-279.75 104.16, -284.85 108.04, -296.86 112.12 M-273.46 103.3 C-278.95 105.44, -284.59 108, -296.86 112.12" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(801.1498254589234 263.1575657921736) rotate(0 -148.39641986341223 55.6850621379798)"><path d="M-273.27 120.4 C-279.73 117.53, -284.87 117.68, -296.86 112.12 M-273.27 120.4 C-278.93 118.36, -284.61 116.73, -296.86 112.12" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g stroke-linecap="round"><g transform="translate(196.73927690760354 54.50014116616944) rotate(0 29.336689349718768 0.02151378544505178)"><path d="M0.67 1.02 C10.56 1.31, 49.51 0.86, 59.12 0.69 M-0.44 0.51 C9.28 0.55, 47.8 -1.21, 57.97 -1.07" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(196.73927690760354 54.50014116616944) rotate(0 29.336689349718768 0.02151378544505178)"><path d="M34.65 7.92 C42.96 4.66, 48.94 0.78, 57.97 -1.07 M34.65 7.92 C41.47 5.72, 48.8 3.07, 57.97 -1.07" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(196.73927690760354 54.50014116616944) rotate(0 29.336689349718768 0.02151378544505178)"><path d="M34.32 -9.17 C42.58 -7.65, 48.65 -6.74, 57.97 -1.07 M34.32 -9.17 C41.4 -6.21, 48.82 -3.71, 57.97 -1.07" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask></svg>
</figure>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Index files are a common feature of bioinformatics file formats; they contain information about the structure of the main file that allows tools like GATK to access a subset of the data without having to read through the whole file.
This is important because of how big these files can get.</p>
</div>
<hr />
<h2 id="0-warmup-test-the-samtools-and-gatk-commands-interactively">0. Warmup: Test the Samtools and GATK commands interactively<a class="headerlink" href="#0-warmup-test-the-samtools-and-gatk-commands-interactively" title="Permanent link">&para;</a></h2>
<p>First we want to try out the commands manually before we attempt to wrap them in a workflow.
The tools we need (Samtools and GATK) are not installed in the GitHub Codespaces environment, so we'll use them via containers (see <a href="../../../hello_nextflow/05_hello_containers/">Hello Containers</a>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Make sure you're in the <code>nf4-science/genomics</code> directory so that the last part of the path shown when you type <code>pwd</code> is <code>genomics</code>.</p>
</div>
<h3 id="01-index-a-bam-input-file-with-samtools">0.1. Index a BAM input file with Samtools<a class="headerlink" href="#01-index-a-bam-input-file-with-samtools" title="Permanent link">&para;</a></h3>
<p>We're going to pull down a Samtools container, spin it up interactively and run the <code>samtools index</code> command on one of the BAM files.</p>
<h4 id="011-pull-the-samtools-container">0.1.1. Pull the Samtools container<a class="headerlink" href="#011-pull-the-samtools-container" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>docker<span class="w"> </span>pull<span class="w"> </span>community.wave.seqera.io/library/samtools:1.20--b5dfbd93de237464
</code></pre></div>
<h4 id="012-spin-up-the-samtools-container-interactively">0.1.2. Spin up the Samtools container interactively<a class="headerlink" href="#012-spin-up-the-samtools-container-interactively" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>docker<span class="w"> </span>run<span class="w"> </span>-it<span class="w"> </span>-v<span class="w"> </span>./data:/data<span class="w"> </span>community.wave.seqera.io/library/samtools:1.20--b5dfbd93de237464
</code></pre></div>
<h4 id="013-run-the-indexing-command">0.1.3. Run the indexing command<a class="headerlink" href="#013-run-the-indexing-command" title="Permanent link">&para;</a></h4>
<p>The <a href="https://www.htslib.org/doc/samtools-index.html">Samtools documentation</a> gives us the command line to run to index a BAM file.</p>
<p>We only need to provide the input file; the tool will automatically generate a name for the output by appending <code>.bai</code> to the input filename.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>samtools<span class="w"> </span>index<span class="w"> </span>/data/bam/reads_mother.bam
</code></pre></div>
<p>This should complete immediately, and you should now see a file called <code>reads_mother.bam.bai</code> in the same directory as the original BAM input file.</p>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="go">data/bam/</span>
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="go">├── reads_father.bam</span>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="go">├── reads_mother.bam</span>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="go">├── reads_mother.bam.bai</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="go">└── reads_son.bam</span>
</code></pre></div>
<h4 id="014-exit-the-samtools-container">0.1.4. Exit the Samtools container<a class="headerlink" href="#014-exit-the-samtools-container" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="nb">exit</span>
</code></pre></div>
<h3 id="02-call-variants-with-gatk-haplotypecaller">0.2. Call variants with GATK HaplotypeCaller<a class="headerlink" href="#02-call-variants-with-gatk-haplotypecaller" title="Permanent link">&para;</a></h3>
<p>We're going to pull down a GATK container, spin it up interactively and run the <code>gatk HaplotypeCaller</code> command on the BAM file we just indexed.</p>
<h4 id="021-pull-the-gatk-container">0.2.1. Pull the GATK container<a class="headerlink" href="#021-pull-the-gatk-container" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>docker<span class="w"> </span>pull<span class="w"> </span>community.wave.seqera.io/library/gatk4:4.5.0.0--730ee8817e436867
</code></pre></div>
<h4 id="022-spin-up-the-gatk-container-interactively">0.2.2. Spin up the GATK container interactively<a class="headerlink" href="#022-spin-up-the-gatk-container-interactively" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>docker<span class="w"> </span>run<span class="w"> </span>-it<span class="w"> </span>-v<span class="w"> </span>./data:/data<span class="w"> </span>community.wave.seqera.io/library/gatk4:4.5.0.0--730ee8817e436867
</code></pre></div>
<h4 id="023-run-the-variant-calling-command">0.2.3. Run the variant calling command<a class="headerlink" href="#023-run-the-variant-calling-command" title="Permanent link">&para;</a></h4>
<p>The <a href="https://gatk.broadinstitute.org/hc/en-us/articles/21905025322523-HaplotypeCaller">GATK documentation</a> gives us the command line to run to perform variant calling on a BAM file.</p>
<p>We need to provide the BAM input file (<code>-I</code>) as well as the reference genome (<code>-R</code>), a name for the output file (<code>-O</code>) and a list of genomic intervals to analyze (<code>-L</code>).</p>
<p>However, we don't need to specify the path to the index file; the tool will automatically look for it in the same directory, based on the established naming and co-location convention.
The same applies to the reference genome's accessory files (index and sequence dictionary files, <code>*.fai</code> and <code>*.dict</code>).</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>gatk<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="w">        </span>-R<span class="w"> </span>/data/ref/ref.fasta<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="w">        </span>-I<span class="w"> </span>/data/bam/reads_mother.bam<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="w">        </span>-O<span class="w"> </span>reads_mother.vcf<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="w">        </span>-L<span class="w"> </span>/data/ref/intervals.bed
</code></pre></div>
<p>The output file <code>reads_mother.vcf</code> is created inside your working directory in the container, so you won't see it in the VS Code file explorer unless you change the output file path.
However, it's a small test file, so you can <code>cat</code> it to open it and view the contents.
If you scroll all the way up to the start of the file, you'll find a header composed of many lines of metadata, followed by a list of variant calls, one per line.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">reads_mother.vcf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-8-26">26</a></span>
<span class="normal"><a href="#__codelineno-8-27">27</a></span>
<span class="normal"><a href="#__codelineno-8-28">28</a></span>
<span class="normal"><a href="#__codelineno-8-29">29</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-8-26" name="__codelineno-8-26"></a><span class="gp">#</span>CHROM<span class="w">	</span>POS<span class="w">	</span>ID<span class="w">	</span>REF<span class="w">	</span>ALT<span class="w">	</span>QUAL<span class="w">	</span>FILTER<span class="w">	</span>INFO<span class="w">	</span>FORMAT<span class="w">	</span>reads_mother
<a id="__codelineno-8-27" name="__codelineno-8-27"></a><span class="go">20_10037292_10066351	3480	.	C	CT	503.03	.	AC=2;AF=1.00;AN=2;DP=23;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=27.95;SOR=1.179	GT:AD:DP:GQ:PL	1/1:0,18:18:54:517,54,0</span>
<a id="__codelineno-8-28" name="__codelineno-8-28"></a><span class="go">20_10037292_10066351	3520	.	AT	A	609.03	.	AC=2;AF=1.00;AN=2;DP=18;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=33.83;SOR=0.693	GT:AD:DP:GQ:PL	1/1:0,18:18:54:623,54,0</span>
<a id="__codelineno-8-29" name="__codelineno-8-29"></a><span class="go">20_10037292_10066351	3529	.	T	A	155.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=-0.544;DP=21;ExcessHet=0.0000;FS=1.871;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=7.78;ReadPosRankSum=-1.158;SOR=1.034	GT:AD:DP:GQ:PL	0/1:12,8:20:99:163,0,328</span>
</code></pre></div></td></tr></table></div>
<p>Each line describes a possible variant identified in the sample's sequencing data. For guidance on interpreting VCF format, see <a href="https://www.ebi.ac.uk/training/online/courses/human-genetic-variation-introduction/variant-identification-and-analysis/understanding-vcf-format/">this helpful article</a>.</p>
<p>The output VCF file is accompanied by an index file called <code>reads_mother.vcf.idx</code> that was automatically created by GATK.
It has the same function as the BAM index file, to allow tools to seek and retrieve subsets of data without loading in the entire file.</p>
<h4 id="024-exit-the-gatk-container">0.2.4. Exit the GATK container<a class="headerlink" href="#024-exit-the-gatk-container" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="nb">exit</span>
</code></pre></div>
<h3 id="takeaway">Takeaway<a class="headerlink" href="#takeaway" title="Permanent link">&para;</a></h3>
<p>You know how to test the Samtools indexing and GATK variant calling commands in their respective containers.</p>
<h3 id="whats-next">What's next?<a class="headerlink" href="#whats-next" title="Permanent link">&para;</a></h3>
<p>Learn how to wrap those same commands into a two-step workflow that uses containers to execute the work.</p>
<hr />
<h2 id="1-write-a-single-stage-workflow-that-runs-samtools-index-on-a-bam-file">1. Write a single-stage workflow that runs Samtools index on a BAM file<a class="headerlink" href="#1-write-a-single-stage-workflow-that-runs-samtools-index-on-a-bam-file" title="Permanent link">&para;</a></h2>
<p>We provide you with a workflow file, <code>genomics-1.nf</code>, that outlines the main parts of the workflow.
It's not functional; its purpose is just to serve as a skeleton that you'll use to write the actual workflow.</p>
<h3 id="11-define-the-indexing-process">1.1. Define the indexing process<a class="headerlink" href="#11-define-the-indexing-process" title="Permanent link">&para;</a></h3>
<p>Let's start by writing a process, which we'll call <code>SAMTOOLS_INDEX</code>, describing the indexing operation.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-10-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-10-10">10</a></span>
<span class="normal"><a href="#__codelineno-10-11">11</a></span>
<span class="normal"><a href="#__codelineno-10-12">12</a></span>
<span class="normal"><a href="#__codelineno-10-13">13</a></span>
<span class="normal"><a href="#__codelineno-10-14">14</a></span>
<span class="normal"><a href="#__codelineno-10-15">15</a></span>
<span class="normal"><a href="#__codelineno-10-16">16</a></span>
<span class="normal"><a href="#__codelineno-10-17">17</a></span>
<span class="normal"><a href="#__codelineno-10-18">18</a></span>
<span class="normal"><a href="#__codelineno-10-19">19</a></span>
<span class="normal"><a href="#__codelineno-10-20">20</a></span>
<span class="normal"><a href="#__codelineno-10-21">21</a></span>
<span class="normal"><a href="#__codelineno-10-22">22</a></span>
<span class="normal"><a href="#__codelineno-10-23">23</a></span>
<span class="normal"><a href="#__codelineno-10-24">24</a></span>
<span class="normal"><a href="#__codelineno-10-25">25</a></span>
<span class="normal"><a href="#__codelineno-10-26">26</a></span>
<span class="normal"><a href="#__codelineno-10-27">27</a></span>
<span class="normal"><a href="#__codelineno-10-28">28</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-10-9" name="__codelineno-10-9"></a><span class="cm">/*</span>
<a id="__codelineno-10-10" name="__codelineno-10-10"></a><span class="cm"> * Generate BAM index file</span>
<a id="__codelineno-10-11" name="__codelineno-10-11"></a><span class="cm"> */</span>
<a id="__codelineno-10-12" name="__codelineno-10-12"></a><span class="n">process</span><span class="w"> </span><span class="n">SAMTOOLS_INDEX</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-10-13" name="__codelineno-10-13"></a>
<a id="__codelineno-10-14" name="__codelineno-10-14"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s1">&#39;community.wave.seqera.io/library/samtools:1.20--b5dfbd93de237464&#39;</span>
<a id="__codelineno-10-15" name="__codelineno-10-15"></a>
<a id="__codelineno-10-16" name="__codelineno-10-16"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">outdir</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;symlink&#39;</span>
<a id="__codelineno-10-17" name="__codelineno-10-17"></a>
<a id="__codelineno-10-18" name="__codelineno-10-18"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-10-19" name="__codelineno-10-19"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">input_bam</span>
<a id="__codelineno-10-20" name="__codelineno-10-20"></a>
<a id="__codelineno-10-21" name="__codelineno-10-21"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-10-22" name="__codelineno-10-22"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.bai&quot;</span>
<a id="__codelineno-10-23" name="__codelineno-10-23"></a>
<a id="__codelineno-10-24" name="__codelineno-10-24"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-10-25" name="__codelineno-10-25"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-10-26" name="__codelineno-10-26"></a><span class="s2">    samtools index &#39;$input_bam&#39;</span>
<a id="__codelineno-10-27" name="__codelineno-10-27"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-10-28" name="__codelineno-10-28"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<p>You should recognize all the pieces from what you learned in Part 1 &amp; Part 2 of this training series; the only notable change is that this time we're using <code>mode: symlink</code> for the <code>publishDir</code> directive, and we're using a parameter to define the <code>publishDir</code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Even though the data files we're using here are very small, in genomics they can get very large. For the purposes of demonstration in the teaching environment, we're using the 'symlink' publishing mode to avoid unnecessary file copies. You shouldn't do this in your final workflows, since you'll lose results when you clean up your <code>work</code> directory.</p>
</div>
<p>This process is going to require us to pass in a file path via the <code>input_bam</code> input, so let's set that up next.</p>
<h3 id="12-add-an-input-and-output-parameter-declaration">1.2. Add an input and output parameter declaration<a class="headerlink" href="#12-add-an-input-and-output-parameter-declaration" title="Permanent link">&para;</a></h3>
<p>At the top of the file, under the <code>Pipeline parameters</code> section, we declare a CLI parameter called <code>reads_bam</code> and give it a default value.
That way, we can be lazy and not specify the input when we type the command to launch the pipeline (for development purposes). We're also going to set <code>params.outdir</code> with a default value for the output directory.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-11-3">3</a></span>
<span class="normal"><a href="#__codelineno-11-4">4</a></span>
<span class="normal"><a href="#__codelineno-11-5">5</a></span>
<span class="normal"><a href="#__codelineno-11-6">6</a></span>
<span class="normal"><a href="#__codelineno-11-7">7</a></span>
<span class="normal"><a href="#__codelineno-11-8">8</a></span>
<span class="normal"><a href="#__codelineno-11-9">9</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-11-3" name="__codelineno-11-3"></a><span class="cm">/*</span>
<a id="__codelineno-11-4" name="__codelineno-11-4"></a><span class="cm"> * Pipeline parameters</span>
<a id="__codelineno-11-5" name="__codelineno-11-5"></a><span class="cm"> */</span>
<a id="__codelineno-11-6" name="__codelineno-11-6"></a>
<a id="__codelineno-11-7" name="__codelineno-11-7"></a><span class="c1">// Primary input</span>
<a id="__codelineno-11-8" name="__codelineno-11-8"></a><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/bam/reads_mother.bam&quot;</span>
<a id="__codelineno-11-9" name="__codelineno-11-9"></a><span class="n">params</span><span class="o">.</span><span class="na">outdir</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;results_genomics&quot;</span>
</code></pre></div></td></tr></table></div>
<p>Now we have a process ready, as well as a parameter to give it an input to run on, so let's wire those things up together.</p>
<!-- TODO get rid of projectDir -->

<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code>${projectDir}</code> is a built-in Nextflow variable that points to the directory where the current Nextflow workflow script (<code>genomics-1.nf</code>) is located.</p>
<p>This makes it easy to reference files, data directories, and other resources included in the workflow repository without hardcoding absolute paths.</p>
</div>
<h3 id="13-add-workflow-block-to-run-samtools_index">1.3. Add workflow block to run SAMTOOLS_INDEX<a class="headerlink" href="#13-add-workflow-block-to-run-samtools_index" title="Permanent link">&para;</a></h3>
<p>In the <code>workflow</code> block, we need to set up a <strong>channel</strong> to feed the input to the <code>SAMTOOLS_INDEX</code> process; then we can call the process itself to run on the contents of that channel.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-12-30">30</a></span>
<span class="normal"><a href="#__codelineno-12-31">31</a></span>
<span class="normal"><a href="#__codelineno-12-32">32</a></span>
<span class="normal"><a href="#__codelineno-12-33">33</a></span>
<span class="normal"><a href="#__codelineno-12-34">34</a></span>
<span class="normal"><a href="#__codelineno-12-35">35</a></span>
<span class="normal"><a href="#__codelineno-12-36">36</a></span>
<span class="normal"><a href="#__codelineno-12-37">37</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-12-30" name="__codelineno-12-30"></a><span class="n">workflow</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-12-31" name="__codelineno-12-31"></a>
<a id="__codelineno-12-32" name="__codelineno-12-32"></a><span class="w">    </span><span class="c1">// Create input channel (single file via CLI parameter)</span>
<a id="__codelineno-12-33" name="__codelineno-12-33"></a><span class="w">    </span><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="o">)</span>
<a id="__codelineno-12-34" name="__codelineno-12-34"></a>
<a id="__codelineno-12-35" name="__codelineno-12-35"></a><span class="w">    </span><span class="c1">// Create index file for input BAM file</span>
<a id="__codelineno-12-36" name="__codelineno-12-36"></a><span class="w">    </span><span class="n">SAMTOOLS_INDEX</span><span class="o">(</span><span class="n">reads_ch</span><span class="o">)</span>
<a id="__codelineno-12-37" name="__codelineno-12-37"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<p>You'll notice we're using the same <code>.fromPath</code> channel factory as we used in <a href="../../../hello_nextflow/02_hello_channels/">Hello Channels</a>.
Indeed, we're doing something very similar.
The difference is that we're telling Nextflow to just load the file path itself into the channel as an input element, rather than reading in its contents.</p>
<h3 id="14-run-the-workflow-to-verify-that-the-indexing-step-works">1.4. Run the workflow to verify that the indexing step works<a class="headerlink" href="#14-run-the-workflow-to-verify-that-the-indexing-step-works" title="Permanent link">&para;</a></h3>
<p>Let's run the workflow! As a reminder, we don't need to specify an input in the command line because we set up a default value for the input when we declared the input parameter.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-1.nf
</code></pre></div>
<p>The command should produce something like this:</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a>
<a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a><span class="go"> ┃ Launching `genomics-1.nf` [reverent_sinoussi] DSL2 - revision: 41d43ad7fe</span>
<a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a>
<a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a><span class="go">executor &gt;  local (1)</span>
<a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a><span class="go">[2a/e69536] SAMTOOLS_INDEX (1) | 1 of 1 ✔</span>
</code></pre></div>
<p>You can check that the index file has been generated correctly by looking in the work directory or in the directory set up with <code>publishDir</code>.</p>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="go">work/2a/e695367b2f60df09cf826b07192dc3</span>
<a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a><span class="go">├── reads_mother.bam -&gt; /workspaces/training/nf4-science/genomics/data/bam/reads_mother.bam</span>
<a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="go">└── reads_mother.bam.bai</span>
</code></pre></div>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="go">results_genomics/</span>
<a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="go">└── reads_mother.bam.bai</span>
</code></pre></div>
<p>There it is!</p>
<h3 id="takeaway_1">Takeaway<a class="headerlink" href="#takeaway_1" title="Permanent link">&para;</a></h3>
<p>You know how to wrap a genomics tool in a single-step Nextflow workflow and have it run using a container.</p>
<h3 id="whats-next_1">What's next?<a class="headerlink" href="#whats-next_1" title="Permanent link">&para;</a></h3>
<p>Add a second step that consumes the output of the first.</p>
<hr />
<h2 id="2-add-a-second-process-to-run-gatk-haplotypecaller-on-the-indexed-bam-file">2. Add a second process to run GATK HaplotypeCaller on the indexed BAM file<a class="headerlink" href="#2-add-a-second-process-to-run-gatk-haplotypecaller-on-the-indexed-bam-file" title="Permanent link">&para;</a></h2>
<p>Now that we have an index for our input file, we can move on to setting up the variant calling step, which is the interesting part of the workflow.</p>
<h3 id="21-define-the-variant-calling-process">2.1. Define the variant calling process<a class="headerlink" href="#21-define-the-variant-calling-process" title="Permanent link">&para;</a></h3>
<p>Let's write a process, which we'll call <code>GATK_HAPLOTYPECALLER</code>, describing the variant calling operation.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-17-30">30</a></span>
<span class="normal"><a href="#__codelineno-17-31">31</a></span>
<span class="normal"><a href="#__codelineno-17-32">32</a></span>
<span class="normal"><a href="#__codelineno-17-33">33</a></span>
<span class="normal"><a href="#__codelineno-17-34">34</a></span>
<span class="normal"><a href="#__codelineno-17-35">35</a></span>
<span class="normal"><a href="#__codelineno-17-36">36</a></span>
<span class="normal"><a href="#__codelineno-17-37">37</a></span>
<span class="normal"><a href="#__codelineno-17-38">38</a></span>
<span class="normal"><a href="#__codelineno-17-39">39</a></span>
<span class="normal"><a href="#__codelineno-17-40">40</a></span>
<span class="normal"><a href="#__codelineno-17-41">41</a></span>
<span class="normal"><a href="#__codelineno-17-42">42</a></span>
<span class="normal"><a href="#__codelineno-17-43">43</a></span>
<span class="normal"><a href="#__codelineno-17-44">44</a></span>
<span class="normal"><a href="#__codelineno-17-45">45</a></span>
<span class="normal"><a href="#__codelineno-17-46">46</a></span>
<span class="normal"><a href="#__codelineno-17-47">47</a></span>
<span class="normal"><a href="#__codelineno-17-48">48</a></span>
<span class="normal"><a href="#__codelineno-17-49">49</a></span>
<span class="normal"><a href="#__codelineno-17-50">50</a></span>
<span class="normal"><a href="#__codelineno-17-51">51</a></span>
<span class="normal"><a href="#__codelineno-17-52">52</a></span>
<span class="normal"><a href="#__codelineno-17-53">53</a></span>
<span class="normal"><a href="#__codelineno-17-54">54</a></span>
<span class="normal"><a href="#__codelineno-17-55">55</a></span>
<span class="normal"><a href="#__codelineno-17-56">56</a></span>
<span class="normal"><a href="#__codelineno-17-57">57</a></span>
<span class="normal"><a href="#__codelineno-17-58">58</a></span>
<span class="normal"><a href="#__codelineno-17-59">59</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-17-30" name="__codelineno-17-30"></a><span class="cm">/*</span>
<a id="__codelineno-17-31" name="__codelineno-17-31"></a><span class="cm"> * Call variants with GATK HaplotypeCaller</span>
<a id="__codelineno-17-32" name="__codelineno-17-32"></a><span class="cm"> */</span>
<a id="__codelineno-17-33" name="__codelineno-17-33"></a><span class="n">process</span><span class="w"> </span><span class="n">GATK_HAPLOTYPECALLER</span><span class="w"> </span><span class="o">{</span>
<a id="__codelineno-17-34" name="__codelineno-17-34"></a>
<a id="__codelineno-17-35" name="__codelineno-17-35"></a><span class="w">    </span><span class="n">container</span><span class="w"> </span><span class="s2">&quot;community.wave.seqera.io/library/gatk4:4.5.0.0--730ee8817e436867&quot;</span>
<a id="__codelineno-17-36" name="__codelineno-17-36"></a>
<a id="__codelineno-17-37" name="__codelineno-17-37"></a><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">outdir</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;symlink&#39;</span>
<a id="__codelineno-17-38" name="__codelineno-17-38"></a>
<a id="__codelineno-17-39" name="__codelineno-17-39"></a><span class="w">    </span><span class="nl">input:</span>
<a id="__codelineno-17-40" name="__codelineno-17-40"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">input_bam</span>
<a id="__codelineno-17-41" name="__codelineno-17-41"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">input_bam_index</span>
<a id="__codelineno-17-42" name="__codelineno-17-42"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">ref_fasta</span>
<a id="__codelineno-17-43" name="__codelineno-17-43"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">ref_index</span>
<a id="__codelineno-17-44" name="__codelineno-17-44"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">ref_dict</span>
<a id="__codelineno-17-45" name="__codelineno-17-45"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="n">interval_list</span>
<a id="__codelineno-17-46" name="__codelineno-17-46"></a>
<a id="__codelineno-17-47" name="__codelineno-17-47"></a><span class="w">    </span><span class="nl">output:</span>
<a id="__codelineno-17-48" name="__codelineno-17-48"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.vcf&quot;</span><span class="w">     </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">vcf</span>
<a id="__codelineno-17-49" name="__codelineno-17-49"></a><span class="w">        </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.vcf.idx&quot;</span><span class="w"> </span><span class="o">,</span><span class="w"> </span><span class="nl">emit:</span><span class="w"> </span><span class="n">idx</span>
<a id="__codelineno-17-50" name="__codelineno-17-50"></a>
<a id="__codelineno-17-51" name="__codelineno-17-51"></a><span class="w">    </span><span class="nl">script:</span>
<a id="__codelineno-17-52" name="__codelineno-17-52"></a><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<a id="__codelineno-17-53" name="__codelineno-17-53"></a><span class="s2">    gatk HaplotypeCaller \</span>
<a id="__codelineno-17-54" name="__codelineno-17-54"></a><span class="s2">        -R ${ref_fasta} \</span>
<a id="__codelineno-17-55" name="__codelineno-17-55"></a><span class="s2">        -I ${input_bam} \</span>
<a id="__codelineno-17-56" name="__codelineno-17-56"></a><span class="s2">        -O ${input_bam}.vcf \</span>
<a id="__codelineno-17-57" name="__codelineno-17-57"></a><span class="s2">        -L ${interval_list}</span>
<a id="__codelineno-17-58" name="__codelineno-17-58"></a><span class="s2">    &quot;&quot;&quot;</span>
<a id="__codelineno-17-59" name="__codelineno-17-59"></a><span class="o">}</span>
</code></pre></div></td></tr></table></div>
<p>You'll notice that we've introduced some new syntax here (<code>emit:</code>) to uniquely name each of our output channels, and the reasons for this will become clear soon.</p>
<p>This command takes quite a few more inputs, because GATK needs more information to perform the analysis compared to a simple indexing job.
But you'll note that there are even more inputs defined in the inputs block than are listed in the GATK command. Why is that?</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The GATK knows to look for the BAM index file and the reference genome's accessory files because it is aware of the conventions surrounding those files.
However, Nextflow is designed to be domain-agnostic and doesn't know anything about bioinformatics file format requirements.</p>
</div>
<p>We need to tell Nextflow explicitly that it has to stage those files in the working directory at runtime; otherwise it won't do it, and GATK will (correctly) throw an error about the index files being missing.</p>
<p>Similarly, we have to list the output VCF's index file (the <code>"${input_bam}.vcf.idx"</code> file) explicitly so that Nextflow will know to keep track of that file in case it's needed in subsequent steps.</p>
<h3 id="22-add-definitions-for-accessory-inputs">2.2. Add definitions for accessory inputs<a class="headerlink" href="#22-add-definitions-for-accessory-inputs" title="Permanent link">&para;</a></h3>
<p>Since our new process expects a handful of additional files to be provided, we set up some CLI parameters for them under the <code>Pipeline parameters</code> section, along with some default values (same reasons as before).</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-18-10">10</a></span>
<span class="normal"><a href="#__codelineno-18-11">11</a></span>
<span class="normal"><a href="#__codelineno-18-12">12</a></span>
<span class="normal"><a href="#__codelineno-18-13">13</a></span>
<span class="normal"><a href="#__codelineno-18-14">14</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-18-10" name="__codelineno-18-10"></a><span class="c1">// Accessory files</span>
<a id="__codelineno-18-11" name="__codelineno-18-11"></a><span class="n">params</span><span class="o">.</span><span class="na">reference</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/ref/ref.fasta&quot;</span>
<a id="__codelineno-18-12" name="__codelineno-18-12"></a><span class="n">params</span><span class="o">.</span><span class="na">reference_index</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/ref/ref.fasta.fai&quot;</span>
<a id="__codelineno-18-13" name="__codelineno-18-13"></a><span class="n">params</span><span class="o">.</span><span class="na">reference_dict</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/ref/ref.dict&quot;</span>
<a id="__codelineno-18-14" name="__codelineno-18-14"></a><span class="n">params</span><span class="o">.</span><span class="na">intervals</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/ref/intervals.bed&quot;</span>
</code></pre></div></td></tr></table></div>
<h3 id="23-create-variables-to-hold-the-accessory-file-paths">2.3. Create variables to hold the accessory file paths<a class="headerlink" href="#23-create-variables-to-hold-the-accessory-file-paths" title="Permanent link">&para;</a></h3>
<p>While main data inputs are streamed dynamically through channels, there are two approaches for handling accessory files. The recommended approach is to create explicit channels, which makes data flow clearer and more consistent. Alternatively, the file() function to create variables can be used for simpler cases, particularly when you need to reference the same file in multiple processes - though be aware this still creates channels implicitly.</p>
<p>Add this to the workflow block (after the <code>reads_ch</code> creation):</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-19-71">71</a></span>
<span class="normal"><a href="#__codelineno-19-72">72</a></span>
<span class="normal"><a href="#__codelineno-19-73">73</a></span>
<span class="normal"><a href="#__codelineno-19-74">74</a></span>
<span class="normal"><a href="#__codelineno-19-75">75</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-19-71" name="__codelineno-19-71"></a><span class="c1">// Load the file paths for the accessory files (reference and intervals)</span>
<a id="__codelineno-19-72" name="__codelineno-19-72"></a><span class="n">ref_file</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reference</span><span class="o">)</span>
<a id="__codelineno-19-73" name="__codelineno-19-73"></a><span class="n">ref_index_file</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reference_index</span><span class="o">)</span>
<a id="__codelineno-19-74" name="__codelineno-19-74"></a><span class="n">ref_dict_file</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reference_dict</span><span class="o">)</span>
<a id="__codelineno-19-75" name="__codelineno-19-75"></a><span class="n">intervals_file</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">intervals</span><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p>This will make the accessory file paths available for providing as input to any processes that need them.</p>
<h3 id="24-add-a-call-to-the-workflow-block-to-run-gatk_haplotypecaller">2.4. Add a call to the workflow block to run GATK_HAPLOTYPECALLER<a class="headerlink" href="#24-add-a-call-to-the-workflow-block-to-run-gatk_haplotypecaller" title="Permanent link">&para;</a></h3>
<p>Now that we've got our second process set up and all the inputs and accessory files are ready and available, we can add a call to the <code>GATK_HAPLOTYPECALLER</code> process in the workflow body.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-20-80">80</a></span>
<span class="normal"><a href="#__codelineno-20-81">81</a></span>
<span class="normal"><a href="#__codelineno-20-82">82</a></span>
<span class="normal"><a href="#__codelineno-20-83">83</a></span>
<span class="normal"><a href="#__codelineno-20-84">84</a></span>
<span class="normal"><a href="#__codelineno-20-85">85</a></span>
<span class="normal"><a href="#__codelineno-20-86">86</a></span>
<span class="normal"><a href="#__codelineno-20-87">87</a></span>
<span class="normal"><a href="#__codelineno-20-88">88</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-20-80" name="__codelineno-20-80"></a><span class="c1">// Call variants from the indexed BAM file</span>
<a id="__codelineno-20-81" name="__codelineno-20-81"></a><span class="n">GATK_HAPLOTYPECALLER</span><span class="o">(</span>
<a id="__codelineno-20-82" name="__codelineno-20-82"></a><span class="w">    </span><span class="n">reads_ch</span><span class="o">,</span>
<a id="__codelineno-20-83" name="__codelineno-20-83"></a><span class="w">    </span><span class="n">SAMTOOLS_INDEX</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
<a id="__codelineno-20-84" name="__codelineno-20-84"></a><span class="w">    </span><span class="n">ref_file</span><span class="o">,</span>
<a id="__codelineno-20-85" name="__codelineno-20-85"></a><span class="w">    </span><span class="n">ref_index_file</span><span class="o">,</span>
<a id="__codelineno-20-86" name="__codelineno-20-86"></a><span class="w">    </span><span class="n">ref_dict_file</span><span class="o">,</span>
<a id="__codelineno-20-87" name="__codelineno-20-87"></a><span class="w">    </span><span class="n">intervals_file</span>
<a id="__codelineno-20-88" name="__codelineno-20-88"></a><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p>You should recognize the <code>*.out</code> syntax from Part 1 of this training series; we are telling Nextflow to take the channel output by <code>SAMTOOLS_INDEX</code> and plugging that into the <code>GATK_HAPLOTYPECALLER</code> process call.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You'll notice that the inputs are provided in the exact same order in the call to the process as they are listed in the input block of the process.
In Nextflow, inputs are positional, meaning you <em>must</em> follow the same order; and of course there have to be the same number of elements.</p>
</div>
<h3 id="25-run-the-workflow-to-verify-that-the-variant-calling-step-works">2.5. Run the workflow to verify that the variant calling step works<a class="headerlink" href="#25-run-the-workflow-to-verify-that-the-variant-calling-step-works" title="Permanent link">&para;</a></h3>
<p>Let's run the expanded workflow with <code>-resume</code> so that we don't have to run the indexing step again.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-1.nf<span class="w"> </span>-resume
</code></pre></div>
<p>Now if we look at the console output, we see the two processes listed:</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a>
<a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a><span class="go"> ┃ Launching `genomics-1.nf` [grave_volta] DSL2 - revision: 4790abc96a</span>
<a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a>
<a id="__codelineno-22-5" name="__codelineno-22-5" href="#__codelineno-22-5"></a><span class="go">executor &gt;  local (1)</span>
<a id="__codelineno-22-6" name="__codelineno-22-6" href="#__codelineno-22-6"></a><span class="go">[2a/e69536] SAMTOOLS_INDEX (1)       | 1 of 1, cached: 1 ✔</span>
<a id="__codelineno-22-7" name="__codelineno-22-7" href="#__codelineno-22-7"></a><span class="go">[53/e18e98] GATK_HAPLOTYPECALLER (1) | 1 of 1 ✔</span>
</code></pre></div>
<p>The first process was skipped thanks to the caching, as expected, whereas the second process was run since it's brand new.</p>
<p>You'll find the output file <code>reads_mother.bam.vcf</code> in the results directory, as well its index file (<code>*.vcf.idx</code>). Both are symbolic links to the original files in the work directory where the process call was executed.</p>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a><span class="go">results_genomics/</span>
<a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a><span class="go">├── reads_mother.bam.bai</span>
<a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a><span class="go">├── reads_mother.bam.vcf -&gt; /workspaces/training/nf4-science/genomics/work/53/e18e987d56c47f59b7dd268649ec01/reads_mother.bam.vcf</span>
<a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a><span class="go">└── reads_mother.bam.vcf.idx -&gt; /workspaces/training/nf4-science/genomics/work/53/e18e987d56c47f59b7dd268649ec01/reads_mother.bam.vcf.idx</span>
</code></pre></div>
<p>If you open the VCF file, you should see the same contents as in the file you generated by running the GATK command directly in the container.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">reads_mother.bam.vcf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-24-26">26</a></span>
<span class="normal"><a href="#__codelineno-24-27">27</a></span>
<span class="normal"><a href="#__codelineno-24-28">28</a></span>
<span class="normal"><a href="#__codelineno-24-29">29</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-24-26" name="__codelineno-24-26"></a><span class="gp">#</span>CHROM<span class="w">	</span>POS<span class="w">	</span>ID<span class="w">	</span>REF<span class="w">	</span>ALT<span class="w">	</span>QUAL<span class="w">	</span>FILTER<span class="w">	</span>INFO<span class="w">	</span>FORMAT<span class="w">	</span>reads_mother
<a id="__codelineno-24-27" name="__codelineno-24-27"></a><span class="go">20_10037292_10066351	3480	.	C	CT	503.03	.	AC=2;AF=1.00;AN=2;DP=23;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=27.95;SOR=1.179	GT:AD:DP:GQ:PL	1/1:0,18:18:54:517,54,0</span>
<a id="__codelineno-24-28" name="__codelineno-24-28"></a><span class="go">20_10037292_10066351	3520	.	AT	A	609.03	.	AC=2;AF=1.00;AN=2;DP=18;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=33.83;SOR=0.693	GT:AD:DP:GQ:PL	1/1:0,18:18:54:623,54,0</span>
<a id="__codelineno-24-29" name="__codelineno-24-29"></a><span class="go">20_10037292_10066351	3529	.	T	A	155.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=-0.544;DP=21;ExcessHet=0.0000;FS=1.871;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=7.78;ReadPosRankSum=-1.158;SOR=1.034	GT:AD:DP:GQ:PL	0/1:12,8:20:99:163,0,328</span>
</code></pre></div></td></tr></table></div>
<p>This is the output we care about generating for each sample in our study.</p>
<h3 id="takeaway_2">Takeaway<a class="headerlink" href="#takeaway_2" title="Permanent link">&para;</a></h3>
<p>You know how to make a very basic two-step workflow that does real analysis work and is capable of dealing with genomics file format idiosyncrasies like the accessory files.</p>
<h3 id="whats-next_2">What's next?<a class="headerlink" href="#whats-next_2" title="Permanent link">&para;</a></h3>
<p>Make the workflow handle multiple samples in bulk.</p>
<hr />
<h2 id="3-adapt-the-workflow-to-run-on-a-batch-of-samples">3. Adapt the workflow to run on a batch of samples<a class="headerlink" href="#3-adapt-the-workflow-to-run-on-a-batch-of-samples" title="Permanent link">&para;</a></h2>
<p>It's all well and good to have a workflow that can automate processing on a single sample, but what if you have 1000 samples?
Do you need to write a bash script that loops through all your samples?</p>
<p>No, thank goodness! Just make a minor tweak to the code and Nextflow will handle that for you too.</p>
<h3 id="31-turn-the-input-parameter-declaration-into-an-array-listing-the-three-samples">3.1. Turn the input parameter declaration into an array listing the three samples<a class="headerlink" href="#31-turn-the-input-parameter-declaration-into-an-array-listing-the-three-samples" title="Permanent link">&para;</a></h3>
<p>Let's turn that default file path in the input BAM file declaration into an array listing file paths for our three test samples, up under the <code>Pipeline parameters</code> section.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-25-7">7</a></span>
<span class="normal"><a href="#__codelineno-25-8">8</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-25-7" name="__codelineno-25-7"></a><span class="c1">// Primary input</span>
<a id="__codelineno-25-8" name="__codelineno-25-8"></a><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/bam/reads_mother.bam&quot;</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-26-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-26-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-26-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-26-10">10</a></span>
<span class="normal"><a href="#__codelineno-26-11">11</a></span>
<span class="normal"><a href="#__codelineno-26-12">12</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-26-7" name="__codelineno-26-7"></a><span class="c1">// Primary input (array of three samples)</span>
<a id="__codelineno-26-8" name="__codelineno-26-8"></a><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>
<a id="__codelineno-26-9" name="__codelineno-26-9"></a><span class="w">    </span><span class="s2">&quot;${projectDir}/data/bam/reads_mother.bam&quot;</span><span class="o">,</span>
<a id="__codelineno-26-10" name="__codelineno-26-10"></a><span class="w">    </span><span class="s2">&quot;${projectDir}/data/bam/reads_father.bam&quot;</span><span class="o">,</span>
<a id="__codelineno-26-11" name="__codelineno-26-11"></a><span class="w">    </span><span class="s2">&quot;${projectDir}/data/bam/reads_son.bam&quot;</span>
<a id="__codelineno-26-12" name="__codelineno-26-12"></a><span class="o">]</span>
</code></pre></div></td></tr></table></div>
<p>And that's actually all we need to do, because the channel factory we use in the workflow body (<code>.fromPath</code>) is just as happy to accept multiple file paths to load into the input channel as it was to load a single one.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Normally, you wouldn't want to hardcode the list of samples into your workflow file, but we're doing that here to keep things simple.
We'll present more elegant ways for handling inputs later in this training series.</p>
</div>
<h3 id="32-run-the-workflow-to-verify-that-it-runs-on-all-three-samples">3.2. Run the workflow to verify that it runs on all three samples<a class="headerlink" href="#32-run-the-workflow-to-verify-that-it-runs-on-all-three-samples" title="Permanent link">&para;</a></h3>
<p>Let's try running the workflow now that the plumbing is set up to run on all three test samples.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-1.nf<span class="w"> </span>-resume
</code></pre></div>
<p>Funny thing: this <em>might work</em>, OR it <em>might fail</em>.
If your workflow run succeeded, run it again until you get an error like this:</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-28-2" name="__codelineno-28-2" href="#__codelineno-28-2"></a>
<a id="__codelineno-28-3" name="__codelineno-28-3" href="#__codelineno-28-3"></a><span class="go"> ┃ Launching `genomics-1.nf` [loving_pasteur] DSL2 - revision: d2a8e63076</span>
<a id="__codelineno-28-4" name="__codelineno-28-4" href="#__codelineno-28-4"></a>
<a id="__codelineno-28-5" name="__codelineno-28-5" href="#__codelineno-28-5"></a><span class="go">executor &gt;  local (4)</span>
<a id="__codelineno-28-6" name="__codelineno-28-6" href="#__codelineno-28-6"></a><span class="go">[01/eea165] SAMTOOLS_INDEX (2)       | 3 of 3, cached: 1 ✔</span>
<a id="__codelineno-28-7" name="__codelineno-28-7" href="#__codelineno-28-7"></a><span class="go">[a5/fa9fd0] GATK_HAPLOTYPECALLER (3) | 1 of 3, cached: 1</span>
<a id="__codelineno-28-8" name="__codelineno-28-8" href="#__codelineno-28-8"></a><span class="go">ERROR ~ Error executing process &gt; &#39;GATK_HAPLOTYPECALLER (2)&#39;</span>
<a id="__codelineno-28-9" name="__codelineno-28-9" href="#__codelineno-28-9"></a>
<a id="__codelineno-28-10" name="__codelineno-28-10" href="#__codelineno-28-10"></a><span class="go">Caused by:</span>
<a id="__codelineno-28-11" name="__codelineno-28-11" href="#__codelineno-28-11"></a><span class="go">  Process `GATK_HAPLOTYPECALLER (2)` terminated with an error exit status (2)</span>
<a id="__codelineno-28-12" name="__codelineno-28-12" href="#__codelineno-28-12"></a>
<a id="__codelineno-28-13" name="__codelineno-28-13" href="#__codelineno-28-13"></a><span class="go">Command executed:</span>
<a id="__codelineno-28-14" name="__codelineno-28-14" href="#__codelineno-28-14"></a>
<a id="__codelineno-28-15" name="__codelineno-28-15" href="#__codelineno-28-15"></a><span class="go">  gatk HaplotypeCaller         -R ref.fasta         -I reads_father.bam         -O reads_father.bam.vcf         -L intervals.bed         -ERC GVCF</span>
<a id="__codelineno-28-16" name="__codelineno-28-16" href="#__codelineno-28-16"></a>
<a id="__codelineno-28-17" name="__codelineno-28-17" href="#__codelineno-28-17"></a><span class="go">Command exit status:</span>
<a id="__codelineno-28-18" name="__codelineno-28-18" href="#__codelineno-28-18"></a><span class="go">  2</span>
<a id="__codelineno-28-19" name="__codelineno-28-19" href="#__codelineno-28-19"></a>
<a id="__codelineno-28-20" name="__codelineno-28-20" href="#__codelineno-28-20"></a><span class="go">Command error:</span>
</code></pre></div>
<p>Further down, buried in the GATK command error output, there will be a line like this:</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a><span class="go">A USER ERROR has occurred: Traversal by intervals was requested but some input files are not indexed.</span>
</code></pre></div>
<p>Well, that's weird, considering we explicitly indexed the BAM files in the first step of the workflow. Could there be something wrong with the plumbing?</p>
<h4 id="321-check-the-work-directories-for-the-relevant-calls">3.2.1. Check the work directories for the relevant calls<a class="headerlink" href="#321-check-the-work-directories-for-the-relevant-calls" title="Permanent link">&para;</a></h4>
<p>Let's take a look inside the work directory for the failed <code>GATK_HAPLOTYPECALLER</code> process call listed in the console output.</p>
<div class="highlight"><span class="filename">Directory contents</span><pre><span></span><code><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a><span class="go">work/a5/fa9fd0994b6beede5fb9ea073596c2</span>
<a id="__codelineno-30-2" name="__codelineno-30-2" href="#__codelineno-30-2"></a><span class="go">├── intervals.bed -&gt; /workspaces/training/nf4-science/genomics/data/ref/intervals.bed</span>
<a id="__codelineno-30-3" name="__codelineno-30-3" href="#__codelineno-30-3"></a><span class="go">├── reads_father.bam.bai -&gt; /workspaces/training/nf4-science/genomics/work/01/eea16597bd6e810fb4cf89e60f8c2d/reads_father.bam.bai</span>
<a id="__codelineno-30-4" name="__codelineno-30-4" href="#__codelineno-30-4"></a><span class="go">├── reads_son.bam -&gt; /workspaces/training/nf4-science/genomics/data/bam/reads_son.bam</span>
<a id="__codelineno-30-5" name="__codelineno-30-5" href="#__codelineno-30-5"></a><span class="go">├── reads_son.bam.vcf</span>
<a id="__codelineno-30-6" name="__codelineno-30-6" href="#__codelineno-30-6"></a><span class="go">├── reads_son.bam.vcf.idx</span>
<a id="__codelineno-30-7" name="__codelineno-30-7" href="#__codelineno-30-7"></a><span class="go">├── ref.dict -&gt; /workspaces/training/nf4-science/genomics/data/ref/ref.dict</span>
<a id="__codelineno-30-8" name="__codelineno-30-8" href="#__codelineno-30-8"></a><span class="go">├── ref.fasta -&gt; /workspaces/training/nf4-science/genomics/data/ref/ref.fasta</span>
<a id="__codelineno-30-9" name="__codelineno-30-9" href="#__codelineno-30-9"></a><span class="go">└── ref.fasta.fai -&gt; /workspaces/training/nf4-science/genomics/data/ref/ref.fasta.fai</span>
</code></pre></div>
<p>Pay particular attention to the names of the BAM file and the BAM index that are listed in this directory: <code>reads_son.bam</code> and <code>reads_father.bam.bai</code>.</p>
<p>What the heck? Nextflow has staged an index file in this process call's work directory, but it's the wrong one. How could this have happened?</p>
<h4 id="321-use-the-view-operator-to-inspect-channel-contents">3.2.1. Use the <a href="https://www.nextflow.io/docs/latest/reference/operator.html#view">view() operator</a> to inspect channel contents<a class="headerlink" href="#321-use-the-view-operator-to-inspect-channel-contents" title="Permanent link">&para;</a></h4>
<p>Add these two lines in the workflow body before the <code>GATK_HAPLOTYPER</code> process call:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-31-84">84</a></span>
<span class="normal"><a href="#__codelineno-31-85">85</a></span>
<span class="normal"><a href="#__codelineno-31-86">86</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-31-84" name="__codelineno-31-84"></a><span class="w">    </span><span class="c1">// temporary diagnostics</span>
<a id="__codelineno-31-85" name="__codelineno-31-85"></a><span class="w">    </span><span class="n">reads_ch</span><span class="o">.</span><span class="na">view</span><span class="o">()</span>
<a id="__codelineno-31-86" name="__codelineno-31-86"></a><span class="w">    </span><span class="n">SAMTOOLS_INDEX</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">view</span><span class="o">()</span>
</code></pre></div></td></tr></table></div>
<p>Then run the workflow command again.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-1.nf
</code></pre></div>
<p>You may need to run it several times for it to fail again.
This error will not reproduce consistently because it is dependent on some variability in the execution times of the individual process calls.</p>
<p>This is what the output of the two <code>.view()</code> calls we added looks like for a failed run:</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a><span class="go">/workspaces/training/nf4-science/genomics/data/bam/reads_mother.bam</span>
<a id="__codelineno-33-2" name="__codelineno-33-2" href="#__codelineno-33-2"></a><span class="go">/workspaces/training/nf4-science/genomics/data/bam/reads_father.bam</span>
<a id="__codelineno-33-3" name="__codelineno-33-3" href="#__codelineno-33-3"></a><span class="go">/workspaces/training/nf4-science/genomics/data/bam/reads_son.bam</span>
<a id="__codelineno-33-4" name="__codelineno-33-4" href="#__codelineno-33-4"></a><span class="go">/workspaces/training/nf4-science/genomics/work/9c/53492e3518447b75363e1cd951be4b/reads_father.bam.bai</span>
<a id="__codelineno-33-5" name="__codelineno-33-5" href="#__codelineno-33-5"></a><span class="go">/workspaces/training/nf4-science/genomics/work/cc/37894fffdf6cc84c3b0b47f9b536b7/reads_son.bam.bai</span>
<a id="__codelineno-33-6" name="__codelineno-33-6" href="#__codelineno-33-6"></a><span class="go">/workspaces/training/nf4-science/genomics/work/4d/dff681a3d137ba7d9866e3d9307bd0/reads_mother.bam.bai</span>
</code></pre></div>
<p>The first three lines correspond to the input channel and the second, to the output channel.
You can see that the BAM files and index files for the three samples are not listed in the same order!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When you call a Nextflow process on a channel containing multiple elements, Nextflow will try to parallelize execution as much as possible, and will collect outputs in whatever order they become available.
The consequence is that the corresponding outputs may be collected in a different order than the original inputs were fed in.</p>
</div>
<p>As currently written, our workflow script assumes that the index files will come out of the indexing step listed in the same mother/father/son order as the inputs were given.
But that is not guaranteed to be the case, which is why sometimes (though not always) the wrong files get paired up in the second step.</p>
<p>To fix this, we need to make sure the BAM files and their index files travel together through the channels.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The <code>view()</code> statements in the workflow code don't do anything, so it's not a problem to leave them in.
However they will clutter up your console output, so we recommend removing them when you're done troubleshooting the issue.</p>
</div>
<h3 id="33-change-the-output-of-the-samtools_index-process-into-a-tuple-that-keeps-the-input-file-and-its-index-together">3.3. Change the output of the SAMTOOLS_INDEX process into a tuple that keeps the input file and its index together<a class="headerlink" href="#33-change-the-output-of-the-samtools_index-process-into-a-tuple-that-keeps-the-input-file-and-its-index-together" title="Permanent link">&para;</a></h3>
<p>The simplest way to ensure a BAM file and its index stay closely associated is to package them together into a tuple coming out of the index task.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A <strong>tuple</strong> is a finite, ordered list of elements that is commonly used for returning multiple values from a function. Tuples are particularly useful for passing multiple inputs or outputs between processes while preserving their association and order.</p>
</div>
<p>First, let's change the output of the <code>SAMTOOLS_INDEX</code> process to include the BAM file in its output declaration.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-34-32">32</a></span>
<span class="normal"><a href="#__codelineno-34-33">33</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-34-32" name="__codelineno-34-32"></a><span class="nl">output:</span>
<a id="__codelineno-34-33" name="__codelineno-34-33"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;${input_bam}.bai&quot;</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-35-32">32</a></span>
<span class="normal"><a href="#__codelineno-35-33">33</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-35-32" name="__codelineno-35-32"></a><span class="nl">output:</span>
<a id="__codelineno-35-33" name="__codelineno-35-33"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">input_bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;${input_bam}.bai&quot;</span><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p>This way, each index file will be tightly coupled with its original BAM file, and the overall output of the indexing step will be a single channel containing pairs of files.</p>
<h3 id="34-change-the-input-to-the-gatk_haplotypecaller-process-to-be-a-tuple">3.4. Change the input to the GATK_HAPLOTYPECALLER process to be a tuple<a class="headerlink" href="#34-change-the-input-to-the-gatk_haplotypecaller-process-to-be-a-tuple" title="Permanent link">&para;</a></h3>
<p>Since we've changed the 'shape' of the output of the first process in the workflow, we need to update the input definition of the second process to match.</p>
<p>Specifically, where we previously declared two separate input paths in the input block of the <code>GATK_HAPLOTYPECALLER</code> process, we now declare a single input matching the structure of the tuple emitted by <code>SAMTOOLS_INDEX</code>.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-36-49">49</a></span>
<span class="normal"><a href="#__codelineno-36-50">50</a></span>
<span class="normal"><a href="#__codelineno-36-51">51</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-36-49" name="__codelineno-36-49"></a><span class="nl">input:</span>
<a id="__codelineno-36-50" name="__codelineno-36-50"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">input_bam</span>
<a id="__codelineno-36-51" name="__codelineno-36-51"></a><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">input_bam_index</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-37-49">49</a></span>
<span class="normal"><a href="#__codelineno-37-50">50</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-37-49" name="__codelineno-37-49"></a><span class="nl">input:</span>
<a id="__codelineno-37-50" name="__codelineno-37-50"></a><span class="w">    </span><span class="n">tuple</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">input_bam</span><span class="o">),</span><span class="w"> </span><span class="n">path</span><span class="o">(</span><span class="n">input_bam_index</span><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p>Of course, since we've now changed the shape of the inputs that <code>GATK_HAPLOTYPECALLER</code> expects, we need to update the process call accordingly in the workflow body.</p>
<h3 id="35-update-the-call-to-gatk_haplotypecaller-in-the-workflow-block">3.5. Update the call to GATK_HAPLOTYPECALLER in the workflow block<a class="headerlink" href="#35-update-the-call-to-gatk_haplotypecaller-in-the-workflow-block" title="Permanent link">&para;</a></h3>
<p>We no longer need to provide the original <code>reads_ch</code> to the <code>GATK_HAPLOTYPECALLER</code> process, since the BAM file is now bundled (in the form of a symlink) into the channel output by <code>SAMTOOLS_INDEX</code>.</p>
<p>As a result, we can simply delete that line.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-38-84">84</a></span>
<span class="normal"><a href="#__codelineno-38-85">85</a></span>
<span class="normal"><a href="#__codelineno-38-86">86</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-38-84" name="__codelineno-38-84"></a><span class="n">GATK_HAPLOTYPECALLER</span><span class="o">(</span>
<a id="__codelineno-38-85" name="__codelineno-38-85"></a><span class="w">    </span><span class="n">reads_ch</span><span class="o">,</span>
<a id="__codelineno-38-86" name="__codelineno-38-86"></a><span class="w">    </span><span class="n">SAMTOOLS_INDEX</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-39-84">84</a></span>
<span class="normal"><a href="#__codelineno-39-85">85</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-39-84" name="__codelineno-39-84"></a><span class="n">GATK_HAPLOTYPECALLER</span><span class="o">(</span>
<a id="__codelineno-39-85" name="__codelineno-39-85"></a><span class="w">    </span><span class="n">SAMTOOLS_INDEX</span><span class="o">.</span><span class="na">out</span><span class="o">,</span>
</code></pre></div></td></tr></table></div>
<p>That is all the re-wiring that is necessary to solve the index mismatch problem.</p>
<h3 id="36-run-the-workflow-to-verify-it-works-correctly-on-all-three-samples-every-time">3.6. Run the workflow to verify it works correctly on all three samples every time<a class="headerlink" href="#36-run-the-workflow-to-verify-it-works-correctly-on-all-three-samples-every-time" title="Permanent link">&para;</a></h3>
<p>Of course, the proof is in the pudding, so let's run the workflow again a few times to make sure this will work reliably going forward.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-40-1" name="__codelineno-40-1" href="#__codelineno-40-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-1.nf
</code></pre></div>
<p>This time (and every time) everything should run correctly:</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-41-1" name="__codelineno-41-1" href="#__codelineno-41-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-41-2" name="__codelineno-41-2" href="#__codelineno-41-2"></a>
<a id="__codelineno-41-3" name="__codelineno-41-3" href="#__codelineno-41-3"></a><span class="go"> ┃ Launching `genomics-1.nf` [special_goldstine] DSL2 - revision: 4cbbf6ea3e</span>
<a id="__codelineno-41-4" name="__codelineno-41-4" href="#__codelineno-41-4"></a>
<a id="__codelineno-41-5" name="__codelineno-41-5" href="#__codelineno-41-5"></a><span class="go">executor &gt;  local (6)</span>
<a id="__codelineno-41-6" name="__codelineno-41-6" href="#__codelineno-41-6"></a><span class="go">[d6/10c2c4] SAMTOOLS_INDEX (1)       | 3 of 3 ✔</span>
<a id="__codelineno-41-7" name="__codelineno-41-7" href="#__codelineno-41-7"></a><span class="go">[88/1783aa] GATK_HAPLOTYPECALLER (2) | 3 of 3 ✔</span>
</code></pre></div>
<p>If you'd like, you can use <code>.view()</code> again to peek at what the contents of the <code>SAMTOOLS_INDEX</code> output channel looks like:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-42-92">92</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-42-92" name="__codelineno-42-92"></a><span class="n">SAMTOOLS_INDEX</span><span class="o">.</span><span class="na">out</span><span class="o">.</span><span class="na">view</span><span class="o">()</span>
</code></pre></div></td></tr></table></div>
<p>You'll see the channel contains the three expected tuples (file paths truncated for readability).</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-43-1" name="__codelineno-43-1" href="#__codelineno-43-1"></a><span class="go">[.../4c/e16099*/reads_son.bam, .../4c/e16099*/reads_son.bam.bai]</span>
<a id="__codelineno-43-2" name="__codelineno-43-2" href="#__codelineno-43-2"></a><span class="go">[.../42/e70b8b*/reads_father.bam, .../42/e70b8b*/reads_father.bam.bai]</span>
<a id="__codelineno-43-3" name="__codelineno-43-3" href="#__codelineno-43-3"></a><span class="go">[.../18/23b4bb*/reads_mother.bam, .../18/23b4bb*/reads_mother.bam.bai]</span>
</code></pre></div>
<p>That will be much safer, going forward.</p>
<h3 id="takeaway_3">Takeaway<a class="headerlink" href="#takeaway_3" title="Permanent link">&para;</a></h3>
<p>You know how to make your workflow run on multiple samples (independently).</p>
<h3 id="whats-next_3">What's next?<a class="headerlink" href="#whats-next_3" title="Permanent link">&para;</a></h3>
<p>Make it easier to handle samples in bulk.</p>
<hr />
<h2 id="4-make-the-workflow-accept-a-text-file-containing-a-batch-of-input-files">4. Make the workflow accept a text file containing a batch of input files<a class="headerlink" href="#4-make-the-workflow-accept-a-text-file-containing-a-batch-of-input-files" title="Permanent link">&para;</a></h2>
<p>A very common way to provide multiple data input files to a workflow is to do it with a text file containing the file paths.
It can be as simple as a text file listing one file path per line and nothing else, or the file can contain additional metadata, in which case it's often called a samplesheet.</p>
<p>Here we are going to show you how to do the simple case.</p>
<h3 id="41-examine-the-provided-text-file-listing-the-input-file-paths">4.1. Examine the provided text file listing the input file paths<a class="headerlink" href="#41-examine-the-provided-text-file-listing-the-input-file-paths" title="Permanent link">&para;</a></h3>
<p>We already made a text file listing the input file paths, called <code>sample_bams.txt</code>, which you can find in the <code>data/</code> directory.</p>
<div class="highlight"><span class="filename">sample_bams.txt</span><pre><span></span><code><a id="__codelineno-44-1" name="__codelineno-44-1" href="#__codelineno-44-1"></a>/workspaces/training/nf4-science/genomics/data/bam/reads_mother.bam
<a id="__codelineno-44-2" name="__codelineno-44-2" href="#__codelineno-44-2"></a>/workspaces/training/nf4-science/genomics/data/bam/reads_father.bam
<a id="__codelineno-44-3" name="__codelineno-44-3" href="#__codelineno-44-3"></a>/workspaces/training/nf4-science/genomics/data/bam/reads_son.bam
</code></pre></div>
<p>As you can see, we listed one file path per line, and they are absolute paths.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The files we are using here are just on your GitHub Codespaces's local filesystem, but we could also point to files in cloud storage.</p>
</div>
<h3 id="42-update-the-parameter-default">4.2. Update the parameter default<a class="headerlink" href="#42-update-the-parameter-default" title="Permanent link">&para;</a></h3>
<p>Let's switch the default value for our <code>reads_bam</code> input parameter to point to the <code>sample_bams.txt</code> file.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-45-7"> 7</a></span>
<span class="normal"><a href="#__codelineno-45-8"> 8</a></span>
<span class="normal"><a href="#__codelineno-45-9"> 9</a></span>
<span class="normal"><a href="#__codelineno-45-10">10</a></span>
<span class="normal"><a href="#__codelineno-45-11">11</a></span>
<span class="normal"><a href="#__codelineno-45-12">12</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-45-7" name="__codelineno-45-7"></a><span class="c1">// Primary input</span>
<a id="__codelineno-45-8" name="__codelineno-45-8"></a><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">[</span>
<a id="__codelineno-45-9" name="__codelineno-45-9"></a><span class="w">    </span><span class="s2">&quot;${projectDir}/data/bam/reads_mother.bam&quot;</span><span class="o">,</span>
<a id="__codelineno-45-10" name="__codelineno-45-10"></a><span class="w">    </span><span class="s2">&quot;${projectDir}/data/bam/reads_father.bam&quot;</span><span class="o">,</span>
<a id="__codelineno-45-11" name="__codelineno-45-11"></a><span class="w">    </span><span class="s2">&quot;${projectDir}/data/bam/reads_son.bam&quot;</span>
<a id="__codelineno-45-12" name="__codelineno-45-12"></a><span class="o">]</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-46-7">7</a></span>
<span class="normal"><a href="#__codelineno-46-8">8</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-46-7" name="__codelineno-46-7"></a><span class="c1">// Primary input (file of input files, one per line)</span>
<a id="__codelineno-46-8" name="__codelineno-46-8"></a><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;${projectDir}/data/sample_bams.txt&quot;</span>
</code></pre></div></td></tr></table></div>
<p>This way we can continue to be lazy, but the list of files no longer lives in the workflow code itself, which is a big step in the right direction.</p>
<h3 id="43-update-the-channel-factory-to-read-lines-from-a-file">4.3. Update the channel factory to read lines from a file<a class="headerlink" href="#43-update-the-channel-factory-to-read-lines-from-a-file" title="Permanent link">&para;</a></h3>
<p>Currently, our input channel factory treats any files we give it as the data inputs we want to feed to the indexing process.
Since we're now giving it a file that lists input file paths, we need to change its behavior to parse the file and treat the file paths it contains as the data inputs.</p>
<p>Fortunately we can do that very simply, just by adding the <a href="https://www.nextflow.io/docs/latest/reference/operator.html#operator-splittext"><code>.splitText()</code> operator</a> to the channel construction step.</p>
<p><em>Before:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-47-68">68</a></span>
<span class="normal"><a href="#__codelineno-47-69">69</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-47-68" name="__codelineno-47-68"></a><span class="c1">// Create input channel (single file via CLI parameter)</span>
<a id="__codelineno-47-69" name="__codelineno-47-69"></a><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="o">)</span>
</code></pre></div></td></tr></table></div>
<p><em>After:</em></p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">genomics-1.nf</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"><a href="#__codelineno-48-68">68</a></span>
<span class="normal"><a href="#__codelineno-48-69">69</a></span></pre></div></td><td class="code"><div><pre><span></span><code><a id="__codelineno-48-68" name="__codelineno-48-68"></a><span class="c1">// Create input channel from a text file listing input file paths</span>
<a id="__codelineno-48-69" name="__codelineno-48-69"></a><span class="n">reads_ch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Channel</span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">reads_bam</span><span class="o">).</span><span class="na">splitText</span><span class="o">()</span>
</code></pre></div></td></tr></table></div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>This is another great opportunity to use the <code>.view()</code> operator to look at what the channel contents look like before and after applying an operator.</p>
</div>
<h3 id="44-run-the-workflow-to-verify-that-it-works-correctly">4.4. Run the workflow to verify that it works correctly<a class="headerlink" href="#44-run-the-workflow-to-verify-that-it-works-correctly" title="Permanent link">&para;</a></h3>
<p>Let's run the workflow one more time.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-49-1" name="__codelineno-49-1" href="#__codelineno-49-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>genomics-1.nf<span class="w"> </span>-resume
</code></pre></div>
<p>This should produce the same result as before, right?</p>
<div class="highlight"><span class="filename">Output</span><pre><span></span><code><a id="__codelineno-50-1" name="__codelineno-50-1" href="#__codelineno-50-1"></a><span class="go"> N E X T F L O W   ~  version 24.10.0</span>
<a id="__codelineno-50-2" name="__codelineno-50-2" href="#__codelineno-50-2"></a>
<a id="__codelineno-50-3" name="__codelineno-50-3" href="#__codelineno-50-3"></a><span class="go"> ┃ Launching `genomics-1.nf` [sick_albattani] DSL2 - revision: 46d84642f6</span>
<a id="__codelineno-50-4" name="__codelineno-50-4" href="#__codelineno-50-4"></a>
<a id="__codelineno-50-5" name="__codelineno-50-5" href="#__codelineno-50-5"></a><span class="go">[18/23b4bb] SAMTOOLS_INDEX (1)       | 3 of 3, cached: 3 ✔</span>
<a id="__codelineno-50-6" name="__codelineno-50-6" href="#__codelineno-50-6"></a><span class="go">[12/f727bb] GATK_HAPLOTYPECALLER (3) | 3 of 3, cached: 3 ✔</span>
</code></pre></div>
<p>Yes! In fact, Nextflow correctly detects that the process calls are exactly the same, and doesn't even bother re-running everything, since we were running with <code>-resume</code>.</p>
<p>And that's it! Our simple variant calling workflow has all the basic features we wanted.</p>
<h3 id="takeaway_4">Takeaway<a class="headerlink" href="#takeaway_4" title="Permanent link">&para;</a></h3>
<p>You know how to make a multi-step linear workflow to index a BAM file and apply per-sample variant calling using GATK.</p>
<p>More generally, you've learned how to use essential Nextflow components and logic to build a simple genomics pipeline that does real work, taking into account the idiosyncrasies of genomics file formats and tool requirements.</p>
<h3 id="whats-next_4">What's next?<a class="headerlink" href="#whats-next_4" title="Permanent link">&para;</a></h3>
<p>Celebrate your success and take an extra long break!</p>
<p>In the next part of this course, you'll learn how to use a few additional Nextflow features (including more channel operators) to apply joint variant calling to the data.</p>









  




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Voltar ao topo
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Rodapé" >
        
          
          <a href="../00_orientation/" class="md-footer__link md-footer__link--prev" aria-label="Anterior: Orientation">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Anterior
              </span>
              <div class="md-ellipsis">
                Orientation
              </div>
            </div>
          </a>
        
        
          
          <a href="../02_joint_calling/" class="md-footer__link md-footer__link--next" aria-label="Próximo: Part 2: Joint calling on a cohort">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Próximo
              </span>
              <div class="md-ellipsis">
                Part 2: Joint calling on a cohort
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  <div>
    Copyright &copy; 2020-2025,
    <a href="https://seqera.io" target="_blank" rel="noopener">Seqera</a>.
  </div>
  <div class="footer_cc_link">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512">
      <!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.-->
      <path
        d="m245.83 214.87-33.22 17.28c-9.43-19.58-25.24-19.93-27.46-19.93-22.13 0-33.22 14.61-33.22 43.84 0 23.57 9.21 43.84 33.22 43.84 14.47 0 24.65-7.09 30.57-21.26l30.55 15.5c-6.17 11.51-25.69 38.98-65.1 38.98-22.6 0-73.96-10.32-73.96-77.05 0-58.69 43-77.06 72.63-77.06 30.72-.01 52.7 11.95 65.99 35.86zm143.05 0-32.78 17.28c-9.5-19.77-25.72-19.93-27.9-19.93-22.14 0-33.22 14.61-33.22 43.84 0 23.55 9.23 43.84 33.22 43.84 14.45 0 24.65-7.09 30.54-21.26l31 15.5c-2.1 3.75-21.39 38.98-65.09 38.98-22.69 0-73.96-9.87-73.96-77.05 0-58.67 42.97-77.06 72.63-77.06 30.71-.01 52.58 11.95 65.56 35.86zM247.56 8.05C104.74 8.05 0 123.11 0 256.05c0 138.49 113.6 248 247.56 248 129.93 0 248.44-100.87 248.44-248 0-137.87-106.62-248-248.44-248zm.87 450.81c-112.54 0-203.7-93.04-203.7-202.81 0-105.42 85.43-203.27 203.72-203.27 112.53 0 202.82 89.46 202.82 203.26-.01 121.69-99.68 202.82-202.84 202.82z"
      ></path>
    </svg>
    <a
      href="https://creativecommons.org/licenses/by-nc-nd/4.0/"
      target="_blank"
      rel="noopener"
      title="Creative Commons CC BY-NC-ND 4.0"
    >
      Creative Commons BY-NC-ND 4.0</a
    >.
  </div>
   Made with
  <a
    href="https://squidfunk.github.io/mkdocs-material/"
    target="_blank"
    rel="noopener"
  >
    Material for MkDocs
  </a>
  
</div>
      
        <div class="md-social">
  <a
    href="https://seqera.io"
    target="_blank"
    rel="noopener"
    class="footer_seqera_logo"
  >
    <img src="/assets/img/seqera_logo_white.png" />
  </a>
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
      <div class="md-consent" data-md-component="consent" id="__consent" hidden>
        <div class="md-consent__overlay"></div>
        <aside class="md-consent__inner">
          <form class="md-consent__form md-grid md-typeset" name="consent">
            


  


<h4>Cookie consent</h4>
<p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our training materials better.</p>
<input class="md-toggle" type="checkbox" id="__settings" >
<div class="md-consent__settings">
  <ul class="task-list">
    
    
      
        
  
  
    
    
  
  <li class="task-list-item">
    <label class="task-list-control">
      <input type="checkbox" name="analytics" checked>
      <span class="task-list-indicator"></span>
      Google Analytics
    </label>
  </li>

      
    
    
      
        
  
  
    
    
  
  <li class="task-list-item">
    <label class="task-list-control">
      <input type="checkbox" name="github" checked>
      <span class="task-list-indicator"></span>
      GitHub
    </label>
  </li>

      
    
    
  </ul>
</div>
<div class="md-consent__controls">
  
    
      <button class="md-button md-button--primary">Aceitar</button>
    
    
    
  
    
    
    
      <label class="md-button" for="__settings">Gerir configurações</label>
    
  
</div>
          </form>
        </aside>
      </div>
      <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout((function(){document.querySelector("[data-md-component=consent]").hidden=!1}),250);var form=document.forms.consent;for(var action of["submit","reset"])form.addEventListener(action,(function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map((function(e){return[e,!0]})))),location.hash="",location.reload()}))</script>
    
    
      
      <script id="__config" type="application/json">{"base": "../../../..", "features": ["content.action.edit", "content.code.annotate", "content.code.copy", "navigation.footer", "navigation.instant", "navigation.top", "navigation.tracking", "search.share", "toc.follow"], "search": "../../../../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copiado para \u00e1rea de transfer\u00eancia", "clipboard.copy": "Copiar para \u00e1rea de transfer\u00eancia", "search.result.more.one": "Mais 1 nesta p\u00e1gina", "search.result.more.other": "Mais # nesta p\u00e1gina", "search.result.none": "Nenhum resultado encontrado", "search.result.one": "1 resultado encontrado", "search.result.other": "# resultados encontrados", "search.result.placeholder": "Digite para iniciar a busca", "search.result.term.missing": "Ausente", "select.version": "Selecione a vers\u00e3o"}, "version": {"alias": true, "provider": "mike"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>