# `nf-core/demo`

[`nf-core/demo`](https://nf-co.re/demo/) is a simple nf-core style pipeline for workshops and demonstrations.

It was created using the full nf-core template and is designed to run and configure quickly.

<figure class="excalidraw">
--8<-- "docs/nf_customize/img/subway.excalidraw.svg"
</figure>

The [`nf-core/demo`](https://nf-co.re/demo/) pipeline consists of three processes:

- ([`FastQC`](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)): Read quality control
- ([`SEQTK_TRIM`](https://github.com/lh3/seqtk)): Trim low quality bases from FASTQ files
- ([`MULTIQC`](http://multiqc.info/)): Present quality control reports for raw reads

[`nf-core/demo`](https://nf-co.re/demo/) takes a samplesheet that contains paths to FASTQ files as an input and will produce four output folders with logs and reports:

- `fastqc/`
    - `*_fastqc.html`: FastQC report containing quality metrics.
    - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.
- `fq/`
    - `*.fastp.html`: Trimmed fq files.
- `multiqc/`
    - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
    - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
    - `multiqc_plots/`: directory containing static images from the report in various formats.
- `pipeline_info/`
    - Reports generated by Nextflow
    - Reports generated by nf-core
    - Parameters file

You can view the code for this pipeline on the [`nf-core/demo` GitHub repository](https://github.com/nf-core/demo).

To help you understand the expectations for running an nf-core pipeline, they come with extensive documentation about its parameters, usage, and outputs.

The documentation for the `nf-core/demo` pipeline can be found [on the nf-core/demo pipelines page](https://nf-co.re/demo/docs/usage).

## Required inputs

Before running any nf-core pipeline you will need to check if any parameters are required.

You can find required parameters on the pipelines parameters page.

The [parameters page of the `nf-core/demo` pipeline](https://nf-co.re/demo/dev/parameters) shows that this pipeline requires two parameters (`--input` and `--outdir`) to run.

![nf-core/demo parameters](img/demo-parameters.png)

Without these, the pipeline will not launch and will throw an error.

### `--input`

The `--input` parameter requires a path to comma-separated file (CSV) containing information about the samples in the experiment:

```bash
--input 'path/to/samplesheet.csv'
```

The [nf-core/demo usage documentation](https://nf-co.re/demo/dev/docs/usage/) describes the required `--input` as a comma-separated file (`.csv`). The `.csv` file must contain 3 columns with the headers `sample`, `fastq_1`, and `fastq_2`.

The samplesheet file may consist of both single- and paired-end data and may look something like the one below:

```csv title="samplesheet.csv" linenums="1"
sample,fastq_1,fastq_2
SAMPLE1_PE,path/to/sample1_R1.fastq.gz,path/to/sample1_R2.fastq.gz
SAMPLE2_PE,path/to/sample2_R1.fastq.gz,path/to/sample2_R2.fastq.gz
SAMPLE3_SE,path/to/sample3_R1.fastq.gz,
```

### `--outdir`

The `--output` parameter is used to name the output directory where the results will be saved. It takes a string as its input:

```bash
--output results
```

!!! note

    You do not need to create the output folder before you run the pipeline.

## Testing `nf-core/demo` with profiles

A profile is a set of configuration attributes that can be added to your execution command by using the `-profile` option:

```bash
-profile <profile name>
```

Configuration profiles are defined using the special scope `profile` within configuration files. Profiles group the attributes that belong to the same profile using a common prefix:

```console title="example.config" linenums="1"
profiles {
  foo {
    process.memory = '2 GB'
  }

  bar {
    process.memory = '4 GB'
  }
}
```

### `-profile test`

Every nf-core pipeline comes with a `test` profile. This is a minimal set of configuration settings for the pipeline to run using a small test dataset that is hosted on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.

The `test` profile is expected to run and can be used to help diagnose local issues before you scale up your analysis.

The `test` profile for `nf-core/demo` is shown below:

```groovy title="conf/test.config" linenums="1"
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Nextflow config file for running minimal tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Defines input files and everything required to run a fast and simple pipeline test.

    Use as follows:
        nextflow run nf-core/demo -profile test,<docker/singularity> --outdir <OUTDIR>

----------------------------------------------------------------------------------------
*/

process {
    resourceLimits = [
        cpus: 4,
        memory: '15.GB',
        time: '1.h'
    ]
}

params {
    config_profile_name        = 'Test profile'
    config_profile_description = 'Minimal test dataset to check pipeline function'

    // Input data
    input  = 'https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/samplesheet/samplesheet_test_illumina_amplicon.csv'
}
```

The `nf-core/demo` `test` profile already contains the input parameter (explained in more detail below). This means that the `--input` parameter does not need to be added to the execution command. However, the `outdir` parameter is not included in the `test` profile and must be added to the execution command using the `--outdir` flag.

```bash
nextflow run nf-core/demo -profile test --outdir results
```

!!! question "Exercise"

    Execute the `nf-core/demo` pipeline with the `test` profile and an output directory named `results`:

    ```bash
    nextflow run nf-core/demo -profile test --outdir results
    ```

    !!! warning "This execution is expected to fail!"

As the software required to run each process (e.g., seqtk) is not available in the Gitpod environment the exercise above is expected to fail:

```console
Caused by:
  Process `NFCORE_DEMO:DEMO:SEQTK_TRIM (SAMPLE2_PE)` terminated with an error exit status (127)
<truncated>
```

Fortunately, nf-core pipelines come packed with directives for containers and environments that can be flexibly enabled using profiles for different software (e.g., `docker`, `singularity`, and `conda`):

```bash
-profile singularity
```

In Gitpod, you can add the `singularity` profile to your execution command and Nextflow will download and enable Singularity software images to run each process.

The singularity profile is defined in the nextflow.config file in the main pipeline repository.

```groovy title="nextflow.config" linenums="100"
singularity {
    singularity.enabled     = true
    singularity.autoMounts  = true
    conda.enabled           = false
    docker.enabled          = false
    podman.enabled          = false
    shifter.enabled         = false
    charliecloud.enabled    = false
    apptainer.enabled       = false
}
```

!!! note "Multiple config files"

    Multiple profiles can be included by separating them with a comma (e.g., `-profile test,singularity`).

!!! question "Exercise"

    Amend your run command by adding the singularity profile:

    ```bash
    nextflow run nf-core/demo -profile test,singularity --outdir results
    ```

The `nf-core/demo` pipeline should now run successfully!

!!! note

    Singularity must be installed for this command to run.

## Using your own data

Instead of using the `test` profile you can use the `--input` parameter to choose your own sample sheet as an input.

As described above, the input is a CSV file with 3 columns and the headers `sample`, `fastq_1`, and `fastq_2`.

The nf-core/demo pipeline will auto-detect whether a sample is single- or paired-end and if a sample has been sequenced more than once using the information provided in the sample sheet by default.

!!! question "Exercise"

    Create a sample sheet for the paired-end reads for gut, liver, and lung samples in the data folder:

    1. Create a CSV file named `samplesheet.csv`:

    ```bash
    code samplesheet.csv
    ```

    2. Add the header line, and, for each sample, an id and the complete paths to the paired-end reads:

    ```csv title="samplesheet.csv" linenums="1"
    sample,fastq_1,fastq_2
    gut,/workspace/gitpod/nf-customize/data/gut_1.fastq.gz,/workspace/gitpod/nf-customize/data/gut_2.fastq.gz
    liver,/workspace/gitpod/nf-customize/data/liver_1.fastq.gz,/workspace/gitpod/nf-customize/data/liver_2.fastq.gz
    lung,/workspace/gitpod/nf-customize/data/lung_1.fastq.gz,/workspace/gitpod/nf-customize/data/lung_2.fastq.gz
    ```

    !!! warning "Make sure you save this file in your working directory (`/workspace/gitpod/nf-customize/`)"

You can use you new samplesheet with the `--input` parameter in your execution command.

In this case, the other parameters in the test profile (e.g., `config_profile_name`) can be ignored as they are not explicitly required by the pipeline or in this Gitpod environment.

!!! question "Exercise"

    Run the `nf-core/demo` pipeline with the `singularity` profile and your newly created samplesheet as your input.

    ```bash
    nextflow run nf-core/demo -profile singularity --input samplesheet.csv --outdir results
    ```

The pipeline should run successfully!

---

Congratulations! You have now run the `nf-core/demo` pipeline using you own dataset!
